From b64b76a5dd287d2ee740b49466abfecd67ecdf03 Mon Sep 17 00:00:00 2001
From: Rajanikant Chirmade <rajanikant_chirmade@xyratex.com>
Date: Wed, 23 Oct 2013 09:08:55 +0530
Subject: [PATCH 007/121]  - Removed db4 related members.

---
 stats/util/stats_util.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/stats/util/stats_util.c b/stats/util/stats_util.c
index 4c665b6..bae1ee9 100644
--- a/stats/util/stats_util.c
+++ b/stats/util/stats_util.c
@@ -55,12 +55,9 @@ enum {
 
 static struct m0_net_xprt        *xprt           = &m0_net_lnet_xprt;
 static struct m0_net_domain       client_net_dom = { };
-static struct m0_dbenv            client_dbenv;
 
 static struct m0_rpc_client_ctx cctx = {
 	.rcx_net_dom               = &client_net_dom,
-	.rcx_db_name               = "m0stats_client_db",
-	.rcx_dbenv                 = &client_dbenv,
 	.rcx_max_rpcs_in_flight    = 1,
 	.rcx_nr_slots              = STATS_SESSION_SLOTS,
 	.rcx_max_rpc_msg_size      = M0_RPC_DEF_MAX_RPC_MSG_SIZE,
-- 
1.8.3.2

