From 5277708e17ab7dfade1bb36b2c639dbb1aeb56a5 Mon Sep 17 00:00:00 2001
From: Rohan Puri <rohan_puri@xyratex.com>
Date: Mon, 15 Jul 2013 16:51:31 +0530
Subject: [PATCH 063/121] Fix comment style

---
 mero/magic.h | 1 -
 1 file changed, 1 deletion(-)

diff --git a/mero/magic.h b/mero/magic.h
index 3c74651..e23b06f 100644
--- a/mero/magic.h
+++ b/mero/magic.h
@@ -117,7 +117,6 @@ enum m0_magic_satchel {
 	/* ADDB monitor list's head (addb hellhole) */
 	M0_ADDB_MONITOR_LIST_HEAD_MAGIC = 0x33addb3704773477,
 
-	/* ADDB monitor (addb shoebill)*/
 	M0_ADDB_MONITOR_LIST_LINK_MAGIC = 0x33addb7719304577,
 
 /* balloc */
-- 
1.8.3.2

