From 77fe5793ffe9807940cb6c5ca249cecee52cf25b Mon Sep 17 00:00:00 2001
From: Rohan Puri <rohan_puri@xyratex.com>
Date: Wed, 31 Jul 2013 18:29:42 +0530
Subject: [PATCH 087/121] Addressed review comments (round 5)

---
 addb/addb_svc.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/addb/addb_svc.c b/addb/addb_svc.c
index e4d27af..8b6de83 100644
--- a/addb/addb_svc.c
+++ b/addb/addb_svc.c
@@ -183,7 +183,6 @@ static int addb_svc_rsto_service_allocate(struct m0_reqh_service **service,
 	m0_cond_init(&svc->as_cond, &(*service)->rs_mutex);
 	addb_svc_bob_init(svc);
 
-
 	M0_POST(addb_svc_invariant(svc));
 
 	return 0;
-- 
1.8.3.2

