From 1cac86e37991607214e54c14f50f1149283bbd05 Mon Sep 17 00:00:00 2001
From: Rohan Puri <rohan_puri@xyratex.com>
Date: Mon, 26 Aug 2013 14:34:42 +0530
Subject: [PATCH 089/121] Addressed reviews comments

---
 addb/addb_monitor.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/addb/addb_monitor.c b/addb/addb_monitor.c
index 19ad726..6e3487c 100644
--- a/addb/addb_monitor.c
+++ b/addb/addb_monitor.c
@@ -367,7 +367,6 @@ M0_INTERNAL int m0_addb_monitor_summaries_post(struct m0_reqh *reqh)
 		m0_mutex_unlock(&sum->asr_mutex);
 	}m0_tl_endfor;
 
-
 	/**
 	 * @todo: Post on global machine too.
 	 */
-- 
1.8.3.2

