From b3237a3550d452b2b2ec9d01e6114a8a9e3d99ec Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Wed, 29 May 2013 13:54:58 +0300
Subject: [PATCH 034/290] be: fixed typo and compiler error

---
 be/seg.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/be/seg.c b/be/seg.c
index 2ca2143..a4c6630 100644
--- a/be/seg.c
+++ b/be/seg.c
@@ -63,7 +63,7 @@ static void seg_header_prepare(struct m0_be_seg *seg, struct m0_stob_io *io)
 	struct be_seg_hdr *header = (struct be_seg_hdr *)seg->bs_addr;
 
 	/* fill segment header */
-	header->bh_addr = BE_SEG_DEAULT_ADDR;
+	header->bh_addr = BE_SEG_DEFAULT_ADDR;
 	header->bh_size = sizeof(struct be_seg_hdr);
 
 	seg_io_prepare(seg, io);
@@ -98,7 +98,7 @@ int m0_be_seg_create(struct m0_be_seg *seg)
 	M0_PRE(seg->bs_stob->so_domain != NULL);
 	M0_PRE(seg->bs_stob->so_state != CSS_EXISTS);
 
-	return m0_stob_create(&seg->bs_stob, NULL) ?: seg_header_write(seg);
+	return m0_stob_create(seg->bs_stob, NULL) ?: seg_header_write(seg);
 }
 
 int m0_be_seg_destroy(struct m0_be_seg *seg)
-- 
1.8.3.2

