From 53bc50f5563def625ca07f20014769e4eea94d0d Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Wed, 29 May 2013 17:30:48 +0300
Subject: [PATCH 047/290] be/seg: pass MAP_FIXED to mmap() and verify that p ==
 seg_addr0

---
 be/seg.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/be/seg.c b/be/seg.c
index 0494e57..eb9e81a 100644
--- a/be/seg.c
+++ b/be/seg.c
@@ -213,9 +213,10 @@ int m0_be_seg_open(struct m0_be_seg *seg)
 
 	/* mmap an area at bh_addr of bh_size. */
 	p = mmap(seg_addr0, seg_size, PROT_READ|PROT_WRITE,
-		 MAP_ANONYMOUS, -1, 0);
+		 MAP_FIXED|MAP_ANONYMOUS, -1, 0);
 	if (p == MAP_FAILED)
 		return -errno;
+	M0_ASSERT(p == seg_addr0);
 
 	/* Read whole segment from storage. */
 	rc = stob_io_single_read(seg_addr0, seg_size, seg->bs_stob,
-- 
1.8.3.2

