From 4ac7bf393df9c2f0d165d11c21d5127d71157184 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Wed, 29 May 2013 18:12:25 +0300
Subject: [PATCH 052/290] be/seg: use EOPNOTSUPP in __KERNEL__ mode, declare *p
 in userspace only

---
 be/seg.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/be/seg.c b/be/seg.c
index 6969e01..d1c21be 100644
--- a/be/seg.c
+++ b/be/seg.c
@@ -187,8 +187,10 @@ int m0_be_seg_open(struct m0_be_seg *seg)
 	struct be_seg_hdr *hdrbuf;      /* seg hdr buffer */
 	m0_bcount_t        hdrblocks;   /* number of sectors */
 	void              *seg_addr0;
-	void              *p;
 	m0_bcount_t        seg_size;
+#ifndef __KERNEL__
+	void              *p;
+#endif
 
 	/* Allocate buffer for segment header. */
 	st_block_shift = seg->bs_stob->so_op->sop_block_shift(seg->bs_stob);
@@ -226,7 +228,7 @@ int m0_be_seg_open(struct m0_be_seg *seg)
 				 0, st_block_shift);
 	return rc;
 #else
-	return -ENOTSUP;
+	return -EOPNOTSUPP;
 #endif
 }
 
-- 
1.8.3.2

