From a1e1b3c662a379b424dd262a25b6453f36f85e14 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Wed, 29 May 2013 18:45:14 +0300
Subject: [PATCH 053/290] be (alloc): add M0_INTERNAL

---
 be/alloc.c | 36 ++++++++++++++++++++----------------
 be/alloc.h | 29 ++++++++++++++++-------------
 2 files changed, 36 insertions(+), 29 deletions(-)

diff --git a/be/alloc.c b/be/alloc.c
index 93853d4..76c9383 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -26,43 +26,47 @@
  * @{
  */
 
-int m0_be_allocator_init(struct m0_be_allocator *a, struct m0_be_seg *seg)
+M0_INTERNAL int
+m0_be_allocator_init(struct m0_be_allocator *a, struct m0_be_seg *seg)
 {
-	return -1;
+	return -1; /*XXX*/
 }
 
-void m0_be_allocator_fini(struct m0_be_allocator *a)
+M0_INTERNAL void m0_be_allocator_fini(struct m0_be_allocator *a)
 {
 }
 
-int m0_be_allocator_create(struct m0_be_allocator *a)
+M0_INTERNAL int m0_be_allocator_create(struct m0_be_allocator *a)
 {
-	return -1;
+	return -1; /*XXX*/
 }
 
-int m0_be_allocator_destroy(struct m0_be_allocator *a)
+M0_INTERNAL int m0_be_allocator_destroy(struct m0_be_allocator *a)
 {
-	return -1;
+	return -1; /*XXX*/
 }
 
-void m0_be_allocator_credit(const struct m0_be_allocator *a,
-			    enum m0_be_allocator_op optype, m0_bcount_t size,
-			    unsigned shift, struct m0_be_tx_credit *accum)
+M0_INTERNAL void m0_be_allocator_credit(const struct m0_be_allocator *a,
+					enum m0_be_allocator_op optype,
+					m0_bcount_t size,
+					unsigned shift,
+					struct m0_be_tx_credit *accum)
 {
 }
 
-void *m0_be_alloc(struct m0_be_allocator *a, struct m0_be_tx *tx,
-		  m0_bcount_t size, unsigned shift)
+M0_INTERNAL void *m0_be_alloc(struct m0_be_allocator *a, struct m0_be_tx *tx,
+			      m0_bcount_t size, unsigned shift)
 {
-	return NULL;
+	return NULL; /*XXX*/
 }
 
-void m0_be_free(struct m0_be_allocator *a, struct m0_be_tx *tx, void *ptr)
+M0_INTERNAL void
+m0_be_free(struct m0_be_allocator *a, struct m0_be_tx *tx, void *ptr)
 {
 }
 
-void m0_be_alloc_stats(const struct m0_be_allocator *a,
-		       struct m0_be_allocator_stats *out)
+M0_INTERNAL void m0_be_alloc_stats(const struct m0_be_allocator *a,
+				   struct m0_be_allocator_stats *out)
 {
 }
 
diff --git a/be/alloc.h b/be/alloc.h
index 3993e42..fcb67d0 100644
--- a/be/alloc.h
+++ b/be/alloc.h
@@ -38,37 +38,40 @@ struct m0_be_allocator {
 	struct m0_be_seg *ba_seg;
 };
 
-int m0_be_allocator_init(struct m0_be_allocator *a, struct m0_be_seg *seg);
+M0_INTERNAL int m0_be_allocator_init(struct m0_be_allocator *a,
+				     struct m0_be_seg *seg);
 
-void m0_be_allocator_fini(struct m0_be_allocator *a);
+M0_INTERNAL void m0_be_allocator_fini(struct m0_be_allocator *a);
 
-int m0_be_allocator_create(struct m0_be_allocator *a);
-int m0_be_allocator_destroy(struct m0_be_allocator *a);
+M0_INTERNAL int m0_be_allocator_create(struct m0_be_allocator *a);
+M0_INTERNAL int m0_be_allocator_destroy(struct m0_be_allocator *a);
 
 enum m0_be_allocator_op {
 	M0_BAO_ALLOC,
 	M0_BAO_FREE
 };
 
-void m0_be_allocator_credit(const struct m0_be_allocator *a,
-			    enum m0_be_allocator_op optype, m0_bcount_t size,
-			    unsigned shift, struct m0_be_tx_credit *accum);
+M0_INTERNAL void m0_be_allocator_credit(const struct m0_be_allocator *a,
+					enum m0_be_allocator_op optype,
+					m0_bcount_t size,
+					unsigned shift,
+					struct m0_be_tx_credit *accum);
 
-void *m0_be_alloc(struct m0_be_allocator *a, struct m0_be_tx *tx,
-		  m0_bcount_t size, unsigned shift);
+M0_INTERNAL void *m0_be_alloc(struct m0_be_allocator *a, struct m0_be_tx *tx,
+			      m0_bcount_t size, unsigned shift);
 
-void m0_be_free(struct m0_be_allocator *a, struct m0_be_tx *tx, void *ptr);
+M0_INTERNAL void m0_be_free(struct m0_be_allocator *a,
+			    struct m0_be_tx *tx, void *ptr);
 
 struct m0_be_allocator_stats {
 	m0_bcount_t bas_free_space;
 };
 
-void m0_be_alloc_stats(const struct m0_be_allocator *a,
-		       struct m0_be_allocator_stats *out);
+M0_INTERNAL void m0_be_alloc_stats(const struct m0_be_allocator *a,
+				   struct m0_be_allocator_stats *out);
 
 /** @} end of be group */
 #endif /* __MERO_BE_ALLOC_H__ */
-
 /*
  *  Local variables:
  *  c-indentation-style: "K&R"
-- 
1.8.3.2

