From ad00b8dc6a347bd0357b5cc216e9f9c2422d694f Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Wed, 29 May 2013 18:51:32 +0300
Subject: [PATCH 055/290] be (list): add M0_INTERNAL

---
 be/list.c |  8 ++------
 be/list.h | 33 ++++++++++++++++++++-------------
 2 files changed, 22 insertions(+), 19 deletions(-)

diff --git a/be/list.c b/be/list.c
index e6fd384..8538198 100644
--- a/be/list.c
+++ b/be/list.c
@@ -18,17 +18,13 @@
  * Original creation date: 29-May-2013
  */
 
-
 /**
- * @addtogroup XXX
+ * @addtogroup be
  *
  * @{
  */
 
-
-
-/** @} end of XXX group */
-
+/** @} end of be group */
 
 /*
  *  Local variables:
diff --git a/be/list.h b/be/list.h
index 94c7468..a032557 100644
--- a/be/list.h
+++ b/be/list.h
@@ -33,23 +33,29 @@ struct m0_be_list {
 	struct m0_be_seg *bl_seg;
 };
 
-void m0_be_list_init(struct m0_be_list *list, struct m0_be_seg *seg);
-void m0_be_list_fini(struct m0_be_list *list);
+M0_INTERNAL void m0_be_list_init(struct m0_be_list *list,
+				 struct m0_be_seg *seg);
 
-void m0_be_list_create(struct m0_be_list *list);
-void m0_be_list_destroy(struct m0_be_list *list);
+M0_INTERNAL void m0_be_list_fini(struct m0_be_list *list);
 
-void m0_be_list_get(const struct m0_be_list *list, m0_bcount_t nelems,
-		    struct m0_be_op *op);
+M0_INTERNAL void m0_be_list_create(struct m0_be_list *list);
+M0_INTERNAL void m0_be_list_destroy(struct m0_be_list *list);
 
-void m0_be_list_put(const struct m0_be_list *list, m0_bcount_t nelems);
+M0_INTERNAL void m0_be_list_get(const struct m0_be_list *list,
+				m0_bcount_t nelems,
+				struct m0_be_op *op);
+
+M0_INTERNAL void m0_be_list_put(const struct m0_be_list *list,
+				m0_bcount_t nelems);
 
 /**
  * Captures (origin - left)-th .. (origin + right)-th elements of the list.
  */
-void m0_be_list_capture(const struct m0_be_list *list, struct m0_be_tx *tx,
-			const struct m0_tlink *origin, m0_bcount_t left,
-			m0_bcount_t right);
+M0_INTERNAL void m0_be_list_capture(const struct m0_be_list *list,
+				    struct m0_be_tx *tx,
+				    const struct m0_tlink *origin,
+				    m0_bcount_t left,
+				    m0_bcount_t right);
 
 enum m0_be_list_op {
 	M0_BLO_INSERT,
@@ -57,9 +63,10 @@ enum m0_be_list_op {
 	M0_BLO_MOVE
 };
 
-void m0_be_list_credit(const struct m0_be_list *list,
-		       enum m0_be_list_op optype, m0_bcount_t nr,
-		       struct m0_be_tx_credit *accum);
+M0_INTERNAL void m0_be_list_credit(const struct m0_be_list *list,
+				   enum m0_be_list_op optype,
+				   m0_bcount_t nr,
+				   struct m0_be_tx_credit *accum);
 
 /** @} end of be group */
 #endif /* __MERO_BE_LIST_H__ */
-- 
1.8.3.2

