From 031f44dabb77fd137d78a5b0dca628bf4540b074 Mon Sep 17 00:00:00 2001
From: Anatoliy Bilenko <anatoliy_bilenko@xyratex.com>
Date: Thu, 30 May 2013 17:02:48 +0300
Subject: [PATCH 061/290] be: removed unused parameter

---
 be/btree.h | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/be/btree.h b/be/btree.h
index c5aa957..0440b63 100644
--- a/be/btree.h
+++ b/be/btree.h
@@ -54,6 +54,7 @@ M0_INTERNAL void m0_be_btree_destroy(struct m0_be_btree *tree,
 				     struct m0_be_op *op);
 
 struct m0_be_btree_cursor {
+	struct m0_be_btree *bc_tree;
 	struct m0_be_op     bc_op;
 	struct m0_buf       bc_current;
 };
@@ -71,9 +72,8 @@ M0_INTERNAL void m0_be_btree_cursor_put(struct m0_be_btree_cursor *cursor);
 M0_INTERNAL void m0_be_btree_cursor_next(struct m0_be_btree_cursor *cursor);
 
 struct m0_be_btree_anchor {
-	struct m0_be_btree *ba_tree;
-	struct m0_buf       ba_buf;
-	bool                ba_write; /**< Is write lock being held? */
+	struct m0_buf ba_buf;
+	bool          ba_write; /**< Is write lock being held? */
 };
 
 M0_INTERNAL void m0_be_btree_maxkey(const struct m0_be_btree *tree,
-- 
1.8.3.2

