From 40188fef32c7f8cf6922f7d9e93b7171f2fca534 Mon Sep 17 00:00:00 2001
From: Andriy Tkachuk <andriy_tkachuk@xyratex.com>
Date: Tue, 4 Jun 2013 15:07:04 +0300
Subject: [PATCH 096/290] be: m0_emap_extent_update() interface updated

---
 db/extmap.h | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/db/extmap.h b/db/extmap.h
index 2d718c2..ad5b556 100644
--- a/db/extmap.h
+++ b/db/extmap.h
@@ -296,15 +296,17 @@ M0_INTERNAL void m0_emap_paste(struct m0_emap_cursor *it,
 M0_INTERNAL void m0_emap_merge(struct m0_emap_cursor *it, m0_bindex_t delta);
 
 /**
-   Updates the segment the cursor is currently positioned at with the given
+   Updates the segment at the specified prefix and offset with the given
    segment having same prefix.
 
-   Asynchronous operation, get status via m0_emap_be_op_get(it)->bo_sm.sm_rc.
-
-   @pre m0_uint128_eq(&it->ec_seg.ee_pre, &es->ee_pre) == true
+   @pre m0_uint128_eq(prefix, &es->ee_pre) == true
  */
-M0_INTERNAL void m0_emap_extent_update(struct m0_emap_cursor *it,
-				       struct m0_emap_seg    *es);
+M0_INTERNAL void m0_emap_extent_update(struct m0_emap     *map,
+				       struct m0_be_tx    *tx,
+				       struct m0_be_op    *op,
+				 const struct m0_uint128  *prefix,
+				 const m0_bindex_t         offset,
+				 const struct m0_emap_seg *es);
 
 /**
    Release the resources associated with the cursor.
-- 
1.8.3.2

