From f054917b3e35eb0df6693fedcb626d135778a203 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Wed, 5 Jun 2013 02:38:15 +0300
Subject: [PATCH 106/290] style: sizeof expr is preferred to sizeof(type)

---
 be/alloc.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/be/alloc.c b/be/alloc.c
index 9d81cc3..8e0ffaa 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -132,7 +132,7 @@ be_alloc_chunk_trymerge(struct be_alloc_chunk *a, struct be_alloc_chunk *b)
 	M0_PRE((char *) a < (char *) b);
 	if (a != NULL && b != NULL &&
 	    a->bac_free && b->bac_free) {
-		a->bac_size += b->bac_size + sizeof(struct be_alloc_chunk);
+		a->bac_size += sizeof(*b) + b->bac_size;
 		chunks_all_tlist_del(b);
 		chunks_free_tlist_del(b);
 		be_alloc_chunk_fini(b);
@@ -185,8 +185,7 @@ M0_INTERNAL int m0_be_allocator_create(struct m0_be_allocator *a)
 	h = a->ba_h;
 	/** @todo GET_PTR h */
 	c = &h->bah_main;
-	overhead   = a->ba_seg->bs_reserved +
-		     sizeof(struct m0_be_allocator_header);
+	overhead   = a->ba_seg->bs_reserved + sizeof(a->ba_h);
 	free_space = a->ba_seg->bs_size - overhead;
 
 	/* check if segment is large enough to allocate at least 1 byte */
-- 
1.8.3.2

