From 7c5c6c5d24e793ea357a29893bf2f7fa222ba0b0 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Tue, 18 Jun 2013 19:09:55 +0300
Subject: [PATCH 170/290] be/tx: use normal tsearch() loop since tx->t_root is
 initialized.

---
 be/tx.c | 12 +++---------
 1 file changed, 3 insertions(+), 9 deletions(-)

diff --git a/be/tx.c b/be/tx.c
index d96b84e..41c431d 100644
--- a/be/tx.c
+++ b/be/tx.c
@@ -181,6 +181,7 @@ M0_INTERNAL void m0_be_tx_init(struct m0_be_tx *tx, uint64_t tid,
 {
 	M0_PRE(tx != NULL);
 	m0_sm_init(&tx->t_sm, &tx_sm_conf, M0_BTS_INIT, tx_grp);
+	tx->t_root       = NULL;
 	tx->t_id         = tid;
 	tx->t_be         = be;
 	tx->t_persistent = persistent;
@@ -266,16 +267,10 @@ m0_be_tx_capture(struct m0_be_tx *tx, const struct m0_be_reg *reg)
 	new->rd_buf = tx->t_reg_area + pos->tc_reg_size;
 	new->rd_reg = *reg;
 
+	M0_LOG(M0_DEBUG, "capture %p: [%p, %p)", new, new->rd_reg.br_addr,
+	       new->rd_reg.br_addr + new->rd_reg.br_size);
 	credit_mod(&tx->t_captured, new, +1);
 
-#if 1
-	/* we assume for now that there's no intersection */
-	memcpy(new->rd_buf,
-	       new->rd_reg.br_addr, new->rd_reg.br_size);
-	credit_mod(&tx->t_used, new, +1);
-	credit_mod(&tx->t_pos,  new, +1);
-
-#else
 	/* Cut pieces off the new region and store them into old regions'
 	 * buffers, until the new region either do not intersect any of the
 	 * old ones or is completely stuffed into old buffers. */
@@ -330,7 +325,6 @@ m0_be_tx_capture(struct m0_be_tx *tx, const struct m0_be_reg *reg)
 		}
 		prev = old;
 	}
-#endif
 	M0_POST(m0_be__tx_invariant(tx));
 }
 
-- 
1.8.3.2

