From e70c30e622a53f6762e993c7d883a5d7a7dcdbf0 Mon Sep 17 00:00:00 2001
From: Anatoliy Bilenko <anatoliy_bilenko@xyratex.com>
Date: Tue, 25 Jun 2013 16:54:13 +0300
Subject: [PATCH 224/290] xcode: fixed 'Use of uninitialized value $xcode in
 concatenation' warning.

---
 xcode/gccxml2xcode | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/xcode/gccxml2xcode b/xcode/gccxml2xcode
index 93e6831..c7bf56a 100755
--- a/xcode/gccxml2xcode
+++ b/xcode/gccxml2xcode
@@ -481,7 +481,7 @@ sub gen_xc_c_type_struct_decl
 {
     my @items = @_;
 
-    my $xcode;
+    my $xcode = "";
 
     for my $item (@items) {
         $xcode .= $item->{'scope'} . 'struct m0_xcode_type *' . $item->{'name'} . "_xc;\n";
@@ -515,7 +515,7 @@ sub gen_xc_c_helper_type_struct_def
 {
     my @items = @_;
 
-    my $xcode;
+    my $xcode = "";
 
     for my $item (@items) {
 
@@ -553,7 +553,7 @@ sub gen_xc_c_compiletime_checks
 {
     my @items = @_;
 
-    my $xcode;
+    my $xcode = "";
 
     for my $item (@items) {
 
-- 
1.8.3.2

