From 454b079d7b3f9f006bc72000a4967f1d0781bfc1 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Wed, 26 Jun 2013 12:47:01 +0300
Subject: [PATCH 228/290] be/tx: deleted obsolete comment

---
 be/tx.h | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/be/tx.h b/be/tx.h
index 372a828..ce298f9 100644
--- a/be/tx.h
+++ b/be/tx.h
@@ -493,11 +493,6 @@ M0_INTERNAL void m0_be_tx_init(struct m0_be_tx *tx, uint64_t tid,
 
 M0_INTERNAL void m0_be_tx_fini(struct m0_be_tx *tx);
 
-/**
- * Frees reg_area, reg_d_area and regions' tree (t_root).
- *
- * XXX DELETEME? Do we really need this function?
- */
 M0_INTERNAL void m0_be_tx_prep(struct m0_be_tx *tx,
 			       const struct m0_be_tx_credit *credit);
 
-- 
1.8.3.2

