From 27b32a45cf7d599b3583fbf50a46c9b5613de553 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Wed, 26 Jun 2013 14:46:14 +0300
Subject: [PATCH 229/290] be/credit: m0_be_tx_credit_mod() removed

---
 be/tx_credit.c | 15 ---------------
 be/tx_credit.h |  5 -----
 2 files changed, 20 deletions(-)

diff --git a/be/tx_credit.c b/be/tx_credit.c
index dff86c9..ba8e67d 100644
--- a/be/tx_credit.c
+++ b/be/tx_credit.c
@@ -62,21 +62,6 @@ M0_INTERNAL bool m0_be_tx_credit_le(const struct m0_be_tx_credit *c0,
 	       c0->tc_reg_size <= c1->tc_reg_size;
 }
 
-#if 0
-/* XXX temprary disabled by max */
-M0_INTERNAL void m0_be_tx_credit_mod(struct m0_be_tx_credit *cr,
-				     const struct m0_be_reg_d *rd, int sgn)
-{
-	M0_PRE(M0_IN(sgn, (+1, -1)));
-
-	cr->tc_reg_nr	+= sgn;
-	cr->tc_reg_size += sgn * rd->rd_reg.br_size;
-
-	M0_ASSERT((int64_t)cr->tc_reg_nr >= 0);
-	M0_ASSERT((int64_t)cr->tc_reg_size >= 0);
-}
-#endif
-
 /** @} end of be group */
 
 /*
diff --git a/be/tx_credit.h b/be/tx_credit.h
index 45ec627..3ee60b0 100644
--- a/be/tx_credit.h
+++ b/be/tx_credit.h
@@ -85,11 +85,6 @@ M0_INTERNAL void m0_be_tx_credit_mac(struct m0_be_tx_credit *c,
 
 M0_INTERNAL bool m0_be_tx_credit_le(const struct m0_be_tx_credit *c0,
 				    const struct m0_be_tx_credit *c1);
-#if 0
-M0_INTERNAL void m0_be_tx_credit_mod(struct m0_be_tx_credit *cr,
-				     const struct m0_be_reg_d *reg, int sgn);
-#endif
-
 #define M0_BE_TX_CREDIT_TYPE(type)		\
 	((struct m0_be_tx_credit) {		\
 		.tc_reg_nr = 1,			\
-- 
1.8.3.2

