From 2a9730009d49f1b967b4f82df62db33ec4701c3b Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Wed, 26 Jun 2013 21:05:01 +0300
Subject: [PATCH 239/290] be/ut: m0_be_ut_seg_storage_fini() enabled

---
 be/ut/helper.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/be/ut/helper.c b/be/ut/helper.c
index 9fc5444..8534727 100644
--- a/be/ut/helper.c
+++ b/be/ut/helper.c
@@ -43,10 +43,8 @@ static struct m0_stob_id be_ut_h_stob_id = {
 
 void m0_be_ut_seg_storage_fini(void)
 {
-	/*
 	int rc = system("rm -rf " BE_UT_H_STORAGE_DIR);
 	M0_ASSERT(rc == 0);
-	*/
 }
 
 void m0_be_ut_seg_storage_init(void)
-- 
1.8.3.2

