From 5b20ea6abd2c5069f0802473c22cd37b87aa9b32 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Thu, 4 Jul 2013 23:59:22 +0300
Subject: [PATCH 281/290] be: refactor m0_be_tx_{,un}capture()

---
 be/tx.c | 36 +++++++++++-------------------------
 1 file changed, 11 insertions(+), 25 deletions(-)

diff --git a/be/tx.c b/be/tx.c
index d7204b9..02d5c19 100644
--- a/be/tx.c
+++ b/be/tx.c
@@ -336,42 +336,28 @@ m0_be_tx_capture(struct m0_be_tx *tx, const struct m0_be_reg *reg)
 	M0_POST(m0_be__tx_invariant(tx));
 }
 #else
-M0_INTERNAL void
-m0_be_tx_capture(struct m0_be_tx *tx, const struct m0_be_reg *reg)
+static void cap_uncap(struct m0_be_tx *tx, const struct m0_be_reg *reg,
+		      void (*op)(struct m0_be_reg_area *ra,
+				 const struct m0_be_reg_d *rd))
 {
 	M0_PRE(m0_be__tx_invariant(tx));
 	M0_PRE(m0_be__tx_state(tx) == M0_BTS_ACTIVE);
 	M0_PRE(tx_is_locked(tx));
 
-	struct m0_be_reg_d new = {
-		.rd_tx	= tx,
-		.rd_buf = NULL,
-		.rd_reg = *reg,
-	};
-
-	M0_LOG(M0_DEBUG, "capture [%p, %p)", new.rd_reg.br_addr,
-	       (char *) new.rd_reg.br_addr + new.rd_reg.br_size);
+	op(&tx->t_reg_area,
+	   &((const struct m0_be_reg_d){ .rd_tx = tx, .rd_reg = *reg }));
+}
 
-	m0_be_reg_area_capture(&tx->t_reg_area, &new);
+M0_INTERNAL void
+m0_be_tx_capture(struct m0_be_tx *tx, const struct m0_be_reg *reg)
+{
+	cap_uncap(tx, reg, m0_be_reg_area_capture);
 }
 
 M0_INTERNAL void
 m0_be_tx_uncapture(struct m0_be_tx *tx, const struct m0_be_reg *reg)
 {
-	M0_PRE(m0_be__tx_invariant(tx));
-	M0_PRE(m0_be__tx_state(tx) == M0_BTS_ACTIVE);
-	M0_PRE(tx_is_locked(tx));
-
-	struct m0_be_reg_d rd = {
-		.rd_tx	= tx,
-		.rd_buf = NULL,
-		.rd_reg = *reg,
-	};
-
-	M0_LOG(M0_DEBUG, "uncapture [%p, %p)", rd.rd_reg.br_addr,
-	       (char *) rd.rd_reg.br_addr + rd.rd_reg.br_size);
-
-	m0_be_reg_area_uncapture(&tx->t_reg_area, &rd);
+	cap_uncap(tx, reg, m0_be_reg_area_uncapture);
 }
 #endif /* XXX */
 
-- 
1.8.3.2

