From 389903c5c8211dda1e8f3239cfb5e2cb88683718 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Fri, 5 Jul 2013 00:24:10 +0300
Subject: [PATCH 282/290] be m0_be__tx_state_set): describe transition in
 M0_ENTRY()

---
 be/tx.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/be/tx.c b/be/tx.c
index 02d5c19..3a71f32 100644
--- a/be/tx.c
+++ b/be/tx.c
@@ -499,6 +499,7 @@ M0_INTERNAL enum m0_be_tx_state m0_be__tx_state(const struct m0_be_tx *tx)
 M0_INTERNAL void
 m0_be__tx_state_set(struct m0_be_tx *tx, enum m0_be_tx_state state)
 {
+	M0_ENTRY("%d --> %d", m0_be__tx_state(tx), state);
 	M0_PRE(m0_be__tx_invariant(tx));
 	M0_PRE(tx_is_locked(tx));
 	M0_PRE(m0_be__tx_state(tx) != state);
@@ -513,6 +514,7 @@ m0_be__tx_state_set(struct m0_be_tx *tx, enum m0_be_tx_state state)
 	tx_tlist_move(&tx_engine(tx)->te_txs[state], tx);
 
 	M0_POST(m0_be__tx_invariant(tx));
+	M0_LEAVE();
 }
 
 static void tx_fail(struct m0_be_tx *tx, int err)
@@ -545,12 +547,12 @@ _tx_state_set(struct m0_sm_group *grp M0_UNUSED, struct m0_sm_ast *ast)
 	struct m0_be_tx    *tx = container_of(ast, struct m0_be_tx, t_ast);
 	enum m0_be_tx_state state = (enum m0_be_tx_state)ast->sa_datum;
 
-	M0_ENTRY("state=%u", state);
+	M0_ENTRY();
 	M0_PRE(IS_IN_ARRAY(state, tx_states));
 
-	/* XXX TODO: bob_of() */
 	m0_be__tx_state_set(tx, state);
 	_txs_set_ready(tx);
+
 	M0_LEAVE();
 }
 
@@ -646,9 +648,11 @@ static bool tx_is_locked(const struct m0_be_tx *tx)
 static int placed_st_in(struct m0_sm *mach)
 {
 	struct m0_be_tx *tx = container_of(mach, struct m0_be_tx, t_sm);
-	M0_ENTRY("t_glob_stable=%d", tx->t_glob_stable);
+	M0_ENTRY("t_glob_stable=%d", !!tx->t_glob_stable);
 
 	M0_LEAVE();
+	/* XXX FIXME: Don't intermix "external" and "chained" styles of state
+	 * transitions. This makes code hard to read. (Reported by Nikita.) */
 	return tx->t_glob_stable ? M0_BTS_DONE : -1;
 }
 
-- 
1.8.3.2

