From e5a713c91460b0c1c051f62db9709bb1e4d35327 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Mon, 18 Mar 2013 13:27:31 +0200
Subject: [PATCH 03/15] mdstore: fill dirent's d_ino field in
 m0_mdstore_readdir()

---
 mdstore/mdstore.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 432ebd2..46981cd 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -767,12 +767,13 @@ M0_INTERNAL int m0_mdstore_readdir(struct m0_mdstore       *md,
 		if (nob >= reclen) {
 			memcpy(ent->d_name, m0_bitstring_buf_get(pos),
 			       m0_bitstring_len_get(pos));
+			ent->d_ino = (ino_t)it.ci_rec.cnr_fid.f_key;
 			ent->d_namelen = m0_bitstring_len_get(pos);
 			ent->d_reclen = reclen;
 			M0_LOG(M0_DEBUG,
-			       "Readdir filled entry \"%.*s\" recsize %d",
+			       "Readdir filled entry \"%.*s\", %u recsize %d",
 			       ent->d_namelen, (char *)ent->d_name,
-			       ent->d_reclen);
+			       (unsigned)ent->d_ino, ent->d_reclen);
 		} else {
 			/*
 			 * If buffer was too small to hold even one record,
-- 
1.8.3.2

