From 6cb39a4179342d17165d3a888a085bb27b90c709 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Fri, 29 Mar 2013 13:27:48 +0200
Subject: [PATCH 05/15] m0t1fs/fid: do not use csb_next_key for f_key
 allocation

---
 m0t1fs/linux_kernel/dir.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/m0t1fs/linux_kernel/dir.c b/m0t1fs/linux_kernel/dir.c
index 07f0215..c73d1ea 100644
--- a/m0t1fs/linux_kernel/dir.c
+++ b/m0t1fs/linux_kernel/dir.c
@@ -165,7 +165,7 @@ static struct m0_fid m0t1fs_fid_alloc(struct m0t1fs_sb *csb)
 	struct m0_fid fid;
 
 	M0_PRE(m0t1fs_fs_is_locked(csb));
-	m0_fid_set(&fid, 0, csb->csb_next_key++);
+	m0_fid_set(&fid, 0, 0);
 
 	return fid;
 }
-- 
1.8.3.2

