From 1fba61d37a50e326669349e0e77f20feb66b875a Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Fri, 29 Mar 2013 17:31:35 +0200
Subject: [PATCH 06/15] m0t1fs/fid: pass allocated fid/f_key to ios' create

---
 m0t1fs/linux_kernel/dir.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/m0t1fs/linux_kernel/dir.c b/m0t1fs/linux_kernel/dir.c
index c73d1ea..70c812a 100644
--- a/m0t1fs/linux_kernel/dir.c
+++ b/m0t1fs/linux_kernel/dir.c
@@ -350,6 +350,7 @@ static int m0t1fs_create(struct inode     *dir,
 	if (rc != 0)
 		goto out;
 
+	ci->ci_fid.f_key = rep->c_body.b_tfid.f_key;
 	if (S_ISREG(mode)) {
 		rc = m0t1fs_component_objects_op(ci, m0t1fs_ios_cob_create);
 		if (rc != 0)
-- 
1.8.3.2

