From 326e8ece67b56fbd7b0278812c1c4adddc137204 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Sun, 31 Mar 2013 21:48:12 +0300
Subject: [PATCH 13/15] cob: format nsrec insertion trace the same as
 oirec/nsrec one

---
 cob/cob.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/cob/cob.c b/cob/cob.c
index ee21ea9..8e83599 100644
--- a/cob/cob.c
+++ b/cob/cob.c
@@ -1254,8 +1254,8 @@ M0_INTERNAL int m0_cob_create(struct m0_cob *cob,
 			 &fabkey, sizeof fabkey, cob->co_fabrec,
 			 m0_cob_fabrec_size(cob->co_fabrec));
 	rc = m0_table_insert(tx, &pair);
-	M0_LOG(M0_DEBUG, "insert fab [%llx:%llx] v%llu.%llu, lnk=%.*s, rc=%d",
-	       (long long unsigned)fabkey.cfb_fid.f_container,
+	M0_LOG(M0_DEBUG, "insert fabrec [%llx:%llx]->v%llu.%llu, lnk=%.*s,"
+	       " rc=%d", (long long unsigned)fabkey.cfb_fid.f_container,
 	       (long long unsigned)fabkey.cfb_fid.f_key,
 	       (long long unsigned)cob->co_fabrec->cfb_version.vn_lsn,
 	       (long long unsigned)cob->co_fabrec->cfb_version.vn_vc,
-- 
1.8.3.2

