From e77ddf6136a9ee2aa6ed1ef033154daabd3b7249 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Mon, 30 Sep 2013 14:11:34 +0300
Subject: [PATCH 2/5] be/tx: size logging should be in critical section

---
 be/tx.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/be/tx.c b/be/tx.c
index e00ba24..70ceb86 100644
--- a/be/tx.c
+++ b/be/tx.c
@@ -228,7 +228,14 @@ M0_INTERNAL void m0_be_tx_open(struct m0_be_tx *tx)
 #define SIZE 100
 	void *buffer[100];
 	char **strings;
+	static bool   initialized;
+	static struct m0_mutex lock;
 
+	if (!initialized) {
+		initialized = true;
+		m0_mutex_init(&lock);
+	}
+	m0_mutex_lock(&lock);
 	dprintf(2, "%lu %lu ", tx->t_prepared.tc_reg_nr, tx->t_prepared.tc_reg_size);
 
 	nptrs = backtrace(buffer, SIZE);
@@ -237,6 +244,7 @@ M0_INTERNAL void m0_be_tx_open(struct m0_be_tx *tx)
 	for (j = 0; j < nptrs; j++)
 		dprintf(2, "\"%s\"%c", strings[j], j == nptrs - 1 ? '\n' : ' ');
 	free(strings);
+	m0_mutex_unlock(&lock);
 
 	M0_ENTRY();
 	M0_PRE(BE_TX_LOCKED_AT_STATE(tx, (M0_BTS_PREPARE)));
-- 
1.8.3.2

