From f1ac431cf7fc9d84e49b8b96008e2f1ba6e29286 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Wed, 16 Oct 2013 18:40:29 +0300
Subject: [PATCH 01/17] scripts: don't clone mero source

---
 scripts/mero_build_test.sh | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/scripts/mero_build_test.sh b/scripts/mero_build_test.sh
index 721577a..42deee1 100755
--- a/scripts/mero_build_test.sh
+++ b/scripts/mero_build_test.sh
@@ -351,9 +351,15 @@ init_dirs()
 
         print_msg "INFO:Cloning Mero source please wait..."
 
+	if [ ! -f $MERO_SOURCE/mero.spec.in ]; then
+		echo "Mero source don't exists!"
+	fi
+
+<< EOF
         # Clone the Mero source
         git clone --recursive $GIT_PROTOCOL://$GIT_USER@$GIT_WEB_ADDRESS/$GIT_REPOSITORY \
                 $MERO_SOURCE >> $TESTROOT/$DIRTIME/mero_build_test.log 2>&1 || return $?
+EOF
     else # [ "$src_dir" != "" ]
         ln -s $src_dir $MERO_SOURCE
     fi
-- 
1.8.3.2

