From 05826e2714b7829c66ac5df6d2575d59516fd0ee Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Wed, 16 Oct 2013 22:19:28 +0300
Subject: [PATCH 02/17] scripts: don't distclean after tests

---
 scripts/mero_build_test.sh | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/scripts/mero_build_test.sh b/scripts/mero_build_test.sh
index 42deee1..ce484a9 100755
--- a/scripts/mero_build_test.sh
+++ b/scripts/mero_build_test.sh
@@ -260,11 +260,14 @@ run_test_automate()
 	    rc=$?
         fi
 
-        print_msg "INFO:$current_branch:Cleanup:Start"
-        cd $MERO_CORE_PATH > /dev/null
-        make distclean >> \
-                $TESTROOT/$DIRTIME/$current_branch/logs/make-clean.log 2>&1
-        print_msg "INFO:$current_branch:Cleanup:End"
+	cd $MERO_CORE_PATH > /dev/null
+	make doc
+
+#	print_msg "INFO:$current_branch:Cleanup:Start"
+#	cd $MERO_CORE_PATH > /dev/null
+#	make distclean >> \
+#	$TESTROOT/$DIRTIME/$current_branch/logs/make-clean.log 2>&1
+#	print_msg "INFO:$current_branch:Cleanup:End"
     done
 
     return $rc
-- 
1.8.3.2

