From 400c23084c2ebf2cae8fd7562a2a4fcc63396578 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Wed, 16 Oct 2013 23:05:18 +0300
Subject: [PATCH 03/17] scripts: `make -j9` to build Mero

---
 scripts/mero_build_test.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/scripts/mero_build_test.sh b/scripts/mero_build_test.sh
index ce484a9..55c9463 100755
--- a/scripts/mero_build_test.sh
+++ b/scripts/mero_build_test.sh
@@ -240,7 +240,7 @@ run_test_automate()
 
         if run_command "$MERO_SOURCE" 'sh autogen.sh'                    && \
            run_command "$MERO_SOURCE" './configure' '--enable-coverage'  && \
-           run_command "$MERO_SOURCE" 'make'; then
+           run_command "$MERO_SOURCE" 'make' '-j9'; then
 
 	    pushd 2>/dev/null $TESTROOT/$DIRTIME/sandbox > /dev/null
             run_command '' "sudo $MERO_SOURCE/utils/ut.sh" '' 1
-- 
1.8.3.2

