From 1809dde6a5488de19159bbd710feef6c4dfe608e Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Wed, 16 Oct 2013 23:41:21 +0300
Subject: [PATCH 04/17] scripts: branch is already checked out, just clean it

---
 scripts/mero_build_test.sh | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/scripts/mero_build_test.sh b/scripts/mero_build_test.sh
index 55c9463..f8f70b3 100755
--- a/scripts/mero_build_test.sh
+++ b/scripts/mero_build_test.sh
@@ -235,8 +235,10 @@ run_test_automate()
             create_dir $TESTROOT/$DIRTIME/$current_branch/$directory || return $?
         done
 
-        git checkout $current_branch >> \
-                $TESTROOT/$DIRTIME/$current_branch/logs/git.log 2>&1
+#	git checkout $current_branch >> \
+#			$TESTROOT/$DIRTIME/$current_branch/logs/git.log 2>&1
+
+	sudo git clean -dfx
 
         if run_command "$MERO_SOURCE" 'sh autogen.sh'                    && \
            run_command "$MERO_SOURCE" './configure' '--enable-coverage'  && \
-- 
1.8.3.2

