From 59f110867d0d2008084486da1a14a0acec670612 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Wed, 16 Oct 2013 23:47:46 +0300
Subject: [PATCH 05/17] scripts: remove all .libs directories owned by root

---
 scripts/mero_build_test.sh | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/scripts/mero_build_test.sh b/scripts/mero_build_test.sh
index f8f70b3..bdc8ce7 100755
--- a/scripts/mero_build_test.sh
+++ b/scripts/mero_build_test.sh
@@ -265,6 +265,8 @@ run_test_automate()
 	cd $MERO_CORE_PATH > /dev/null
 	make doc
 
+	find `pwd` -name .libs | xargs sudo rm -rf
+
 #	print_msg "INFO:$current_branch:Cleanup:Start"
 #	cd $MERO_CORE_PATH > /dev/null
 #	make distclean >> \
-- 
1.8.3.2

