From 95a986fef960a1f8afc3026d584f025b492a3848 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Sun, 27 Oct 2013 19:30:35 +0200
Subject: [PATCH 12/17] st/jenkins: added missing ;;

---
 st/jenkins.sh | 1 +
 1 file changed, 1 insertion(+)

diff --git a/st/jenkins.sh b/st/jenkins.sh
index 3a0da6f..68e6e9c 100755
--- a/st/jenkins.sh
+++ b/st/jenkins.sh
@@ -33,6 +33,7 @@ main() {
 		cd $MERO_SRC_ROOT
 		sudo chown --recursive `id -nu`:`id -ng` "$MERO_SRC_ROOT"
 		git clean -dfx
+		;;
 	"help")
 		echo "See st/jenkins.sh" # TODO improve
 		;;
-- 
1.8.3.2

