From 3a8085bdcf414d2602ebce89cc70e89ceb59b440 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Sun, 27 Oct 2013 19:41:16 +0200
Subject: [PATCH 13/17] st/pm/net-test: run script in debug mode

---
 st/pm/net-test/st.sh | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/st/pm/net-test/st.sh b/st/pm/net-test/st.sh
index fb99b8d..2366204 100755
--- a/st/pm/net-test/st.sh
+++ b/st/pm/net-test/st.sh
@@ -1,4 +1,6 @@
 #!/bin/bash
+set -eux
+export PS4='+ ${FUNCNAME[0]:+${FUNCNAME[0]}():}line ${LINENO}: '
 
 # TODO move net-test scripts here
 "$MERO_SRC_ROOT/net/test/st/st.sh" "$@"
-- 
1.8.3.2

