From 96c2ea922dbc11e4a189844cc7e6ebb1b55a8857 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Sun, 27 Oct 2013 20:19:02 +0200
Subject: [PATCH 16/17] configure.ac: echo configure command line

---
 configure.ac | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/configure.ac b/configure.ac
index 7cfe3e1..c425462 100644
--- a/configure.ac
+++ b/configure.ac
@@ -682,6 +682,8 @@ AS_IF([test "x$LIBS" != x],
 echo ""
 echo "Mero version :  $PACKAGE_VERSION"
 echo ""
+echo "CONFIGURE_OPT:  $BUILD_CONFIGURE_OPTS"
+echo ""
 echo "M0_CPPFLAGS  :  \"$M0_CPPFLAGS\""
 echo "M0_CFLAGS    :  \"$M0_CFLAGS\""
 echo "M0_KCPPFLAGS :  \"$M0_KCPPFLAGS\""
-- 
1.8.3.2

