From b29db5b4ede3d32d5b31f3f44380671c041e4b35 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Fri, 25 Jan 2013 18:54:45 +0200
Subject: [PATCH 07/99] added description for M0_NET_TEST_CMD_SIZE_MAX

---
 net/test/commands.h | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/net/test/commands.h b/net/test/commands.h
index f6bef65..135be06 100644
--- a/net/test/commands.h
+++ b/net/test/commands.h
@@ -41,8 +41,13 @@
  */
 
 enum {
-	/** @todo 16k, change */
-	/** @todo send size, ack size, send command */
+	/*
+	 * It is a size of command network buffer.
+	 * Command will be serialized to this buffer, so it should
+	 * be large enough to contain entire serialized command.
+	 * @note There can be a problem with large number of test nodes,
+	 * because M0_NET_TEST_CMD_INIT have list of node endpoints inside.
+	 */
 	M0_NET_TEST_CMD_SIZE_MAX     = 16384,
 };
 
-- 
1.8.3.2

