From 406265193c10e592dbf364116c43a54c3df3434f Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Fri, 25 Jan 2013 18:58:08 +0200
Subject: [PATCH 08/99] improved description for RTT in STATUS_DATA command

---
 net/test/commands.h | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/net/test/commands.h b/net/test/commands.h
index 135be06..b71b3e6 100644
--- a/net/test/commands.h
+++ b/net/test/commands.h
@@ -177,11 +177,12 @@ struct m0_net_test_cmd_status_data {
 	struct m0_net_test_mps	  ntcsd_mps_send;
 	/** 'receive' messages per second statistics with 1s interval */
 	struct m0_net_test_mps	  ntcsd_mps_recv;
-	/** RTT statistics (without lost messages) (test client only) */
+	/**
+	 * RTT statistics (without lost messages) (test client only)
+	 * @note Only RTT can be measured without very precise
+	 * time synchronization between test client and test server.
+	 */
 	struct m0_net_test_stats  ntcsd_rtt;
-	/** @todo send/recv RTT
-	   (needs time synchronization over clients/servers) */
-	/** @todo msg stats for bulk test */
 };
 
 /**
-- 
1.8.3.2

