From 0cd0f8e1cf3d8a21470a8e009761fc1e055df4a3 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Fri, 25 Jan 2013 20:25:06 +0200
Subject: [PATCH 18/99] fixed memory leak in node_cmd_wait()

---
 net/test/node.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/net/test/node.c b/net/test/node.c
index 352474f..83df2f3 100644
--- a/net/test/node.c
+++ b/net/test/node.c
@@ -736,7 +736,6 @@ static int node_cmd_get(struct m0_net_test_cmd_ctx *cmd_ctx,
 	return rc;
 }
 
-/** @todo memory leak for all commands with different type from parameter */
 static int node_cmd_wait(struct m0_net_test_node_ctx *ctx,
 			 struct m0_net_test_cmd *cmd,
 			 enum m0_net_test_cmd_type type)
@@ -752,6 +751,8 @@ static int node_cmd_wait(struct m0_net_test_node_ctx *ctx,
 		rc = node_cmd_get(&ctx->ntnc_cmd, cmd, deadline);
 		if (rc != 0 && rc != -ETIMEDOUT)
 			return rc;	/** @todo add retry count */
+		if (rc == 0 && cmd->ntc_type != type)
+			m0_net_test_commands_received_free(cmd);
 	} while (!(rc == 0 && cmd->ntc_type == type) && !ctx->ntnc_exit_flag);
 	return 0;
 }
-- 
1.8.3.2

