From dd992005172d3141ae7402cbf2edc5791e64ca92 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Thu, 31 Jan 2013 16:05:05 +0200
Subject: [PATCH 25/99] client-server UT now have 20s timeouts

---
 net/test/ut/client_server.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/net/test/ut/client_server.c b/net/test/ut/client_server.c
index 1a639f6..bab2e62 100644
--- a/net/test/ut/client_server.c
+++ b/net/test/ut/client_server.c
@@ -42,16 +42,16 @@ enum {
 	NTCS_NODES_MAX		  = 128,
 	NTCS_NODE_ADDR_MAX	  = 0x100,
 	/** @todo 20min for debugging in gdb */
+	/*
 	NTCS_TIMEOUT_CMD_MS	  = 1200000,
 	NTCS_TIMEOUT_SEND_MS	  = 1200000,
 	NTCS_TIMEOUT_RECV_MS	  = 1200000,
 	NTCS_TIMEOUT_BULK_MS	  = 1200000,
-	/*
+	*/
 	NTCS_TIMEOUT_CMD_MS	  = 20000,
 	NTCS_TIMEOUT_SEND_MS	  = 20000,
 	NTCS_TIMEOUT_RECV_MS	  = 20000,
 	NTCS_TIMEOUT_BULK_MS	  = 20000,
-	*/
 	NTCS_TMID_CONSOLE4CLIENTS = 2998,
 	NTCS_TMID_CONSOLE4SERVERS = 2999,
 	NTCS_TMID_NODES		  = 3000,
-- 
1.8.3.2

