From 83ae9c574e0a317f16831b743cb8e21f4ecd8d1b Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Sun, 3 Feb 2013 19:11:32 +0200
Subject: [PATCH 33/99] LogD 1874: s/maximim/maximum/

---
 net/test/network.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/test/network.h b/net/test/network.h
index efa1ea1..f0e82bd 100644
--- a/net/test/network.h
+++ b/net/test/network.h
@@ -86,7 +86,7 @@ struct m0_net_test_network_cfg {
 	m0_bcount_t				    ntncfg_buf_size_bulk;
 	/** number of bulk buffers */
 	uint32_t				    ntncfg_buf_bulk_nr;
-	/** maximim number of endpoints in context */
+	/** maximum number of endpoints in context */
 	uint32_t				    ntncfg_ep_max;
 	/**
 	 * Timeouts for every type of network buffer.
-- 
1.8.3.2

