From ecc65fe6456a4b2f15594e35f4a8af78f34e0478 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Sun, 3 Feb 2013 19:22:58 +0200
Subject: [PATCH 37/99] LogD 1875, part 2: extra blank line

---
 net/test/node_bulk.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/net/test/node_bulk.c b/net/test/node_bulk.c
index 6daf100..3f014ef 100644
--- a/net/test/node_bulk.c
+++ b/net/test/node_bulk.c
@@ -1314,7 +1314,6 @@ static void node_bulk_worker(struct node_bulk_ctx *ctx)
 	m0_clink_del(&tm_clink);
 	m0_clink_fini(&tm_clink);
 	m0_chan_fini(&tm_chan);
-
 }
 
 static int node_bulk_test_init_fini(struct node_bulk_ctx *ctx,
-- 
1.8.3.2

