From 775322ab148c01bcbb82b06ee81aeb8d67b254a5 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Tue, 5 Feb 2013 15:41:20 +0200
Subject: [PATCH 50/99] LogD 1888: removed leading * in commands.h comment

---
 net/test/commands.h | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/net/test/commands.h b/net/test/commands.h
index ee73064..15c6be8 100644
--- a/net/test/commands.h
+++ b/net/test/commands.h
@@ -242,9 +242,9 @@ typedef void (*m0_net_test_commands_send_cb_t)(struct m0_net_test_cmd_ctx *ctx,
  */
 struct m0_net_test_cmd_ctx {
 	/**
-	 * Network context for this command context.
-	 * First ntcc_ep_nr ping buffers are used for commands sending,
-	 * other ntcc_ep_nr * 2 ping buffers are used for commands receiving.
+	   Network context for this command context.
+	   First ntcc_ep_nr ping buffers are used for commands sending,
+	   other ntcc_ep_nr * 2 ping buffers are used for commands receiving.
 	 */
 	struct m0_net_test_network_ctx	   ntcc_net;
 	/**
-- 
1.8.3.2

