From f2146153df4f4382d33f087ef1bf3a8e5021743b Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Wed, 13 Feb 2013 20:38:45 +0200
Subject: [PATCH 64/99] net-test/demo: add support for console init/fini

---
 net/test/demo/demo.exp | 26 +++++++++++++++++++-------
 1 file changed, 19 insertions(+), 7 deletions(-)

diff --git a/net/test/demo/demo.exp b/net/test/demo/demo.exp
index 87ab9f6..c972513 100755
--- a/net/test/demo/demo.exp
+++ b/net/test/demo/demo.exp
@@ -33,7 +33,11 @@ proc exec_cmd {pid sid cmd {tmout 10} {e 0}} {
 
 # initialize a host and return [pid, spawn id] of ssh
 proc init_host {addr {tmout 10}} {
-    set pid [spawn ssh $addr]
+    if {"z$addr" == "zlocalhost"} {
+        set pid [spawn bash]
+    } else {
+        set pid [spawn ssh $addr]
+    }
     set sid $spawn_id
     set t0 [timestamp]
     expect -i $sid -timeout $tmout -re {^[^\r\n]*\r*\n} {
@@ -52,13 +56,19 @@ proc init_host {addr {tmout 10}} {
     } eof {error eof}
 }
 
-# returns 2 lists and 3 flattened arrays:
-# ihosts ehosts pid_al sid_al nid_al
-proc init_hosts {shosts chosts ntopts_al} {
+# returns 2 lists and 5 flattened arrays:
+# 0. ihosts - initialized hosts
+# 1. ehosts - error hosts
+# 2. pid_al - ssh pid
+# 3. sid_al - ssh spawn_id
+# 4. nid_al - NID
+# 5. cid_al - command EP: 
+# 6. 
+proc init_hosts {shosts chosts consoles ntopts_al} {
     array set ntopts $ntopts_al
     # create list of unique addresses:
     set hosts ""
-    foreach h [concat $shosts $chosts] {
+    foreach h [concat $shosts $chosts $consoles] {
         if {! [info exists u($h)]} {
             set u($h) 1
             lappend hosts $h
@@ -124,7 +134,7 @@ proc main {argv0 argv} {
     set usg "\[options\] srv1\[,srv2...\] clnt1\[,clnt2...\]"
     if {[catch {
         array set ntopts [::cmdline::getoptions argv $optdesc $usg]
-        if {[llength $argv] != 2} {
+        if {[llength $argv] < 2 || [llength $argv] > 3} {
             error [::cmdline::usage $optdesc $usg]}
     } err]} {
         send_error "USAGE: $err"
@@ -132,7 +142,9 @@ proc main {argv0 argv} {
     }
     set shosts [split [lindex $argv 0] ,]
     set chosts [split [lindex $argv 1] ,]
-    set r [init_hosts $shosts $chosts [array get ntopts]]
+    if {[llength $argv] == 3} {set console [lindex $argv 2]} {
+        set console localhost}
+    set r [init_hosts $shosts $chosts [list $console] [array get ntopts]]
     fini_hosts [lindex $r 0] [lindex $r 2] [lindex $r 3] $ntopts(w)
 }
 
-- 
1.8.3.2

