From 19579bab41de0c2eceaf92cc2a81d6cc1fb312fa Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Mon, 25 Feb 2013 13:13:03 +0200
Subject: [PATCH 86/99] added net-test ST to 'm0 run-st'

---
 scripts/m0 | 1 +
 1 file changed, 1 insertion(+)

diff --git a/scripts/m0 b/scripts/m0
index 8102df1..00e3dac 100755
--- a/scripts/m0
+++ b/scripts/m0
@@ -46,6 +46,7 @@ run_kut() {
 run_st() {
     _reset
     $SUDO "$SRC/m0t1fs/linux_kernel/st/st"
+    $SUDO "$SRC/net/test/st/st.sh"
     $SUDO "$SRC/m0t1fs/linux_kernel/st/m0t1fs_test.sh"
 }
 
-- 
1.8.3.2

