From 53ee935403b34e9dba3d07da84e8a0e102ac8d63 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Mon, 4 Mar 2013 03:43:06 +0200
Subject: [PATCH 92/99] enabled 3d graphs

---
 net/test/demo/demo.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/test/demo/demo.sh b/net/test/demo/demo.sh
index c0b4f72..a71a919 100755
--- a/net/test/demo/demo.sh
+++ b/net/test/demo/demo.sh
@@ -1249,7 +1249,7 @@ gnuplot_run()
 draw_graphs()
 {
 	draw_graphs_2d
-	# draw_graphs_3d
+	draw_graphs_3d
 	gnuplot_run
 }
 
-- 
1.8.3.2

