From 994b446828ebc9891a9526727154e699d292c9ee Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Tue, 5 Mar 2013 03:41:35 +0200
Subject: [PATCH 94/99] removed flipped tables from demo script

---
 net/test/demo/demo.sh | 15 ++-------------
 1 file changed, 2 insertions(+), 13 deletions(-)

diff --git a/net/test/demo/demo.sh b/net/test/demo/demo.sh
index a71a919..f115a61 100755
--- a/net/test/demo/demo.sh
+++ b/net/test/demo/demo.sh
@@ -35,8 +35,6 @@ DIR_TABLES="$DIR_RESULT/tables"
 #	0Y - test concurrency
 DIR_TABLES_UNIFORM="$DIR_RESULT/tables-uniform"
 #	contains tables without axis ticks
-DIR_TABLES_FLIPPED="$DIR_RESULT/tables-flipped"
-#	tables are flipped (0X and 0Y are swapped) and without axis ticks
 # DIR_TABLES_*/
 #	[bulk|ping]-[kernel|user]-[kernel|user]-\
 #		[Bandwidth|RTT|MPS]-[min|max|avg|stddev]
@@ -369,12 +367,12 @@ console_configure2()
 dirs_create()
 {
 	$MKDIR "$DIR_SETUP" "$DIR_DATA" "$DIR_RESULT" "$DIR_PLOT_SCRIPTS"
-	$MKDIR "$DIR_TABLES" "$DIR_TABLES_UNIFORM" "$DIR_TABLES_FLIPPED"
+	$MKDIR "$DIR_TABLES" "$DIR_TABLES_UNIFORM"
 }
 
 files_clean()
 {
-	$RM_RF "$DIR_TABLES" "$DIR_TABLES_UNIFORM" "$DIR_TABLES_FLIPPED"
+	$RM_RF "$DIR_TABLES" "$DIR_TABLES_UNIFORM"
 	$RM_RF "$DIR_SETUP" "$DIR_DATA" "$DIR_RESULT" "$DIR_PLOT_SCRIPTS"
 }
 
@@ -962,12 +960,6 @@ table_make_uniform()
 	sed 1d | awk '{$1=""; print}' | sed -e 's/^[ \t]*//'
 }
 
-# TODO
-table_make_flipped()
-{
-	cat
-}
-
 # make gnuplot uniform tables (without axis ticks)
 table_additional_generate()
 {
@@ -978,11 +970,8 @@ table_additional_generate()
 		       $(test_table_prefix $dir_name))"
 	local table_uniform="$(table_filename $measurement $m_type \
 			$(test_table_prefix $dir_name $DIR_TABLES_UNIFORM))"
-	local table_flipped="$(table_filename $measurement $m_type \
-			$(test_table_prefix $dir_name $DIR_TABLES_FLIPPED))"
 
 	table_make_uniform < "$table" > "$table_uniform"
-	table_make_flipped < "$table_uniform" > "$table_flipped"
 }
 
 plot_header()
-- 
1.8.3.2

