From dec61a6cda54c61c13b4531d779646328ab51d8d Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Tue, 26 Feb 2013 00:47:53 +0800
Subject: [PATCH 08/24] add -G and -L command line arguments

---
 m0t1fs/linux_kernel/st/st | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/m0t1fs/linux_kernel/st/st b/m0t1fs/linux_kernel/st/st
index d5de11a..f2e8bc3 100755
--- a/m0t1fs/linux_kernel/st/st
+++ b/m0t1fs/linux_kernel/st/st
@@ -28,6 +28,8 @@ export_lnet_nid_endpoints() {
     export IOS0_ENDPOINT="$LNET_NID:12345:34:1"
     export IOS1_ENDPOINT="$LNET_NID:12345:34:2"
     export MDS_ENDPOINT="$IOS0_ENDPOINT"
+    export CLI2MDS_ENDPOINT1="$LNET_NID:12345:34:201"
+    export CLI2MDS_ENDPOINT2="$LNET_NID:12345:34:202"
     export CONFD_ENDPOINT="$IOS0_ENDPOINT"
 }
 
@@ -80,6 +82,7 @@ modules_insert() {
 }
 
 modules_remove() {
+    sleep 20
     rmmod m0mero
     rmmod galois
 }
@@ -132,6 +135,7 @@ EOF`"
 -r -p -D $SANDBOX_DIR/db -T linux -S $SANDBOX_DIR/stobs
 -A $SANDBOX_DIR/addb-stobs
 -e lnet:$IOS0_ENDPOINT $CONFD_SPEC -s mdservice -s addb -s ioservice
+-G lnet:$MDS_ENDPOINT -L lnet:$CLI2MDS_ENDPOINT1
  -m $MAX_RPC_MSG_SIZE -q $TM_MIN_RECV_QUEUE_LEN
 EOF`"
     [ "$MODE" = autoconf ] && OPTIONS="-C $CONFD_ENDPOINT -P prof-ac"
@@ -200,6 +204,7 @@ _init() {
 _fini() {
     services_stop
     sandbox_fini
+    lsmod || grep m0mero
     modules_remove
     exit ${1:-$?}
 }
-- 
1.8.3.2

