From 266849ebf9316d21647ee9a430f6a327237f18b2 Mon Sep 17 00:00:00 2001
From: Amit Jambure <Amit_Jambure@xyratex.com>
Date: Thu, 23 May 2013 15:33:55 +0530
Subject: [PATCH 18/60] rpc: Fold m0_rpc_rcv_session_establish() in its caller.

---
 rpc/session_foms.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/rpc/session_foms.c b/rpc/session_foms.c
index 18c5871..7c8d4d7 100644
--- a/rpc/session_foms.c
+++ b/rpc/session_foms.c
@@ -398,7 +398,6 @@ M0_INTERNAL int m0_rpc_fom_session_establish_tick(struct m0_fom *fom)
 out:
 	reply->rser_sender_id  = request->rse_sender_id;
 	reply->rser_rc         = rc;
-
 	if (rc != 0) {
 		reply->rser_session_id = SESSION_ID_INVALID;
 		if (session != NULL)
-- 
1.8.3.2

