From 0728891a8a57570f2b8202cd93148335a8abfb29 Mon Sep 17 00:00:00 2001
From: Amit Jambure <Amit_Jambure@xyratex.com>
Date: Thu, 23 May 2013 15:41:41 +0530
Subject: [PATCH 19/60] rpc: Simplify m0_rpc_rcv_[session|conn]_terminate.

---
 rpc/conn.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/rpc/conn.c b/rpc/conn.c
index 4902178..f949469 100644
--- a/rpc/conn.c
+++ b/rpc/conn.c
@@ -902,6 +902,7 @@ M0_INTERNAL int m0_rpc_rcv_conn_terminate(struct m0_rpc_conn *conn)
 		m0_rpc_conn_cleanup_all_sessions(conn);
 	deregister_all_item_sources(conn);
 	conn_state_set(conn, M0_RPC_CONN_TERMINATED);
+
 	M0_ASSERT(m0_rpc_conn_invariant(conn));
 	/* In-core state will be cleaned up by
 	   m0_rpc_conn_terminate_reply_sent() */
-- 
1.8.3.2

