From 6df85b3ba5b07af88d164f4eb1ac6b2c46a5c16b Mon Sep 17 00:00:00 2001
From: Andriy Tkachuk <andriy_tkachuk@xyratex.com>
Date: Tue, 4 Feb 2014 15:36:18 +0200
Subject: [PATCH 33/60] m0_rpc_item_send() fix: take reference before formation

---
 rpc/item.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/rpc/item.c b/rpc/item.c
index f862e09..51fef39 100644
--- a/rpc/item.c
+++ b/rpc/item.c
@@ -730,8 +730,8 @@ M0_INTERNAL void m0_rpc_item_send(struct m0_rpc_item *item)
 	 * See rpc/frmops.c:item_sent() and m0_rpc_item_failed()
 	 */
 	m0_rpc_session_hold_busy(item->ri_session);
-	m0_rpc_frm_enq_item(&item->ri_session->s_conn->c_rpcchan->rc_frm, item);
 	m0_rpc_item_get(item);
+	m0_rpc_frm_enq_item(&item->ri_session->s_conn->c_rpcchan->rc_frm, item);
 	M0_LEAVE();
 }
 
-- 
1.8.3.2

