From f48372603e5f2a5a2b6ebfd42b9e31306fe3d691 Mon Sep 17 00:00:00 2001
From: Andriy Tkachuk <andriy_tkachuk@xyratex.com>
Date: Thu, 6 Feb 2014 00:59:31 +0200
Subject: [PATCH 35/60] enable UTs which pass

---
 scripts/m0mount |  4 ++--
 utils/ut_main.c | 20 ++++++++++----------
 2 files changed, 12 insertions(+), 12 deletions(-)

diff --git a/scripts/m0mount b/scripts/m0mount
index 0e8de58..a73284e 100755
--- a/scripts/m0mount
+++ b/scripts/m0mount
@@ -141,9 +141,9 @@ M0_TRACE_LEVEL=warn+
 M0_TRACE_PRINT_CONTEXT=short
 
 # kernel space tracing parameters
-MERO_MODULE_TRACE_MASK=fop,m0t1fs
+MERO_MODULE_TRACE_MASK=m0t1fs,fop,rpc,formation
 MERO_TRACE_PRINT_CONTEXT=short
-MERO_TRACE_LEVEL=notice+
+MERO_TRACE_LEVEL=warn+
 
 # Local mount data
 MP=/mnt/m0
diff --git a/utils/ut_main.c b/utils/ut_main.c
index f6f0e63..029cb1c 100644
--- a/utils/ut_main.c
+++ b/utils/ut_main.c
@@ -65,7 +65,7 @@ extern const struct m0_test_suite fol_ut;
 //extern const struct m0_test_suite frm_ut;
 extern const struct m0_test_suite ios_bufferpool_ut;
 extern const struct m0_test_suite item_ut;
-//extern const struct m0_test_suite item_source_ut;
+extern const struct m0_test_suite item_source_ut;
 extern const struct m0_test_suite layout_ut;
 extern const struct m0_test_suite m0_addb_ut;
 extern const struct m0_test_suite m0_fop_lock_ut;
@@ -79,16 +79,16 @@ extern const struct m0_test_suite m0_net_test_ut;
 extern const struct m0_test_suite m0_net_tm_prov_ut;
 extern const struct m0_test_suite m0d_ut;
 extern const struct m0_test_suite mdservice_ut;
-//extern const struct m0_test_suite packet_encdec_ut;
+extern const struct m0_test_suite packet_encdec_ut;
 extern const struct m0_test_suite parity_math_ut;
 extern const struct m0_test_suite poolmach_ut;
 extern const struct m0_test_suite reqh_ut;
 extern const struct m0_test_suite reqh_service_ut;
 extern const struct m0_test_suite rm_ut;
-//extern const struct m0_test_suite rpc_mc_ut;
+extern const struct m0_test_suite rpc_mc_ut;
 //extern const struct m0_test_suite rpc_rcv_session_ut;
-//extern const struct m0_test_suite rpc_service_ut;
-//extern const struct m0_test_suite rpclib_ut;
+extern const struct m0_test_suite rpc_service_ut;
+extern const struct m0_test_suite rpclib_ut;
 //extern const struct m0_test_suite session_ut;
 extern const struct m0_test_suite sm_ut;
 extern const struct m0_test_suite sns_cm_repair_ut;
@@ -133,7 +133,7 @@ void add_uts(void)
 //	m0_ut_add(&frm_ut);
 	m0_ut_add(&ios_bufferpool_ut);
 	m0_ut_add(&item_ut);
-//	m0_ut_add(&item_source_ut);
+	m0_ut_add(&item_source_ut);
 	m0_ut_add(&layout_ut);
 	m0_ut_add(&m0_addb_ut);
 	m0_ut_add(&m0_fop_lock_ut);
@@ -146,16 +146,16 @@ void add_uts(void)
 	m0_ut_add(&m0_net_test_ut);
 	m0_ut_add(&m0_net_tm_prov_ut);
 	m0_ut_add(&m0d_ut);
-//	m0_ut_add(&packet_encdec_ut);
+	m0_ut_add(&packet_encdec_ut);
 	m0_ut_add(&parity_math_ut);
 	m0_ut_add(&poolmach_ut);
 	m0_ut_add(&reqh_ut);
 	m0_ut_add(&reqh_service_ut);
 	m0_ut_add(&rm_ut);
-//	m0_ut_add(&rpc_mc_ut);
+	m0_ut_add(&rpc_mc_ut);
 //	m0_ut_add(&rpc_rcv_session_ut);
-//	m0_ut_add(&rpc_service_ut);
-//	m0_ut_add(&rpclib_ut);
+	m0_ut_add(&rpc_service_ut);
+	m0_ut_add(&rpclib_ut);
 //	m0_ut_add(&session_ut);
 	m0_ut_add(&sm_ut);
 	m0_ut_add(&snscm_xform_ut);
-- 
1.8.3.2

