From 6ffcb187fc852c0fe0e9905064e10f58ebbe4157 Mon Sep 17 00:00:00 2001
From: Andriy Tkachuk <andriy_tkachuk@xyratex.com>
Date: Tue, 11 Feb 2014 23:53:31 +0200
Subject: [PATCH 46/60] m0_rpc_session_item_failed() fixed: rio_replied() was
 not called

---
 rpc/session.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/rpc/session.c b/rpc/session.c
index ddfee56..0b7d5b1 100644
--- a/rpc/session.c
+++ b/rpc/session.c
@@ -765,6 +765,11 @@ M0_INTERNAL void m0_rpc_session_item_failed(struct m0_rpc_item *item)
 	M0_PRE(item != NULL && item->ri_error != 0);
 	M0_PRE(item->ri_sm.sm_state == M0_RPC_ITEM_FAILED);
 
+	if (m0_rpc_item_is_request(item)) {
+		M0_ASSERT(item->ri_error != 0);
+		if (item->ri_ops != NULL && item->ri_ops->rio_replied != NULL)
+			item->ri_ops->rio_replied(item);
+	}
 }
 
 #undef M0_TRACE_SUBSYSTEM
-- 
1.8.3.2

