From bf88431de079d5a1d1c9fbdbef0d4af807321d37 Mon Sep 17 00:00:00 2001
From: Andriy Tkachuk <andriy_tkachuk@xyratex.com>
Date: Wed, 12 Feb 2014 12:19:02 +0200
Subject: [PATCH 50/60] Kbuild.in: enable rpc/ut

---
 Kbuild.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Kbuild.in b/Kbuild.in
index a1a8d9b..fead693 100644
--- a/Kbuild.in
+++ b/Kbuild.in
@@ -123,7 +123,7 @@ include $(src)/net/lnet/ut/Kbuild.sub
 include $(src)/net/test/ut/Kbuild.sub # TODO: factor out net-test into a separate module
 include $(src)/reqh/ut/Kbuild.sub
 include $(src)/rm/ut/Kbuild.sub
-#include $(src)/rpc/ut/Kbuild.sub
+include $(src)/rpc/ut/Kbuild.sub
 include $(src)/sm/ut/Kbuild.sub
 include $(src)/ut/Kbuild.sub
 include $(src)/xcode/ut/Kbuild.sub
-- 
1.8.3.2

