From 0fca53912a87fd5d7662ac5cf50660053a5339e0 Mon Sep 17 00:00:00 2001
From: Andriy Tkachuk <andriy_tkachuk@xyratex.com>
Date: Fri, 14 Feb 2014 14:50:43 +0200
Subject: [PATCH 52/60] rpc/formation: one-way items queue dropped

---
 rpc/formation2.c          | 34 ++++++++++++----------------------
 rpc/formation2_internal.h |  6 ++----
 2 files changed, 14 insertions(+), 26 deletions(-)

diff --git a/rpc/formation2.c b/rpc/formation2.c
index 1643e64..3bf4095 100644
--- a/rpc/formation2.c
+++ b/rpc/formation2.c
@@ -74,10 +74,8 @@ static bool
 constraints_are_valid(const struct m0_rpc_frm_constraints *constraints);
 
 static const char *str_qtype[] = {
-	[FRMQ_URGENT_NORMAL]    = "URGENT_NORMAL",
-	[FRMQ_URGENT_ONEWAY]   = "URGENT_ONEWAY",
-	[FRMQ_WAITING_NORMAL]   = "WAITING_NORMAL",
-	[FRMQ_WAITING_ONEWAY]  = "WAITING_ONEWAY"
+	[FRMQ_URGENT]   = "URGENT",
+	[FRMQ_WAITING]  = "WAITING",
 };
 
 M0_BASSERT(ARRAY_SIZE(str_qtype) == FRMQ_NR_QUEUES);
@@ -218,9 +216,8 @@ M0_INTERNAL void m0_rpc_frm_init(struct m0_rpc_frm *frm,
 	frm->f_constraints = *constraints; /* structure instance copy */
 	frm->f_magic       =  M0_RPC_FRM_MAGIC;
 
-	for_each_itemq_in_frm(q, frm) {
+	for_each_itemq_in_frm(q, frm)
 		itemq_tlist_init(q);
-	}
 
 	chan    = frm_rchan(frm);
 	addb_mc = REQH_ADDB_MC_CONFIGURED(chan->rc_rpc_machine->rm_reqh) ?
@@ -255,9 +252,8 @@ M0_INTERNAL void m0_rpc_frm_fini(struct m0_rpc_frm *frm)
 	drop_all_oneway_items(frm);
 	M0_ASSERT(frm->f_state == FRM_IDLE);
 	m0_addb_ctx_fini(&frm->f_addb_ctx);
-	for_each_itemq_in_frm(q, frm) {
+	for_each_itemq_in_frm(q, frm)
 		itemq_tlist_fini(q);
-	}
 
 	frm->f_state = FRM_UNINITIALISED;
 	frm->f_magic = 0;
@@ -272,7 +268,7 @@ static void drop_all_oneway_items(struct m0_rpc_frm *frm)
 	int                 i;
 
 	for (i = 0; i < FRMQ_NR_QUEUES; i++) {
-		if (M0_IN(i, (FRMQ_URGENT_ONEWAY, FRMQ_WAITING_ONEWAY))) {
+		if (M0_IN(i, (FRMQ_URGENT, FRMQ_WAITING))) {
 			q = &frm->f_itemq[i];
 			m0_tl_for(itemq, q, item) {
 				M0_ASSERT(m0_rpc_item_is_oneway(item));
@@ -353,8 +349,7 @@ up:
 M0_INTERNAL bool item_is_in_waiting_queue(const struct m0_rpc_item *item,
 					  const struct m0_rpc_frm *frm)
 {
-	return  M0_IN(item->ri_itemq, (&frm->f_itemq[FRMQ_WAITING_NORMAL],
-				       &frm->f_itemq[FRMQ_WAITING_ONEWAY]));
+	return  M0_IN(item->ri_itemq, (&frm->f_itemq[FRMQ_WAITING]));
 }
 
 /**
@@ -365,27 +360,23 @@ static enum m0_rpc_frm_itemq_type
 frm_which_qtype(struct m0_rpc_frm *frm, const struct m0_rpc_item *item)
 {
 	enum m0_rpc_frm_itemq_type qtype;
-	bool                       oneway;
 	bool                       deadline_passed;
 
 	M0_ENTRY("item: %p", item);
 	M0_PRE(item != NULL);
 
-	oneway          = m0_rpc_item_is_oneway(item);
 	deadline_passed = m0_time_now() >= item->ri_deadline;
 
 	M0_LOG(M0_DEBUG,
-	       "deadline: [%llu:%llu] oneway: %s deadline_passed: %s",
+	       "deadline: [%llu:%llu] deadline_passed: %s",
 	       (unsigned long long)m0_time_seconds(item->ri_deadline),
 	       (unsigned long long)m0_time_nanoseconds(item->ri_deadline),
-	       m0_bool_to_str(oneway), m0_bool_to_str(deadline_passed));
+	       m0_bool_to_str(deadline_passed));
 
 	if (deadline_passed)
-		qtype = oneway ? FRMQ_URGENT_ONEWAY
-			       : FRMQ_URGENT_NORMAL;
+		qtype = FRMQ_URGENT;
 	else
-		qtype = oneway ? FRMQ_WAITING_ONEWAY
-			       : FRMQ_WAITING_NORMAL;
+		qtype = FRMQ_WAITING;
 	M0_LEAVE("qtype: %s", str_qtype[qtype]);
 	return qtype;
 }
@@ -500,7 +491,7 @@ static void frm_balance(struct m0_rpc_frm *frm)
  * FRM_BALANCE_NOTE_1
  * This case can arise if:
  * - Accumulated bytes are >= max_nr_bytes_accumulated,
- *   hence frm is READY AND
+ *   hence frm is READY
  */
 
 /**
@@ -519,8 +510,7 @@ static bool frm_is_ready(const struct m0_rpc_frm *frm)
 	if (M0_FI_ENABLED("ready"))
 		return true;
 	has_urgent_items =
-		!itemq_tlist_is_empty(&frm->f_itemq[FRMQ_URGENT_NORMAL]) ||
-		!itemq_tlist_is_empty(&frm->f_itemq[FRMQ_URGENT_ONEWAY]);
+		!itemq_tlist_is_empty(&frm->f_itemq[FRMQ_URGENT]);
 
 	c = &frm->f_constraints;
 	return frm->f_nr_packets_enqed < c->fc_max_nr_packets_enqed &&
diff --git a/rpc/formation2_internal.h b/rpc/formation2_internal.h
index 4a73bec..6a8341b 100644
--- a/rpc/formation2_internal.h
+++ b/rpc/formation2_internal.h
@@ -148,10 +148,8 @@ enum frm_state {
    before its deadline is passed.
  */
 enum m0_rpc_frm_itemq_type {
-	FRMQ_URGENT_NORMAL,
-	FRMQ_URGENT_ONEWAY,
-	FRMQ_WAITING_NORMAL,
-	FRMQ_WAITING_ONEWAY,
+	FRMQ_URGENT,
+	FRMQ_WAITING,
 	FRMQ_NR_QUEUES
 };
 
-- 
1.8.3.2

