From 1bf611557a6823b95d79ef885f8af46088e29304 Mon Sep 17 00:00:00 2001
From: Andriy Tkachuk <andriy_tkachuk@xyratex.com>
Date: Tue, 25 Feb 2014 20:37:00 +0200
Subject: [PATCH 57/60] rpc/formation: few style fixes by inspection

---
 rpc/formation2.c | 39 +++++++++++++++------------------------
 1 file changed, 15 insertions(+), 24 deletions(-)

diff --git a/rpc/formation2.c b/rpc/formation2.c
index 7e438e6..c61a579 100644
--- a/rpc/formation2.c
+++ b/rpc/formation2.c
@@ -68,7 +68,7 @@ static bool item_will_exceed_packet_size(struct m0_rpc_item         *item,
 
 static bool item_supports_merging(const struct m0_rpc_item *item);
 
-static void drop_all_oneway_items(struct m0_rpc_frm *frm);
+static void drop_all_items(struct m0_rpc_frm *frm);
 
 static bool
 constraints_are_valid(const struct m0_rpc_frm_constraints *constraints);
@@ -249,7 +249,7 @@ M0_INTERNAL void m0_rpc_frm_fini(struct m0_rpc_frm *frm)
 	 * Choosing later one for its simplicity.
 	 * In future iff needed this can be changed to implement first option.
 	 */
-	drop_all_oneway_items(frm);
+	drop_all_items(frm);
 	M0_ASSERT(frm->f_state == FRM_IDLE);
 	m0_addb_ctx_fini(&frm->f_addb_ctx);
 	for_each_itemq_in_frm(q, frm)
@@ -261,24 +261,22 @@ M0_INTERNAL void m0_rpc_frm_fini(struct m0_rpc_frm *frm)
 	M0_LEAVE();
 }
 
-static void drop_all_oneway_items(struct m0_rpc_frm *frm)
+static void drop_all_items(struct m0_rpc_frm *frm)
 {
 	struct m0_rpc_item *item;
 	struct m0_tl       *q;
 	int                 i;
 
 	for (i = 0; i < FRMQ_NR_QUEUES; i++) {
-		if (M0_IN(i, (FRMQ_URGENT, FRMQ_WAITING))) {
-			q = &frm->f_itemq[i];
-			m0_tl_for(itemq, q, item) {
-				M0_ASSERT(m0_rpc_item_is_oneway(item));
-				m0_rpc_item_get(item);
-				frm_remove(frm, item);
-				m0_rpc_item_failed(item, -ECANCELED);
-				m0_rpc_item_put(item);
-			} m0_tl_endfor;
-			M0_ASSERT(itemq_tlist_is_empty(q));
-		}
+		q = &frm->f_itemq[i];
+		m0_tl_for(itemq, q, item) {
+			M0_ASSERT(m0_rpc_item_is_oneway(item));
+			m0_rpc_item_get(item);
+			frm_remove(frm, item);
+			m0_rpc_item_failed(item, -ECANCELED);
+			m0_rpc_item_put(item);
+		} m0_tl_endfor;
+		M0_ASSERT(itemq_tlist_is_empty(q));
 	}
 }
 
@@ -349,7 +347,7 @@ up:
 M0_INTERNAL bool item_is_in_waiting_queue(const struct m0_rpc_item *item,
 					  const struct m0_rpc_frm *frm)
 {
-	return  M0_IN(item->ri_itemq, (&frm->f_itemq[FRMQ_WAITING]));
+	return item->ri_itemq == &frm->f_itemq[FRMQ_WAITING];
 }
 
 /**
@@ -373,10 +371,7 @@ frm_which_qtype(struct m0_rpc_frm *frm, const struct m0_rpc_item *item)
 	       (unsigned long long)m0_time_nanoseconds(item->ri_deadline),
 	       m0_bool_to_str(deadline_passed));
 
-	if (deadline_passed)
-		qtype = FRMQ_URGENT;
-	else
-		qtype = FRMQ_WAITING;
+	qtype = deadline_passed ? FRMQ_URGENT : FRMQ_WAITING;
 	M0_LEAVE("qtype: %s", str_qtype[qtype]);
 	return qtype;
 }
@@ -683,8 +678,6 @@ static void frm_try_merging_item(struct m0_rpc_frm  *frm,
  */
 static int frm_packet_ready(struct m0_rpc_frm *frm, struct m0_rpc_packet *p)
 {
-	int rc;
-
 	M0_ENTRY("frm: %p packet %p", frm, p);
 
 	M0_PRE(frm != NULL && p != NULL && !m0_rpc_packet_is_empty(p));
@@ -694,9 +687,7 @@ static int frm_packet_ready(struct m0_rpc_frm *frm, struct m0_rpc_packet *p)
 
 	p->rp_frm = frm;
 	/* See packet_ready() in rpc/frmops.c */
-	rc = frm->f_ops->fo_packet_ready(p);
-
-	M0_RETURN(rc);
+	M0_RETURN(frm->f_ops->fo_packet_ready(p));
 }
 
 M0_INTERNAL void m0_rpc_frm_run_formation(struct m0_rpc_frm *frm)
-- 
1.8.3.2

