From 3ed1433b61f6b9fae55299fc5e5fb3e7f90b3e56 Mon Sep 17 00:00:00 2001
From: Mandar Sawant <mandar_sawant@xyratex.com>
Date: Wed, 12 Mar 2014 15:46:47 +0530
Subject: [PATCH 8/8] be.conv.confx, - Code align.

---
 conf/db.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/conf/db.c b/conf/db.c
index 27ee360..7216642 100644
--- a/conf/db.c
+++ b/conf/db.c
@@ -178,7 +178,6 @@ static int confx_obj_dup(struct m0_confx_obj **dest, struct m0_confx_obj *src,
 M0_INTERNAL int m0_confx_to_db(struct m0_confx *conf, struct m0_be_seg *seg,
 			       struct m0_be_tx *tx, struct m0_confx **out)
 {
-<<<<<<< HEAD
 	struct m0_be_btree btree = { .bb_seg = seg };
 	int                rc = 0;
 	int                i;
-- 
1.8.3.2

