From 67a1b2289f268a2cef7a42e0056b8e4757b765a0 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Wed, 2 Apr 2014 00:12:34 +0400
Subject: [PATCH 08/13] - unroll not needed changes

---
 rm/rm.c | 42 ++++++++++++++++++++----------------------
 1 file changed, 20 insertions(+), 22 deletions(-)

diff --git a/rm/rm.c b/rm/rm.c
index 81bd491..cbba37c 100644
--- a/rm/rm.c
+++ b/rm/rm.c
@@ -346,47 +346,45 @@ m0_rm_resource_type_lookup(const struct m0_rm_domain *dom,
 }
 
 M0_INTERNAL void m0_rm_resource_add(struct m0_rm_resource_type *rtype,
-				    struct m0_rm_resource      *res)
+                                    struct m0_rm_resource      *res)
 {
-	M0_ENTRY("res-type: %p resource : %p", rtype, res);
-
-	m0_mutex_lock(&rtype->rt_lock);
-	M0_PRE(resource_type_invariant(rtype));
-	M0_PRE(res->r_ref == 0);
-	M0_PRE(m0_rm_resource_find(rtype, res) == NULL);
-	res->r_type = rtype;
+        M0_ENTRY("res-type: %p resource : %p", rtype, res);
 
+        m0_mutex_lock(&rtype->rt_lock);
+        M0_PRE(resource_type_invariant(rtype));
+        M0_PRE(res->r_ref == 0);
+        M0_PRE(m0_rm_resource_find(rtype, res) == NULL);
+        res->r_type = rtype;
         res_tlink_init_at(res, &rtype->rt_resources);
         m0_remotes_tlist_init(&res->r_remote);
-        M0_POST(res_tlist_contains(&rtype->rt_resources, res));
+        m0_rm_resource_bob_init(res);
         M0_CNT_INC(rtype->rt_nr_resources);
+        M0_POST(res_tlist_contains(&rtype->rt_resources, res));
         M0_POST(resource_type_invariant(rtype));
-	m0_rm_resource_bob_init(res);
-	m0_mutex_unlock(&rtype->rt_lock);
-	M0_POST(res->r_type == rtype);
-	M0_LEAVE();
+        m0_mutex_unlock(&rtype->rt_lock);
+        M0_POST(res->r_type == rtype);
+        M0_LEAVE();
 }
 M0_EXPORTED(m0_rm_resource_add);
 
 M0_INTERNAL void m0_rm_resource_del(struct m0_rm_resource *res)
 {
-	struct m0_rm_resource_type *rtype = res->r_type;
+        struct m0_rm_resource_type *rtype = res->r_type;
 
-	M0_ENTRY("resource : %p", res);
-	m0_mutex_lock(&rtype->rt_lock);
+        M0_ENTRY("resource : %p", res);
+        m0_mutex_lock(&rtype->rt_lock);
         M0_PRE(res_tlist_contains(&rtype->rt_resources, res));
         M0_PRE(m0_remotes_tlist_is_empty(&res->r_remote));
         M0_PRE(resource_type_invariant(rtype));
 
         res_tlink_del_fini(res);
-
         M0_CNT_DEC(rtype->rt_nr_resources);
-        M0_POST(!res_tlist_contains(&rtype->rt_resources, res));
-        M0_POST(resource_type_invariant(rtype));
 
-	m0_rm_resource_bob_fini(res);
-	m0_mutex_unlock(&rtype->rt_lock);
-	M0_LEAVE();
+        M0_POST(resource_type_invariant(rtype));
+        M0_POST(!res_tlist_contains(&rtype->rt_resources, res));
+        m0_rm_resource_bob_fini(res);
+        m0_mutex_unlock(&rtype->rt_lock);
+        M0_LEAVE();
 }
 M0_EXPORTED(m0_rm_resource_del);
 
-- 
1.8.3.2

