From 6a0ad5e73a18eea6f753500a4226c0306da1fd0e Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Wed, 2 Apr 2014 17:36:28 +0400
Subject: [PATCH 11/13] - comment fixes

---
 m0t1fs/linux_kernel/m0t1fs.h | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/m0t1fs/linux_kernel/m0t1fs.h b/m0t1fs/linux_kernel/m0t1fs.h
index f423d77..58a0aa8 100644
--- a/m0t1fs/linux_kernel/m0t1fs.h
+++ b/m0t1fs/linux_kernel/m0t1fs.h
@@ -87,12 +87,14 @@
       Client allocates fid for a file during its creation. It assigns
       <0, $fid_start> for first file, <0, $fid_start+1> for second
       and so on.
-      mdstore uses constant fid - M0_MDSERVICE_SLASH_FID <1, 1> - for root.
-      Couple more are reserved for sessions and other things.
+      mdstore uses constant fid - M0_MDSERVICE_SLASH_FID <1, 65535>
+      for namespace root exposed to user.
 
-      One more - for .mero dir <1, 4> and one more for .mero/fid dir <1, 5>
-      Hence fid_start must be greater than 5. See M0_DOT_MERO_FID and
-      M0_DOT_MERO_FID_FID
+      Fids below 65535 are reserved for special objects such as: storage root,
+      ".mero", ".mero/fid" and potentially many others that will live in ".mero"
+      directory.
+
+      Hence fid_start must be greater than 65535.
 
    'device' argument of mount command is ignored.
 
-- 
1.8.3.2

