From ab48d7958b654bdd12b7f11d08f7ec0f38f9b7ef Mon Sep 17 00:00:00 2001
From: "jinshan.xiong" <jinshan.xiong@clusterstor.com>
Date: Thu, 15 Jul 2010 21:43:55 -0600
Subject: [PATCH 07/34] - move c2_dbuf to adt.h

---
 lib/adt.h | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/lib/adt.h b/lib/adt.h
index bed023a..e516563 100644
--- a/lib/adt.h
+++ b/lib/adt.h
@@ -28,6 +28,34 @@ struct c2_buf {
 	c2_bcount_t b_nob;
 };
 
+/**
+  data buffer description in C2 service.
+
+  Whenever client wants to send a data buffer to server, the data buffer
+  must be encoded as a c2_dbuf. The most important field is @alignment which
+  indicates the alignment requirement of data.
+ */
+struct c2_dbuf {
+        int alignment; /**< alignment of data buffer */
+        int buflen;
+        char *buffer;
+};
+
+/**
+  xdr function for c2_dbuf.
+
+  @param[in] xdrs - 
+  @param[in] dbuf - the c2_dbuf will be encoded or decoded
+  @param[in] xdr  - this is a callback function to encode/decode XDR stream;
+                    for rdma, this function must be xdr_rdma_chunks;
+                    otherwise NULL
+
+  @retval TRUE    - success
+  @retval FALSE   - something wrong
+ */
+bool_t   c2_xdr_dbuf   (XDRS *xdrs, struct c2_dbuf *dbuf,
+                        bool_t (*xdr)(XDRS *, char *, int));
+
 /** @} end of adt group */
 
 
-- 
1.8.3.2

