From 95af6b71c72b00ede5086b2633783c81fbae595f Mon Sep 17 00:00:00 2001
From: "jinshan.xiong" <jinshan.xiong@clusterstor.com>
Date: Mon, 19 Jul 2010 11:34:33 -0600
Subject: [PATCH 08/34] - small changes applied

---
 net/usunrpc/rfc5666/svc_rdma.h | 17 +----------------
 1 file changed, 1 insertion(+), 16 deletions(-)

diff --git a/net/usunrpc/rfc5666/svc_rdma.h b/net/usunrpc/rfc5666/svc_rdma.h
index 6dbed77..382caee 100644
--- a/net/usunrpc/rfc5666/svc_rdma.h
+++ b/net/usunrpc/rfc5666/svc_rdma.h
@@ -23,19 +23,6 @@ struct svcrdma_config {
         int max_buf_size;
 };
 
-/**
-  data buffer description in C2 service.
-
-  Whenever client wants to send a data buffer to server, the data buffer
-  must be encoded as a c2_dbuf. The most important field is @alignment which
-  indicates the alignment requirement of data.
- */
-struct c2_dbuf {
-        int alignment; /**< alignment of data buffer */
-        int buflen;
-        char *buffer;
-};
-
 /* external interfaces */
 /**
   svcrdma_create tries to create a service on RDMA.
@@ -49,9 +36,7 @@ struct c2_dbuf {
  */
 SVCXPRT *svcrdma_create(struct sockaddr_in *sa, struct svcrdma_config *cfg);
 /**
- */
-bool_t   xdr_dbuf      (XDRS *xdrs, struct c2_dbuf *dbuf);
-/**
+  a standard xdr function for c2_dbuf in rdma case.
  */
 bool_t   xdr_rdma_chunks(XDRS *xdrs, char *buffer, int buflen);
 
-- 
1.8.3.2

