From f078455891a82e145e6e6d3e165d6bfef1e7b832 Mon Sep 17 00:00:00 2001
From: "jinshan.xiong" <jinshan.xiong@clusterstor.com>
Date: Fri, 6 Aug 2010 14:27:12 -0600
Subject: [PATCH 30/34] - further fix mentioned by nikita

---
 configure.ac     | 5 -----
 stob/ut/server.c | 3 ++-
 2 files changed, 2 insertions(+), 6 deletions(-)

diff --git a/configure.ac b/configure.ac
index 8eda013..d1943af 100755
--- a/configure.ac
+++ b/configure.ac
@@ -289,11 +289,6 @@ AC_ARG_ENABLE([db-o_direct],
         [enable_db_o_direct=no]
 )
 
-AC_ARG_ENABLE([rdma],
-        [AC_HELP_STRING([--enable-rdma], [enable rdma support])], [],
-        [enable_rdma=no]
-)
-
 AC_MSG_CHECKING([for operating system])
 AC_MSG_RESULT([$host_os])
 case $host_os in
diff --git a/stob/ut/server.c b/stob/ut/server.c
index 02ae930..a406db8 100644
--- a/stob/ut/server.c
+++ b/stob/ut/server.c
@@ -434,7 +434,8 @@ int main(int argc, char **argv)
 	if (strcmp(trans, "tcp") ||
 	    strcmp(trans, "udp") ||
 	    strcmp(trans, "rdma"))
-		fprintf(stderr, "unknown transport %s\n", trans);
+		fprintf(stderr, "unknown transport %s, it must be one"
+			"of tcp or rdma\n", trans);
 		return -1;
 
 	C2_ASSERT(strlen(path) < ARRAY_SIZE(opath) - 8);
-- 
1.8.3.2

