From 2eba1ff9aba6ced43eb818a4def300831d2484b7 Mon Sep 17 00:00:00 2001
From: Carl Braganza <Carl_Braganza@us.xyratex.com>
Date: Wed, 2 Mar 2011 14:09:19 -0800
Subject: [PATCH 31/34] removed duplicate rule for io_getevents - restored to
 master version

---
 configure.ac | 18 ++++--------------
 1 file changed, 4 insertions(+), 14 deletions(-)

diff --git a/configure.ac b/configure.ac
index d1943af..178a799 100755
--- a/configure.ac
+++ b/configure.ac
@@ -446,24 +446,15 @@ if test x$enable_gssrpc = xyes; then
         fi
 fi
 
-if test x$OS = xlinux; then
-        OLD_LIBS=$LIBS
-        LIBS=""
-        AC_SEARCH_LIBS([io_getevents], [c aio], [], [AC_MSG_ERROR([io_getevents cannot be found!])])
-        AIO_LIBS=$LIBS
-        AC_SUBST([AIO_LIBS])
-        LIBS=$OLD_LIBS
-fi
-
 AC_CHECK_LIB([rdmacm], [rdma_create_id], [],
-             [AC_MSG_ERROR([cannot find ramdcm lib])])
+             [AC_MSG_ERROR([cannot find rdmacm lib])])
 AC_CHECK_LIB([ibverbs], [ibv_open_device], [],
              [AC_MSG_ERROR([cannot find ibverbs])])
 AC_CHECK_HEADERS([rdma/rdma_cma.h infiniband/verbs.h], [],
                  [AC_MSG_ERROR([cannot find rdma header files])])
-if test x$OS = xlinux; then
-         AC_MSG_CHECKING([if kernel supports sunrpc over rdma])
-         COLIBRI_LINUX_TRY_COMPILE([
+
+AC_MSG_CHECKING([if kernel supports sunrpc over rdma])
+COLIBRI_LINUX_TRY_COMPILE([
                 #include <linux/sunrpc/xprt.h>
          ],[
                 int abc = XPRT_TRANSPORT_RDMA;
@@ -472,7 +463,6 @@ if test x$OS = xlinux; then
          ],[
                 AC_MSG_ERROR([no kernel rdma, sorry!])
          ])
-fi
 
 OLD_LIBS=$LIBS
 LIBS=""
-- 
1.8.3.2

