From 06ca81d0ad4ff488a5a539dd28adbe427aba6268 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Thu, 3 Mar 2011 03:38:22 -0700
Subject: [PATCH 20/91] - some changes proposed by Nikita in inspection time.

---
 cob/cob.h         | 18 +++++++++++-------
 mdstore/mdstore.c | 12 ++++--------
 2 files changed, 15 insertions(+), 15 deletions(-)

diff --git a/cob/cob.h b/cob/cob.h
index 97848a0..3f7671b 100644
--- a/cob/cob.h
+++ b/cob/cob.h
@@ -79,7 +79,10 @@ struct c2_cob_domain {
         struct c2_table         cd_namespace;
         struct c2_table         cd_fileattr_basic;
         struct c2_table         cd_fileattr_omg;
-        /** an ADDB context for events related to this domain. */
+
+        /** 
+           an ADDB context for events related to this domain. 
+        */
         struct c2_addb_ctx      cd_addb;
 };
 
@@ -110,8 +113,10 @@ struct c2_cob_nsrec {
         uint32_t          cnr_nlink;   /**< number of hard links. */
         uint64_t          cnr_omgid;   /**< uid/gid/mode slot reference */
 
-        uint32_t          cnr_rdev;    /**< device ID (if special file). */
-        uint64_t          cnr_size;    /**< total size, in bytes. */
+        union {
+                uint32_t  cnr_rdev;    /**< device ID (if special file). */
+                uint64_t  cnr_size;    /**< total size, in bytes. */
+        } u;
         uint32_t          cnr_blksize; /**< blocksize for filesystem I/O. */
         uint32_t          cnr_blocks;  /**< number of blocks allocated. */
         uint32_t          cnr_atime;   /**< time of last access. */
@@ -138,10 +143,9 @@ struct c2_cob_oikey {
    from smaller name-space entries.
  */
 struct c2_cob_fabrec {
-        struct c2_verno   cfb_version; /**< version from last fop */
-        uint32_t          cfb_sid;     /**< stream id performed modification. */
-        uint32_t          cfb_nid;     /**< node id performed modification. */
-        /* add ACL, layout_ref, any other md not needed for stat(2) */
+        struct c2_verno   cfb_version;  /**< version from last fop */
+        uint64_t          cfb_layoutid; /**< reference to layout. */
+        /* add ACL, any other md not needed for stat(2) */
 };
 
 struct c2_cob_omgrec {
diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 3f552c4..92d126a 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -122,13 +122,14 @@ int c2_md_store_create(struct c2_md_store *md, struct c2_fop *fop,
          * @todo: do we need rdev support for specil files in
          * replicator? 
          */
-        nsrec.cnr_rdev = 0;
+        nsrec.u.cnr_rdev = 0;
         
         /*
          * @todo: these need to be fixed. Before data replication
          * we don't need them.
          */
-        nsrec.cnr_size = 0;
+        nsrec.u.cnr_size = 0;
+
         nsrec.cnr_blksize = 0;
         nsrec.cnr_blocks = 0;
 
@@ -142,12 +143,9 @@ int c2_md_store_create(struct c2_md_store *md, struct c2_fop *fop,
         nsrec.cnr_mtime = 0;
         nsrec.cnr_ctime = 0;
 
+        /* @todo: Fill fabrec */
         C2_SET0(&fabrec);
         
-        /* @todo: Update fabrec version. */
-        fabrec.cfb_sid = body->b_sid;
-        fabrec.cfb_nid = body->b_nid;
-
         /* @todo: Fill omgrec */
         C2_SET0(&omgrec);
 
@@ -478,8 +476,6 @@ int c2_md_store_setattr(struct c2_md_store *md, struct c2_fop *fop,
 
         /* @todo: Update fabrec version? */
         fabrec = cob->co_fabrec;
-        fabrec.cfb_sid = body->b_sid;
-        fabrec.cfb_nid = body->b_nid;
         
         /* @todo: Fill omgrec */
         C2_SET0(&omgrec);
-- 
1.8.3.2

