From d25bf9eded956107f6f0df856c4cfe2a2f727111 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Mon, 18 Apr 2011 04:12:17 -0600
Subject: [PATCH 67/91] - check for error in tx_init

---
 reqh/reqh.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/reqh/reqh.c b/reqh/reqh.c
index 8680cef..e3d3770 100644
--- a/reqh/reqh.c
+++ b/reqh/reqh.c
@@ -46,8 +46,10 @@ int c2_reqh_fop_handle(struct c2_reqh *reqh, struct c2_fop *fop, void *cookie)
         C2_SET0(&ctx);
         C2_SET0(&env);
 
-        c2_db_tx_init(&tx.tx_dbtx, 
-                      reqh->rh_site->s_mdstore->md_dom.cd_dbenv, 0);
+        result = c2_db_tx_init(&tx.tx_dbtx, 
+                               reqh->rh_site->s_mdstore->md_dom.cd_dbenv, 0);
+        if (result != 0)
+                return result;
 
         ctx.fc_site = reqh->rh_site;
         ctx.fc_fol  = reqh->rh_fol;
-- 
1.8.3.2

