From 2d64c00e43bbd8a04904fd365b5be90aa556e847 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Thu, 21 Apr 2011 05:58:17 -0600
Subject: [PATCH 78/91] - fixed with nlink mgmt during unlink in mdstore.

---
 mdstore/mdstore.c | 19 +++++++++++--------
 1 file changed, 11 insertions(+), 8 deletions(-)

diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index fce2c20..4b090ec 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -136,6 +136,8 @@ int c2_md_store_create_attr(struct c2_md_store *md,
          * is saved.
          */
         rc = c2_cob_update(cob, &cob->co_nsrec, NULL, NULL, tx);
+        if (md->md_root && !c2_fid_eq(cob->co_fid, md->md_root->co_fid))
+                C2_ASSERT(cob->co_nsrec.cnr_nlink <= 2);
         return rc;
 }
 
@@ -310,20 +312,21 @@ int c2_md_store_unlink(struct c2_md_store *md,
                 c2_free(nskey);
                 if (rc)
                         goto out;
-                
-                cob->co_nsrec.cnr_nlink--;
+                pcob->co_nsrec.cnr_nlink--;
+
                 c2_cob_make_nskey(&nskey, cob->co_fid, ".", 1);
                 rc = c2_cob_del_name(cob, nskey, &ctx->fc_tx->tx_dbtx);
                 c2_free(nskey);
                 if (rc)
                         goto out;
-
                 cob->co_nsrec.cnr_nlink--;
                 
                 /*
-                 * No hardlinks on directories, nlink should be 0 now.
+                 * No hardlinks on directories, nlink should be 1 now
+                 * as we just killed "." and only one ref left - name
+                 * itself.
                  */
-//                C2_ASSERT(cob->co_nsrec.cnr_nlink == 0);
+                C2_ASSERT(cob->co_nsrec.cnr_nlink == 1);
         } else {
                 c2_cob_make_nskey(&nskey, pcob->co_fid, 
                                   req->u_name.n_name,
@@ -368,9 +371,9 @@ int c2_md_store_unlink(struct c2_md_store *md,
                         }
                 }
                 c2_free(nskey);
-                cob->co_nsrec.cnr_nlink--;
         }
 
+        cob->co_nsrec.cnr_nlink--;
         if (cob->co_nsrec.cnr_nlink == 0) {
                 /* 
                  * Last nlink, let's delete everything left. 
@@ -394,8 +397,8 @@ int c2_md_store_unlink(struct c2_md_store *md,
         pcob->co_nsrec.cnr_mtime = time(&now);
                          
         /*
-         * Update on-store cob so that updated time
-         * is saved.
+         * Update on-store cob so that updated time and nlink
+         * are saved.
          */
         rc = c2_cob_update(pcob, &pcob->co_nsrec,  NULL, NULL, 
                            &ctx->fc_tx->tx_dbtx);
-- 
1.8.3.2

