From a2b0ee26a9b9951c4ea7576e1dd91a9003c34fd0 Mon Sep 17 00:00:00 2001
From: Rajanikant Chirmade <rajanikant_chirmade@xyratex.com>
Date: Tue, 11 Mar 2014 16:47:09 +0530
Subject: [PATCH 09/20]  - Added precondition for arm_timer code.

---
 fop/fom.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/fop/fom.c b/fop/fom.c
index a007ec1..ecd353b 100644
--- a/fop/fom.c
+++ b/fop/fom.c
@@ -806,6 +806,7 @@ static void loc_ast_rearm_timer(struct m0_sm_group *grp, struct m0_sm_ast *ast)
 						   fl_post_stats_ast);
 	int                     result;
 
+	M0_PRE(m0_locality_invariant(loc));
 	m0_timer_stop(&loc->fl_fop_rate_timer);
 	m0_timer_fini(&loc->fl_fop_rate_timer);
 
-- 
1.8.3.2

