From e4dea1536ee13c9665ca5bb908837e5a68e22114 Mon Sep 17 00:00:00 2001
From: Rajanikant Chirmade <rajanikant_chirmade@xyratex.com>
Date: Mon, 24 Mar 2014 19:37:22 +0530
Subject: [PATCH 15/20]  - removed trailing whitespaces

---
 fop/fop_rate_monitor.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fop/fop_rate_monitor.h b/fop/fop_rate_monitor.h
index a54d3ee..2a1590d 100644
--- a/fop/fop_rate_monitor.h
+++ b/fop/fop_rate_monitor.h
@@ -56,7 +56,7 @@ struct m0_fop_rate_monitor {
  * It initialise fop rate monitor.
  * @param loc pointer to fom locality.
  * @retval 0 if sucess
- * 	   -ENOMEM if failed to allocate summary record.
+ *         -ENOMEM if failed to allocate summary record.
  */
 M0_INTERNAL
 int m0_fop_rate_monitor_init(struct m0_fom_locality *loc);
-- 
1.8.3.2

