From 390d00768372ffda41123b2cc61aa5858e8a54f1 Mon Sep 17 00:00:00 2001
From: Rajanikant Chirmade <rajanikant_chirmade@xyratex.com>
Date: Wed, 2 Apr 2014 12:30:32 +0530
Subject: [PATCH 20/20]  - Removed unnecessary veriable.

---
 fop/fom.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/fop/fom.c b/fop/fom.c
index 2be1dde..ca25c22 100644
--- a/fop/fom.c
+++ b/fop/fom.c
@@ -862,14 +862,12 @@ static void loc_ast_post_stats(struct m0_sm_group *grp, struct m0_sm_ast *ast)
 	struct m0_fom_locality     *loc = container_of(ast,
 						       struct m0_fom_locality,
 						       fl_post_stats_ast);
-	struct m0_fop_rate_monitor *fop_rate_monitor =
-		m0_fop_rate_monitor_get(loc);
 	struct m0_addb_ctx         *cv[] = {&loc->fl_addb_ctx, NULL};
 	struct m0_addb_mc          *mc = &loc->fl_dom->fd_reqh->rh_addb_mc;
 
 	m0_addb_post_cntr(mc, cv, &loc->fl_stat_run_times);
 	m0_addb_post_cntr(mc, cv, &loc->fl_stat_sched_wait_times);
-	m0_addb_post_cntr(mc, cv, &fop_rate_monitor->frm_addb_ctr);
+	m0_addb_post_cntr(mc, cv, &m0_fop_rate_monitor_get(loc)->frm_addb_ctr);
 
 	M0_ADDB_POST(mc, &m0_addb_rt_fl_runq_nr, cv, loc->fl_runq_nr);
 	M0_ADDB_POST(mc, &m0_addb_rt_fl_wail_nr, cv, loc->fl_wail_nr);
-- 
1.8.3.2

