From 04d3c8c860c3ae70524e9c2d518f98f5d65d33e3 Mon Sep 17 00:00:00 2001
From: Nikita Danilov <nikita_danilov@xyratex.com>
Date: Thu, 13 Sep 2012 15:39:50 +0400
Subject: [PATCH 2/3] Add missing file to the branch.

---
 lib/string.h | 50 ++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 50 insertions(+)
 create mode 100644 lib/string.h

diff --git a/lib/string.h b/lib/string.h
new file mode 100644
index 0000000..81c7914
--- /dev/null
+++ b/lib/string.h
@@ -0,0 +1,50 @@
+/* -*- C -*- */
+/*
+ * COPYRIGHT 2012 XYRATEX TECHNOLOGY LIMITED
+ *
+ * THIS DRAWING/DOCUMENT, ITS SPECIFICATIONS, AND THE DATA CONTAINED
+ * HEREIN, ARE THE EXCLUSIVE PROPERTY OF XYRATEX TECHNOLOGY
+ * LIMITED, ISSUED IN STRICT CONFIDENCE AND SHALL NOT, WITHOUT
+ * THE PRIOR WRITTEN PERMISSION OF XYRATEX TECHNOLOGY LIMITED,
+ * BE REPRODUCED, COPIED, OR DISCLOSED TO A THIRD PARTY, OR
+ * USED FOR ANY PURPOSE WHATSOEVER, OR STORED IN A RETRIEVAL SYSTEM
+ * EXCEPT AS ALLOWED BY THE TERMS OF XYRATEX LICENSES AND AGREEMENTS.
+ *
+ * YOU SHOULD HAVE RECEIVED A COPY OF XYRATEX'S LICENSE ALONG WITH
+ * THIS RELEASE. IF NOT PLEASE CONTACT A XYRATEX REPRESENTATIVE
+ * http://www.xyratex.com/contact
+ *
+ * Original author: Nikita Danilov <nikita_danilov@xyratex.com>
+ * Original creation date: 01-Sep-2012
+ */
+
+#pragma once
+
+#ifndef __COLIBRI_LIB_STRING_H__
+#define __COLIBRI_LIB_STRING_H__
+
+/*
+ * Define standard string manipulation functions (strcat, strlen, strcmp, &c.)
+ * together with sprintf(3) and snprintf(3).
+ */
+
+#ifndef __KERNEL__
+# include <stdio.h>
+# include <string.h>
+#else
+# include <linux/kernel.h>
+# include <linux/string.h>
+#endif
+
+/* __COLIBRI_LIB_STRING_H__ */
+#endif
+
+/*
+ *  Local variables:
+ *  c-indentation-style: "K&R"
+ *  c-basic-offset: 8
+ *  tab-width: 8
+ *  fill-column: 80
+ *  scroll-step: 1
+ *  End:
+ */
-- 
1.8.3.2

