From 34022e7c1a8c0f434a9af3fd1f29dd1229af9b8c Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Mon, 22 Apr 2013 17:08:48 +0800
Subject: [PATCH 02/14] events to be handled by master node.

---
 pool/pool.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/pool/pool.c b/pool/pool.c
index 93fd440..898cebb 100644
--- a/pool/pool.c
+++ b/pool/pool.c
@@ -91,6 +91,9 @@
    node gets this message, it should decode update event from the fop and
    update its pool machine state accordingly.
 
+   The pool machine update event should be delivered to master node. Master
+   node will handle the update event, and send update FOPs to other replica.
+
    <hr>
    @section pool_mach_store_replica-lspec Logical Specification
 
-- 
1.8.3.2

