From f40c219ad2b7f8731b790785d8b984c1b7c3425a Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Mon, 10 Mar 2014 12:17:06 +0800
Subject: [PATCH 18/26] better coding style for bool comparism.

---
 lib/bitmap.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/bitmap.c b/lib/bitmap.c
index 5c8eab0..205adef 100644
--- a/lib/bitmap.c
+++ b/lib/bitmap.c
@@ -107,9 +107,10 @@ M0_EXPORTED(m0_bitmap_get);
 M0_INTERNAL size_t m0_bitmap_ffz(const struct m0_bitmap *map)
 {
 	size_t idx;
+
 	/* use linux find_first_zero_bit() ? */
 	for (idx = 0; idx < map->b_nr; idx++) {
-		if (m0_bitmap_get(map, idx) == false)
+		if (!m0_bitmap_get(map, idx))
 			return idx;
 	}
 	return -1;
-- 
1.8.3.2

