From 0f9646bca5631f22ee1c321e2a4fb01346201479 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Thu, 7 Mar 2013 17:56:38 +0200
Subject: [PATCH 05/63] console: minor style fixes

---
 console/console.c   | 5 +++--
 console/st/server.c | 2 +-
 2 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/console/console.c b/console/console.c
index 57f3f1a..2fc32b6 100644
--- a/console/console.c
+++ b/console/console.c
@@ -39,8 +39,9 @@
    @{
  */
 
-uint32_t timeout;
-bool     verbose;
+bool verbose;
+
+static uint32_t timeout;
 
 /**
  * @brief Iterate over FOP and print names of its members.
diff --git a/console/st/server.c b/console/st/server.c
index 5c18fc5..b1f70a0 100644
--- a/console/st/server.c
+++ b/console/st/server.c
@@ -42,7 +42,6 @@
 #define SERVER_DB_FILE_NAME   "cons_server.db"
 #define SERVER_STOB_FILE_NAME "cons_server.stob"
 #define SERVER_LOG_FILE_NAME  "cons_server.log"
-#define CONSOLE_STR_LEN       16
 
 static int signaled = 0;
 
@@ -54,6 +53,7 @@ static void sig_handler(int num)
 /** @brief Test server for m0console */
 int main(int argc, char **argv)
 {
+	enum { CONSOLE_STR_LEN = 16 };
 	char     tm_len[CONSOLE_STR_LEN];
 	char     rpc_size[CONSOLE_STR_LEN];
 	int      result;
-- 
1.8.3.2

