From c9915c4975b5c8e116050ab71de3f6116babcfe5 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Wed, 13 Mar 2013 13:16:56 +0200
Subject: [PATCH 19/63] fop (Makefile.sub): fix an error in CLEANFILES

---
 fop/Makefile.sub | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/fop/Makefile.sub b/fop/Makefile.sub
index 5dda9cc..af6b803 100644
--- a/fop/Makefile.sub
+++ b/fop/Makefile.sub
@@ -21,7 +21,7 @@ EXTRA_DIST += fop/linux_kernel/fop_kernel.h
 
 CLEANFILES += fop/fop_xc.h \
               fop/fop_xc.c \
-              fop/fop_xc.gccxml \
+              fop/fop.gccxml \
               fop/fom_generic_xc.h \
               fop/fom_generic_xc.c \
-	      fop/fom_generic.gccxml
+              fop/fom_generic.gccxml
-- 
1.8.3.2

