From dd223c3587e7552893433c6a95aea32d4153dc06 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Thu, 14 Mar 2013 09:55:08 +0200
Subject: [PATCH 24/63] utils/ub_main.c: use `-o' for options

---
 utils/ub_main.c | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/utils/ub_main.c b/utils/ub_main.c
index 01579ea..93d3f8a 100644
--- a/utils/ub_main.c
+++ b/utils/ub_main.c
@@ -49,15 +49,15 @@ extern struct m0_ub_set m0_trace_ub;
 #define UB_SANDBOX "./ub-sandbox"
 
 struct ub_args {
-	bool        ua_ub_list;
-	uint32_t    ua_rounds;
-	char       *ua_name;
-	char       *ua_args;
+	uint32_t ua_rounds;
+	char    *ua_name;
+	char    *ua_opts;
+	bool     ua_ub_list;
 };
 
 static void ub_args_fini(struct ub_args *args)
 {
-	m0_free(args->ua_args);
+	m0_free(args->ua_opts);
 	m0_free(args->ua_name);
 }
 
@@ -65,7 +65,7 @@ static int ub_args_parse(int argc, char *argv[], struct ub_args *out)
 {
 	out->ua_rounds = 1;
 	out->ua_name = NULL;
-	out->ua_args = NULL;
+	out->ua_opts = NULL;
 	out->ua_ub_list = false;
 
 	return M0_GETOPTS("ub", argc, argv,
@@ -82,9 +82,10 @@ static int ub_args_parse(int argc, char *argv[], struct ub_args *out)
 			       LAMBDA(void, (const char *str) {
 					       out->ua_name = strdup(str);
 				       })),
-		  M0_STRINGARG('a', "Benchmark test args",
+		  M0_STRINGARG('o', "Benchmark options:"
+			       " OPTION=VALUE[,OPTION=VALUE]...",
 			       LAMBDA(void, (const char *str) {
-					       out->ua_args = strdup(str);
+					       out->ua_opts = strdup(str);
 				       }))
 		);
 }
-- 
1.8.3.2

