From 7d38b27fb7bf366e25414c482917254a0b720a20 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Fri, 15 Mar 2013 15:43:25 +0200
Subject: [PATCH 35/63] rpc-ub: add `-w 10' to m0d CLI options

---
 rpc/ub/ub.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/rpc/ub/ub.c b/rpc/ub/ub.c
index 59c1f0e..4de81af 100644
--- a/rpc/ub/ub.c
+++ b/rpc/ub/ub.c
@@ -67,7 +67,7 @@ static struct ub_rpc_client g_client[3];
 
 #define NAME(ext) "rpc-ub" ext
 static char *g_argv[] = {
-	NAME(""), "-r", "-p", "-T", "AD", "-D", NAME(".db"),
+	NAME(""), "-w", "10", "-r", "-p", "-T", "AD", "-D", NAME(".db"),
 	"-S", NAME(".stob"), "-A", NAME(".addb-stob"), "-e", SERVER_ENDPOINT,
 	"-s", "ds1"
 };
-- 
1.8.3.2

