From a3c363d2391025c8acd38e2104b616a0b4237642 Mon Sep 17 00:00:00 2001
From: Anatoliy Bilenko <anatoliy_bilenko@xyratex.com>
Date: Mon, 18 Mar 2013 15:21:55 +0200
Subject: [PATCH 38/63] scripts/m0: Handled return status in build() function
 explicitly.

---
 scripts/m0 | 1 +
 1 file changed, 1 insertion(+)

diff --git a/scripts/m0 b/scripts/m0
index bdc31bd..2a8be46 100755
--- a/scripts/m0
+++ b/scripts/m0
@@ -34,6 +34,7 @@ build() {
     local _OUT=`mktemp`
     trap "rm $_OUT" 0
     make $MAKE_OPTS 2>&1 | tee $_OUT
+    [ ${PIPESTATUS[0]} -eq 0 ] || exit ${PIPESTATUS[0]}
     if grep 'WARNING:' $_OUT; then
 	die 'WARNING in make output is treated as error'
     fi
-- 
1.8.3.2

