From eed47cc37afddc10ed4d44f2c30aa61a0a931546 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Wed, 20 Mar 2013 22:55:18 +0200
Subject: [PATCH 46/63] rpc-ub: make the number of connections configurable

---
 rpc/ub/ub.c | 34 ++++++++++++++++++++++------------
 1 file changed, 22 insertions(+), 12 deletions(-)

diff --git a/rpc/ub/ub.c b/rpc/ub/ub.c
index bc2e7fd..ae570b0 100644
--- a/rpc/ub/ub.c
+++ b/rpc/ub/ub.c
@@ -54,7 +54,7 @@ static void args_init(struct args *args)
 	args->a_use_lnet = false;
 }
 
-enum {
+enum { /* XXX TODO: Revise these values. */
 	MAX_NR_CONNS = 10000,
 	MAX_NR_SLOTS = 100,
 	MAX_MSG_LEN  = 8192
@@ -141,7 +141,7 @@ struct ub_rpc_client {
 	struct m0_dbenv          rc_dbenv;
 };
 
-static struct ub_rpc_client g_client[1]; /* XXX CONFIGUREME */
+static struct ub_rpc_client *g_clients;
 
 #define NAME(ext) "rpc-ub" ext
 static char *g_argv[] = {
@@ -221,28 +221,38 @@ static int _start(const char *opts)
 		return rc;
 
 	rc = m0_rpc_server_start(&g_sctx);
-	if (rc != 0) {
-		m0_net_xprt_fini(g_xprt);
-		return rc;
+	if (rc != 0)
+		goto xprt_fini;
+
+	M0_ALLOC_ARR(g_clients, g_args.a_nr_conns);
+	if (g_clients == NULL) {
+		rc = -ENOMEM;
+		goto server_stop;
 	}
 
-	for (i = 0; i < ARRAY_SIZE(g_client); ++i) {
+	for (i = 0; i < g_args.a_nr_conns; ++i) {
 		snprintf(db, ARRAY_SIZE(db), "rpc-ub.client-db_%d", i);
 		snprintf(ep, ARRAY_SIZE(ep), CLIENT_ENDPOINT_FMT,
 			 2 + i); /* 1 is server EP, so we start from 2 */
-		_client_start(&g_client[i], CLIENT_COB_DOM_ID + i, db, ep);
+		_client_start(&g_clients[i], CLIENT_COB_DOM_ID + i, db, ep);
 	}
+
 	m0_rpc_ub_fops_init();
 	return 0;
+server_stop:
+	m0_rpc_server_stop(&g_sctx);
+xprt_fini:
+	m0_net_xprt_fini(g_xprt);
+	return rc;
 }
 
 static void _stop(void)
 {
 	int i;
 
-	for (i = 0; i < ARRAY_SIZE(g_client); ++i)
-		_client_stop(&g_client[i]);
-
+	for (i = 0; i < g_args.a_nr_conns; ++i)
+		_client_stop(&g_clients[i]);
+	m0_free(g_clients);
 	m0_rpc_server_stop(&g_sctx);
 	m0_net_xprt_fini(g_xprt);
 	m0_rpc_ub_fops_fini();
@@ -308,8 +318,8 @@ static void run(int iter M0_UNUSED)
 	   Needs investigation!
 	 */
 	for (i = 0; i < 3; ++i) {
-		for (j = 0; j < ARRAY_SIZE(g_client); ++j) {
-			session = &g_client[j].rc_ctx.rcx_session;
+		for (j = 0; j < g_args.a_nr_conns; ++j) {
+			session = &g_clients[j].rc_ctx.rcx_session;
 			fop_send(session, i);
 		}
 	}
-- 
1.8.3.2

