From ae2f26cdd4ca8216504085ef5ed218af9d1fd5f5 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Thu, 21 Mar 2013 17:27:20 +0200
Subject: [PATCH 53/63] rpc-ub (args_format_error): update the error message

---
 rpc/ub/ub.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/rpc/ub/ub.c b/rpc/ub/ub.c
index fa70eed..fce7416 100644
--- a/rpc/ub/ub.c
+++ b/rpc/ub/ub.c
@@ -64,7 +64,8 @@ enum { /* XXX TODO: Revise these values. */
 static int args_format_error(void)
 {
 	fprintf(stderr, "**ERROR** Invalid format of the string with options.\n"
-		"  Expected: nr_conns=NUM,nr_slots=NUM,msg_len=NUM\n");
+		"  Expected:"
+		" nr_conns=NUM,nr_slots=NUM,nr_msgs=NUM,msg_len=NUM\n");
 	return -EINVAL;
 }
 
-- 
1.8.3.2

