From 2c2d4b2e4a6590b0a81fbb11211a2e45d227067a Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Sun, 24 Mar 2013 16:06:16 +0200
Subject: [PATCH 56/63] rpc-ub: implement token-wise parsing of options

---
 rpc/ub/ub.c | 49 ++++++++++++++++++++++++++++++++++++++++---------
 1 file changed, 40 insertions(+), 9 deletions(-)

diff --git a/rpc/ub/ub.c b/rpc/ub/ub.c
index c6e42cc..6fe7696 100644
--- a/rpc/ub/ub.c
+++ b/rpc/ub/ub.c
@@ -42,7 +42,7 @@ struct args {
 	unsigned int a_nr_conns;
 	unsigned int a_nr_slots;
 	unsigned int a_nr_msgs;
-	unsigned int a_msg_len; /* XXX USEME */
+	unsigned int a_msg_len;
 };
 
 /** Assigns default values to the arguments. */
@@ -91,14 +91,45 @@ static int args_check_limits(const struct args *args)
 	return -EINVAL;
 }
 
+struct match {
+	const char   *m_pattern;
+	unsigned int *m_dest;
+};
+
+static bool token_matches(const char *token, const struct match *tbl)
+{
+	for (; tbl->m_pattern != NULL; ++tbl) {
+		int rc = sscanf(token, tbl->m_pattern, tbl->m_dest);
+		if (rc == 1)
+			return true;
+	}
+	return false;
+}
+
 static int args_parse(const char *src, struct args *dest)
 {
 	if (src == NULL || *src == 0)
 		return 0;
-	int rc = sscanf(src, "nr_conns=%u,nr_slots=%u,nr_msgs=%u,msg_len=%u",
-			&dest->a_nr_conns, &dest->a_nr_slots, &dest->a_nr_msgs,
-			&dest->a_msg_len);
-	return rc == 4 ? args_check_limits(dest) : args_format_error();
+
+	char *s;
+	char *token;
+	struct match match_tbl[] = {
+		{ "nr_conns=%u", &dest->a_nr_conns },
+		{ "nr_slots=%u", &dest->a_nr_slots },
+		{ "nr_msgs=%u",  &dest->a_nr_msgs },
+		{ "msg_len=%u",  &dest->a_msg_len },
+		{ NULL, NULL }
+	};
+
+	s = strdupa(src);
+	if (s == NULL)
+		return -ENOMEM;
+
+	while ((token = strsep(&s, ",")) != NULL) {
+		if (!token_matches(token, match_tbl))
+			return args_format_error();
+	}
+	return args_check_limits(dest);
 }
 
 static struct args g_args;
@@ -206,8 +237,8 @@ static int _start(const char *opts)
 {
 	int  i;
 	int  rc;
-	char db[40] = "";
-	char ep[40] = "";
+	char db[40];
+	char ep[40];
 
 	args_init(&g_args);
 	rc = args_parse(opts, &g_args) ?: m0_net_xprt_init(g_xprt);
@@ -225,8 +256,8 @@ static int _start(const char *opts)
 	}
 
 	for (i = 0; i < g_args.a_nr_conns; ++i) {
-		snprintf(db, ARRAY_SIZE(db), "rpc-ub.client-db_%d", i);
-		snprintf(ep, ARRAY_SIZE(ep), CLIENT_ENDPOINT_FMT,
+		snprintf(db, sizeof db, "rpc-ub.client-db_%d", i);
+		snprintf(ep, sizeof ep, CLIENT_ENDPOINT_FMT,
 			 2 + i); /* 1 is server EP, so we start from 2 */
 		_client_start(&g_clients[i], CLIENT_COB_DOM_ID + i, db, ep);
 	}
-- 
1.8.3.2

