From 1a240379544a0bf6bedf7b430b762aa6f2848e69 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Sun, 24 Mar 2013 22:54:29 +0200
Subject: [PATCH 57/63] ub: slightly improve help messages

---
 lib/ub.c        |  4 ++--
 utils/ub_main.c | 12 ++++++------
 2 files changed, 8 insertions(+), 8 deletions(-)

diff --git a/lib/ub.c b/lib/ub.c
index 7da0510..74e6f19 100644
--- a/lib/ub.c
+++ b/lib/ub.c
@@ -191,8 +191,8 @@ M0_INTERNAL int m0_ub_run(uint32_t rounds, const char *opts)
 	}
 end:
 	if (rc != 0)
-		fprintf(stderr, "*ERROR* Benchmark initialisation failed:"
-			" set=`%s' rc=%d\n", set->us_name, rc);
+		fprintf(stderr, "Benchmark initialisation failed:"
+			" `%s', rc=%d\n", set->us_name, rc);
 	return rc;
 }
 
diff --git a/utils/ub_main.c b/utils/ub_main.c
index aee1731..2a884aa 100644
--- a/utils/ub_main.c
+++ b/utils/ub_main.c
@@ -67,20 +67,20 @@ static int ub_args_parse(int argc, char *argv[], struct ub_args *out)
 
 	return M0_GETOPTS("ub", argc, argv,
 		  M0_HELPARG('h'),
-		  M0_NUMBERARG('r', "Number of rounds UB has to be run",
-			       LAMBDA(void, (int64_t rounds) {
-					       out->ua_rounds = rounds;
-				       })),
 		  M0_VOIDARG('l', "List available benchmarks and exit",
 			     LAMBDA(void, (void) {
 					     out->ua_ub_list = true;
 				     })),
+		  M0_NUMBERARG('r', "Number of rounds a benchmark has to run",
+			       LAMBDA(void, (int64_t rounds) {
+					       out->ua_rounds = rounds;
+				       })),
 		  M0_STRINGARG('t', "Benchmark to run",
 			       LAMBDA(void, (const char *str) {
 					       out->ua_name = strdup(str);
 				       })),
-		  M0_STRINGARG('o', "Benchmark options:"
-			       " NAME[=VALUE][,NAME[=VALUE]]...",
+		  M0_STRINGARG('o', "Optional parameters (ignored by most"
+			       " benchmarks)",
 			       LAMBDA(void, (const char *str) {
 					       out->ua_opts = strdup(str);
 				       }))
-- 
1.8.3.2

