From ebfc678bae3ae7cdb90c807c843fb881d8e61bce Mon Sep 17 00:00:00 2001
From: Anatoliy Bilenko <abilenko@RG044C.xyus.xyratex.com>
Date: Mon, 1 Apr 2013 06:22:19 -0700
Subject: [PATCH 63/63] Fix: compile error

---
 rpc/rpclib.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/rpc/rpclib.h b/rpc/rpclib.h
index ce385a8..3635811 100644
--- a/rpc/rpclib.h
+++ b/rpc/rpclib.h
@@ -243,6 +243,7 @@ int m0_rpc_client_call(struct m0_fop *fop,
 		       m0_time_t deadline);
 
 
+struct m0_rpc_server_ctx; //fwd
 struct m0_rpc_machine *
 m0_rpc_server_ctx_get_rmachine(struct m0_rpc_server_ctx *sctx);
 
-- 
1.8.3.2

