From 163975f1503e270c117b2701188d8294c49df3e7 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Mon, 19 Nov 2012 17:36:15 +0200
Subject: [PATCH 06/58] disabled debug printf() in net-test console

---
 net/test/console.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/net/test/console.c b/net/test/console.c
index 761bb84..629077a 100644
--- a/net/test/console.c
+++ b/net/test/console.c
@@ -25,10 +25,10 @@
 
 /* @todo debug only, remove it */
 #ifndef __KERNEL__
-#define LOGD(format, ...) printf(format, ##__VA_ARGS__)
 /*
-#define LOGD(format, ...) do {} while (0)
+#define LOGD(format, ...) printf(format, ##__VA_ARGS__)
 */
+#define LOGD(format, ...) do {} while (0)
 #else
 #define LOGD(format, ...) do {} while (0)
 #endif
-- 
1.8.3.2

