From a8133181a90da55ce50067018cdc1810951ffb2e Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Mon, 19 Nov 2012 18:48:37 +0200
Subject: [PATCH 11/58] added space before ] in bash script

---
 net/test/st/run-1x1.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/test/st/run-1x1.sh b/net/test/st/run-1x1.sh
index ba535ce..5eda5dc 100644
--- a/net/test/st/run-1x1.sh
+++ b/net/test/st/run-1x1.sh
@@ -20,7 +20,7 @@ source "$ST_COMMON"
 
 eval_kill_pid() {
 	eval pid=\${$1-xxx}
-	if [ "$pid" != "xxx" -a -f /proc/$pid/exe]; then
+	if [ "$pid" != "xxx" -a -f /proc/$pid/exe ]; then
 		echo killing $pid
 		KILL_PID+="$pid "
 	fi
-- 
1.8.3.2

