From 4a8f237f7392ddb8257e16fb8a8f9972827e8f0d Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Mon, 26 Nov 2012 03:53:18 +0200
Subject: [PATCH 14/58] LogD 1485: removed unused #if 0 ... #endif block

---
 net/test/network.c | 10 ----------
 1 file changed, 10 deletions(-)

diff --git a/net/test/network.c b/net/test/network.c
index e9bf038..33b6b5a 100644
--- a/net/test/network.c
+++ b/net/test/network.c
@@ -57,16 +57,6 @@
 #endif
 #endif
 
-#if 0
-/* see buf_desc_decode() and c2_net_test_network_bd_encode() */
-C2_BASSERT(sizeof(unsigned)	 == sizeof(uint32_t));
-C2_BASSERT(sizeof(unsigned long) == sizeof(c2_bcount_t));
-
-enum {
-	C2_NET_TEST_STRLEN_NBD_HEADER = 32,
-};
-#endif
-
 enum {
 	C2_NET_TEST_NETWORK_BD_MAGIC = 1,	/** @todo fix */
 	C2_NET_TEST_NETWORK_BDS_MAGIC = 2,
-- 
1.8.3.2

