From 66b0fe029ba60028e14e62e5ef8fa1ef931fa172 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Mon, 26 Nov 2012 04:14:47 +0200
Subject: [PATCH 15/58] LogD 1486: count down using "i" directly instead of
 adding "j"

---
 net/test/network.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/net/test/network.c b/net/test/network.c
index 33b6b5a..2166acd 100644
--- a/net/test/network.c
+++ b/net/test/network.c
@@ -439,7 +439,6 @@ int c2_net_test_network_ep_add_slist(struct c2_net_test_network_ctx *ctx,
 {
 	int    rc = 0;
 	size_t i;
-	size_t j;
 
 	C2_PRE(c2_net_test_network_ctx_invariant(ctx));
 	C2_PRE(c2_net_test_slist_invariant(eps));
@@ -451,7 +450,7 @@ int c2_net_test_network_ep_add_slist(struct c2_net_test_network_ctx *ctx,
 	}
 	if (rc < 0) {
 		/** c2_net_end_point_put() for last i endpoints */
-		for (j = 0; j < i; ++j)
+		for (; i != 0; --i)
 			c2_net_end_point_put(ctx->ntc_ep[--ctx->ntc_ep_nr]);
 	}
 	C2_POST(c2_net_test_network_ctx_invariant(ctx));
-- 
1.8.3.2

