From cedbb1d82dab506a2684595ea2c4d81bebbe598e Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Mon, 26 Nov 2012 09:56:59 +0200
Subject: [PATCH 32/58] LogD 1502: lctl network up after modprobe in ST

---
 net/test/st/run-1x1.sh | 1 +
 1 file changed, 1 insertion(+)

diff --git a/net/test/st/run-1x1.sh b/net/test/st/run-1x1.sh
index 5eda5dc..c730c0d 100644
--- a/net/test/st/run-1x1.sh
+++ b/net/test/st/run-1x1.sh
@@ -40,6 +40,7 @@ unload_all() {
 trap unload_all EXIT
 
 modprobe_lnet
+lctl network up > /dev/null
 modload_galois
 modload || exit $?
 
-- 
1.8.3.2

