From 20eb0af4b39ed25f0bfd4a64156b37a5cb83118e Mon Sep 17 00:00:00 2001
From: Mandar Sawant <mandar_sawant@xyratex.com>
Date: Mon, 23 Dec 2013 16:44:22 +0530
Subject: [PATCH 4/7] be.conv.conf, - Revert log level to M0_DEBUG.

---
 be/tx.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/be/tx.c b/be/tx.c
index 1bdc765..2e01a5d 100644
--- a/be/tx.c
+++ b/be/tx.c
@@ -228,9 +228,9 @@ M0_INTERNAL void m0_be_tx_open(struct m0_be_tx *tx)
 	rc = rc ?: m0_be_reg_area_init(&tx->t_reg_area, &tx->t_prepared, true);
 
 	if (rc == -EINVAL) {
-		M0_LOG(M0_FATAL, "tx = %p: tx credit is invalid", tx);
+		M0_LOG(M0_DEBUG, "tx = %p: tx credit is invalid", tx);
 	} else if (rc == -ENOMEM) {
-		M0_LOG(M0_FATAL, "tx = %p: there is not enough memory "
+		M0_LOG(M0_DEBUG, "tx = %p: there is not enough memory "
 		       "to allocate using prepared credit "BETXCR_F,
 		       tx, BETXCR_P(&tx->t_prepared));
 	}
-- 
1.8.3.2

