From 9408d67d121cda4287294dfa6237599a973f37d7 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Thu, 5 Dec 2013 02:19:17 +0200
Subject: [PATCH 2/2] db: seg_map_tryload() shouldn't lock seg_map lock

---
 db/db.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/db/db.c b/db/db.c
index 6fd62e5..a7d6a36 100644
--- a/db/db.c
+++ b/db/db.c
@@ -187,9 +187,7 @@ static void seg_map_tryload(void)
 	M0_ASSERT_INFO(seg_map_size == 0,
 		       "This function should be called before the first "
 		       "m0_dbenv_init() call. seg_map_size = %d", seg_map_size);
-	seg_map_lock();
 	m0_be_ut_state_tryload(seg_map_file, &seg_map_tryload_item);
-	seg_map_unlock();
 }
 
 static void dbenv_seg_init(struct m0_be_ut_seg *ut_seg,
-- 
1.8.3.2

