From 36fe368bb93f5f046d9b44926a18cb0f897f2458 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Wed, 21 Mar 2012 18:15:35 +0200
Subject: [PATCH 4/9] use C2_ALLOC_PTR() instead of c2_alloc() if possible

---
 rpc/it/ping_fom.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/rpc/it/ping_fom.c b/rpc/it/ping_fom.c
index d35b1fb..72ca09d 100644
--- a/rpc/it/ping_fom.c
+++ b/rpc/it/ping_fom.c
@@ -95,7 +95,7 @@ static int ping_fop_fom_create(struct c2_fop *fop, struct c2_fom **m)
         C2_PRE(fop != NULL);
         C2_PRE(m != NULL);
 
-        fom_obj = c2_alloc(sizeof *fom_obj);
+	C2_ALLOC_PTR(fom_obj);
         if (fom_obj == NULL)
                 return -ENOMEM;
 	fom_obj->fp_fop = fop;
-- 
1.8.3.2

