From 5ba93c70f3c0d4846630075f64e07863508dd663 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Wed, 21 Mar 2012 18:22:41 +0200
Subject: [PATCH 5/9] moved c2_fom_type_ops_default after c2_fom_create_dummy()

---
 fop/fop_format.c | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/fop/fop_format.c b/fop/fop_format.c
index dfc42ed..c623218 100644
--- a/fop/fop_format.c
+++ b/fop/fop_format.c
@@ -39,17 +39,17 @@ static struct c2_fop_decorator *decorators[C2_FOP_DECORATOR_MAX];
 static size_t decorators_nr = 0;
 static bool decoration_used = false;
 
-/** Default fom type ops */
-const struct c2_fom_type_ops c2_fom_type_ops_default = {
-	.fto_create = c2_fom_create_dummy
-};
-
 static int c2_fom_create_dummy(struct c2_fop *fop, struct c2_fom **fom)
 {
 	*fom = NULL;
 	return 0;
 }
 
+/** Default fom type ops */
+const struct c2_fom_type_ops c2_fom_type_ops_default = {
+	.fto_create = c2_fom_create_dummy
+};
+
 void c2_fop_field_type_unprepare(struct c2_fop_field_type *ftype)
 {
 	int    i;
-- 
1.8.3.2

