From ceb1aae232a97c6f29a896b8a8203338d85a7e4e Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Thu, 14 Nov 2013 23:40:06 +0200
Subject: [PATCH 04/26] be/tx: comment about memory allocation added

---
 be/tx.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/be/tx.c b/be/tx.c
index 7455861..c88f5ae 100644
--- a/be/tx.c
+++ b/be/tx.c
@@ -225,6 +225,7 @@ M0_INTERNAL void m0_be_tx_open(struct m0_be_tx *tx)
 
 	rc = m0_be_tx_credit_eq(&tx->t_prepared,
 				&m0_be_tx_credit_invalid) ? -EINVAL : 0;
+	/** XXX @todo allocate memory after log reservation */
 	rc = rc ?: m0_be_reg_area_init(&tx->t_reg_area, &tx->t_prepared, true);
 
 	be_tx_state_move(tx, rc == 0 ? M0_BTS_OPENING : M0_BTS_FAILED, rc);
-- 
1.8.3.2

