From 58293a8345718fcc41949d7c838f748793fdb1dd Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Thu, 14 Nov 2013 23:40:42 +0200
Subject: [PATCH 05/26] be/tx_credit: M0_BE_TX_CREDIT_PTR() added

---
 be/tx_credit.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/be/tx_credit.h b/be/tx_credit.h
index 46ec1cd..dfede40 100644
--- a/be/tx_credit.h
+++ b/be/tx_credit.h
@@ -75,6 +75,7 @@ extern const struct m0_be_tx_credit m0_be_tx_credit_invalid;
 	(struct m0_be_tx_credit){ .tc_reg_nr = (nr), .tc_reg_size = (size) }
 
 #define M0_BE_TX_CREDIT_TYPE(type) M0_BE_TX_CREDIT(1, sizeof(type))
+#define M0_BE_TX_CREDIT_PTR(ptr) M0_BE_TX_CREDIT(1, sizeof(*ptr))
 
 /** Format for the printf() family functions. @see BETXCR_P */
 #define BETXCR_F "(%lu, %lu)"
-- 
1.8.3.2

