From cb05bf9cf07c3ab02e277e9f04f38397e5f87a06 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Sun, 17 Nov 2013 22:02:20 +0200
Subject: [PATCH 11/26] be/seg_dict: m0_be_allocator_credit can use btree
 segment

---
 be/seg_dict.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/be/seg_dict.c b/be/seg_dict.c
index 0eb07df..5e60d1d 100644
--- a/be/seg_dict.c
+++ b/be/seg_dict.c
@@ -158,10 +158,16 @@ M0_INTERNAL void m0_be_seg_dict_init(struct m0_be_seg *seg)
 M0_INTERNAL void m0_be_seg_dict_create_credit(const struct m0_be_seg *seg,
 					      struct m0_be_tx_credit *accum)
 {
-	const struct m0_be_btree *tree = dict_get_const(seg);
+	struct m0_be_btree tree;
 
 	M0_PRE(m0_be_seg__invariant(seg));
-	m0_be_btree_create_credit(tree, 1, accum);
+
+	/**
+	 * XXX undefined behavior here: workaround for
+	 * m0_be_btree_create_credit()
+	 */
+	tree.bb_seg = (struct m0_be_seg *)seg;
+	m0_be_btree_create_credit(&tree, 1, accum);
 }
 
 M0_INTERNAL void m0_be_seg_dict_create(struct m0_be_seg *seg,
-- 
1.8.3.2

