From 22679cb845394596da4dff951cce06ba68bb6ed8 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Sun, 17 Nov 2013 22:26:40 +0200
Subject: [PATCH 12/26] be/allocator: destroy() credit fixed

---
 be/alloc.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/be/alloc.c b/be/alloc.c
index 28e2e9a..7c21f08 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -768,6 +768,7 @@ M0_INTERNAL void m0_be_allocator_credit(struct m0_be_allocator *a,
 					struct m0_be_tx_credit *accum)
 {
 	struct m0_be_tx_credit cred_list_create = {};
+	struct m0_be_tx_credit cred_list_destroy = {};
 	struct m0_be_tx_credit capture_around_credit = {};
 	struct m0_be_tx_credit chunk_add_after_credit = {};
 	struct m0_be_tx_credit chunk_del_fini_credit = {};
@@ -782,7 +783,9 @@ M0_INTERNAL void m0_be_allocator_credit(struct m0_be_allocator *a,
 	shift = max_check(shift, (unsigned) M0_BE_ALLOC_SHIFT_MIN);
 	mem_zero_credit = M0_BE_TX_CREDIT(1, size * 2);
 
-	m0_be_list_credit(NULL, M0_BLO_CREATE, 1, &cred_list_create);
+	m0_be_list_credit(NULL, M0_BLO_CREATE,	1, &cred_list_create);
+	m0_be_list_credit(NULL, M0_BLO_DESTROY, 1, &cred_list_destroy);
+
 	m0_be_tx_credit_add(&capture_around_credit, &header_credit);
 	m0_be_tx_credit_mac(&capture_around_credit, &chunk_credit, 3);
 
@@ -807,8 +810,7 @@ M0_INTERNAL void m0_be_allocator_credit(struct m0_be_allocator *a,
 			break;
 		case M0_BAO_DESTROY:
 			m0_be_tx_credit_add(accum, &chunk_del_fini_credit);
-			/* tlist_fini x2 */
-			m0_be_tx_credit_mac(accum, &header_credit, 2);
+			m0_be_tx_credit_mac(accum, &cred_list_destroy, 2);
 			break;
 		case M0_BAO_ALLOC_ALIGNED:
 			m0_be_tx_credit_add(accum, &chunk_del_fini_credit);
-- 
1.8.3.2

