From fecf7c8b313c411a87eb776718db643e8c565fb1 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Thu, 23 Feb 2012 14:13:08 +0200
Subject: [PATCH 013/370] confc.c: explain what path_status() does

---
 conf/confc.c | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/conf/confc.c b/conf/confc.c
index 327c573..e20111c 100644
--- a/conf/confc.c
+++ b/conf/confc.c
@@ -92,6 +92,28 @@ request_fill(struct c2_confop_fetch *req, const struct c2_conf_obj *path_org,
 static int
 path_status(struct c2_confc_ctx *ctx, enum c2_conf_status *status)
 {
+	/*
+	 * XXX[remove this comment after DLDINSP]
+	 *
+	 * Refer to confc state diagram in the HLD.
+	 *
+	 * path_status() follows ctx->fc_path and checks the statuses
+	 * of met objects.  Upon success, the function sets *status to:
+	 *
+	 * - C2_CS_READY:   path target is reachable
+	 *
+	 * - C2_CS_MISSING: At least one of the intermediate objects
+	 *                  or the target itself is C2_CS_MISSING.
+	 *                  path_status() changes statuses of such
+	 *                  objects to C2_CS_LOADING and fills
+	 *                  ctx->fc_req.
+	 *
+	 * - C2_CS_LOADING: Neither path target nor missing objects
+	 *                  can be reached because of C2_CS_LOADING
+	 *                  object(s) blocking the path.
+	 *                  path_status() registers ctx->fc_clink with
+	 *                  the channel(s) of loading object(s).
+	 */
 	struct c2_conf_obj *obj = ctx->fc_path.p_origin ?: confc_cache->cc_root;
 	struct c2_conf_pathcomp *comp = ctx->fc_path.p_comps;
 	size_t len = ctx->fc_path.p_len; /* number of components to go */
-- 
1.8.3.2

