From 9a4ae2af5099a4db14b73f0aeb7497a4b2fd0d02 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Tue, 28 Feb 2012 11:38:48 +0200
Subject: [PATCH 022/370] add the skeleton of functional specification

---
 conf/confc.h | 50 ++++++++++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 48 insertions(+), 2 deletions(-)

diff --git a/conf/confc.h b/conf/confc.h
index 8e269cd..69e3158 100644
--- a/conf/confc.h
+++ b/conf/confc.h
@@ -34,7 +34,45 @@
  * Confc obtains configuration data from network-accessible
  * configuration server (confd) and caches this data in memory.
  *
- * @section conf-confc-setup Initialization and termination
+ * - @ref confc-dld-fspec-data
+ * - @ref confc-dld-fspec-sub
+ *   - @ref confc-dld-fspec-sub-tors
+ *   - @ref confc-dld-fspec-sub-accessors
+ *   - @ref confc-dld-fspec-sub-modifiers
+ * - @ref confc-dld-fspec-recipes
+ *   - @ref confc-dld-fspec-recipes-setup
+ *   - @ref confc-dld-fspec-recipes-open
+ *   - XXX
+ * - @ref confc_dfspec "Detailed Functional Specification"
+ *
+ * <hr> <!------------------------------------------------------------->
+ * @section confc-dld-fspec-data  Data Structures
+ *
+ * XXX
+ *
+ * <hr> <!------------------------------------------------------------->
+ * @section confc-dld-fspec-sub  Subroutines
+ *
+ * XXX
+ *
+ * @subsection confc-dld-fspec-sub-tors  Constructors and Destructors
+ *
+ * XXX
+ *
+ * @subsection confc-dld-fspec-sub-accessors  Accessors and Invariants
+ *
+ * XXX
+ *
+ * @subsection confc-dld-fspec-sub-modifiers  Operational Interfaces
+ *
+ * XXX
+ *
+ * <hr> <!------------------------------------------------------------->
+ * @section confc-dld-fspec-recipes  Recipes
+ *
+ * XXX
+ *
+ * @subsection confc-dld-fspec-recipes-setup Initialization and termination
  *
  * Prior to accessing configuration, the application (aka
  * configuration consumer) should initialize configuration client by
@@ -63,7 +101,7 @@
  * }
  * @endcode
  *
- * @section conf-confc-opening Accessing configuration objects
+ * @subsection confc-dld-fspec-recipes-open Accessing configuration objects
  *
  * The application gets access to configuration data by "opening"
  * configuration objects. c2_confc_open() is used to obtain a pointer
@@ -136,8 +174,15 @@
  * 	return c2_confc_open_sync(NULL, path, ARRAY_SIZE(path));
  * }
  * @endcode
+ *
+ * @see confc_dfspec
  */
 
+/**
+ * @defgroup confc_dfspec  Configuration Client (confc)
+ *
+ * @{
+ */
 struct c2_confc_dir;
 
 /**
@@ -289,4 +334,5 @@ void * c2_confc_dirnext(struct c2_confc_dir *dir);
 
 void c2_confc_dirclose(struct c2_confc_dir *dir);
 
+/** @} confc_dfspec */
 #endif /* __COLIBRI_CONF_CONFC_H__ */
-- 
1.8.3.2

