From f4395d75d72a15a9580018848cf42bbc376825d5 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Tue, 28 Feb 2012 12:41:58 +0200
Subject: [PATCH 023/370] fspec-data: c2_confc_ctx described

---
 conf/confc.h | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/conf/confc.h b/conf/confc.h
index 69e3158..ab42728 100644
--- a/conf/confc.h
+++ b/conf/confc.h
@@ -48,7 +48,14 @@
  * <hr> <!------------------------------------------------------------->
  * @section confc-dld-fspec-data  Data Structures
  *
- * XXX
+ * c2_confc_ctx --- configuration retrieval context.  This object is
+ * passed as an argument to asynchronous "open" functions
+ * (c2_confc_open(), c2_confc_diropen()).  Only two of its members are
+ * supposed to be used by configuration consumers:
+ * - @c fc_complete: this channel is signaled when the retrieval
+ *   operation completes;
+ * - @c fc_result: result of the retrieval operation --- pointer to
+ *   the requested configuration object or directory.
  *
  * <hr> <!------------------------------------------------------------->
  * @section confc-dld-fspec-sub  Subroutines
-- 
1.8.3.2

