From bec8f55664e253a426dfddc585d90a90d95d7232 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Wed, 7 Mar 2012 10:26:15 +0200
Subject: [PATCH 041/370] xdata.ff: fix type name

---
 conf/xdata.ff | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/conf/xdata.ff b/conf/xdata.ff
index 6f5e748..7b478cf 100644
--- a/conf/xdata.ff
+++ b/conf/xdata.ff
@@ -45,8 +45,8 @@ record {
 		c2_conf_buf o_objkey
 	} ff_origin; /* Configuration object the path originates from. */
 	sequence {
-		u32               c_nr;
-		c2_confx_pathcomp c_data
+		u32              c_nr;
+		c2_conf_pathcomp c_data
 	} ff_comps   /* Path components. */
 } c2_conf_fetch;
 
@@ -72,7 +72,7 @@ record {
 	/* Parent filesystem. */
 	c2_conf_buf fs_filesystem;
 
-	/* Service type. See c2_cfg_service_type for values. */
+	/* Service type.  See c2_cfg_service_type for values. */
 	u32         fs_type;
 
 	/* End-points from which this service is reachable. */
@@ -141,7 +141,7 @@ record {
 record {
 	c2_conf_buf fo_key;
 	union {
-		u8                  v_type;
+		u8                  v_type; /* see c2_conf_objtype for values */
 		c2_confx_profile    v_profile;
 		c2_confx_filesystem v_filesystem;
 		c2_confx_service    v_service;
-- 
1.8.3.2

