From 564ed7fa3e30a63c82f46bee9eef7a97222ffe70 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Sun, 11 Mar 2012 18:05:44 +0200
Subject: [PATCH 050/370] DLD: explain threading and concurrency

---
 conf/confc.c | 25 ++++++++++++++++++++++++-
 conf/confc.h |  2 +-
 2 files changed, 25 insertions(+), 2 deletions(-)

diff --git a/conf/confc.c b/conf/confc.c
index 3cc4ee7..3155372 100644
--- a/conf/confc.c
+++ b/conf/confc.c
@@ -41,6 +41,8 @@
  *   - @ref confc-dld-lspec-comps
  *   - @ref confc-dld-lspec-state
  *   - @ref confc-dld-lspec-thread
+ *   - @ref confc-dld-lspec-numa
+ * - @ref confc-dld-conformance
  * - XXX
  *
  * <hr> <!------------------------------------------------------------->
@@ -210,8 +212,29 @@
  *
  * @subsection confc-dld-lspec-thread Threading and Concurrency Model
  *
- * XXX
+ * When a confc instance is started (c2_confc_init(), it gets
+ * associated with a state machine group (c2_sm_group).  A confc state
+ * machine undergoes a transition when the state machine group is
+ * locked.
+ *
+ * A user managing the state machine group is responsible for making
+ * sure c2_sm_asts_run() is called when c2_sm_group::s_clink is
+ * signaled.  See @ref sm (search for `"ast" thread'.)
+ *
+ * A confc instance locks c2_conf_cache::cc_lock when performing any
+ * of the following operations:
+ * - updating c2_conf_cache::cc_registry;
+ * - updating c2_conf_obj::co_status;
+ * - updating c2_conf_obj::co_nrefs.
+ *
+ * @subsection confc-dld-lspec-numa NUMA Optimizations
  *
+ * XXX Have no idea what to say here. Your suggestions are welcome.
+ *
+ * <hr> <!------------------------------------------------------------->
+ * @section confc-dld-conformance Conformance
+ *
+ * XXX
  */
 
 static struct {
diff --git a/conf/confc.h b/conf/confc.h
index 7a93aaa..e2990ba 100644
--- a/conf/confc.h
+++ b/conf/confc.h
@@ -93,7 +93,7 @@
  * A confc instance is associated with a state machine group
  * (c2_sm_group). A user managing this group is responsible for making
  * sure c2_sm_asts_run() is called when the group's channel is
- * signalled; "AST" section of @ref sm has more details on this topic.
+ * signaled; "AST" section of @ref sm has more details on this topic.
  *
  * Initiated confc must be eventually terminated with c2_confc_fini().
  *
-- 
1.8.3.2

