From 3d4d8eeda7ca383bf569d7873c0aea002cb37525 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Sat, 17 Mar 2012 01:33:51 +0200
Subject: [PATCH 071/370] improve wording

---
 conf/_confc_offline.h | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/conf/_confc_offline.h b/conf/_confc_offline.h
index 165d8a5..7138949 100644
--- a/conf/_confc_offline.h
+++ b/conf/_confc_offline.h
@@ -44,10 +44,10 @@
  *
  * An object is represented by a sequence of attributes.  The first
  * attribute is object's identity.  Set of remaining attributes
- * depends on object's type.  Attributes are ordered in accordance
- * with fields of confx_* structure that corresponds to object's type.
+ * depends on object's type.  Attributes are ordered similarly to the
+ * fields of confx_* structure that corresponds to object's type.
  *
- * For example, this configuration string
+ * E.g., this configuration string
  *
 @verbatim
 (prof:"test","c2t1fs") (fs:"c2t1fs",[11,22][50,60,70],[])
-- 
1.8.3.2

