From f6339f13ac504475666cb6f9b199f71e1ca800cf Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Mon, 19 Mar 2012 09:57:54 +0200
Subject: [PATCH 077/370] downgrade conf/rpc.c

---
 conf/rpc.c | 99 ++++++--------------------------------------------------------
 1 file changed, 9 insertions(+), 90 deletions(-)

diff --git a/conf/rpc.c b/conf/rpc.c
index 962d8dd..a51396f 100644
--- a/conf/rpc.c
+++ b/conf/rpc.c
@@ -22,99 +22,18 @@
 #  include <config.h>
 #endif
 #include "conf/rpc.h"
-#include "conf/xdata.h"  /* c2_conf_fetch_xc */
-#include "xcode/xcode.h"
-#include "rpc/rpc2.h"    /* C2_RPC_ITEM_TYPE_DEF */
+#include "xcode/bufvec_xcode.h"  /* c2_xcode_fop_size_get */
+#include "fop/fop_base.h"        /* c2_fop_type_ops */
+#include "fop/fop_format.h"      /* C2_FOP_TYPE_DECLARE */
+#include "rpc/rpc_base.h"        /* C2_RPC_ITEM_TYPE_{REQUEST,REPLY} */
+#include "rpc/rpc_opcodes.h"
 
 static const struct c2_fop_type_ops fopt_ops = {
-	/* XXX c2_xcode_fop_size_get() may be inappropriate here */
 	.fto_size_get = c2_xcode_fop_size_get
 };
 
-static struct c2_rpc_item_type_ops fetch_ops = {
-	/* return c2_xcode_length() */
-	.rito_item_size = XXX,
-	.rito_encode    = request_encode,
-	/* XXX TODO should be set for confd */
-	.rito_decode    = NULL
-};
-
-/*
- * Defines `struct c2_fop_type c2_conf_fetch_fopt'.
- *
- * XXX Dependency: struct c2_fop_type_format c2_conf_fetch_tfmt
- */
-C2_FOP_TYPE_DECLARE_OPS(c2_conf_fetch, "c2_conf_fetch", &fopt_ops,
-			C2_CONF_FETCH_OPCODE, C2_RPC_ITEM_TYPE_REQUEST,
-			&fetch_ops);
-
-C2_RPC_ITEM_TYPE_DEF(fetch_itype, C2_CONF_FETCH_OPCODE,
-		     C2_RPC_ITEM_TYPE_REQUEST, &fetch_ops);
-
-static struct c2_rpc_item_type_ops fetch_resp_ops = {
-	/* return c2_xcode_length() */
-	.rito_item_size = XXX,
-	/* XXX TODO should be set for confd */
-	.rito_encode    = NULL,
-	.rito_decode    = response_decode
-};
-
-C2_FOP_TYPE_DECLARE_OPS(c2_conf_fetch_resp, "c2_conf_fetch_resp", &ftype_ops,
-			C2_CONF_FETCH_RESP_OPCODE, C2_RPC_ITEM_TYPE_REPLY,
-			&fetch_resp_ops);
-
-C2_RPC_ITEM_TYPE_DEF(fetch_resp_itype, C2_CONF_FETCH_RESP_OPCODE,
-		     C2_RPC_ITEM_TYPE_REPLY, &fetch_resp_ops);
-
-/**
- * Encodes configuration request into a network buffer provided by the
- * rpc layer.
- */
-static int
-request_encode(struct c2_rpc_item_type *item_type __attribute__((unused)),
-	       struct c2_rpc_item *item, struct c2_bufvec_cursor *cur)
-{
-	/*
-	 * struct c2_xcode_ctx xtx;
-	 * struct c2_xcode_obj xobj = {
-	 *     .xo_type = c2_conf_fetch_xc,
-	 *     .xo_ptr  = c2_fop_data(c2_rpc_item_to_fop(item))
-	 * };
-	 *
-	 * C2_PRE(item->ri_type == &fetch_itype);
-	 *
-	 * Ensure by bob-checking that xobj.xo_ptr is of c2_conf_fetch type.
-	 *
-	 * Validate the request (req):
-	 *   - req->ff_origin.o_objtype < C2_CO_NR;
-	 *   - req->ff_origin belongs the cache;
-	 *   - equi(req->ff_comps.c_nr == 0, req->ff_comps.c_data == NULL);
-	 *   - C2_ASSERT(c2_conf_path_validate
-	 *               (req->ff_origin.o_objtype, req->ff_origin.o_objkey,
-	 *                req->ff_comps.c_data, req->ff_comps.c_nr, NULL) == 0);
-	 *
-	 * // See c2_fop_item_type_default_encode().
-	 * Encode item->ri_type->rit_opcode.
-	 *
-	 * c2_xcode_ctx_init(&xtx, &xobj);
-	 * xtx.xtx_buf = *cur;
-	 * return c2_xcode_encode(&xtx);
-	 */
-	XXX;
-}
+C2_FOP_TYPE_DECLARE(c2_conf_fetch, "c2_conf_fetch", &fopt_ops,
+		    C2_CONF_FETCH_OPCODE, C2_RPC_ITEM_TYPE_REQUEST);
 
-static int response_decode(struct c2_rpc_item_type *item_type,
-			   struct c2_rpc_item **item,
-			   struct c2_bufvec_cursor *cur)
-{
-	/*
-	 * C2_PRE(item_type == &fetch_resp_itype);
-	 *
-	 * - Allocate and initialize c2_fop.
-	 * - c2_xcode_decode() serialized configuration data, pointed
-	 *   to by `cur' ==> dynamically allocated c2_conf_fetch_resp.
-	 * - Set c2_fop_data() to the address of c2_conf_fetch_resp.
-	 * - Set *item.
-	 */
-	XXX;
-}
+C2_FOP_TYPE_DECLARE(c2_conf_fetch_resp, "c2_conf_fetch_resp", &fopt_ops,
+		    C2_CONF_FETCH_RESP_OPCODE, C2_RPC_ITEM_TYPE_REPLY);
-- 
1.8.3.2

