From 172cf6fd71d0f75eb9a7c009bdaa64e324b8900c Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Fri, 30 Mar 2012 00:38:18 +0300
Subject: [PATCH 104/370] finalize ctx->fc_extra clinks

---
 conf/confc.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/conf/confc.c b/conf/confc.c
index 503851b..b244b18 100644
--- a/conf/confc.c
+++ b/conf/confc.c
@@ -1162,17 +1162,27 @@ static int cache__preload(struct c2_conf_map *cache, const char *conf_str)
 void c2_confc_ctx_init(struct c2_confc_ctx *ctx, struct c2_confc *client)
 {
 	C2_PRE(confc_is_initialised(confc));
+
 	ctx->fc_client = confc;
 	c2_sm_init(&ctx->fc_mach, &confc_states_conf, S_INITIAL,
 		   confc->cc_group, XXX /* *c2_addb_ctx */);
 	c2_chan_init(&ctx->fc_complete);
 	c2_clink_init(&ctx->fc_clink, on_object_update);
+	ctx->fc_extra = NULL;
+	ctx->fc_nr_extra = 0;
+
 	C2_POST(ctx_invariant(ctx));
 }
 
 void c2_confc_ctx_fini(struct c2_confc_ctx *ctx)
 {
+	size_t i;
 	C2_PRE(ctx_invariant(ctx));
+
+	for (i = 0; i < ctx->fc_nr_extra; ++i)
+		c2_clink_fini(ctx->fc_extra + i);
+	ctx->fc_extra = NULL;
+	ctx->fc_nr_extra = 0;
 	c2_clink_fini(&ctx->fc_clink);
 	c2_chan_fini(&ctx->fc_complete);
 	c2_sm_fini(&ctx->fc_mach);
@@ -1186,7 +1196,8 @@ static bool ctx_invariant(struct c2_confc_ctx *ctx)
 		item->ri_type == &request_item_type &&
 		item->ri_ops != NULL &&
 		item->ri_ops->rio_replied == on_replied &&
-		c2_fop_data(ctx->fc_fop) == &ctx->fc_req;
+		c2_fop_data(ctx->fc_fop) == &ctx->fc_req &&
+		equi(ctx->fc_nr_extra == 0, ctx->fc_extra == NULL);
 }
 
 static enum c2_conf_status
-- 
1.8.3.2

