From ac8583beea4c9d62041104eb51edf3b909968721 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Tue, 3 Apr 2012 00:26:33 +0300
Subject: [PATCH 116/370] system tests described

---
 conf/confc.c | 50 +++++++++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 49 insertions(+), 1 deletion(-)

diff --git a/conf/confc.c b/conf/confc.c
index 85d4f5f..63f26e4 100644
--- a/conf/confc.c
+++ b/conf/confc.c
@@ -505,7 +505,55 @@
  * <hr> <!------------------------------------------------------------->
  * @section confc-dld-st System Tests
  *
- * XXX
+ * System testing procedure consists of three steps:
+ *
+ *   -# Generate configuration.
+ *   -# Run @ref confc-dld-fspec-recipes "use cases".
+ *   -# Compare actual results with expected.
+ *
+ * A special user space utility -- @b confg, configuration generator --
+ * performs `Generate' step.  It generates random configuration data,
+ * queries to be executed upon this data, and results that are
+ * expected to be obtained by running these queries.
+ *
+ * `Run' step is performed by a confc application that takes generated
+ * configuration and queries as input parameters and executes the
+ * queries, producing output.
+ *
+ * Actual output is compared with expected at `Compare' step.
+ *
+ * - Configuration data
+ *
+ * confg generates "configuration string" --- an ASCII string
+ * introduced in @ref confc-offline.
+ *
+ * As soon as confd is able to read configuration database and handle
+ * confc's requests, confg will need to be changed to generate
+ * configuration database.
+ *
+ * - Queries
+ *
+ * System tests cover three use cases (see @ref
+ * confc-dld-fspec-recipes):
+ *
+ *   -# Getting [configuration] data on a filesystem.
+ *   -# Getting data on a service given its type.
+ *   -# Getting list of devices used by specific service on specific node.
+ *
+ * Case 1 takes no input parameters (apart from profile name which
+ * does not count).  Case 2 takes service type as an input parameter.
+ * Case 3 takes service type and node identifier.
+ *
+ * A query is number of use case and values of input parameters.
+ *
+ * - Output
+ *
+ * When the confc application from `Run' step (see above) runs a
+ * [configuration] query, it produces a string with textual
+ * representation of results of this query.  Sequences in such
+ * representations (e.g., a list of filesystem parameters, a list of
+ * partitions) should be sorted, otherwise two semantically equivalent
+ * sequences may have differing representations ("[a,b]" != "[b,a]").
  *
  * <hr> <!------------------------------------------------------------->
  * @section confc-dld-O Analysis
-- 
1.8.3.2

