From e253501f047688da1a731d06585605c328df5ddc Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Wed, 18 Apr 2012 14:47:22 +0300
Subject: [PATCH 136/370] conf/path.h: fill `Subroutines' section

---
 conf/path.h | 29 +++++++++++++++++++----------
 1 file changed, 19 insertions(+), 10 deletions(-)

diff --git a/conf/path.h b/conf/path.h
index 129e768..88b394b 100644
--- a/conf/path.h
+++ b/conf/path.h
@@ -149,14 +149,16 @@ Legend:
  * <hr> <!------------------------------------------------------------>
  * @section conf-fspec-path-sub Subroutines
  *
- * - c2_conf_path_is_valid()
- * - c2_conf_eop()
- * - c2_conf_downlink()
- * - c2_conf_pathcomp_is_last()
- * - c2_conf_path_len()
- * - c2_conf_pathcomp_dup()
- *
- * XXX @todo describe
+ * - c2_conf_path_is_valid(): performs sanity check of a path.
+ * - c2_conf_downlink(): returns description of a downlink
+ *   (c2_conf_relation), given types of source and target objects.
+ * - c2_conf_eop(): checks whether given path component is an
+ *   end-of-path marker.
+ * - c2_conf_pathcomp_is_last(): returns true iff given path component
+ *   precedes end-of-path marker (cf. c2_conf_eop()).
+ * - c2_conf_path_len(): returns the number of path components,
+ *   excluding end-of-path marker.
+ * - c2_conf_pathcomp_dup(): duplicates an array of path components.
  *
  * <hr> <!------------------------------------------------------------>
  * @section conf-fspec-path-macros Convenience Macros
@@ -183,7 +185,7 @@ struct c2_conf_obj;
 struct c2_conf_pathcomp;
 
 /**
- * Checks validity of a path.
+ * Performs sanity check of a path.
  *
  * An empty path is valid.  Non-empty path is valid iff the following
  * properties hold:
@@ -193,6 +195,10 @@ struct c2_conf_pathcomp;
  *   component;
  * - a key is preceded by a one-to-many downlink.
  *
+ * @note  c2_conf_path_is_valid() does not verify keys of `path' (i.e.,
+ *        it does not check the existence of objects referred to by
+ *        keys).
+ *
  * @param origin      Path origin.
  * @param path        Path components.
  * @param[out] isdir  If the path is valid and isdir != NULL, then
@@ -266,9 +272,12 @@ const struct c2_conf_relation * c2_conf_downlink(uint32_t src, uint32_t tgt);
 bool c2_conf_eop(const struct c2_conf_pathcomp *pc);
 
 /**
- * Returns true iff this is the last component before C2_CONF_EOP.
+ * Returns true iff given path component precedes end-of-path marker
+ * (C2_CONF_EOP).
  *
  * @pre  !c2_conf_eop(pc)
+ *
+ * @see c2_conf_eop()
  */
 bool c2_conf_pathcomp_is_last(const struct c2_conf_pathcomp *pc);
 
-- 
1.8.3.2

