From 5468e9d0618ee9d84e0d7b62df53974d95e7351b Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Fri, 20 Apr 2012 18:22:20 +0300
Subject: [PATCH 143/370] confc.c: minor stylistic improvements

---
 conf/confc.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/conf/confc.c b/conf/confc.c
index a5566e4..560a027 100644
--- a/conf/confc.c
+++ b/conf/confc.c
@@ -834,8 +834,7 @@ static void grow_cache_st_in(struct c2_sm *mach)
 	 * resp = c2_fop_data(c2_rpc_item_to_fop(item->ri_reply));
 	 * C2_ASSERT(`resp' bob_check()s);
 	 *
-	 * rc = resp->fr_rc;
-	 * rc = rc ?: cache_grow(&ctx->fc_client->cc_registry, resp);
+	 * rc = resp->fr_rc ?: cache_grow(&ctx->fc_client->cc_registry, resp);
 	 *
 	 * // Let rpc layer free the memory allocated for response.
 	 * c2_rpc_item_put(item->ri_reply);
@@ -922,7 +921,7 @@ static const struct c2_sm_state_descr confc_states[S_NR] = {
 		.sd_ex        = NULL,
 		.sd_invariant = NULL,
 		.sd_allowed   = 1 << S_WAIT_REPLY | 1 << S_WAIT_STATUS
-		| 1 << S_TERMINAL | 1 << S_FAILURE
+				| 1 << S_TERMINAL | 1 << S_FAILURE
 	},
 	[S_WAIT_REPLY] = {
 		.sd_flags     = 0,
-- 
1.8.3.2

