From 4618d761dc4457f0315260cdec437888d8384d32 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Sat, 12 May 2012 02:04:34 +0300
Subject: [PATCH 174/370] remove outdated identifier

---
 conf/confc.c | 1 -
 conf/obj.c   | 4 ++--
 2 files changed, 2 insertions(+), 3 deletions(-)

diff --git a/conf/confc.c b/conf/confc.c
index 90fe42f..9c431e7 100644
--- a/conf/confc.c
+++ b/conf/confc.c
@@ -22,7 +22,6 @@
 #  include <config.h>
 #endif
 #include "conf/confc.h"
-#include "conf/cache.h"  /* c2_conf_cache */
 #include "lib/cdefs.h"   /* container_of, C2_HAS_TYPE, equi */
 #include "lib/arith.h"   /* C2_CNT_INC, C2_CNT_DEC */
 
diff --git a/conf/obj.c b/conf/obj.c
index 9ce8d02..14c4a47 100644
--- a/conf/obj.c
+++ b/conf/obj.c
@@ -237,8 +237,8 @@
  *   counter.  c2_confc_fini() asserts (with C2_PRE()) that no objects
  *   are pinned when the cache is being destroyed.
  * - @b i.conf.cache.unique-objects
- *   A registry of cached objects (c2_conf_cache::cc_registry) is used
- *   to achieve uniqueness of configuration object identities.
+ *   A registry of cached objects (c2_conf_map) is used to achieve
+ *   uniqueness of configuration object identities.
  *
  * <hr> <!------------------------------------------------------------>
  * @section conf-ut Unit Tests
-- 
1.8.3.2

