From eee69c84fe593f3d02e71e6eb81880f5ebb5a6f7 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Sat, 18 Aug 2012 00:31:25 +0300
Subject: [PATCH 318/370] conf/obj_ops.h: introduce c2_conf_obj_ops::coo_fini()

---
 conf/obj_ops.h | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/conf/obj_ops.h b/conf/obj_ops.h
index 65d2abb..8c3c675 100644
--- a/conf/obj_ops.h
+++ b/conf/obj_ops.h
@@ -96,6 +96,16 @@ struct c2_conf_obj_ops {
 	 * object in case of C2_CONF_DIRNEXT.
 	 */
 	int (*coo_readdir)(struct c2_conf_obj *dir, struct c2_conf_obj **pptr);
+	/**
+	 * Destructor of configuration object.
+	 *
+	 * The intended usage of ->coo_fini() is to free the allocated
+	 * concrete object (i.e., the ambient object that `obj' is
+	 * embedded in).
+	 *
+	 * NULL is a noop.
+	 */
+	void (*coo_fini)(struct c2_conf_obj *obj);
 };
 
 /**
-- 
1.8.3.2

