From e8de979d0834dd6f64ffd7bf2351d7874b7ceeab Mon Sep 17 00:00:00 2001
From: Anatoliy Bilenko <anatoliy_bilenko@xyratex.com>
Date: Sat, 1 Sep 2012 14:12:22 +0300
Subject: [PATCH 354/370] preload: Remove useless conditions, add assertions.

---
 conf/preload.c | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/conf/preload.c b/conf/preload.c
index 44af51d..2be2323 100644
--- a/conf/preload.c
+++ b/conf/preload.c
@@ -30,7 +30,7 @@
 #include <ctype.h>		/* isspace */
 #include <stdlib.h>		/* strto{ull,ll,ul,l} @todo: move to misc.h */
 
-/* #define HUNK_DBG */
+#define HUNK_DBG
 #ifdef HUNK_DBG
 #include <stdio.h>
 #endif
@@ -275,13 +275,13 @@ static int unit_parse(struct hunk *unit, char ss)
 		if (is_in_braces(&braces))
 			continue;
 
-		if (*p == ss || p == unit->end) {
-			unit->end = (*p == ss) ? --p : p;
-			return unit_trim(unit);
+		if (*p == ss) {
+			unit->end = --p;
+			break;
 		}
 	}
 
-	return -ENOENT;
+	return unit_trim(unit);
 }
 
 static int unit_next(struct hunk *unit, const struct hunk *buffer, char ss)
@@ -293,15 +293,15 @@ static int unit_next(struct hunk *unit, const struct hunk *buffer, char ss)
 	unit->begin = unit->end;
 	unit->end = buffer->end;
 
-	++unit->begin;
+	unit->begin++;
 	if (!hunk_invariant(unit))
 		return -ENOENT;
 
 	if ((rc = unit_trim(unit)) != 0)
 		return rc;
 
-	if (*unit->begin++ != ss)
-		return -EINVAL;
+	C2_ASSERT(*unit->begin == ss);
+	unit->begin++;
 
 	C2_POST(hunk_invariant(unit));
 	return 0;
-- 
1.8.3.2

