From 938a050607de411a6075f6b0988bbc6d26a1769f Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Sat, 6 Apr 2013 00:10:13 +0800
Subject: [PATCH 16/18] fix merge issue

---
 reqh/reqh.c | 25 +++++++++----------------
 1 file changed, 9 insertions(+), 16 deletions(-)

diff --git a/reqh/reqh.c b/reqh/reqh.c
index 6a017af..709b50d 100644
--- a/reqh/reqh.c
+++ b/reqh/reqh.c
@@ -83,26 +83,13 @@ M0_INTERNAL int m0_reqh_init(struct m0_reqh *reqh,
 
 	M0_PRE(reqh != NULL);
 
-	result = m0_fom_domain_init(&reqh->rh_fom_dom);
+	result = m0_layout_domain_init(&reqh->rh_ldom, reqh->rh_dbenv);
 	if (result != 0)
 		return result;
-	reqh->rh_dtm             = reqh_args->rhia_dtm;
-	reqh->rh_dbenv           = reqh_args->rhia_db;
-	reqh->rh_svc             = reqh_args->rhia_svc;
-	reqh->rh_mdstore         = reqh_args->rhia_mdstore;
-	reqh->rh_fol             = reqh_args->rhia_fol;
-	reqh->rh_shutdown        = false;
-	reqh->rh_fom_dom.fd_reqh = reqh;
 
-	result = m0_layout_domain_init(&reqh->rh_ldom, reqh->rh_dbenv);
-	if (result != 0) {
-		m0_fom_domain_fini(&reqh->rh_fom_dom);
-		return result;
-	}
 	result = m0_layout_standard_types_register(&reqh->rh_ldom);
 	if (result != 0) {
 		m0_layout_domain_fini(&reqh->rh_ldom);
-		m0_fom_domain_fini(&reqh->rh_fom_dom);
 		return result;
 	}
 
@@ -130,8 +117,11 @@ M0_INTERNAL int m0_reqh_init(struct m0_reqh *reqh,
 						addb_stob_seg_size,
 						addb_stob_size,
 						addb_stob_timeout);
-		if (result != 0)
+		if (result != 0) {
+			m0_layout_standard_types_unregister(&reqh->rh_ldom);
+			m0_layout_domain_fini(&reqh->rh_ldom);
 			return result;
+		}
 
 		m0_addb_mc_configure_pt_evmgr(&reqh->rh_addb_mc);
 		if (!m0_addb_mc_is_fully_configured(&m0_addb_gmc))
@@ -148,8 +138,11 @@ M0_INTERNAL int m0_reqh_init(struct m0_reqh *reqh,
 
 	reqh->rh_fom_dom.fd_reqh = reqh;
 	result = m0_fom_domain_init(&reqh->rh_fom_dom);
-	if (result != 0)
+	if (result != 0) {
+		m0_layout_standard_types_unregister(&reqh->rh_ldom);
+		m0_layout_domain_fini(&reqh->rh_ldom);
 		return result;
+	}
 	reqh->rh_dtm             = reqh_args->rhia_dtm;
 	reqh->rh_dbenv           = reqh_args->rhia_db;
 	reqh->rh_svc             = reqh_args->rhia_svc;
-- 
1.8.3.2

