From fbfbb671dca3cf341985b3dae59c328bcf900eb2 Mon Sep 17 00:00:00 2001
From: Andriy Tkachuk <andriy_tkachuk@xyratex.com>
Date: Fri, 31 May 2013 16:25:08 +0300
Subject: [PATCH 081/228] be/seg: dictionary primitives added

---
 be/seg.c | 25 +++++++++++++++++++++++++
 be/seg.h | 15 +++++++++++++++
 2 files changed, 40 insertions(+)

diff --git a/be/seg.c b/be/seg.c
index cbc3eba..b76f485 100644
--- a/be/seg.c
+++ b/be/seg.c
@@ -35,10 +35,17 @@
 #define BE_SEG_DEFAULT_ADDR   ((void *)0x400000000000)
 #define BE_SEG_HEADER_OFFSET  (0ULL)
 
+struct m0_be_seg_dict {
+	char   bsd_name[M0_BSD_MAXNAME];
+	void  *bsd_ptr;
+};
+
 /* "On-disk" header for segment, stored in STOB at zero offset */
 struct be_seg_hdr {
 	void       *bh_addr;  /* segment address in RAM */
 	m0_bcount_t bh_size;  /* segment size */
+	/** Segment dictionary */
+	struct m0_be_seg_dict  bs_dict[M0_SEG_DICT_SIZE];
 };
 
 static int stob_io_single_write(void *buf, m0_bcount_t bufsize,
@@ -296,6 +303,24 @@ M0_INTERNAL int m0_be__reg_write(struct m0_be_seg *seg, struct m0_be_op *op,
 	return -1; /*XXX*/
 }
 
+M0_INTERNAL int m0_be_seg_dict_lookup(struct m0_be_seg *seg, const char *name,
+					void **out)
+{
+	*out = NULL;
+	return 0;
+}
+
+M0_INTERNAL int m0_be_seg_dict_insert(struct m0_be_seg *seg, const char *name,
+					void *value)
+{
+	return 0;
+}
+
+M0_INTERNAL int m0_be_seg_dict_delete(struct m0_be_seg *seg, const char *name)
+{
+	return 0;
+}
+
 /** @} end of be group */
 /*
  *  Local variables:
diff --git a/be/seg.h b/be/seg.h
index 7f52e53..e32315e 100644
--- a/be/seg.h
+++ b/be/seg.h
@@ -34,6 +34,13 @@ struct m0_be_op;
  * @{
  */
 
+enum {
+	/** Dictionary entity name maximum string size */
+	M0_BSD_MAXNAME = 80,
+	/** Maximum number of entities in dictionary */
+	M0_SEG_DICT_SIZE = 32,
+};
+
 enum m0_be_seg_states {
 	M0_BSS_INIT,
 	M0_BSS_OPENED,
@@ -74,6 +81,14 @@ struct m0_be_reg {
 	void             *br_addr;
 };
 
+M0_INTERNAL int m0_be_seg_dict_lookup(struct m0_be_seg *seg, const char *name,
+					void **out);
+
+M0_INTERNAL int m0_be_seg_dict_insert(struct m0_be_seg *seg, const char *name,
+					void *value);
+
+M0_INTERNAL int m0_be_seg_dict_delete(struct m0_be_seg *seg, const char *name);
+
 #define M0_BE_REG(seg, size, addr) \
 	((struct m0_be_reg) {      \
 		.br_seg  = (seg),  \
-- 
1.8.3.2

