From c382da7c708f18fbf913fcac534a80ec3bdca987 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Thu, 6 Jun 2013 18:10:27 +0300
Subject: [PATCH 118/228] be/list: account for pointers only, use separate reg
 for each one.

---
 be/list.c | 21 ++++++++++++---------
 1 file changed, 12 insertions(+), 9 deletions(-)

diff --git a/be/list.c b/be/list.c
index 82246f1..904338b 100644
--- a/be/list.c
+++ b/be/list.c
@@ -126,22 +126,25 @@ M0_INTERNAL void m0_be_list_credit(const struct m0_be_list *list,
 	struct m0_be_tx_credit k;
 	M0_PRE(list != NULL && list->bl_descr != NULL);
 
+	/* Count each ll_next/ll_prev/l_head/l_tail pointer as separate
+	 * region. */
 	switch (optype) {
 	case M0_BLO_INSERT:
 	case M0_BLO_DELETE:
-		k.tc_reg_nr = 4;
-		k.tc_reg_size = 4*sizeof(void *) +
-			list->bl_descr->td_container_size;
-		m0_be_tx_credit_add(accum, &k);
-		break;
-	case M0_BLO_MOVE:
 	default:
+		/* left->ll_next, self->ll_prev, self->ll_next,
+		 * right->ll_prev and possibly l_head and l_tail */
 		k.tc_reg_nr = 6;
-		k.tc_reg_size = 6*sizeof(void *) +
-			list->bl_descr->td_container_size;
-		m0_be_tx_credit_add(accum, &k);
+		break;
+	case M0_BLO_MOVE:
+		/* oldleft->ll_next, self->ll_prev, self->ll_next,
+		 * oldright->ll_prev, newleft->ll_next, newright->ll_prev,
+		 * and possibly l_head and l_tail */
+		k.tc_reg_nr = 8;
 		break;
 	}
+	k.tc_reg_size = k.tc_reg_nr * sizeof(void *);
+	m0_be_tx_credit_add(accum, &k);
 }
 
 /*
-- 
1.8.3.2

