From 945f8d94eae257db86374e0e6c36845f55763c7a Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Fri, 21 Jun 2013 16:04:16 +0300
Subject: [PATCH 205/228] lib: extend M0_BITS and M0_COUNT_PARAMS to 19
 parameters.

---
 lib/cdefs.h |  8 +++++---
 lib/misc.h  | 11 +++++++++++
 2 files changed, 16 insertions(+), 3 deletions(-)

diff --git a/lib/cdefs.h b/lib/cdefs.h
index 1b98680..7c4e8b5 100644
--- a/lib/cdefs.h
+++ b/lib/cdefs.h
@@ -89,7 +89,7 @@ M0_FIELD_IS(type, field, uint64_t) ?				\
 	 sizeof(char [M0_FIELD_IS(type, field, uint64_t) - 1])
 
 /**
- * Returns the number of parameters given to this variadic macro (up to 9
+ * Returns the number of parameters given to this variadic macro (up to 19
  * parameters are supported)
  * @note M0_COUNT_PARAMS() returns max(number_of_parameters - 1, 0)
  *     e.g. M0_COUNT_PARAMS()        -> 0
@@ -98,8 +98,10 @@ M0_FIELD_IS(type, field, uint64_t) ?				\
  *          M0_COUNT_PARAMS(x, y, z) -> 2
  */
 #define M0_COUNT_PARAMS(...) \
-	M0_COUNT_PARAMS2(__VA_ARGS__, 9,8,7,6,5,4,3,2,1,0)
-#define M0_COUNT_PARAMS2(_0,_1,_2,_3,_4,_5,_6,_7,_8,_9,_, ...) _
+	M0_COUNT_PARAMS2(__VA_ARGS__, 19,18,17,16,15,14,13,12,11,10,\
+			 9,8,7,6,5,4,3,2,1,0)
+#define M0_COUNT_PARAMS2(_0,_1,_2,_3,_4,_5,_6,_7,_8,_9,\
+			 _10,_11,_12,_13,_14,_15,_16,_17,_18,_19,_, ...) _
 
 /**
  * Concatenates two arguments to produce a single token.
diff --git a/lib/misc.h b/lib/misc.h
index 17cf116..a15eb7f 100644
--- a/lib/misc.h
+++ b/lib/misc.h
@@ -149,6 +149,17 @@ M0_INTERNAL uint64_t m0_round_down(uint64_t val, uint64_t size);
 #define __M0_BITS_6(i, ...)  ((1ULL << (i)) | __M0_BITS_5(__VA_ARGS__))
 #define __M0_BITS_7(i, ...)  ((1ULL << (i)) | __M0_BITS_6(__VA_ARGS__))
 #define __M0_BITS_8(i, ...)  ((1ULL << (i)) | __M0_BITS_7(__VA_ARGS__))
+#define __M0_BITS_9(i, ...)  ((1ULL << (i)) | __M0_BITS_8(__VA_ARGS__))
+#define __M0_BITS_10(i, ...) ((1ULL << (i)) | __M0_BITS_9(__VA_ARGS__))
+#define __M0_BITS_11(i, ...) ((1ULL << (i)) | __M0_BITS_10(__VA_ARGS__))
+#define __M0_BITS_12(i, ...) ((1ULL << (i)) | __M0_BITS_11(__VA_ARGS__))
+#define __M0_BITS_13(i, ...) ((1ULL << (i)) | __M0_BITS_12(__VA_ARGS__))
+#define __M0_BITS_14(i, ...) ((1ULL << (i)) | __M0_BITS_13(__VA_ARGS__))
+#define __M0_BITS_15(i, ...) ((1ULL << (i)) | __M0_BITS_14(__VA_ARGS__))
+#define __M0_BITS_16(i, ...) ((1ULL << (i)) | __M0_BITS_15(__VA_ARGS__))
+#define __M0_BITS_17(i, ...) ((1ULL << (i)) | __M0_BITS_16(__VA_ARGS__))
+#define __M0_BITS_18(i, ...) ((1ULL << (i)) | __M0_BITS_17(__VA_ARGS__))
+#define __M0_BITS_19(i, ...) ((1ULL << (i)) | __M0_BITS_18(__VA_ARGS__))
 
 M0_INTERNAL const char *m0_bool_to_str(bool b);
 
-- 
1.8.3.2

