From dbbf26d44748a45177da7ff3dac9229ca379125b Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Sun, 30 Jun 2013 11:26:58 +0300
Subject: [PATCH 220/228] be/tx.c: also free tx buffers when tx fails.

---
 be/tx.c | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/be/tx.c b/be/tx.c
index cf1729c..da3777f 100644
--- a/be/tx.c
+++ b/be/tx.c
@@ -184,8 +184,10 @@ static int tx_fom_tick(struct m0_fom *fom)
 		if (rc == 0) {
 			M0_LOG(M0_DEBUG, "launched tx%p i/o", tx);
 			tx_state_set(tx, M0_BTS_PLACE);
-		} else
+		} else {
+			m0_be_tx_free(tx);  /* free region and i/o buffers */
 			tx_state_move(tx, rc, M0_BTS_FAILED);
+		}
 		return M0_FSO_WAIT;
 
 	case M0_BTS_PLACE:
@@ -214,9 +216,9 @@ static int tx_fom_tick(struct m0_fom *fom)
 }
 
 static const struct m0_sm_conf tx_sm_conf = {
-	.scf_name      = "m0_be_tx::bo_fom",
-	.scf_nr_states = M0_BTS_NR,
-	.scf_state     = tx_states
+	.scf_name         = "m0_be_tx::bo_fom",
+	.scf_nr_states    = M0_BTS_NR,
+	.scf_state        = tx_states
 };
 
 static const struct m0_fom_ops tx_fom_ops = {
@@ -227,7 +229,7 @@ static const struct m0_fom_ops tx_fom_ops = {
 };
 
 static const struct m0_fom_type_ops tx_fom_type_ops = {
-	.fto_create = NULL
+	.fto_create       = NULL
 };
 
 static struct m0_fom_type tx_fom_type;
-- 
1.8.3.2

