From 514f14510c3d7dd0cd3393e93cc535607fb00a79 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Sun, 30 Jun 2013 17:50:49 +0300
Subject: [PATCH 223/228] be/tx: tx_link(): move tx between tg lists when tx is
 part of a group.

---
 be/tx.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/be/tx.c b/be/tx.c
index 50b5a26..8e497f8 100644
--- a/be/tx.c
+++ b/be/tx.c
@@ -702,6 +702,8 @@ static void tx_link(struct m0_be_tx *tx)
 {
 	M0_PRE(m0_be__tx_invariant(tx));
 	tx_tlist_move(&tx_engine(tx)->te_txs[tx_state(tx)], tx);
+	if (tx->t_group != NULL)
+		gr_tlist_move(&tx->t_group->tg_txs[tx_state(tx)], tx);
 	M0_POST(m0_be__tx_invariant(tx));
 }
 
-- 
1.8.3.2

