From 7d081a4f6545879c97f09834976dff1d6e7992df Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Thu, 7 Feb 2013 21:15:26 +0800
Subject: [PATCH 09/11] fix spaces

---
 m0t1fs/linux_kernel/super.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/m0t1fs/linux_kernel/super.c b/m0t1fs/linux_kernel/super.c
index 2985079..dd8b38e 100644
--- a/m0t1fs/linux_kernel/super.c
+++ b/m0t1fs/linux_kernel/super.c
@@ -699,7 +699,7 @@ static int m0t1fs_root_alloc(struct super_block *sb)
 
 	/* create the new layout on mds: -EEXIST is OK */
 	rc = m0t1fs_layout_op(csb, M0_LAYOUT_OP_ADD, csb->csb_layout_id,
-				NULL);
+			      NULL);
 	if (rc != 0 && rc != -EEXIST)
 		M0_RETURN(rc);
 
-- 
1.8.3.2

