From 3074adb6f87c52573ef17a5797355fe3a9273627 Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Fri, 8 Feb 2013 18:40:03 +0800
Subject: [PATCH 11/11] revert temprary code

---
 utils/ut_main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/utils/ut_main.c b/utils/ut_main.c
index 51f7769..10a7672 100644
--- a/utils/ut_main.c
+++ b/utils/ut_main.c
@@ -128,7 +128,7 @@ void add_uts(void)
 	m0_ut_add(&parity_math_ut);
 	m0_ut_add(&poolmach_ut);
 	m0_ut_add(&reqh_ut);
-//	m0_ut_add(&rm_ut);
+	m0_ut_add(&rm_ut);
 	m0_ut_add(&rpc_mc_ut);
 	m0_ut_add(&rpc_service_ut);
 	m0_ut_add(&rpclib_ut);
-- 
1.8.3.2

