From 7bcb1495451c07a6c58eb59fb2acb4ed26411d6a Mon Sep 17 00:00:00 2001
From: "anand.vidwansa" <anand_vidwansa@xyratex.com>
Date: Fri, 5 Jul 2013 02:44:00 -0700
Subject: [PATCH 07/50] bigmemalloc: - Removed commented code.

---
 lib/ut/genarray.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/lib/ut/genarray.c b/lib/ut/genarray.c
index 5246398..1d29672 100644
--- a/lib/ut/genarray.c
+++ b/lib/ut/genarray.c
@@ -97,7 +97,6 @@ void test_genarray(void)
 	M0_UT_ASSERT(arr.ga_levels[GA_LEVEL0] != 0);
 	M0_UT_ASSERT(arr.ga_levels[GA_LEVEL1] != 0);
 
-	//ele_nr_in_page = PAGE_CACHE_SIZE / sizeof(unsigned long);
 	ele_nr_in_page = M0_0VEC_ALIGN / sizeof(unsigned long);
 	for (id = 0; id < ele_nr_in_page; ++id) {
 		tier1 = (unsigned long *)arr.ga_levels[GA_LEVEL1] + id;
-- 
1.8.3.2

