From 499ea685e42b8aa5e3e5a559fe0822f97817e87e Mon Sep 17 00:00:00 2001
From: Rohan Puri <rohan_puri@xyratex.com>
Date: Thu, 18 Jul 2013 16:38:26 +0530
Subject: [PATCH 019/159] Removed whitespace

---
 addb/addb_monitor.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/addb/addb_monitor.c b/addb/addb_monitor.c
index d4e0067..109f4cb 100644
--- a/addb/addb_monitor.c
+++ b/addb/addb_monitor.c
@@ -183,7 +183,7 @@ M0_INTERNAL int m0_addb_monitor_init(struct m0_addb_monitor        *monitor,
 					    const struct m0_reqh         *reqh),
       struct m0_addb_sum_rec * (*am_sum_get) (const struct m0_addb_monitor *m,
 					      const struct m0_reqh *reqh),
- 			  int (*am_sum_alloc) (const struct m0_addb_monitor *m,
+			  int (*am_sum_alloc) (const struct m0_addb_monitor *m,
 					       const struct m0_reqh *reqh,
 					       const struct m0_addb_rec_type *r))
 {
-- 
1.8.3.2

