From 919934cc00e9d96e8ce3270cbdb5ee8d16ed6380 Mon Sep 17 00:00:00 2001
From: Rajanikant Chirmade <rajanikant_chirmade@xyratex.com>
Date: Wed, 13 Nov 2013 17:42:12 +0530
Subject: [PATCH 154/159]  - Fixed fom_monitor registration from client as well
 as server issue.

---
 fop/fom.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/fop/fom.c b/fop/fom.c
index 7025111..eef1b29 100644
--- a/fop/fom.c
+++ b/fop/fom.c
@@ -265,7 +265,7 @@ static int fop_rate_monitor_init(struct m0_reqh         *reqh,
 		M0_RETURN(-ENOMEM);
 
 	m0_addb_monitor_init(monitor, &fop_rate_monitor_ops);
-	m0_addb_monitor_sum_rec_init(sum_rec, &m0_addb_rt_fop_rate,
+	m0_addb_monitor_sum_rec_init(sum_rec, &m0_addb_rt_fop_rate_cntr,
 				     (uint64_t *)&fop_rate_stats_sum,
 				     FOP_RATE_STATS_NR);
 
@@ -1081,7 +1081,6 @@ static void __fom_domain_fini(struct m0_fom_domain *dom)
 
 	M0_ASSERT(m0_fom_domain_invariant(dom));
 
-	fop_rate_monitor_fini(dom->fd_reqh, &fop_rate_monitor);
 	fd_loc_nr = dom->fd_localities_nr;
 	while (fd_loc_nr > 0) {
 		loc_fini(&dom->fd_localities[fd_loc_nr - 1]);
-- 
1.8.3.2

