From 663b2d8fb4d70222448bc6e6a233fa6d9ca89f5e Mon Sep 17 00:00:00 2001
From: Rajanikant Chirmade <rajanikant_chirmade@xyratex.com>
Date: Mon, 18 Nov 2013 14:19:28 +0530
Subject: [PATCH 158/159]  - Fixed some of review commets.

---
 fop/fom.h | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/fop/fom.h b/fop/fom.h
index c961805..1932306 100644
--- a/fop/fom.h
+++ b/fop/fom.h
@@ -260,10 +260,6 @@ struct m0_fom_locality {
 	 */
 	unsigned                     fl_foms;
 
-	/** fop rate stats */
-	uint64_t                     fl_fop_rate_count;
-	m0_time_t		     fl_fop_rate_next_update;
-
 	/** State Machine (SM) group for AST call-backs */
 	struct m0_sm_group	     fl_group;
 
@@ -300,7 +296,10 @@ struct m0_fom_locality {
 	   fom_dequeue() counting by m0_fom::fo_sched_epoch.
 	 */
 	struct m0_addb_counter       fl_stat_sched_wait_times;
-	/** FOP rate conter. It is fop execution per sec. */
+
+	/** FOP rate counter. It is fop executed per sec. */
+	uint64_t                     fl_fop_rate_count;
+	m0_time_t		     fl_fop_rate_next_update;
 	struct m0_addb_counter       fl_stat_fop_rate;
 
 	/** FOP rate counter. It is fop executed per sec. */
-- 
1.8.3.2

