From e676304c1c4f6634eb003c82e9fbab6a8fba4968 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Sat, 9 Feb 2013 20:47:11 +0200
Subject: [PATCH 06/22] m0t1fs/st: allow overriding TRACE_LEVEL and TRACE_MASK

---
 m0t1fs/linux_kernel/st/st | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/m0t1fs/linux_kernel/st/st b/m0t1fs/linux_kernel/st/st
index 34822ca..fa0f9d6 100755
--- a/m0t1fs/linux_kernel/st/st
+++ b/m0t1fs/linux_kernel/st/st
@@ -1,14 +1,14 @@
 #!/bin/bash
 set -e
-set -u
 
 ## CAUTION: This path will be removed by superuser.
 SANDBOX_DIR=~devvm/_m0-sandbox
 
 # TRACE_MASK=all
-TRACE_MASK='!rpc,formation,fop,memory'
-TRACE_LEVEL='info+'
+[ "z$TRACE_MASK" = "z" ] && TRACE_MASK='!rpc,formation,fop,memory'
+[ "z$TRACE_LEVEL" = "z" ] && TRACE_LEVEL='info+'
 # TRACE_CONTEXT=full
+set -u
 
 MAX_RPC_MSG_SIZE=163840
 TM_MIN_RECV_QUEUE_LEN=2
-- 
1.8.3.2

