From 50dd90ae6ffb82b47e7a816f1986777b32e237cd Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Sun, 10 Feb 2013 00:14:31 +0200
Subject: [PATCH 10/22] m0t1fs/readir: use DT_UNKNOWN for files of unknown type

---
 m0t1fs/linux_kernel/dir.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/m0t1fs/linux_kernel/dir.c b/m0t1fs/linux_kernel/dir.c
index 8ed381b..79c0a4d 100644
--- a/m0t1fs/linux_kernel/dir.c
+++ b/m0t1fs/linux_kernel/dir.c
@@ -424,11 +424,11 @@ static int m0t1fs_readdir(struct file *f,
 				type = DT_DIR;
 			} else {
 				/**
-				   TODO: Entry type is hardcoded to regular
-				   files, should be fixed later.
+				   TODO: Entry type is unknown and ino is
+				   pretty random, should be fixed later.
 				 */
 				ino = ++i;
-				type = DT_REG;
+				type = DT_UNKNOWN;
 			}
 
 			M0_LOG(M0_DEBUG, "filled off %lu ino %lu name \"%*s\"",
-- 
1.8.3.2

