From e72a48ed1606126617f4de3424e7705d0dfa02c2 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Sun, 10 Feb 2013 14:26:45 +0200
Subject: [PATCH 14/22] cob/mkfs: root directory starts with 2 nlinks in POSIX
 systems

---
 cob/cob.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cob/cob.c b/cob/cob.c
index daf6d55..e8336b4 100644
--- a/cob/cob.c
+++ b/cob/cob.c
@@ -439,7 +439,7 @@ M0_INTERNAL int m0_cob_domain_mkfs(struct m0_cob_domain *dom,
         nsrec.cnr_omgid = 0;
         nsrec.cnr_fid = *rootfid;
 
-        nsrec.cnr_nlink = 1;
+        nsrec.cnr_nlink = 2;
         nsrec.cnr_size = MKFS_ROOT_SIZE;
         nsrec.cnr_blksize = MKFS_ROOT_BLKSIZE;
         nsrec.cnr_blocks = MKFS_ROOT_BLOCKS;
-- 
1.8.3.2

