From 771a847d9ad04f747d6936f5f7f4a4f3366fc331 Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Thu, 17 Oct 2013 21:13:45 +0800
Subject: [PATCH 8/8] use @verbatim instead of @code.

---
 cm/cm.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/cm/cm.c b/cm/cm.c
index d112183..a00dea2 100644
--- a/cm/cm.c
+++ b/cm/cm.c
@@ -279,7 +279,7 @@
 
    The call sequence of these interfaces is:
 
-   @code
+   @verbatim
                                          |
                                          |
                                          V
@@ -322,7 +322,7 @@
                                                    |returns -ENOENT after this|
                                                    |call.                     |
                                                    ----------------------------
-   @endcode
+   @endverbatim
 
    @subsection CMDLD-lspec-cm-stop Copy machine stop
    Once operation completes successfully, copy machine performs required tasks,
-- 
1.8.3.2

