From e451257b51b1227cd047dc6ae8aa72b193c9f9c9 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Mon, 26 Nov 2012 04:27:35 +0200
Subject: [PATCH 16/71] LogD 1487: len_total = len as a separate statement

---
 net/test/network.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/net/test/network.c b/net/test/network.c
index 2166acd..e3fb3d4 100644
--- a/net/test/network.c
+++ b/net/test/network.c
@@ -592,11 +592,12 @@ static c2_bcount_t network_bd_serialize(enum c2_net_test_serialize_op op,
 	bd.ntnbd_buf_size = buf->nb_length;
 	bd.ntnbd_len	  = buf->nb_desc.nbd_len;
 
-	len_total = len = c2_net_test_serialize(op, &bd,
-					USE_TYPE_DESCR(net_test_network_bd),
-						bv, bv_offset);
+	len = c2_net_test_serialize(op, &bd,
+				    USE_TYPE_DESCR(net_test_network_bd),
+				    bv, bv_offset);
 	if (len == 0)
 		return 0;
+	len_total = len;
 
 	/* optimizing memory allocation */
 	if (op == C2_NET_TEST_DESERIALIZE &&
-- 
1.8.3.2

