From 5bf7b464f5fbf855c09f93583867ebad5c4620d9 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Mon, 26 Nov 2012 08:51:18 +0200
Subject: [PATCH 25/71] LogD 1496: too much whitespace

---
 net/test/node_bulk.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/net/test/node_bulk.c b/net/test/node_bulk.c
index 4def747..307868a 100644
--- a/net/test/node_bulk.c
+++ b/net/test/node_bulk.c
@@ -551,9 +551,9 @@ void node_bulk_state_transition_auto(struct node_bulk_ctx *ctx,
 
 void node_bulk_state_transition_auto_all(struct node_bulk_ctx *ctx)
 {
-	size_t			   bs_index;
-	size_t			   i;
-	size_t			   nr;
+	size_t bs_index;
+	size_t i;
+	size_t nr;
 
 	C2_PRE(ctx != NULL);
 
-- 
1.8.3.2

