From 1dbefe439adef74de2b3a73827689bc42c86fd02 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Mon, 26 Nov 2012 09:06:49 +0200
Subject: [PATCH 26/71] LogD 1497: explained number of serialized buffer
 descriptors decreasing

---
 net/test/node_bulk.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/net/test/node_bulk.c b/net/test/node_bulk.c
index 307868a..a58541f 100644
--- a/net/test/node_bulk.c
+++ b/net/test/node_bulk.c
@@ -1012,6 +1012,15 @@ static c2_bcount_t client_bds_serialize2(struct c2_net_test_network_ctx *net,
 					       msg_buf_index,
 					       offset + len_total);
 	if (len == 0) {
+		/*
+		 * If first buffer descriptor serializing succeed,
+		 * then number of serialized network buffer descriptors
+		 * is increased. But if second serializing fails, then
+		 * number of network buffer descriptors inside ping
+		 * buffer should be decreased because these two
+		 * descriptors should be added both or should not be
+		 * added at all.
+		 */
 		nr = c2_net_test_network_bd_nr_get(net, msg_buf_index);
 		--nr;
 		c2_net_test_network_bd_nr_set(net, msg_buf_index, nr);
-- 
1.8.3.2

