From c587cef1237f916b1534157dc58713a8a24edf66 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Thu, 29 Nov 2012 09:43:46 +0200
Subject: [PATCH 35/71] LogD 1511: removed time_min()/time_max() layer

---
 net/test/console.c | 19 +++++--------------
 1 file changed, 5 insertions(+), 14 deletions(-)

diff --git a/net/test/console.c b/net/test/console.c
index b7a6e45..5cbcebf 100644
--- a/net/test/console.c
+++ b/net/test/console.c
@@ -187,16 +187,6 @@ static void status_data_reset(struct c2_net_test_cmd_status_data *sd)
 	sd->ntcsd_time_finish = C2_MKTIME(0, 0);
 }
 
-static c2_time_t time_min(c2_time_t t1, c2_time_t t2)
-{
-	return min_check(t1, t2);
-}
-
-static c2_time_t time_max(c2_time_t t1, c2_time_t t2)
-{
-	return max_check(t1, t2);
-}
-
 static void status_data_add(struct c2_net_test_cmd_status_data *all_sd,
 			    const struct c2_net_test_cmd_status_data *cmd_sd)
 {
@@ -239,10 +229,11 @@ static void status_data_add(struct c2_net_test_cmd_status_data *all_sd,
 		c2_net_test_stats_add_stats(stats[i].s_all, stats[i].s_node);
 	all_sd->ntcsd_finished &= cmd_sd->ntcsd_finished;
 	if (cmd_sd->ntcsd_finished) {
-		all_sd->ntcsd_time_start = time_min(all_sd->ntcsd_time_start,
-						    cmd_sd->ntcsd_time_start);
-		all_sd->ntcsd_time_finish = time_max(all_sd->ntcsd_time_finish,
-						     cmd_sd->ntcsd_time_finish);
+		all_sd->ntcsd_time_start = min_check(all_sd->ntcsd_time_start,
+						     cmd_sd->ntcsd_time_start);
+		all_sd->ntcsd_time_finish =
+			max_check(all_sd->ntcsd_time_finish,
+				  cmd_sd->ntcsd_time_finish);
 	}
 }
 
-- 
1.8.3.2

