From 5bcae66790539a1790949d1aa9ca435734596884 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Thu, 29 Nov 2012 18:00:05 +0200
Subject: [PATCH 45/71] LogD 1521: clarified "almost added to network buffer
 queue"

---
 net/test/node.c | 14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

diff --git a/net/test/node.c b/net/test/node.c
index 40bcaae..accbdf1 100644
--- a/net/test/node.c
+++ b/net/test/node.c
@@ -346,8 +346,9 @@ finished.up();\l", shape=box];
    - UNUSED - buffer pair is not used in buffer operations now.
      It is initial state of buffer pair.
    - QUEUED - buffer pair is queued or almost added to network bulk queue.
+     See @ref net-test-lspec-bulk-buf-states "RECEIVING state".
    - BD_SENT - buffer descriptors for bulk buffer pair are sent or almost sent
-     to the test server.
+     to the test server.  See @ref net-test-lspec-bulk-buf-states "RECEIVING state".
    - CB_LEFT2(CB_LFET1) - there are 2 (or 1) network buffer callback(s) left
      for this buffer pair (including network buffer callback for the message
      with buffer descriptor).
@@ -451,15 +452,18 @@ finished.up();\l", shape=box];
 
    @see @ref net-test-lspec-algo-sm-legend
 
+   @anchor net-test-lspec-bulk-buf-states
    Bulk buffer states
    - UNUSED - bulk buffer isn't currently used in network operations and
      can be used when passive bulk buffer decriptors arrive.
    - BD_RECEIVED - message with buffer descriptors was received from the test
      client.
-   - RECEIVING - bulk buffer added (or almost added) to the active bulk
-     receive queue.
-   - SENDING - bulk buffer added (or almost added) to the active bulk
-     send queue.
+   - RECEIVING - bulk buffer added to the active bulk receive queue. Bulk
+     buffer enters this state just before adding to the network queue because
+     network buffer callback may be executed before returning from
+     'add to network buffer queue' function.
+   - SENDING - bulk buffer added to the active bulk send queue. Bulk buffer
+     enters this state as well as for the RECEIVING state.
    - TRANSFERRED - bulk buffer was successfully received and sent.
    - FAILED - some operation failed.
    - BADMSG - message with buffer descriptors contains invalid data.
-- 
1.8.3.2

