From e9c93992cd9fe28f14e82b77833e578f767fdbf9 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Fri, 30 Nov 2012 07:31:38 +0200
Subject: [PATCH 53/71] LogD 1530: explained RTT measuring in DLD for bulk
 testing

---
 net/test/node.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/net/test/node.c b/net/test/node.c
index 35ee0ec..909d7c1 100644
--- a/net/test/node.c
+++ b/net/test/node.c
@@ -222,6 +222,11 @@
 
    @subsubsection net-test-lspec-bulk Bulk Test
 
+   RTT is measured as length of time interval
+   [time of transition to transfer start state,
+    time of transition to transfer finish state]. See
+   @ref net-test-lspec-algo-client-bulk, @ref net-test-lspec-algo-server-bulk.
+
    One test message travel:
    @msc
    c [label = "Test Client"],
@@ -361,6 +366,10 @@ finished.up();\l", shape=box];
 
    Final states: TRANSFERRED, FAILED.
 
+   Transfer start state: QUEUED.
+
+   Transfer finish state: TRANSFERRED.
+
    Bulk buffer pair state transitions
    - UNUSED -> QUEUED
      - client_process_unused_bulk()
@@ -472,6 +481,10 @@ finished.up();\l", shape=box];
 
    Final states: TRANSFERRED, FAILED, BADMSG.
 
+   Transfer start state: RECEIVING.
+
+   Transfer finish state: TRANSFERRED.
+
    Bulk buffer state transitions
 
    - UNUSED -> BD_RECEIVED
-- 
1.8.3.2

