From eadedbe46315c8308409d94135c0b094014e9c29 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Fri, 30 Nov 2012 07:47:14 +0200
Subject: [PATCH 55/71] LogD 1532: changed ' ' to '\t' in DLD test console
 diagram

---
 net/test/node.c | 50 +++++++++++++++++++++++++-------------------------
 1 file changed, 25 insertions(+), 25 deletions(-)

diff --git a/net/test/node.c b/net/test/node.c
index 909d7c1..a790c5c 100644
--- a/net/test/node.c
+++ b/net/test/node.c
@@ -531,34 +531,34 @@ finished.up();\l", shape=box];
    servers [label="Test Servers"];
 
    |||;
-   clients rbox clients [label = "Listening for console commands"],
+   clients rbox clients	[label = "Listening for console commands"],
    server rbox server	[label = "Listening for console commands"];
    console => servers	[label = "INIT command"];
    servers => console	[label = "INIT DONE response"];
-   ---                	[label = "waiting for all servers"];
-   console => clients  	[label = "INIT command"];
-   clients => console  	[label = "INIT DONE response"];
-   ---                	[label = "waiting for all clients"];
-   console => servers  	[label = "START command"];
-   servers => console  	[label = "START ACK response"];
-   ---                	[label = "waiting for all servers"];
-   console => clients  	[label = "START command"];
-   clients => console  	[label = "START ACK response"];
-   ---                	[label = "waiting for all clients"];
-   ---                	[label = "running..."];
-   console => clients  	[label = "STATUS command"];
-   clients => console  	[label = "STATUS DATA response"];
-   console => servers  	[label = "STATUS command"];
-   servers => console  	[label = "STATUS DATA response"];
-   ---                	[label = "console wants to stop client&server"];
-   console => clients  	[label = "STOP command"];
-   clients => console  	[label = "STOP DONE response"];
-   clients rbox clients [label = "clients cleanup"];
-   ---                	[label = "waiting for all clients"];
-   console => servers  	[label = "STOP command"];
-   servers => console  	[label = "STOP DONE response"];
-   servers rbox servers [label = "servers cleanup"];
-   ---                	[label = "waiting for all servers"];
+   ---			[label = "waiting for all servers"];
+   console => clients	[label = "INIT command"];
+   clients => console	[label = "INIT DONE response"];
+   ---			[label = "waiting for all clients"];
+   console => servers	[label = "START command"];
+   servers => console	[label = "START ACK response"];
+   ---			[label = "waiting for all servers"];
+   console => clients	[label = "START command"];
+   clients => console	[label = "START ACK response"];
+   ---			[label = "waiting for all clients"];
+   ---			[label = "running..."];
+   console => clients	[label = "STATUS command"];
+   clients => console	[label = "STATUS DATA response"];
+   console => servers	[label = "STATUS command"];
+   servers => console	[label = "STATUS DATA response"];
+   ---			[label = "console wants to stop client&server"];
+   console => clients	[label = "STOP command"];
+   clients => console	[label = "STOP DONE response"];
+   clients rbox clients	[label = "clients cleanup"];
+   ---			[label = "waiting for all clients"];
+   console => servers	[label = "STOP command"];
+   servers => console	[label = "STOP DONE response"];
+   servers rbox servers	[label = "servers cleanup"];
+   ---			[label = "waiting for all servers"];
    @endmsc
 
    @subsubsection net-test-lspec-misc Misc
-- 
1.8.3.2

