From 4f748289d3341052a9031bf4bf9fd03bf07df505 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Mon, 14 Jan 2013 08:26:33 +0200
Subject: [PATCH 66/71] wait channel now _fini()'d in node_bulk_worker

---
 net/test/node_bulk.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/net/test/node_bulk.c b/net/test/node_bulk.c
index ba30c9b..901f387 100644
--- a/net/test/node_bulk.c
+++ b/net/test/node_bulk.c
@@ -1280,6 +1280,12 @@ static void node_bulk_worker(struct node_bulk_ctx *ctx)
 	c2_net_buffer_event_deliver_all(ctx->nbc_net.ntc_tm);
 	node_bulk_state_transition_auto_all(ctx);
 	c2_net_test_nh_sd_copy_locked(&ctx->nbc_nh);
+	/* transfer machine should't signal to tm_chan */
+	c2_net_buffer_event_notify(ctx->nbc_net.ntc_tm, NULL);
+	c2_clink_del(&tm_clink);
+	c2_clink_fini(&tm_clink);
+	c2_chan_fini(&tm_chan);
+
 }
 
 static int node_bulk_test_init_fini(struct node_bulk_ctx *ctx,
-- 
1.8.3.2

