From 5862fd9a72df62e7e33cbe190f4f0ff534648e03 Mon Sep 17 00:00:00 2001
From: Carl Braganza <carl_braganza@xyratex.com>
Date: Fri, 11 Jan 2013 15:55:55 -0800
Subject: [PATCH 6/6] Fixed block comment style.

---
 addb/user_space/addb_pfom.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/addb/user_space/addb_pfom.c b/addb/user_space/addb_pfom.c
index 0587077..8f89a8e 100644
--- a/addb/user_space/addb_pfom.c
+++ b/addb/user_space/addb_pfom.c
@@ -179,7 +179,8 @@ static void addb_pfom_fo_fini(struct m0_fom *fom)
 	m0_fom_timeout_fini(&pfom->pf_timeout);
 	addb_post_fom_bob_fini(pfom);
 
-	/* Mustn't free as the fom is embedded in the service object, but
+	/*
+	 * Mustn't free as the fom is embedded in the service object, but
 	 * notify UT waiters.
 	 */
 	m0_mutex_lock(&rsvc->rs_mutex);
@@ -365,7 +366,8 @@ static void addb_pfom_stop(struct addb_svc *svc)
 		pfom->pf_ast.sa_cb = addb_pfom_stop_cb;
 		m0_sm_ast_post(&fom->fo_loc->fl_group, &pfom->pf_ast);
 
-		/* We could wait for termination on pf_running but
+		/*
+		 * We could wait for termination on pf_running but
 		 * there is no need to as the request handler will
 		 * wait for FOM termination.
 		 */
-- 
1.8.3.2

