From df103bbb5c9c4c61f7cc928cd0e55b83489ee789 Mon Sep 17 00:00:00 2001
From: Rohan Puri <rohan_puri@xyratex.com>
Date: Thu, 18 Jul 2013 16:07:17 +0530
Subject: [PATCH 078/107] Addressed review comments

---
 addb/addb_monitor.c | 1 -
 addb/addb_monitor.h | 1 -
 2 files changed, 2 deletions(-)

diff --git a/addb/addb_monitor.c b/addb/addb_monitor.c
index a99aa63..19ad726 100644
--- a/addb/addb_monitor.c
+++ b/addb/addb_monitor.c
@@ -374,7 +374,6 @@ M0_INTERNAL int m0_addb_monitor_summaries_post(struct m0_reqh *reqh)
 	m0_free(sum_recs_fop);
 	M0_RETURN(rc);
 }
-
 /*
  *  Local variables:
  *  c-indentation-style: "K&R"
diff --git a/addb/addb_monitor.h b/addb/addb_monitor.h
index 1c73ec8..b2098dc 100644
--- a/addb/addb_monitor.h
+++ b/addb/addb_monitor.h
@@ -303,7 +303,6 @@ M0_INTERNAL void m0_addb_monitors_fini(struct m0_reqh *reqh);
 M0_INTERNAL int m0_addb_monitor_summaries_post(struct m0_reqh *reqh);
 
 #endif /* __MERO_ADDB_ADDB_MONITOR_H__ */
-
 /*
  *  Local variables:
  *  c-indentation-style: "K&R"
-- 
1.8.3.2

