From 56db4af92b1c4b7f20530ba72615d272702b40cb Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Sat, 23 Feb 2013 23:29:01 +0200
Subject: [PATCH 05/17] cob: remove , from m0_cob_iterator_get()/next() exit
 traces

---
 cob/cob.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/cob/cob.c b/cob/cob.c
index c4b8cc6..d47859b 100644
--- a/cob/cob.c
+++ b/cob/cob.c
@@ -1013,7 +1013,7 @@ M0_INTERNAL int m0_cob_iterator_get(struct m0_cob_iterator *it)
 	if (rc == 0 && !m0_fid_eq(&it->ci_key->cnk_pfid, it->ci_cob->co_fid))
 		rc = -ENOENT;
 
-	M0_LEAVE("[%lx:%lx]/%.*s, rc: %d", 
+	M0_LEAVE("[%lx:%lx]/%.*s rc: %d", 
 		 (long)it->ci_key->cnk_pfid.f_container,
 		 (long)it->ci_key->cnk_pfid.f_key,
 		 m0_bitstring_len_get(&it->ci_key->cnk_name),
@@ -1035,7 +1035,7 @@ M0_INTERNAL int m0_cob_iterator_next(struct m0_cob_iterator *it)
 	if (rc == 0 && !m0_fid_eq(&it->ci_key->cnk_pfid, it->ci_cob->co_fid))
 		rc = -ENOENT;
 
-	M0_LEAVE("[%lx:%lx]/%.*s, rc: %d", 
+	M0_LEAVE("[%lx:%lx]/%.*s rc: %d", 
 		 (long)it->ci_key->cnk_pfid.f_container,
 		 (long)it->ci_key->cnk_pfid.f_key,
 		 m0_bitstring_len_get(&it->ci_key->cnk_name),
-- 
1.8.3.2

