From 2438ec58c8e73984a41cead8117e2e8b6b429913 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Sun, 24 Feb 2013 12:23:16 +0200
Subject: [PATCH 07/17] mds: get rid of cob iterator loop in
 m0_dir_empty_check()

---
 mdstore/mdstore.c | 10 ++++------
 1 file changed, 4 insertions(+), 6 deletions(-)

diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index c3ca9f9..e9179fa 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -328,8 +328,8 @@ M0_INTERNAL int m0_mdstore_dir_empty_check(struct m0_mdstore *md,
 		M0_LOG(M0_DEBUG, "iterator init: %d", rc);
 		goto out;
 	}
-	for (rc = m0_cob_iterator_get(&it); rc == 0;
-	     rc = m0_cob_iterator_next(&it)) {
+	rc = m0_cob_iterator_get(&it);
+	if (rc == 0) {
 		M0_LOG(M0_DEBUG, "[%lx:%lx]/%.*s contains [%lx:%lx]/%.*s",
 		       cob->co_nskey->cnk_pfid.f_container,
 		       cob->co_nskey->cnk_pfid.f_key,
@@ -340,12 +340,10 @@ M0_INTERNAL int m0_mdstore_dir_empty_check(struct m0_mdstore *md,
 		       m0_bitstring_len_get(&it.ci_key->cnk_name),
 		       (char *)m0_bitstring_buf_get(&it.ci_key->cnk_name));
 		rc = -ENOTEMPTY;
-		break;
-	}
+	} else if (rc == -ENOENT)
+		rc = 0;
 	m0_cob_iterator_fini(&it);
 out:
-	if (rc == -ENOENT)
-		rc = 0;
 	M0_LEAVE("rc: %d", rc);
 	return rc;
 }
-- 
1.8.3.2

