From 7f012ae0ad8a4acba0e2f9f6f52da6d2595f2225 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Sun, 24 Feb 2013 13:31:24 +0200
Subject: [PATCH 08/17] mds/style: de-doxygenize comments inside
 m0_mdstore_readdir()'s body

---
 mdstore/mdstore.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index e9179fa..d51c500 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -727,7 +727,7 @@ M0_INTERNAL int m0_mdstore_readdir(struct m0_mdstore       *md,
 		goto out_end;
 	}
 
-	/** Emulate "." and ".." entries at the 1st and 2nd positions. */
+	/* Emulate "." and ".." entries at the 1st and 2nd positions. */
 	if (pos_len == 0 || (pos_len == 1 && pos_buf[0] == '.')) {
 		dot = 1;
 	} else if (pos_len == 2 && strncmp(pos_buf, "..", 2) == 0) {
@@ -742,7 +742,7 @@ M0_INTERNAL int m0_mdstore_readdir(struct m0_mdstore       *md,
 		goto out;
 	}
 
-	/**
+	/*
 	 * Positions iterator to the closest key greater or equal
 	 * to the starting position. Returns -ENOENT if no such
 	 * key exists.
@@ -771,7 +771,7 @@ M0_INTERNAL int m0_mdstore_readdir(struct m0_mdstore       *md,
 			       ent->d_namelen, (char *)ent->d_name,
 			       ent->d_reclen);
 		} else {
-			/**
+			/*
 			 * If buffer was too small to hold even one record,
 			 * return -EINVAL. Otherwise return 0.
 			 */
@@ -789,7 +789,7 @@ M0_INTERNAL int m0_mdstore_readdir(struct m0_mdstore       *md,
 out_end:
 	m0_cob_iterator_fini(&it);
 	if (last != NULL) {
-		last->d_reclen = 0;  /** The last record indicator. */
+		last->d_reclen = 0;  /* The last record indicator. */
 		pos_buf = (char *)m0_bitstring_buf_get(rdpg->r_pos);
 		pos_len = m0_bitstring_len_get(rdpg->r_pos);
 		rdpg->r_end = m0_bitstring_alloc(pos_buf, pos_len);
-- 
1.8.3.2

