From d06c7d0b5d8cd4095c0188394ccb21dc411183b9 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Tue, 26 Feb 2013 12:00:38 +0200
Subject: [PATCH 13/17] mds/readdir: rename pos_buf/pos_len to s_buf/s_len

---
 mdstore/mdstore.c | 23 +++++++++++------------
 1 file changed, 11 insertions(+), 12 deletions(-)

diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 360ae75..9de1ca6 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -705,20 +705,19 @@ M0_INTERNAL int m0_mdstore_readdir(struct m0_mdstore       *md,
 	int                            reclen;
 	int                            rc;
 	int                            dot;
-	char                          *pos_buf;
-	uint32_t                       pos_len;
+	char                          *s_buf;
+	uint32_t                       s_len;
 	struct m0_bitstring            empty = {.b_len = 0};
 
 	M0_ENTRY();
 	M0_ASSERT(cob != NULL && rdpg != NULL);
 
-	pos_buf = (char *)m0_bitstring_buf_get(rdpg->r_pos);
-	pos_len = m0_bitstring_len_get(rdpg->r_pos);
+	s_buf = (char *)m0_bitstring_buf_get(rdpg->r_pos);
+	s_len = m0_bitstring_len_get(rdpg->r_pos);
 
 	M0_LOG(M0_DEBUG,
 	       "Readdir on object [%lx:%lx] starting from \"%.*s\"",
-	       cob->co_fid->f_container, cob->co_fid->f_key, pos_len,
-	       pos_buf);
+	       cob->co_fid->f_container, cob->co_fid->f_key, s_len, s_buf);
 
 	ent = rdpg->r_buf.b_addr;
 	nob = rdpg->r_buf.b_nob;
@@ -728,9 +727,9 @@ M0_INTERNAL int m0_mdstore_readdir(struct m0_mdstore       *md,
 	}
 
 	/* Emulate "." and ".." entries at the 1st and 2nd positions. */
-	if (pos_len == 1 && pos_buf[0] == '.') {
+	if (s_len == 1 && s_buf[0] == '.') {
 		dot = 1;
-	} else if (pos_len == 2 && strncmp(pos_buf, "..", 2) == 0) {
+	} else if (s_len == 2 && strncmp(s_buf, "..", 2) == 0) {
 		dot = 2;
 	} else {
 		dot = 0;
@@ -791,11 +790,11 @@ out_end:
 	m0_cob_iterator_fini(&it);
 	if (last != NULL) {
 		last->d_reclen = 0;  /* The last record indicator. */
-		pos_buf = (char *)m0_bitstring_buf_get(rdpg->r_pos);
-		pos_len = m0_bitstring_len_get(rdpg->r_pos);
-		rdpg->r_end = m0_bitstring_alloc(pos_buf, pos_len);
+		s_buf = (char *)m0_bitstring_buf_get(rdpg->r_pos);
+		s_len = m0_bitstring_len_get(rdpg->r_pos);
+		rdpg->r_end = m0_bitstring_alloc(s_buf, s_len);
 		M0_LOG(M0_DEBUG, "%s entry: \"%.*s\"", rc ? "last" : "next",
-		       pos_len, pos_buf);
+		       s_len, s_buf);
 		rc = ENOENT;
 	}
 	if (rc == -ENOENT)
-- 
1.8.3.2

