From b476a17e2decbd5633b464bf7cc03e47a480560a Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Tue, 26 Feb 2013 12:04:35 +0200
Subject: [PATCH 14/17] mds/style: enforce col 80 in m0_mdstore_readdir()

---
 mdstore/mdstore.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 9de1ca6..e188ae3 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -753,9 +753,9 @@ M0_INTERNAL int m0_mdstore_readdir(struct m0_mdstore       *md,
 		} else if (dot == 2) {
 			m0_bitstring_copy(rdpg->r_pos, "..", 2);
 		} else {
-			m0_bitstring_copy(rdpg->r_pos,
-					  m0_bitstring_buf_get(&it.ci_key->cnk_name),
-					  m0_bitstring_len_get(&it.ci_key->cnk_name));
+			s_buf = m0_bitstring_buf_get(&it.ci_key->cnk_name);
+			s_len = m0_bitstring_len_get(&it.ci_key->cnk_name);
+			m0_bitstring_copy(rdpg->r_pos, s_buf, s_len);
 		}
 
 		reclen = m0_align(sizeof(*ent) +
-- 
1.8.3.2

