From 0b1840f1c7098cb9c62351d5287d0689bd6908c9 Mon Sep 17 00:00:00 2001
From: Rajesh Bhalerao <Rajesh_Bhalerao@xyratex.com>
Date: Thu, 22 Aug 2013 11:56:17 +0530
Subject: [PATCH 06/24] Exported few functions.

---
 rm/rm.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/rm/rm.c b/rm/rm.c
index ed43d0f..73b14bb 100644
--- a/rm/rm.c
+++ b/rm/rm.c
@@ -1061,6 +1061,7 @@ M0_INTERNAL int m0_rm_loan_proxy_add(struct m0_rm_loan   *loan,
 	}
 	M0_RETURN(rc);
 }
+M0_EXPORTED(m0_rm_loan_proxy_add);
 
 M0_INTERNAL void m0_rm_loan_proxy_del(struct m0_rm_remote_link *rrl)
 {
@@ -1071,6 +1072,7 @@ M0_INTERNAL void m0_rm_loan_proxy_del(struct m0_rm_remote_link *rrl)
 	m0_proxies_tlink_fini(rrl);
 	m0_free(rrl);
 }
+M0_EXPORTED(m0_rm_loan_proxy_del);
 
 M0_INTERNAL void m0_rm_loan_proxies_remove(struct m0_rm_loan *loan)
 {
@@ -1080,6 +1082,7 @@ M0_INTERNAL void m0_rm_loan_proxies_remove(struct m0_rm_loan *loan)
 		m0_rm_loan_proxy_del(rrl);
 	}  m0_tl_endfor;
 }
+M0_EXPORTED(m0_rm_loan_proxies_remove);
 
 M0_INTERNAL int m0_rm_loan_init(struct m0_rm_loan         *loan,
 				const struct m0_rm_credit *credit)
@@ -1588,6 +1591,7 @@ M0_INTERNAL void m0_rm_credit_get(struct m0_rm_incoming *in)
 	m0_rm_owner_unlock(owner);
 	M0_LEAVE();
 }
+M0_EXPORTED(m0_rm_credit_get);
 
 M0_INTERNAL void m0_rm_credit_put(struct m0_rm_incoming *in)
 {
@@ -1607,6 +1611,7 @@ M0_INTERNAL void m0_rm_credit_put(struct m0_rm_incoming *in)
 	rm_addb_credit_counter_update(&in->rin_want);
 	M0_LEAVE();
 }
+M0_EXPORTED(m0_rm_credit_put);
 
 /*
  * After successful completion of incoming request, move OWOS_CACHED credits
@@ -2003,6 +2008,7 @@ M0_INTERNAL void m0_rm_outgoing_complete(struct m0_rm_outgoing *og)
 	owner_balance(owner);
 	M0_LEAVE();
 }
+M0_EXPORTED(m0_rm_outgoing_complete);
 
 /**
  * Helper function called when an incoming request processing completes.
@@ -2245,6 +2251,7 @@ M0_INTERNAL int m0_rm_owner_loan_debit(struct m0_rm_owner  *owner,
 	m0_rm_ur_tlist_fini(&remove_list);
 	M0_RETURN(rc);
 }
+M0_EXPORTED(m0_rm_owner_loan_debit);
 
 M0_INTERNAL int m0_rm_loan_settle(struct m0_rm_owner  *owner,
 				  struct m0_rm_loan   *loan,
@@ -2271,6 +2278,7 @@ M0_INTERNAL int m0_rm_loan_settle(struct m0_rm_owner  *owner,
 
 	return rc;
 }
+M0_EXPORTED(m0_rm_loan_settle);
 
 M0_INTERNAL bool m0_rm_loan_pay_n_check(struct m0_rm_owner *owner,
 					struct m0_rm_loan  *paid_loan,
-- 
1.8.3.2

