From 5600e360a629687a800f0b2648a36438fd134ae5 Mon Sep 17 00:00:00 2001
From: Rajesh Bhalerao <Rajesh_Bhalerao@xyratex.com>
Date: Wed, 18 Sep 2013 11:12:33 +0530
Subject: [PATCH 23/24] Changed m0_cookie_is_eq() function.

---
 lib/cookie.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/lib/cookie.c b/lib/cookie.c
index f125192..90c16e5f 100644
--- a/lib/cookie.c
+++ b/lib/cookie.c
@@ -46,6 +46,12 @@
  * @{
  */
 
+#ifdef __KERNEL__
+#include <linux/string.h>
+#else
+#include <string.h>
+#endif
+
 #include "lib/types.h"
 #include "lib/errno.h" /* -EPROTO */
 #include "lib/cookie.h"
@@ -116,8 +122,7 @@ M0_INTERNAL bool m0_cookie_is_null(const struct m0_cookie cookie)
 M0_INTERNAL bool m0_cookie_is_eq(const struct m0_cookie *cookie1,
 				 const struct m0_cookie *cookie2)
 {
-	return cookie1->co_generation == cookie2->co_generation &&
-		cookie1->co_addr == cookie2->co_addr;
+	return memcmp(cookie1, cookie2, sizeof *cookie1) == 0;
 }
 
 M0_INTERNAL void m0_cookie_global_fini(void)
-- 
1.8.3.2

