From e3e97554f59d8eabdec7c6170a26cab64f7febd9 Mon Sep 17 00:00:00 2001
From: "trupti.patil" <trupti_patil@xyratex.com>
Date: Tue, 30 Oct 2012 00:24:35 +0530
Subject: [PATCH 014/172] Taking out a few commented out code lines

---
 layout/ut/layout.c | 44 +-------------------------------------------
 1 file changed, 1 insertion(+), 43 deletions(-)

diff --git a/layout/ut/layout.c b/layout/ut/layout.c
index f520832..5a593b7 100644
--- a/layout/ut/layout.c
+++ b/layout/ut/layout.c
@@ -796,11 +796,6 @@ static int test_build_pdclust(uint32_t enum_id, uint64_t lid,
  * Builds num_sublayouts number of sub-layouts and stores those in
  * sub_layout_list.
  */
-#if 0 //todo rm
-#ifndef __KERNEL__
-#include <stdlib.h>      /* printf() */ //todo
-#endif
-#endif
 static void sub_layouts_build(uint64_t composite_lid,
 			      struct c2_tl **sub_layout_list,
 			      uint32_t num_sublayouts)
@@ -861,18 +856,6 @@ static void sub_layouts_build(uint64_t composite_lid,
 		total               = total + delta;
 		where               = sl->csl_ext.e_end + 1;
 		c2_sub_layout_tlink_init_at_tail(sl, *sub_layout_list);
-
-#if 0 //todo rm
-#ifndef __KERNEL__
-		printf("sl_lid %llu, e_start %llu, e_end %llu, "
-			"total %llu, where %llu \n",
-			(unsigned long long)sublayout_lid,
-			(unsigned long long)sl->csl_ext.e_start,
-			(unsigned long long)sl->csl_ext.e_end,
-			(unsigned long long)total,
-			(unsigned long long)where);
-#endif
-#endif
 	}
 
 	/*
@@ -881,18 +864,6 @@ static void sub_layouts_build(uint64_t composite_lid,
 	 */
 	sl->csl_ext.e_end = C2_BINDEX_MAX;
 	total = total - delta + sl->csl_ext.e_end - sl->csl_ext.e_start;
-#if 0 //todo rm
-#ifndef __KERNEL__
-		printf("Revised last segment\n sl_lid %llu, "
-			"e_start %llu, e_end %llu, "
-			"total %llu, where %llu \n\n",
-			(unsigned long long)sublayout_lid,
-			(unsigned long long)sl->csl_ext.e_start,
-			(unsigned long long)sl->csl_ext.e_end,
-			(unsigned long long)total,
-			(unsigned long long)where);
-#endif
-#endif
 	C2_UT_ASSERT(total == C2_BINDEX_MAX);
 	C2_UT_ASSERT(!c2_sub_layout_tlist_is_empty(*sub_layout_list));
 }
@@ -934,18 +905,6 @@ static void composite_layout_verify(struct c2_layout *l,
 				      USER_COUNT_INCREMENTED);
 		C2_UT_ASSERT(sl->csl_l->l_user_count == 1);
 
-#if 0 //todo rm
-#ifndef __KERNEL__
-		printf("sl_lid %llu, e_start %llu, e_end %llu, "
-			"delta %llu, where %llu \n",
-			(unsigned long long)sublayout_lid,
-			(unsigned long long)sl->csl_ext.e_start,
-			(unsigned long long)sl->csl_ext.e_end,
-			(unsigned long long)delta,
-			(unsigned long long)where);
-#endif
-#endif
-
 		/* Verify sl->csl_ext. */
 		C2_UT_ASSERT(sl->csl_ext.e_start == where);
 		if (i == num_sublayouts - 1)
@@ -3412,8 +3371,7 @@ static int test_lookup_composite(uint64_t lid,
 		for (i = 0; i < 10; ++i) {
 			sublayout_lid = lid * 100 + i;
 			l = c2_layout_find(&domain, sublayout_lid);
-			//C2_UT_ASSERT(c2_ref_read(&l->l_ref) == 0); //todo
-				// Check, elsewhere it says l_user_count
+			C2_UT_ASSERT(l->l_user_count == 0);
 			c2_layout_put(l);
 			c2_layout_put(l);
 		}
-- 
1.8.3.2

