From 0ddcc4ee1d5b0ffa62fbfd8c19a83d7cd5a38f4d Mon Sep 17 00:00:00 2001
From: "trupti.patil" <trupti_patil@xyratex.com>
Date: Fri, 26 Apr 2013 19:58:28 +0530
Subject: [PATCH 126/172] Updating doc

---
 layout/layout_db.c | 25 ++++---------------------
 1 file changed, 4 insertions(+), 21 deletions(-)

diff --git a/layout/layout_db.c b/layout/layout_db.c
index cb757c6..9effd35 100644
--- a/layout/layout_db.c
+++ b/layout/layout_db.c
@@ -153,7 +153,6 @@
  * The layout schema for the Layout DB module consists of the following tables.
  * - @ref Layout-DB-lspec-schema-layouts
  * - @ref Layout-DB-lspec-schema-cob_lists
- * - @ref Layout-DB-lspec-schema-comp_layout_layer_sublayout
  * - @ref Layout-DB-lspec-schema-comp_layout_layer_emap
  *
  * Key-Record structures for these tables are described below.
@@ -177,6 +176,8 @@
  * - In case of a layout with LIST enum type, an array of m0_fid
  *   structure with size LDB_MAX_INLINE_COB_ENTRIES is used to store a few COB
  *   entries inline into the layouts table itself.
+ * - In case of COMPOSITE layout, 'number of layers' and 'sublayout id
+ *   associated with each layer' are stored in the layouts table.
  * - It is possible that some layouts do not need to store any layout type or
  *   layout enum type specific data in the layouts table. For example, a
  *   layout with COMPOSITE layout type.
@@ -201,22 +202,6 @@
  * index from the array of m0_fid structures stored inline in the layouts
  * table.
  *
- * @subsection Layout-DB-lspec-schema-comp_layout_layer_sublayout
- * Table comp_layout_layer_sublayout
- *
- * @verbatim
- * Table Name: comp_layout_layer_sublayout
- * Key
- *    - composite_layout_id
- *    - composite_layer_index
- * Record
- *    - sublayout_id
- *
- * @endverbatim
- *
- * composite_layout_id is the layout_id for the COMPOSITE type of a layout,
- * stored as a key in the layouts table.
- *
  * @subsection Layout-DB-lspec-schema-comp_layout_layer_emap
  * Table comp_layout_layer_emap
  *
@@ -266,10 +251,8 @@
  * L13 respectively.
  *
  * In this example, for the composite layout CL1, the
- * comp_layout_layer_sublayout table stores following three entries:
- * key: (L1, 0), rec: L11
- * key: (L1, 1), rec: L12
- * key: (L1, 2), rec: L13
+ * layouts stores number of layers as 3 and the three sublayout ids viz.
+ * L11, L12 and L13.
  *
  * The comp_layout_layer_emap may have entries indicating information like the
  * following:
-- 
1.8.3.2

