From 70a95469c39ee2084387fff5d24d0f76704e5614 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Fri, 20 Dec 2013 02:11:24 -0800
Subject: [PATCH 1/2] m0mount: fixes for demo

---
 scripts/m0mount | 62 ++++++++++++++++++++++++++++++++++++++++++++-------------
 1 file changed, 48 insertions(+), 14 deletions(-)

diff --git a/scripts/m0mount b/scripts/m0mount
index 06eb439..e0362a2 100755
--- a/scripts/m0mount
+++ b/scripts/m0mount
@@ -106,16 +106,45 @@ Where:
 OPTIONS_STRING="aln:d:k:p:u:qhLs"
 
 # This example puts 3 ioservices on 3 nodes
+# SERVICES=(
+# 	#sjt02-c1 172.18.50.40@o2ib:12345:41:101
+# 	sjt02-c2 172.18.50.45@o2ib:12345:41:101
+# 	#sjt00-c1 172.18.50.161@o2ib:12345:41:101
+# )
+
+# pdsh> sudo lctl list_nids | grep o2ib
+# mmedved@sjSC-40: 172.18.49.206@o2ib
+# mmedved@sjT04-s2: 172.18.51.110@o2ib
+# mmedved@sjSC-47: 172.18.49.220@o2ib
+# mmedved@sjSC-26: 172.18.49.178@o2ib
+# mmedved@sjT04-s1: 172.18.48.190@o2ib
+# mmedved@sjSC-48: 172.18.49.222@o2ib
+# mmedved@sjT14-s1: 172.18.51.155@o2ib
+# mmedved@sjT07-s2: 172.18.50.104@o2ib
+# mmedved@sjT07-s1: 172.18.49.76@o2ib
+
 SERVICES=(
-	#sjt02-c1 172.18.50.40@o2ib:12345:41:101
-	sjt02-c2 172.18.50.45@o2ib:12345:41:101
-	#sjt00-c1 172.18.50.161@o2ib:12345:41:101
+	sjT04-s1 172.18.48.190@o2ib:12345:41:101
+	sjT04-s1 172.18.48.190@o2ib:12345:41:102
+	sjT04-s2 172.18.51.110@o2ib:12345:41:101
+	sjT07-s1 172.18.49.76@o2ib:12345:41:101
+	sjT07-s2 172.18.50.104@o2ib:12345:41:101
+	sjT14-s1 172.18.51.155@o2ib:12345:41:101
 )
 
 declare -A NODE_UUID
 NODE_UUID[sjt02-c1]=30ab1a00-8085-40d1-a557-8996e2369a7a
 NODE_UUID[sjt02-c2]=6485c5f9-fbde-4e39-8c6a-bb9d46b3bf8a
 NODE_UUID[sjt00-c1]=54b0a56a-56ba-41a8-9caf-3f3981cfdf69
+NODE_UUID[sjT04-s1]=54b0a56a-56ba-41a8-9caf-3f3981cfdf6a
+NODE_UUID[sjT04-s2]=54b0a56a-56ba-41a8-9caf-3f3981cfdf6b
+NODE_UUID[sjT07-s1]=54b0a56a-56ba-41a8-9caf-3f3981cfdf6c
+NODE_UUID[sjT07-s2]=54b0a56a-56ba-41a8-9caf-3f3981cfdf6d
+NODE_UUID[sjT14-s1]=54b0a56a-56ba-41a8-9caf-3f3981cfdf6e
+NODE_UUID[sjSC-26]=54b0a56a-56ba-41a8-9caf-3f3981cfdf6f
+NODE_UUID[sjSC-40]=54b0a56a-56ba-41a8-9caf-3f3981cfdf79
+NODE_UUID[sjSC-47]=54b0a56a-56ba-41a8-9caf-3f3981cfdf89
+NODE_UUID[sjSC-48]=54b0a56a-56ba-41a8-9caf-3f3981cfdf99
 
 # to make command output parsing predictable
 LC_MESSAGES=C
@@ -223,8 +252,8 @@ setup_local_params()
 function r_run () {
 	H=$1
 	shift
-	echo "# ssh root@$H $*" >/dev/tty
-	ssh root@$H $*
+	echo "# ssh $H sudo $*" >/dev/tty
+	ssh $H sudo $*
 }
 
 function setup_host () {
@@ -351,7 +380,7 @@ DISKS_SH_NR=$DISKS_SH_NR
 
 i=$dev_id; j=0; f=0;
 
-echo "Device:" > disks.conf
+echo "Device:" > $WORK_ARENA/disks.conf
 
 devs=\`ls $DISKS_PATTERN | grep -v part\`
 
@@ -361,12 +390,12 @@ for dev in \$devs; do
 	# and it doesn't contain any partition table (output of partprobe is empy)
 	if [[ \$? -eq 0 && -z \$partitions ]]; then
 		if [ \$j -eq 0 ]; then
-			echo "Device:" > disks\$f.conf
-			echo "   - id: 0" >> disks\$f.conf
-			echo "     filename: \$dev" >> disks\$f.conf
+			echo "Device:" > $WORK_ARENA/disks\$f.conf
+			echo "   - id: 0" >> $WORK_ARENA/disks\$f.conf
+			echo "     filename: \$dev" >> $WORK_ARENA/disks\$f.conf
 		else
-			echo "   - id: \$i" | tee -a disks.conf >> disks\$f.conf
-			echo "     filename: \$dev" | tee -a disks.conf >> disks\$f.conf
+			echo "   - id: \$i" | tee -a disks.conf >> $WORK_ARENA/disks\$f.conf
+			echo "     filename: \$dev" | tee -a disks.conf >> $WORK_ARENA/disks\$f.conf
 			i=\`expr \$i + 1\`
 		fi
 		j=\`expr \$j + 1\`
@@ -378,6 +407,8 @@ exit 0
 EOF
 
 	if [ $H != $THIS_HOST ]; then
+		r_run $H touch $DISKS_SH
+		r_run $H chmod 777 $DISKS_SH
 		l_run scp $SF $H:$DISKS_SH
 	else
 		$RUN cp $SF $DISKS_SH
@@ -387,7 +418,7 @@ EOF
 		return 1
 	fi
 
-	$RUN "(cd $WORK_ARENA && sh $DISKS_SH)"
+	$RUN "sh $DISKS_SH"
 	if [ $? -ne 0 ]; then
 		echo ERROR: Failed to get disks list on $H
 		return 1
@@ -414,6 +445,7 @@ function start_server () {
 	local DDIR=$SDIR
 	$RUN rm -rf $SDIR
 	$RUN mkdir -p $SDIR
+	$RUN chmod 777 $SDIR	# XXX SECURITY BREACH
 	local DF=$SDIR/m0d.sh
 	local DISKS_SH_FILE=$WORK_ARENA/disks$dcf_id.conf
 	local STOB_PARAMS="-T linux"
@@ -455,14 +487,16 @@ function start_server () {
 		SNAME="-s mdservice -s rmservice -s stats $SNAME"
 	fi
 
+	$RUN "touch ${SLOG}$I.log"
+	$RUN "chmod 777 ${SLOG}$I.log"
 	$RUN "echo rotated > ${SLOG}$I.log"
-	$RUN "cd $DDIR && \
+	$RUN "bash -c \"cd $DDIR && \
 M0_TRACE_IMMEDIATE_MASK=$M0_TRACE_IMMEDIATE_MASK \
 M0_TRACE_LEVEL=$M0_TRACE_LEVEL \
 M0_TRACE_PRINT_CONTEXT=$M0_TRACE_PRINT_CONTEXT \
 $M0D -r -p $STOB_PARAMS -D $DDIR/db -S $DDIR/stobs -A $DDIR/stobs \
 -w $POOL_WIDTH -G $XPT:$MDS_EP -R $XPT:$STATS_EP
--e $XPT:$EP $IOS_EPs $SNAME $XPT_SETUP" > ${SLOG}$I.log &
+-e $XPT:$EP $IOS_EPs $SNAME $XPT_SETUP\"" > ${SLOG}$I.log &
 	if [ $? -ne 0 ]; then
 		echo ERROR: Failed to start remote server on $H
 		return 1
-- 
1.8.3.2

