From 36e8f440a5c8c326c790112981bcf398803dfaf2 Mon Sep 17 00:00:00 2001
From: Anatoliy Bilenko <anatoliy_bilenko@xyratex.com>
Date: Wed, 17 Apr 2013 09:51:56 +0300
Subject: [PATCH 80/94] Loader updated.

---
 be/ut/be.c | 41 +++++++++++++++++++++++------------------
 1 file changed, 23 insertions(+), 18 deletions(-)

diff --git a/be/ut/be.c b/be/ut/be.c
index 7bf64b6..0305b0e 100644
--- a/be/ut/be.c
+++ b/be/ut/be.c
@@ -47,7 +47,7 @@
 #include "be/reg.h"
 #include "be/tx.h"
 
-#define RVM_CREATE 0
+#define RVM_CREATE 1
 
 static struct m0_be_seg         *seg;
 static struct m0_sm_group       *sm_group;
@@ -307,7 +307,7 @@ static void be_tx_fini(struct m0_be_tx *tx)
 
 M0_UNUSED static void test_be_tree_create(void)
 {
-	struct m0_be_hs_opdata  *bh_opdata;
+	//struct m0_be_hs_opdata  *bh_opdata;
 	struct bt_key_val       *bkv;
 	struct m0_be_tx         *tx;
 	struct btree	        *tree;
@@ -348,22 +348,27 @@ M0_UNUSED static void test_be_tree_create(void)
 	/*
 	 * Save handle in static part of the segment
 	 */
-	M0_ALLOC_PTR(bh_opdata);
-	M0_UT_ASSERT(bh_opdata != NULL);
-
-	M0_ALLOC_PTR(bh_opdata->bho_segid_offset);
-	M0_UT_ASSERT(bh_opdata->bho_segid_offset != NULL);
-
-	bh_opdata->bho_segid_offset->u_hi = 1;
-	bh_opdata->bho_segid_offset->u_lo = 88779911231;
-	bh_opdata->bho_index		  = 0;
-	bh_opdata->bho_seg		  = seg;
-
-	m0_be_hs_update(bh_opdata);
-	ptree = (struct btree **)m0_be_hs_get(seg, 0);
-	*ptree = tree;
-	/* XXX: For now... My task is not to write a good loader here */
-	sleep(5);
+	{
+		struct m0_be_reg *reg;
+
+		tx = be_tx_init();
+		M0_ALLOC_PTR(reg);
+		M0_ASSERT(reg != NULL);
+
+		m0_be_static_reg_capture_buf(0, seg, reg, tx);
+		ptree = (struct btree **)m0_be_static_reg_get_handle_addr(seg, 0);
+		*ptree = tree;
+
+		m0_be_tx_start(tx);
+		m0_be_tx_done(tx);
+
+		m0_sm_group_lock(sm_group);
+		m0_sm_timedwait(&tx->bt_sm, (1 << M0_BETX_DONE) | (1 << M0_BETX_FAILED),
+                        M0_TIME_NEVER);
+		m0_sm_group_unlock(sm_group);
+
+		be_tx_fini(tx);
+	}
 }
 
 M0_UNUSED static void test_be_tree_load(void)
-- 
1.8.3.2

