From bb5b0a9102264a7df6fe581d08ab4621df297b65 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Thu, 24 Jan 2013 15:13:35 +0200
Subject: [PATCH 03/79] added missing file to net/test/Makefile.sub

---
 net/test/Makefile.sub | 57 +++++++++++++++++++++++++++++----------------------
 1 file changed, 32 insertions(+), 25 deletions(-)

diff --git a/net/test/Makefile.sub b/net/test/Makefile.sub
index bb7e1a5..81e4974 100644
--- a/net/test/Makefile.sub
+++ b/net/test/Makefile.sub
@@ -1,29 +1,36 @@
 net_test_libmero_net_test_la_SOURCES += net/test/commands.c \
-                                           net/test/console.c \
-                                           net/test/network.c \
-                                           net/test/node_bulk.c \
-                                           net/test/node.c \
-                                           net/test/node_ping.c \
-                                           net/test/ringbuf.c \
-                                           net/test/serialize.c \
-                                           net/test/service.c \
-                                           net/test/slist.c \
-                                           net/test/stats.c \
-                                           net/test/str.c \
-                                           net/test/commands.h \
-                                           net/test/console.h \
-                                           net/test/network.h \
-                                           net/test/node_bulk.h \
-                                           net/test/node.h \
-                                           net/test/node_ping.h \
-                                           net/test/ringbuf.h \
-                                           net/test/serialize.h \
-                                           net/test/service.h \
-                                           net/test/slist.h \
-                                           net/test/stats.h \
-                                           net/test/str.h \
-                                           net/test/user_space/stats_u.c \
-                                           net/test/user_space/stats_u.h
+                                        net/test/console.c \
+                                        net/test/initfini.c \
+                                        net/test/network.c \
+                                        net/test/node.c \
+                                        net/test/node_bulk.c \
+                                        net/test/node_helper.c \
+                                        net/test/node_ping.c \
+                                        net/test/node_stub.c \
+                                        net/test/ringbuf.c \
+                                        net/test/serialize.c \
+                                        net/test/service.c \
+                                        net/test/slist.c \
+                                        net/test/stats.c \
+                                        net/test/str.c \
+                                        net/test/commands.h \
+                                        net/test/console.h \
+                                        net/test/debug.h \
+                                        net/test/initfini.h \
+                                        net/test/network.h \
+                                        net/test/node.h \
+                                        net/test/node_bulk.h \
+                                        net/test/node_helper.h \
+                                        net/test/node_ping.h \
+                                        net/test/node_stub.h \
+                                        net/test/ringbuf.h \
+                                        net/test/serialize.h \
+                                        net/test/service.h \
+                                        net/test/slist.h \
+                                        net/test/stats.h \
+                                        net/test/str.h \
+                                        net/test/user_space/stats_u.c \
+                                        net/test/user_space/stats_u.h
 
 EXTRA_DIST += net/test/test-user.sh.in \
               net/test/linux_kernel/node_k.c
-- 
1.8.3.2

