From 0b83d108d263a164b53d13f42b785282113d1ca7 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Thu, 24 Jan 2013 15:15:15 +0200
Subject: [PATCH 04/79] corrected M0_GETOPTS() parameters order

---
 net/test/user_space/console_u.c | 15 +++++++--------
 1 file changed, 7 insertions(+), 8 deletions(-)

diff --git a/net/test/user_space/console_u.c b/net/test/user_space/console_u.c
index 85be0d9..d8cb198 100644
--- a/net/test/user_space/console_u.c
+++ b/net/test/user_space/console_u.c
@@ -272,7 +272,13 @@ static int configure(int argc, char *argv[],
 		}))
 
 	M0_GETOPTS("m0netperf", argc, argv,
-		M0_STRINGARG('t', "Test type, {ping|bulk}",
+		ASCIIZ_ARG(addr_console4servers),
+		ASCIIZ_ARG(addr_console4clients),
+		SLIST_ARG(servers),
+		SLIST_ARG(clients),
+		SLIST_ARG(data_servers),
+		SLIST_ARG(data_clients),
+		M0_STRINGARG(opt_test_type, msg_test_type,
 		LAMBDA(void, (const char *type) {
 			if (strncmp(type, "ping", 5) == 0)
 				cfg->ntcc_test_type = M0_NET_TEST_TYPE_PING;
@@ -281,13 +287,6 @@ static int configure(int argc, char *argv[],
 			else
 				success = false;
 		})),
-		ASCIIZ_ARG(addr_console4servers),
-		ASCIIZ_ARG(addr_console4clients),
-		SLIST_ARG(servers),
-		SLIST_ARG(clients),
-		SLIST_ARG(data_servers),
-		SLIST_ARG(data_clients),
-		M0_STRINGARG(opt_test_type, msg_test_type,
 		NUMBER_ARG(msg_nr),
 		SCALED_ARG(msg_size),
 		NUMBER_ARG(bd_buf_nr_server),
-- 
1.8.3.2

