From 5b09dc4efc367d98731496d88f6f8fcaa2ec0c4c Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Thu, 31 Jan 2013 16:01:06 +0200
Subject: [PATCH 23/79] removed comment from kernel semaphore_timeddown()

---
 lib/linux_kernel/semaphore.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/lib/linux_kernel/semaphore.c b/lib/linux_kernel/semaphore.c
index 0f92b3e..7ff31d5 100644
--- a/lib/linux_kernel/semaphore.c
+++ b/lib/linux_kernel/semaphore.c
@@ -80,10 +80,6 @@ M0_INTERNAL bool m0_semaphore_timeddown(struct m0_semaphore *semaphore,
 	ts.tv_sec  = m0_time_seconds(reltime);
 	ts.tv_nsec = m0_time_nanoseconds(reltime);
 	reljiffies = timespec_to_jiffies(&ts);
-	/*
-	if (reljiffies == 0)
-		reljiffies = 1;
-	*/
 
 	return down_timeout(&semaphore->s_sem, reljiffies) == 0;
 }
-- 
1.8.3.2

