From 72cf24377009624df0e44a0734e2e1736b988f42 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Sun, 3 Feb 2013 19:37:44 +0200
Subject: [PATCH 40/79] LogD 1875, part 3: extra blank line

---
 net/test/node_ping.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/net/test/node_ping.c b/net/test/node_ping.c
index 1acff70..a94a3ab 100644
--- a/net/test/node_ping.c
+++ b/net/test/node_ping.c
@@ -480,12 +480,10 @@ static void node_ping_msg_cb(struct m0_net_test_network_ctx *net_ctx,
 	if (q == M0_NET_QT_MSG_RECV && ev->nbe_status == 0)
 		LOGD(", ev->nbe_ep->nep_addr = %s", ev->nbe_ep->nep_addr);
 
-
 	if (ev->nbe_status == -ECANCELED) {
 		m0_semaphore_up(&ctx->npc_buf_q_sem);
 		return;
 	}
-
 	/* save buffer event */
 	bs->bs_ev = *ev;
 	/* save endpoint from successfully received buffer */
-- 
1.8.3.2

