From 61bb80180785655b54ad41fb1b1520f8bc249782 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Sun, 3 Feb 2013 20:00:32 +0200
Subject: [PATCH 42/79] LogD 1880: added @verbatim to the test node CLI doc

---
 net/test/user_space/node_u.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/net/test/user_space/node_u.c b/net/test/user_space/node_u.c
index 8f4bc8d..f4bd244 100644
--- a/net/test/user_space/node_u.c
+++ b/net/test/user_space/node_u.c
@@ -34,6 +34,7 @@
 
 /**
    @page net-test-fspec-cli-node-user Test node command line pamameters
+   @verbatim
    $ net/test/m0nettestd.sh -?
    Usage: m0nettestd options...
 
@@ -44,6 +45,7 @@
 	 -v           : Verbose output
 	 -l           : List available LNET interfaces
 	 -?           : display this help and exit
+   @endverbatim
  */
 
 /**
-- 
1.8.3.2

