From 1fab8361c78e3ee1129c0d405a22af26c5edf975 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Sun, 3 Feb 2013 20:45:35 +0200
Subject: [PATCH 45/79] LogD 1883: && belongs at end of line

---
 net/test/ut/commands.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/net/test/ut/commands.c b/net/test/ut/commands.c
index 21ae9a2..3bf09eb 100644
--- a/net/test/ut/commands.c
+++ b/net/test/ut/commands.c
@@ -635,8 +635,8 @@ static void commands_console_loop(struct net_test_cmd_node *node,
 		m0_net_test_commands_send_wait_all(&node->ntcn_ctx);
 		cmd_rcvd = 0;
 		deadline = timeout_get_abs();
-		while (cmd_rcvd != ARRAY_SIZE(nodes) - 1
-		       && m0_time_now() <= deadline) {
+		while (cmd_rcvd != ARRAY_SIZE(nodes) - 1 &&
+		       m0_time_now() <= deadline) {
 			rc = m0_net_test_commands_recv(&node->ntcn_ctx, cmd,
 						       deadline);
 			M0_UT_ASSERT(rc == 0);
-- 
1.8.3.2

