From 8401483cd6ff49a8cb6ec203b4d2ec1a7602525f Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Tue, 5 Feb 2013 16:13:10 +0200
Subject: [PATCH 52/79] excluded net/test/debug.h from check-include-guards
 script

---
 scripts/check-include-guards | 1 +
 1 file changed, 1 insertion(+)

diff --git a/scripts/check-include-guards b/scripts/check-include-guards
index e3dc7af..7520c93 100755
--- a/scripts/check-include-guards
+++ b/scripts/check-include-guards
@@ -138,6 +138,7 @@ sub check_include_guards
                                 ->not( File::Find::Rule->symlink() )
                                 ->name('*.h')
                                 ->not( File::Find::Rule->name('config.h') )
+				->not( File::Find::Rule->name('debug.h') )
                                 ->in($root_dir_name);
 
     FILE:
-- 
1.8.3.2

