From 8c81f24a9561d6778f57467b61a8ab82140bb87f Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Thu, 14 Feb 2013 14:35:10 +0200
Subject: [PATCH 68/79] disabled debug messages in net-test st

---
 net/test/st/run-1x1.sh | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/net/test/st/run-1x1.sh b/net/test/st/run-1x1.sh
index 37e9e3d..e57c4e1 100644
--- a/net/test/st/run-1x1.sh
+++ b/net/test/st/run-1x1.sh
@@ -49,10 +49,10 @@ modload || exit $?
 # default, to prevent cluttering of UT output with "fake" error messages,
 # generated while testing various error paths using fault injection;
 # this can be overridden with '-e' CLI option.
-# export M0_TRACE_LEVEL='fatal+'
-
+export M0_TRACE_LEVEL='fatal+'
+# export M0_TRACE_LEVEL=debug+
 export M0_TRACE_IMMEDIATE_MASK=all
-export M0_TRACE_LEVEL=debug+
+
 
 "$NTN" -a $ADDR_CMD_SERVER -c $ADDR_CONSOLE4SERVERS &
 PID_SERVER=$!
-- 
1.8.3.2

