From 943b7615e2dc90b6edac118cf2664b034d58ab2f Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Mon, 18 Feb 2013 15:43:48 +0200
Subject: [PATCH 72/79] net-test/demo: remove log_file 0, it writes to file "0"

---
 net/test/demo/demo-init.exp | 1 -
 1 file changed, 1 deletion(-)

diff --git a/net/test/demo/demo-init.exp b/net/test/demo/demo-init.exp
index 8a478f2..e00d6c1 100755
--- a/net/test/demo/demo-init.exp
+++ b/net/test/demo/demo-init.exp
@@ -33,7 +33,6 @@ proc exec_cmd {pid sid cmd {tmout 10} {e 0}} {
 
 # initialize a host and return [pid, spawn id] of ssh
 proc init_host {addr {tmout 10}} {
-    log_file 0
     log_user 0
     if {"z$addr" == "zlocalhost"} {
 	set pid [spawn bash]
-- 
1.8.3.2

