From 1b85083a44eab2102964627b92a75067cfeaa361 Mon Sep 17 00:00:00 2001
From: Rohan Puri <rohan_puri@xyratex.com>
Date: Tue, 29 Oct 2013 15:10:31 +0530
Subject: [PATCH 08/10] Master merged

---
 lib/trace.h | 41 +++++++++++++++++++++--------------------
 1 file changed, 21 insertions(+), 20 deletions(-)

diff --git a/lib/trace.h b/lib/trace.h
index 4888be6..b3f5b8e 100644
--- a/lib/trace.h
+++ b/lib/trace.h
@@ -210,26 +210,27 @@ M0_INTERNAL int m0_trace_set_level(const char *level);
   M0_TRACE_SUBSYS(CM,         6) \
   M0_TRACE_SUBSYS(COB,        7) \
   M0_TRACE_SUBSYS(CONF,       8) \
-  M0_TRACE_SUBSYS(FOP,        9) \
-  M0_TRACE_SUBSYS(FORMATION, 10) \
-  M0_TRACE_SUBSYS(IOSERVICE, 11) \
-  M0_TRACE_SUBSYS(LAYOUT,    12) \
-  M0_TRACE_SUBSYS(LNET,      13) \
-  M0_TRACE_SUBSYS(M0D,       14) \
-  M0_TRACE_SUBSYS(M0T1FS,    15) \
-  M0_TRACE_SUBSYS(MEMORY,    16) \
-  M0_TRACE_SUBSYS(MGMT,      17) \
-  M0_TRACE_SUBSYS(NET,       18) \
-  M0_TRACE_SUBSYS(POOL,      19) \
-  M0_TRACE_SUBSYS(RM,        20) \
-  M0_TRACE_SUBSYS(RPC,       21) \
-  M0_TRACE_SUBSYS(SM,        22) \
-  M0_TRACE_SUBSYS(SNS,       23) \
-  M0_TRACE_SUBSYS(SNSCM,     24) \
-  M0_TRACE_SUBSYS(STOB,      25) \
-  M0_TRACE_SUBSYS(BTREE,     26) \
-  M0_TRACE_SUBSYS(EXTMAP,    27) \
-  M0_TRACE_SUBSYS(STATS,     28)
+  M0_TRACE_SUBSYS(DTM ,       9) \
+  M0_TRACE_SUBSYS(FOP,       10) \
+  M0_TRACE_SUBSYS(FORMATION, 11) \
+  M0_TRACE_SUBSYS(IOSERVICE, 12) \
+  M0_TRACE_SUBSYS(LAYOUT,    13) \
+  M0_TRACE_SUBSYS(LNET,      14) \
+  M0_TRACE_SUBSYS(M0D,       15) \
+  M0_TRACE_SUBSYS(M0T1FS,    16) \
+  M0_TRACE_SUBSYS(MEMORY,    17) \
+  M0_TRACE_SUBSYS(MGMT,      18) \
+  M0_TRACE_SUBSYS(NET,       19) \
+  M0_TRACE_SUBSYS(POOL,      20) \
+  M0_TRACE_SUBSYS(RM,        21) \
+  M0_TRACE_SUBSYS(RPC,       22) \
+  M0_TRACE_SUBSYS(SM,        23) \
+  M0_TRACE_SUBSYS(SNS,       24) \
+  M0_TRACE_SUBSYS(SNSCM,     25) \
+  M0_TRACE_SUBSYS(STOB,      26) \
+  M0_TRACE_SUBSYS(BTREE,     27) \
+  M0_TRACE_SUBSYS(EXTMAP,    28) \
+  M0_TRACE_SUBSYS(STATS,     29)
 
 #define M0_TRACE_SUBSYS(name, value) M0_TRACE_SUBSYS_ ## name = (1 << value),
 /** The subsystem bitmask definitions */
-- 
1.8.3.2

