From 3469f2408936c2eb9f079a76e7010e38a0a3126e Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Wed, 6 Apr 2011 08:48:17 -0600
Subject: [PATCH 041/158] - cleanups

---
 mdstore/mdstore.c | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index c9ec4ea..36f5266 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -32,8 +32,8 @@ static const struct c2_addb_loc mdstore_addb_loc = {
 };
 
 struct c2_fid C2_MD_ROOT_FID = {
-        .f_container = 0x1024ULL, 
-        .f_key       = 0x1ULL
+        .f_container = 1024ULL, 
+        .f_key       = 1ULL
 };
 
 void c2_md_store_make_oikey(struct c2_cob_oikey *oikey, 
@@ -45,14 +45,15 @@ void c2_md_store_make_oikey(struct c2_cob_oikey *oikey,
 
 void c2_md_store_make_nskey(struct c2_cob_nskey **keyh, 
                             struct c2_fid *pfid,
-                            const char *name, int nlen)
+                            const char *name, 
+                            int namelen)
 {
         struct c2_cob_nskey *key;
 
-        key = c2_alloc(sizeof(*key) + nlen);
+        key = c2_alloc(sizeof(*key) + namelen);
         key->cnk_pfid = *pfid;
-        memcpy(c2_bitstring_buf_get(&key->cnk_name), name, nlen);
-        c2_bitstring_len_set(&key->cnk_name, nlen);
+        memcpy(c2_bitstring_buf_get(&key->cnk_name), name, namelen);
+        c2_bitstring_len_set(&key->cnk_name, namelen);
         *keyh = key;
 }
 
@@ -593,13 +594,13 @@ int c2_md_store_locate(struct c2_md_store *md, struct c2_fid *fid,
    Find cob by parent and name.
 */
 int c2_md_store_lookup(struct c2_md_store *md, struct c2_cob *pcob,
-                       const char *name, int nlen, struct c2_cob *cob, 
+                       const char *name, int namelen, struct c2_cob *cob, 
                        struct c2_db_tx *tx)
 {
         struct c2_cob_nskey *nskey;
         int                  rc;
         
-        c2_md_store_make_nskey(&nskey, &pcob->co_fid, name, nlen);
+        c2_md_store_make_nskey(&nskey, &pcob->co_fid, name, namelen);
         rc = c2_cob_lookup(&md->md_dom, nskey, CA_NSKEY_FREE, &cob, tx);
 
         return rc;
-- 
1.8.3.2

