From 6f36ac910c4b82ee809f31c114955256f444976c Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Mon, 18 Apr 2011 14:06:44 -0600
Subject: [PATCH 070/158] - added one more check into test_add_name

---
 cob/ut/cob.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/cob/ut/cob.c b/cob/ut/cob.c
index db0ab7c..d537ff4 100644
--- a/cob/ut/cob.c
+++ b/cob/ut/cob.c
@@ -90,6 +90,7 @@ static void test_add_name(void)
         struct c2_fid        pfid;
         struct c2_db_tx      tx;
         char                 add_name[] = "add name";
+        char                 wrong_name[] = "wrong name";
 
         /* pfid, filename */
         pfid.f_container = 0x123;
@@ -118,6 +119,16 @@ static void test_add_name(void)
         c2_cob_put(cob);
         c2_free(key);
 
+        /* add new name to existing cob */
+        pfid.f_container = 0x123;
+        pfid.f_key = 0x456;
+        c2_cob_make_nskey(&key, &pfid, wrong_name, strlen(wrong_name));
+
+        /* lookup for wrong name, should fail. */
+        rc = c2_cob_lookup(&dom, key, 0, &cob, &tx);
+        C2_UT_ASSERT(rc != 0);
+        c2_free(key);
+
         c2_db_tx_commit(&tx);
 }
 
-- 
1.8.3.2

