From 00f1d645387cd6958b29c33d75cd33f05a940ab3 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Tue, 31 May 2011 05:43:18 -0600
Subject: [PATCH 102/158] - fixed two memory leaks

---
 mdservice/md_foms.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/mdservice/md_foms.c b/mdservice/md_foms.c
index c75b9f8..6a28834 100644
--- a/mdservice/md_foms.c
+++ b/mdservice/md_foms.c
@@ -447,6 +447,7 @@ static int c2_md_rename_fom_state(struct c2_fom *fom)
                 name_src = c2_bitstring_alloc(req->r_sname.n_name, 
                                               req->r_sname.n_len);
                 if (!name_src) {
+                        c2_cob_put(cob);
                         rc = -ENOMEM;
                         goto out;
                 }
@@ -455,6 +456,7 @@ static int c2_md_rename_fom_state(struct c2_fom *fom)
                                               req->r_tname.n_len);
                 if (!name_tgt) {
                         c2_bitstring_free(name_src);
+                        c2_cob_put(cob);
                         rc = -ENOMEM;
                         goto out;
                 }
-- 
1.8.3.2

