From 45ef47d070a27b8ae2506dab6dacb0af2b3afc0e Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Mon, 13 Jun 2011 07:59:50 -0600
Subject: [PATCH 128/158] - turn on addb WARN level.

---
 addb/addb.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/addb/addb.c b/addb/addb.c
index 5cedfff..7c24d34 100644
--- a/addb/addb.c
+++ b/addb/addb.c
@@ -47,7 +47,7 @@
 /*
  * This can be changed.
  */
-enum c2_addb_ev_level c2_addb_level_default = AEL_NOTE;
+enum c2_addb_ev_level c2_addb_level_default = AEL_WARN;
 C2_EXPORTED(c2_addb_level_default);
 
 /**
-- 
1.8.3.2

