From d1e93a1158659f232a2ad9b6ab4b453294ac0d64 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Mon, 13 Jun 2011 11:25:40 -0600
Subject: [PATCH 129/158] - added missed rename case.

---
 mdservice/md_foms.c | 58 +++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 58 insertions(+)

diff --git a/mdservice/md_foms.c b/mdservice/md_foms.c
index a1fa32b..592ef19 100644
--- a/mdservice/md_foms.c
+++ b/mdservice/md_foms.c
@@ -618,6 +618,64 @@ static int c2_md_rename_check_src(struct c2_md_store *md,
                                                               1, tx);
                                 }
                         }
+                } else if (rc == 0) {
+                        rc = c2_md_store_locate(md, tfid_src, &scob, 
+                                                C2_MD_LOCATE_STORED, tx);
+                        if (rc == -ENOENT) {
+                                /*
+                                 * No src object at all, let's create one.
+                                 */
+                                rc = c2_md_store_alloc(md, tfid_src, &scob);
+                                if (rc)
+                                        return rc;
+                                rc = c2_md_store_create(md, pfid_src,
+                                                        scob, sattr, tx);
+                                if (rc) {
+                                        c2_cob_put(scob);
+                                        return rc;
+                                }
+                                                
+                                /*
+                                 * Both object and names exist, let's do
+                                 * normal rename.
+                                 */
+                                rc = c2_md_rename(md, pfid_tgt, pfid_src, 
+                                                  tfid_tgt, tfid_src,
+                                                  tattr, sattr, tcob,
+                                                  scob, tx);
+                                if (rc) {
+                                        c2_cob_put(scob);
+                                        return rc;
+                                }
+                                                
+                                if (scan) {
+                                        /*
+                                         * Src name is gone now but we want to
+                                         * make sure that scan will not re-create
+                                         * it so let's add some zombie name that
+                                         * will be killed after scan.
+                                         */
+                                        rc = c2_md_store_link(md, pfid_src, scob,
+                                                              sattr->ca_name,
+                                                              sattr->ca_namelen,
+                                                              1, tx);
+                                }
+                        } else if (rc == 0) {
+                                /*
+                                 * Both object and names exist, let's do
+                                 * normal rename.
+                                 */
+                                rc = c2_md_rename(md, pfid_tgt, pfid_src, 
+                                                  tfid_tgt, tfid_src,
+                                                  tattr, sattr, tcob,
+                                                  scob, tx);
+                                if (rc) {
+                                        c2_cob_put(sncob);
+                                        c2_cob_put(scob);
+                                        return rc;
+                                }
+                        }
+                        c2_cob_put(sncob);
                 }
         }
         if (scob) {
-- 
1.8.3.2

