From 828f645b6d5e47ca4aac0be4964dcde8d777ae2f Mon Sep 17 00:00:00 2001
From: Andrew Perepechko <andrew_perepechko@xyratex.com>
Date: Wed, 15 Jun 2011 23:05:24 +0400
Subject: [PATCH 134/158] free keys only when they are initialised

---
 mdstore/mdstore.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 6e29169..e518c47 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -380,8 +380,8 @@ int c2_md_store_rename(struct c2_md_store       *md,
                        int                       snamelen,
                        struct c2_db_tx          *tx)
 {
-        struct c2_cob_nskey  *srckey;
-        struct c2_cob_nskey  *tgtkey;
+        struct c2_cob_nskey  *srckey = NULL;
+        struct c2_cob_nskey  *tgtkey = NULL;
         struct c2_cob        *tncob = NULL;
         time_t                now;
         int                   rc;
@@ -415,8 +415,10 @@ int c2_md_store_rename(struct c2_md_store       *md,
 
         rc = c2_cob_update_name(cob_src, srckey, tgtkey, tx);
 out:
-        c2_free(srckey);
-        c2_free(tgtkey);
+        if (srckey != NULL)
+                c2_free(srckey);
+        if (tgtkey != NULL)
+                c2_free(tgtkey);
 
         C2_ADDB_ADD(&md->md_addb, &mdstore_addb_loc, 
                     c2_addb_func_fail, "md_rename", rc);
-- 
1.8.3.2

