From 435fe8974c16adbd62d2f1ac91ab2801672cf44e Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Thu, 16 Jun 2011 10:32:35 -0600
Subject: [PATCH 142/158] - some fixes in rename and create to handle racer
 craziness nicely.

---
 mdservice/md_foms.c | 66 +++++++++++++++++++++++++++++++----------------------
 1 file changed, 39 insertions(+), 27 deletions(-)

diff --git a/mdservice/md_foms.c b/mdservice/md_foms.c
index 20f6bd7..a89e219 100644
--- a/mdservice/md_foms.c
+++ b/mdservice/md_foms.c
@@ -180,29 +180,15 @@ restart:
                 }
         } else if (rc == 0) {
                 /*
-                 * Ok, there is name, let's check versions to see what to
-                 * do with it.
+                 * Different object, this may be hardlink or result of racer
+                 * work.
                  */
-                rc = c2_md_store_locate(md, tfid, &scob, 
-                                        C2_MD_LOCATE_STORED, tx);
-                if (rc) {
-                        /*
-                         * No statdata name, this means broken db.
-                         */
-                        c2_cob_put(ncob);
-                        return rc;
-                }
-
-                /*
-                 * Different object, this must be hardlink.
-                 */
-                if (!c2_fid_eq(ncob->co_fid, scob->co_fid)) {
-                        struct c2_cob *ecob;
+                if (!c2_fid_eq(ncob->co_fid, tfid)) {
+                        struct c2_cob *ecob = NULL;
                         
                         rc = c2_md_store_locate(md, ncob->co_fid, &ecob, 
                                                 C2_MD_LOCATE_STORED, tx);
                         if (rc) {
-                                c2_cob_put(scob);
                                 c2_cob_put(ncob);
                                 return rc;
                         }
@@ -222,9 +208,19 @@ restart:
                         c2_cob_put(ecob);
                 }
 
-                if (attr->ca_version > scob->co_nsrec.cnr_version)
-                        rc = c2_md_store_setattr(md, scob, attr, tx);
-                        
+                /*
+                 * Let's see if we have stat data name and create one
+                 * if not.
+                 */
+                rc = c2_md_store_locate(md, tfid, &scob, 
+                                        C2_MD_LOCATE_STORED, tx);
+                if (rc == -ENOENT) {
+                        C2_ASSERT(attr->ca_nlink > 0);
+                        rc = c2_md_store_create(md, pfid, attr, &scob, tx);
+                } else if (rc == 0) {
+                        if (attr->ca_version > scob->co_nsrec.cnr_version)
+                                rc = c2_md_store_setattr(md, scob, attr, tx);
+                }
                 c2_cob_put(ncob);
         }
         if (scob)
@@ -512,7 +508,7 @@ static int c2_md_rename_check_src(struct c2_md_store *md,
         int rc;
         
         C2_ASSERT(tcob != NULL);
-        
+
         if (c2_fid_eq(tfid_tgt, tfid_src)) {
                 /*
                  * Same object, let's check the name.
@@ -780,12 +776,28 @@ static int c2_md_rename_fom_state(struct c2_fom *fom)
                                                     &tattr, &sattr, tcob, 
                                                     scan, tx);
         } else if (rc == 0) {
-                rc = c2_md_store_locate(md, &tfid_tgt, &tcob, 
-                                        C2_MD_LOCATE_STORED, tx);
-                if (rc == -ENOENT) {
-                        rc = c2_md_store_create(md, &pfid_tgt, &tattr, 
-                                                &tcob, tx);
+                if (c2_fid_eq(tncob->co_fid, &tfid_tgt)) {
+                        rc = c2_md_store_locate(md, &tfid_tgt, &tcob, 
+                                                C2_MD_LOCATE_STORED, tx);
+                        if (rc == -ENOENT) {
+                                /*
+                                 * Target seems to be valid just missing
+                                 * (scan did not find it yet?). Let's create
+                                 * it.
+                                 */
+                                C2_ASSERT(tattr.ca_nlink > 0);
+                                rc = c2_md_store_create(md, &pfid_tgt, &tattr, 
+                                                        &tcob, tx);
+                        }
+                } else {
+                        /*
+                         * Target was just killed but there is some name.
+                         * Let's find it stat data and use as tcob.
+                         */
+                        rc = c2_md_store_locate(md, tncob->co_fid, &tcob, 
+                                                C2_MD_LOCATE_STORED, tx);
                 }
+
                 if (rc == 0)
                         rc = c2_md_rename_check_src(md, &pfid_tgt, &pfid_src, 
                                                     &tfid_tgt, &tfid_src,
-- 
1.8.3.2

