From f9650c3ab19f58a5c576d2edda43cb458cfd178f Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Sun, 19 Jun 2011 11:12:57 -0600
Subject: [PATCH 145/158] - set addb level to notes.

---
 addb/addb.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/addb/addb.c b/addb/addb.c
index 7c24d34..5cedfff 100644
--- a/addb/addb.c
+++ b/addb/addb.c
@@ -47,7 +47,7 @@
 /*
  * This can be changed.
  */
-enum c2_addb_ev_level c2_addb_level_default = AEL_WARN;
+enum c2_addb_ev_level c2_addb_level_default = AEL_NOTE;
 C2_EXPORTED(c2_addb_level_default);
 
 /**
-- 
1.8.3.2

