From 645826b04cc6ef20701db2429aa4ebdff75c098c Mon Sep 17 00:00:00 2001
From: Alexander Zarochentsev <alexander_zarochentsev@xyratex.com>
Date: Thu, 23 Jun 2011 23:59:58 +0400
Subject: [PATCH 152/158] c2_md_store_path: zeroing pointer to freed memory.

---
 mdstore/mdstore.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 51af570..a9c6341 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -759,6 +759,7 @@ out:
                         goto restart;
                 }
                 c2_free(*path);
+                *path = NULL;
         } else {
                 c2_db_tx_commit(&tx);
         }
-- 
1.8.3.2

