From 65d4dec3998422bc632601e15c9939c1a29f284e Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Mon, 1 Jul 2013 11:48:07 +0800
Subject: [PATCH 09/14] add poolmach query/set testing into st.

---
 scripts/m0 | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/scripts/m0 b/scripts/m0
index b90f1f5..13b9692 100755
--- a/scripts/m0
+++ b/scripts/m0
@@ -74,6 +74,9 @@ run_st() {
 
     ## SNS repair
     $SUDO "$SRC/m0t1fs/linux_kernel/st/m0t1fs_sns_repair.sh"
+
+    ## poolmachine query/set testing
+    $SUDO "$SRC/m0t1fs/linux_kernel/st/m0t1fs_poolmach.sh"
 }
 
 dist_check() {
-- 
1.8.3.2

