From e4587ee0300d228481cb58b5a0de82c83149cd51 Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Wed, 3 Jul 2013 11:25:01 +0800
Subject: [PATCH 12/14] use m0_fop_opcode()

---
 pool/pool_foms.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/pool/pool_foms.c b/pool/pool_foms.c
index 865a41f..d965c62 100644
--- a/pool/pool_foms.c
+++ b/pool/pool_foms.c
@@ -48,11 +48,10 @@ static int poolmach_fom_create(struct m0_fop *fop, struct m0_fom **out,
 	if (fom == NULL)
 		return -ENOMEM;
 
-	if (fop->f_item.ri_type->rit_opcode == M0_POOLMACHINE_QUERY_OPCODE) {
+	if (m0_fop_opcode(fop) == M0_POOLMACHINE_QUERY_OPCODE) {
 		rep_fop = m0_fop_alloc(&m0_fop_poolmach_query_rep_fopt, NULL);
 		M0_LOG(M0_DEBUG, "create Query fop");
-	} else if (fop->f_item.ri_type->rit_opcode ==
-		   M0_POOLMACHINE_SET_OPCODE) {
+	} else if (m0_fop_opcode(fop) == M0_POOLMACHINE_SET_OPCODE) {
 		rep_fop = m0_fop_alloc(&m0_fop_poolmach_set_rep_fopt, NULL);
 		M0_LOG(M0_DEBUG, "create set fop");
 	} else {
@@ -88,7 +87,7 @@ static int poolmach_fom_tick(struct m0_fom *fom)
 	req_fop = fom->fo_fop;
 	rep_fop = fom->fo_rep_fop;
 
-	switch (req_fop->f_item.ri_type->rit_opcode) {
+	switch (m0_fop_opcode(req_fop)) {
 	case M0_POOLMACHINE_QUERY_OPCODE: {
 		struct m0_fop_poolmach_query     *query_fop;
 		struct m0_fop_poolmach_query_rep *query_fop_rep;
-- 
1.8.3.2

