From d03e9613ab4da136c0992176937eb10894bcb0b2 Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Tue, 9 Jul 2013 13:53:00 +0800
Subject: [PATCH 14/14] const char string

---
 pool/pool_trigger.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/pool/pool_trigger.c b/pool/pool_trigger.c
index 9022a56..5ec5161 100644
--- a/pool/pool_trigger.c
+++ b/pool/pool_trigger.c
@@ -51,7 +51,7 @@ enum {
 	MAX_SERVERS        = 1024
 };
 
-char *poolmach_state_name[] = {
+const char *poolmach_state_name[] = {
 	[M0_PNDS_ONLINE]          = "Online",
 	[M0_PNDS_FAILED]          = "Failed",
 	[M0_PNDS_OFFLINE]         = "Offline",
@@ -362,7 +362,7 @@ int main(int argc, char *argv[])
 		if (op[0] == 'Q' || op[0] == 'q') {
 			struct m0_fop_poolmach_query_rep *query_fop_rep;
 			struct m0_fop *rep;
-			char          *name;
+			const char    *name;
 			uint32_t       state;
 			rep = m0_rpc_item_to_fop(req->f_item.ri_reply);
 			query_fop_rep = m0_fop_data(rep);
-- 
1.8.3.2

