From d9b73afd8cf7ddf38c961d2471446c457c011356 Mon Sep 17 00:00:00 2001
From: Anatoliy Bilenko <anatoliy_bilenko@xyratex.com>
Date: Wed, 29 Jan 2014 15:20:51 +0200
Subject: [PATCH 3/9] seg0: dld updated.

---
 be/seg.h | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/be/seg.h b/be/seg.h
index 73e0b10..47e8a7a 100644
--- a/be/seg.h
+++ b/be/seg.h
@@ -214,6 +214,14 @@ struct m0_be_segobj_opt {
  * @note this structure is stored as a value for keys in seg0 dictionary.
  */
 struct m0_be_segobj_desc {
+	/**
+	 * After be_obj_up() is called on the desc, this pointer has to be set
+	 * up on initialized be object. Different tiers of hierarchy depend one
+	 * from other and such pointers can be needed to start up a dependent
+	 * be-object.
+	 */
+	void                    *sd_obj;
+
 	enum m0_be_segobj_type   sd_type;
 	unsinged                 sd_opts_nr;
 	struct m0_be_segobj_opt  sd_opts[];
@@ -335,7 +343,7 @@ M0_INTERNAL void m0_be_objects_down(struct m0_be_domain *dom);
  *
  * <b>Pseudocode bringing BE-object up.</b>
  * @code
- * int be_obj_up(struct m0_be_domain *dom, const struct m0_be_segobj_desc *desc)
+ * int be_obj_up(struct m0_be_domain *dom, struct m0_be_segobj_desc *desc)
  * {
  * 	enum m0_be_segobj_type    type = desc->sd_type;
  * 	struct m0_be_segobj_opt  *opts = &desc->sd_opts;
-- 
1.8.3.2

