From b9d822dffa40d97c4546a0307e915bfd9720f788 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Thu, 16 May 2013 12:13:56 +0300
Subject: [PATCH 2/4] rm: clarify documentation of m0_rm_incoming_{init,fini}()

---
 rm/rm.h | 26 +++++++++-----------------
 1 file changed, 9 insertions(+), 17 deletions(-)

diff --git a/rm/rm.h b/rm/rm.h
index 263d0ff..c4ce728 100644
--- a/rm/rm.h
+++ b/rm/rm.h
@@ -1070,8 +1070,10 @@ enum m0_rm_incoming_policy {
 };
 
 /**
- * Flags controlling incoming usage credit request processing. These flags are
- * stored in m0_rm_incoming::rin_flags and analysed in m0_rm_credit_get().
+ * Flags controlling incoming usage credit request processing.
+ *
+ * m0_rm_incoming_init() stores these flags in m0_rm_incoming::rin_flags,
+ * m0_rm_credit_get() analyses them.
  */
 enum m0_rm_incoming_flags {
 	/**
@@ -1662,16 +1664,11 @@ M0_INTERNAL int
 m0_rm_credit_copy(struct m0_rm_credit *dst, const struct m0_rm_credit *src);
 
 /**
- * Initialises the fields of incoming structure.
- * This creates an incoming request with an empty m0_rm_incoming::rin_want
- * credit.
- *
- * @param in - incoming credit request structure
- * @param owner - for which incoming request is intended.
- * @param type - incoming request type
- * @param policy - applicable policy
- * @param flags - type of request (borrow, revoke, local)
- * @see m0_rm_incoming_fini
+ * Initialises m0_rm_incoming structure.
+ *
+ * Creates an incoming request with an empty m0_rm_incoming::rin_want credit.
+ *
+ * @see m0_rm_incoming_flags for the values to compose `flags' parameter of.
  */
 M0_INTERNAL void m0_rm_incoming_init(struct m0_rm_incoming *in,
 				     struct m0_rm_owner *owner,
@@ -1679,11 +1676,6 @@ M0_INTERNAL void m0_rm_incoming_init(struct m0_rm_incoming *in,
 				     enum m0_rm_incoming_policy policy,
 				     uint64_t flags);
 
-/**
- * Finalises the fields of
- * @param in
- * @see Dual to m0_rm_incoming_init().
- */
 M0_INTERNAL void m0_rm_incoming_fini(struct m0_rm_incoming *in);
 
 /**
-- 
1.8.3.2

