From aad3f9c670ec3f8d5772baac6f09150cb5589932 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Tue, 20 Aug 2013 14:25:02 +0300
Subject: [PATCH 07/34] be/engine: temporary solution to fix stop() deadlock

---
 be/engine.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/be/engine.c b/be/engine.c
index bd1a9f4..7707b75 100644
--- a/be/engine.c
+++ b/be/engine.c
@@ -345,7 +345,13 @@ static void be_engine_group_stop_nr(struct m0_be_engine *en, size_t nr)
 	size_t i;
 
 	for (i = 0; i < nr; ++i) {
+		/*
+		 * XXX engine lock-unlock is temporary solution
+		 * to prevent deadlock.
+		 */
+		be_engine_unlock(en);
 		m0_be_tx_group_stop(&en->eng_group[i]);
+		be_engine_lock(en);
 		egr_tlist_del(&en->eng_group[i]);
 	}
 }
-- 
1.8.3.2

