From 57b1c15089b201a59541f531501e8b9c6ff2535d Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Tue, 20 Aug 2013 20:17:43 +0300
Subject: [PATCH 10/34] be/ut/main: m0_be_ut_seg_init_fini() no longer exists

---
 be/ut/main.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/be/ut/main.c b/be/ut/main.c
index f2c2bdd..df40b95 100644
--- a/be/ut/main.c
+++ b/be/ut/main.c
@@ -33,7 +33,6 @@ extern void m0_be_ut_log_store_reserve(void);
 extern void m0_be_ut_log_store_io(void);
 extern void m0_be_ut_log(void);
 
-extern void m0_be_ut_seg_init_fini(void);
 extern void m0_be_ut_seg_create_destroy(void);
 extern void m0_be_ut_seg_open_close(void);
 extern void m0_be_ut_seg_io(void);
-- 
1.8.3.2

