From 058e71401c992b2cc98ec47641971e6f058a30c5 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Wed, 21 Aug 2013 03:38:00 +0300
Subject: [PATCH 14/34] be/ut: header fix

---
 be/ut/btree.c | 1 +
 be/ut/seg.c   | 2 ++
 2 files changed, 3 insertions(+)

diff --git a/be/ut/btree.c b/be/ut/btree.c
index 663c6ab..7ce8a96 100644
--- a/be/ut/btree.c
+++ b/be/ut/btree.c
@@ -27,6 +27,7 @@
 #include "lib/memory.h"    /* M0_ALLOC_PTR */
 #include "be/ut/helper.h"
 #include "ut/ut.h"
+#include <stdlib.h>	   /* atoi */
 
 extern void btree_dbg_print(struct m0_be_btree *tree);
 
diff --git a/be/ut/seg.c b/be/ut/seg.c
index fded7e7..e8f1ce5 100644
--- a/be/ut/seg.c
+++ b/be/ut/seg.c
@@ -27,6 +27,8 @@
 #include "ut/ut.h"		/* M0_UT_ASSERT */
 #include "be/ut/helper.h"	/* m0_be_ut_seg_helper */
 
+#include <stdlib.h>		/* rand_r */
+
 enum {
 	BE_UT_SEG_SIZE	  = 0x20000,
 	BE_UT_SEG_IO_ITER = 0x400,
-- 
1.8.3.2

