From 1ebe946de0b80b06cb93afe600020ad2b9e73d5a Mon Sep 17 00:00:00 2001
From: Anatoliy Bilenko <anatoliy_bilenko@xyratex.com>
Date: Wed, 21 Aug 2013 15:31:57 +0300
Subject: [PATCH 24/34] db: finished with cursor functions

---
 db/db.c | 33 ++++++++++++++++++++++++++++-----
 1 file changed, 28 insertions(+), 5 deletions(-)

diff --git a/db/db.c b/db/db.c
index 131583a..70e9efb 100644
--- a/db/db.c
+++ b/db/db.c
@@ -386,18 +386,41 @@ M0_INTERNAL int m0_db_cursor_last(struct m0_db_cursor *cursor_,
 M0_INTERNAL int m0_db_cursor_set(struct m0_db_cursor *cursor_,
 				 struct m0_db_pair *pair)
 {
-	return -1;
+	M0_BE_OP_SYNC(op, m0_be_btree_update(pair->dp_table->t_i.i_tree,
+					     cursor_->c_tx->dt_i.dt_txn,
+					     &op,
+					     &pair->dp_key.db_i.db_dbt,
+					     &pair->dp_rec.db_i.db_dbt));
+	return 0;
 }
 
-M0_INTERNAL int m0_db_cursor_add(struct m0_db_cursor *cursor,
+M0_INTERNAL int m0_db_cursor_add(struct m0_db_cursor *cursor_,
 				 struct m0_db_pair *pair)
 {
-	return -1;
+	M0_BE_OP_SYNC(op,
+		      m0_be_btree_insert(pair->dp_table->t_i.i_tree,
+					 cursor_->c_tx->dt_i.dt_txn,
+					 &op,
+					 &pair->dp_key.db_i.db_dbt,
+					 &pair->dp_rec.db_i.db_dbt));
+
+	return m0_db_cursor_get(cursor_, pair);
 }
 
-M0_INTERNAL int m0_db_cursor_del(struct m0_db_cursor *cursor)
+M0_INTERNAL int m0_db_cursor_del(struct m0_db_cursor *cursor_)
 {
-	return -1;
+	struct m0_be_btree_cursor *cursor = cursor_->c_i.c_i;
+	struct m0_buf key;
+	struct m0_buf val;
+
+	m0_be_btree_cursor_kv_get(cursor, &key, &val);
+
+	M0_BE_OP_SYNC(op,
+		      m0_be_btree_delete(cursor->bc_tree,
+					 cursor_->c_tx->dt_i.dt_txn,
+					 &op, &key));
+
+	return m0_be_btree_cursor_get_sync(cursor_->c_i.c_i, &key, true);
 }
 
 /** @} end of db group */
-- 
1.8.3.2

