From 90ca78344e11221b75cb24a16036514b0a534849 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Wed, 21 Aug 2013 16:01:26 +0300
Subject: [PATCH 30/34] mero/setup: call m0_dtx_init() for AD stob only

---
 mero/setup.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/mero/setup.c b/mero/setup.c
index d929c76..62e5f6c 100644
--- a/mero/setup.c
+++ b/mero/setup.c
@@ -985,12 +985,12 @@ static int cs_storage_init(const char *stob_type, const char *stob_path,
 	if (rc != 0 && errno != EEXIST)
 		goto out;
 
-	m0_dtx_init(&tx);
-	rc = m0_dtx_open(&tx, db) ?:
-	    cs_linux_stob_init(stob_path, stob);
-	if (rc == 0 && strcasecmp(stob_type, m0_cs_stypes[M0_AD_STOB]) == 0)
-		rc = cs_ad_stob_init(stob, &tx, db);
-	m0_dtx_done(&tx);
+	cs_linux_stob_init(stob_path, stob);
+	if (rc == 0 && strcasecmp(stob_type, m0_cs_stypes[M0_AD_STOB]) == 0) {
+		m0_dtx_init(&tx);
+		rc = m0_dtx_open(&tx, db) ?: cs_ad_stob_init(stob, &tx, db);
+		m0_dtx_done(&tx);
+	}
 
 out:
 	m0_free(objpath);
-- 
1.8.3.2

