From 0fc7e44faf0d6cbccc1146ee6621edb845fb8a0b Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Wed, 21 Aug 2013 17:05:29 +0300
Subject: [PATCH 31/34] mero/setup: don't create m0_dtx for linux stob

---
 mero/setup.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/mero/setup.c b/mero/setup.c
index 62e5f6c..a0da064 100644
--- a/mero/setup.c
+++ b/mero/setup.c
@@ -1251,15 +1251,15 @@ static int cs_addb_storage_init(struct m0_reqh_context *rctx)
 	if (rc != 0)
 		return rc;
 
-	m0_dtx_init(&tx);
-	rc = m0_dtx_open(&tx, &rctx->rc_db);
-	if (rc != 0)
-		goto out;
 	if (strcasecmp(rctx->rc_stype, m0_cs_stypes[M0_LINUX_STOB]) == 0) {
 		rc = m0_stob_create_helper(rctx->rc_addb_stob.cas_stobs.s_ldom,
 					   &tx, &m0_addb_stob_id,
 					   &addb_stob->cas_stob);
 	} else {
+		m0_dtx_init(&tx);
+		rc = m0_dtx_open(&tx, &rctx->rc_db);
+		if (rc != 0)
+			goto out;
 		M0_ASSERT(!m0_tlist_is_empty(&astob_tl,
 					     &addb_stob->cas_stobs.s_adoms));
 		ad_stob = astob_tlist_head(&addb_stob->cas_stobs.s_adoms);
@@ -1268,9 +1268,9 @@ static int cs_addb_storage_init(struct m0_reqh_context *rctx)
 		rc = m0_stob_create_helper(ad_stob->as_dom,
 					   &tx, &m0_addb_stob_id,
 					   &addb_stob->cas_stob);
-	}
 out:
-	m0_dtx_done(&tx);
+		m0_dtx_done(&tx);
+	}
 	return rc;
 }
 
-- 
1.8.3.2

