From 8ab334fd28bb483ff376c5b6a7d5a91e18eb17cc Mon Sep 17 00:00:00 2001
From: Mandar Sawant <mandar_sawant@xyratex.com>
Date: Wed, 2 Apr 2014 12:01:47 +0530
Subject: [PATCH 4/5] conf-prealloc, - Remove extra spaces.

---
 conf/db.c     | 10 +++++-----
 xcode/xcode.c |  1 -
 2 files changed, 5 insertions(+), 6 deletions(-)

diff --git a/conf/db.c b/conf/db.c
index 21bc6a3..7242d4f 100644
--- a/conf/db.c
+++ b/conf/db.c
@@ -111,9 +111,9 @@ static int confx_obj_measure(struct m0_confx_obj *xobj)
  * releasing the memory as part of the db destruction.
  */
 struct confx_allocator {
-	void                *a_chunk;
-	m0_bcount_t          a_total;
-	m0_bcount_t          a_used;
+	void        *a_chunk;
+	m0_bcount_t  a_total;
+	m0_bcount_t  a_used;
 };
 
 /**
@@ -418,8 +418,8 @@ static int __confdb_free(struct m0_be_btree *btree, struct m0_be_seg *seg,
 M0_INTERNAL int m0_confdb_destroy(struct m0_be_seg *seg,
 				  struct m0_be_tx *tx)
 {
-	struct m0_be_btree        *btree;
-	int                        rc;
+	struct m0_be_btree *btree;
+	int                 rc;
 
 	M0_ENTRY();
 
diff --git a/xcode/xcode.c b/xcode/xcode.c
index c6669ef..8cd74b6 100644
--- a/xcode/xcode.c
+++ b/xcode/xcode.c
@@ -391,7 +391,6 @@ M0_INTERNAL void *m0_xcode_alloc(struct m0_xcode_cursor *it, size_t nob)
 	return m0_alloc(nob);
 }
 
-
 static void __xcode_free(struct m0_xcode_cursor *it)
 {
 	struct m0_xcode_cursor_frame *top = m0_xcode_cursor_top(it);
-- 
1.8.3.2

