From 920b88f13c6e0de8f76abb8c6928477a8605db9e Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Thu, 14 Feb 2013 11:22:03 +0200
Subject: [PATCH 05/10] m0t1fs/xattr: return -ENODATA when xattr not found

---
 m0t1fs/linux_kernel/dir.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/m0t1fs/linux_kernel/dir.c b/m0t1fs/linux_kernel/dir.c
index 360b4bd..c61a82d 100644
--- a/m0t1fs/linux_kernel/dir.c
+++ b/m0t1fs/linux_kernel/dir.c
@@ -225,7 +225,8 @@ ssize_t m0t1fs_getxattr(struct dentry *dentry, const char *name,
 			memcpy(buffer, rep->g_value.s_buf, rep->g_value.s_len);
 		}
 		rc = rep->g_value.s_len; /* return xattr length */
-	}
+	} else if (rc == -ENOENT)
+		rc = -ENODATA;
 out:
 	m0t1fs_fs_unlock(csb);
 	M0_LEAVE("rc: %d", rc);
-- 
1.8.3.2

