From 3961832bb387bb8bd3cf1b3b61d3f03c3f88c817 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Thu, 14 Feb 2013 11:41:00 +0200
Subject: [PATCH 06/10] mdstore/xattr: added xattr name to get/setxattr debug
 traces

---
 mdservice/md_foms.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/mdservice/md_foms.c b/mdservice/md_foms.c
index a7534fc..a9994f2 100644
--- a/mdservice/md_foms.c
+++ b/mdservice/md_foms.c
@@ -878,7 +878,8 @@ static int m0_md_tick_getxattr(struct m0_fom *fom)
         M0_ASSERT(fop_rep != NULL);
         rep = m0_fop_data(fop_rep);
 
-        M0_LOG(M0_DEBUG, "Getxattr for [%lx:%lx] started",
+        M0_LOG(M0_DEBUG, "Getxattr %.*s for [%lx:%lx] started",
+	       req->g_key.s_len, (char *)req->g_key.s_buf,
                body->b_tfid.f_container, body->b_tfid.f_key);
 
         /**
@@ -963,7 +964,9 @@ static int m0_md_tick_setxattr(struct m0_fom *fom)
         M0_ASSERT(fop_rep != NULL);
         rep = m0_fop_data(fop_rep);
 
-        M0_LOG(M0_DEBUG, "Setxattr for [%lx:%lx] started",
+        M0_LOG(M0_DEBUG, "Setxattr %.*s=%.*s for [%lx:%lx] started",
+	       req->s_key.s_len, (char *)req->s_key.s_buf,
+	       req->s_value.s_len, (char *)req->s_value.s_buf,
                body->b_tfid.f_container, body->b_tfid.f_key);
 
         /**
-- 
1.8.3.2

