From b71c6e8bf72b00de0ca43da2dba5f67a4724ec82 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Thu, 14 Feb 2013 11:42:17 +0200
Subject: [PATCH 07/10] cob: added ca_eakey/ca_eaval members to struct
 m0_cob_attr

---
 cob/cob.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/cob/cob.h b/cob/cob.h
index 80edb5e..c360a8d 100644
--- a/cob/cob.h
+++ b/cob/cob.h
@@ -337,6 +337,8 @@ struct m0_cob_attr {
 	uint64_t      ca_lid;     /**< layout id */
 	struct m0_buf ca_name;    /**< object name */
 	struct m0_buf ca_link;    /**< symlink body */
+	struct m0_buf ca_eakey;   /**< xattr name */
+	struct m0_buf ca_eaval;   /**< xattr value */
 };
 
 /**
-- 
1.8.3.2

