From cc85066ff6e9b04cd54c9600359df82ee2f53bd5 Mon Sep 17 00:00:00 2001
From: anatoliy <anatoliy@nogapok.localdomain>
Date: Sat, 11 Sep 2010 17:04:53 +0300
Subject: [PATCH] fix. wrong parameter type in function max64u()

---
 lib/arith.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/arith.h b/lib/arith.h
index f5bc233..90672b6 100644
--- a/lib/arith.h
+++ b/lib/arith.h
@@ -47,7 +47,7 @@ static inline uint64_t min64u(uint64_t a, uint64_t b)
 	return a < b ? a : b;
 }
 
-static inline uint64_t max64u(uint32_t a, uint64_t b)
+static inline uint64_t max64u(uint64_t a, uint64_t b)
 {
 	return a > b ? a : b;
 }
-- 
1.8.3.2

