From 4f9ede3ec1a39a5087348ea248c10299baa7101b Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Sun, 16 Jun 2013 12:17:42 +0800
Subject: [PATCH 09/12] use ops string.

---
 ioservice/cob_foms.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ioservice/cob_foms.c b/ioservice/cob_foms.c
index 65bbc93..63c45a2 100644
--- a/ioservice/cob_foms.c
+++ b/ioservice/cob_foms.c
@@ -282,7 +282,7 @@ static int cob_ops_fom_tick(struct m0_fom *fom)
 					  M0_FOPH_FAILURE);
 			M0_LOG(M0_ERROR, "%s cob %lu:%lu on failed device: "
 					 "state = %d rc=%d",
-					 fop_is_create? "Create": "Delete",
+					 ops,
 					 common->c_cobfid.f_container,
 					 common->c_cobfid.f_key,
 					 device_state, rc);
-- 
1.8.3.2

