From b9af9961e77006bd300eaeca6d5f94cbd7ef59ce Mon Sep 17 00:00:00 2001
From: Nikita Danilov <nikita_danilov@xyratex.com>
Date: Mon, 12 Nov 2012 16:50:09 +0400
Subject: [PATCH 12/15] Add new stob files to Makefile.

---
 stob/Makefile.sub | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/stob/Makefile.sub b/stob/Makefile.sub
index dcae74a..365440a 100644
--- a/stob/Makefile.sub
+++ b/stob/Makefile.sub
@@ -1,12 +1,17 @@
 nobase_colibri_include_HEADERS += stob/ad.h \
+                                  stob/cache.h \
                                   stob/linux_getevents.h \
                                   stob/linux.h \
                                   stob/linux_internal.h \
+                                  stob/mirror.h \
                                   stob/stob.h \
                                   stob/stob_id.h
 
-colibri_libcolibri_la_SOURCES  += stob/stob.c \
+colibri_libcolibri_la_SOURCES  += \
+                                  stob/ad.c \
+                                  stob/cache.c \
                                   stob/linux.c \
                                   stob/linux_adieu.c \
-                                  stob/ad.c \
+                                  stob/mirror.c \
+                                  stob/stob.c \
                                   stob/stob_id.c
-- 
1.8.3.2

