From 458417553d648eb0b96526c55b4b91480eec64c3 Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Fri, 27 Dec 2013 18:55:49 +0800
Subject: [PATCH 11/11] remove unused variable.

---
 sns/cm/iter.c | 1 -
 sns/cm/iter.h | 1 -
 2 files changed, 2 deletions(-)

diff --git a/sns/cm/iter.c b/sns/cm/iter.c
index d8affa5..1acb9e3 100644
--- a/sns/cm/iter.c
+++ b/sns/cm/iter.c
@@ -898,7 +898,6 @@ M0_INTERNAL int m0_sns_cm_iter_init(struct m0_sns_cm_iter *it)
 	M0_PRE(it != NULL);
 
 	cm = &scm->sc_base;
-	it->si_dbenv = cm->cm_service.rs_reqh->rh_dbenv;
 	it->si_beseg = cm->cm_service.rs_reqh->rh_beseg;
         rc = m0_ios_cdom_get(cm->cm_service.rs_reqh, &it->si_cob_dom);
         if (rc != 0) {
diff --git a/sns/cm/iter.h b/sns/cm/iter.h
index ad0ee78..940fd6d 100644
--- a/sns/cm/iter.h
+++ b/sns/cm/iter.h
@@ -116,7 +116,6 @@ struct m0_sns_cm_iter {
 	/** Iterator state machine. */
 	struct m0_sm                     si_sm;
 
-	struct m0_dbenv                 *si_dbenv;
 	struct m0_be_seg                *si_beseg;
 
 	struct m0_cob_domain            *si_cob_dom;
-- 
1.8.3.2

