From 7faf8cc4717ed3a1f9dd125b48a06c59938af001 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Fri, 12 Jul 2013 13:20:47 +0300
Subject: [PATCH 02/14] lnet/ut: fix "unused variable" (NLX_DEBUG mode)

---
 net/lnet/ut/lnet_ut.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/net/lnet/ut/lnet_ut.c b/net/lnet/ut/lnet_ut.c
index fe82b3b..5657747 100644
--- a/net/lnet/ut/lnet_ut.c
+++ b/net/lnet/ut/lnet_ut.c
@@ -382,9 +382,9 @@ static void ut_test_framework_dom_cleanup(struct ut_data *td,
 #ifdef NLX_DEBUG
 static void ut_describe_buf(const struct m0_net_buffer *nb)
 {
+#ifdef __KERNEL__
 	struct nlx_xo_buffer       *bp = nb->nb_xprt_private;
 	struct nlx_core_buffer  *lcbuf = &bp->xb_core;
-#ifdef __KERNEL__
 	struct nlx_kcore_buffer   *kcb = lcbuf->cb_kpvt;
 
 	NLXP("\txo:%p lcbuf:%p kcb:%p\n",
@@ -394,9 +394,9 @@ static void ut_describe_buf(const struct m0_net_buffer *nb)
 
 static void ut_describe_tm(const struct m0_net_transfer_mc *tm)
 {
+#ifdef __KERNEL__
 	struct nlx_xo_transfer_mc      *tp = tm->ntm_xprt_private;
 	struct nlx_core_transfer_mc  *lctm = &tp->xtm_core;
-#ifdef __KERNEL__
 	struct nlx_kcore_transfer_mc *kctm = lctm->ctm_kpvt;
 
 	NLXP("\txo:%p lctm:%p kctm:%p\n",
-- 
1.8.3.2

