From 9a47efbe19fee49be747191458e4619ef4183c33 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Fri, 12 Jul 2013 17:00:56 +0300
Subject: [PATCH 05/14] lnet: get rid of dead nlx_tm_stats_report()

---
 net/lnet/lnet_tm.c | 21 +--------------------
 1 file changed, 1 insertion(+), 20 deletions(-)

diff --git a/net/lnet/lnet_tm.c b/net/lnet/lnet_tm.c
index d763fa7..6486bdb 100644
--- a/net/lnet/lnet_tm.c
+++ b/net/lnet/lnet_tm.c
@@ -31,24 +31,6 @@ static inline bool all_tm_queues_are_empty(const struct m0_net_transfer_mc *tm)
 }
 
 /**
-   Logs periodic statistical data using ADDB.
-   @param tm the transfer machine to report about
- */
-static void nlx_tm_stats_report(struct m0_net_transfer_mc *tm)
-{
-#if 0
-	/* Disabled hacky posting of network statistics from
-	 * within the transport.
-	 * On server side replaced by higher level scheduling
-	 * of ADDB statistical posting.
-	 * Code retained here for reference.
-	 */
-	M0_PRE(tm != NULL && nlx_tm_invariant(tm));
-	m0_net__tm_stats_post_addb(tm);
-#endif
-}
-
-/**
    Cancel buffer operations if they have timed out.
    @param tm The transfer machine concerned.
    @param now The current time.
@@ -224,7 +206,6 @@ static void nlx_tm_ev_worker(struct m0_net_transfer_mc *tm)
 		next_stat_time = m0_time_add(last_stat_time,
 					     tp->xtm_stat_interval);
 		if (now >= next_stat_time) {
-			nlx_tm_stats_report(tm);
 			last_stat_time = now;
 			next_stat_time = m0_time_add(last_stat_time,
 						     tp->xtm_stat_interval);
@@ -239,11 +220,11 @@ static void nlx_tm_ev_worker(struct m0_net_transfer_mc *tm)
 		/* termination processing */
 		if (tm->ntm_state == M0_NET_TM_STOPPING) {
 			bool must_stop = false;
+
 			m0_mutex_lock(&tm->ntm_mutex);
 			if (all_tm_queues_are_empty(tm) &&
 			    tm->ntm_callback_counter == 0) {
 				nlx_core_tm_stop(cd, ctp);
-				nlx_tm_stats_report(tm);
 				must_stop = true;
 			}
 			m0_mutex_unlock(&tm->ntm_mutex);
-- 
1.8.3.2

