From d50a0983d1a2fd9efc578c5bcf294e74cee217d5 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Thu, 28 Nov 2013 00:13:33 +0200
Subject: [PATCH 08/13] scripts/be_allocator: print min/max/avg/etc.

---
 scripts/be_allocator.stp | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/scripts/be_allocator.stp b/scripts/be_allocator.stp
index 266df98..90990a1 100644
--- a/scripts/be_allocator.stp
+++ b/scripts/be_allocator.stp
@@ -22,7 +22,13 @@ function print_allocator_stats()
 {
 	printf("%s\n", tz_ctime(gettimeofday_s()))
 	foreach (ftime in func_time_stats) {
-		printf("timing histogram for %s, ns\n", ftime)
+		printf("%s(): min/max/avg/count/sum %d/%d/%d/%d/%d ns\n",
+		       ftime,
+		       @min(func_time_stats[ftime]),
+		       @max(func_time_stats[ftime]),
+		       @avg(func_time_stats[ftime]),
+		       @count(func_time_stats[ftime]),
+		       @sum(func_time_stats[ftime]));
 		print(@hist_log(func_time_stats[ftime]))
 	}
 }
-- 
1.8.3.2

