From f0d6c55f877873d29d51535fb19b398ad9fae5cb Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Thu, 28 Nov 2013 07:35:59 -0800
Subject: [PATCH 12/13] scripts/m0mount.sh: removed unused services

---
 scripts/m0mount.sh | 2 --
 1 file changed, 2 deletions(-)

diff --git a/scripts/m0mount.sh b/scripts/m0mount.sh
index f389546..4d1abdc 100755
--- a/scripts/m0mount.sh
+++ b/scripts/m0mount.sh
@@ -111,8 +111,6 @@ OPTIONS_STRING="aln:d:k:p:u:qhLs"
 # This example puts 3 ioservices on 3 nodes
 SERVICES=(
 	sjt02-c1 172.18.50.40@o2ib:12345:41:101
-	sjt02-c2 172.18.50.45@o2ib:12345:41:101
-	sjt00-c1 172.18.50.161@o2ib:12345:41:101
 )
 
 declare -A NODE_UUID
-- 
1.8.3.2

