From 22e8075d7780c66da8d0c079bb0904d8acee3b7a Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Thu, 14 Mar 2013 13:12:13 +0200
Subject: [PATCH 2/4] mds/rename: add tracing and TODO item for descendency
 check

---
 mdservice/md_foms.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/mdservice/md_foms.c b/mdservice/md_foms.c
index b75046d..1e59ffc 100644
--- a/mdservice/md_foms.c
+++ b/mdservice/md_foms.c
@@ -435,6 +435,17 @@ static int m0_md_tick_rename(struct m0_fom *fom)
 	if (rc != 0)
 		goto out;
 
+	/* After m0_md_fop_init we have r_spath and r_tpath, so we can
+	 * check that target path isn't descendant of source fid. */
+	M0_LOG(M0_DEBUG, "validating rename %.*s/%.*s to %.*s/%.*s",
+	       req->r_spath.s_len, req->r_spath.s_buf,
+	       req->r_sname.s_len, req->r_sname.s_buf,
+	       req->r_tpath.s_len, req->r_tpath.s_buf,
+	       req->r_tname.s_len, req->r_tname.s_buf
+	      );
+	/* TODO: perform the check and return -EINVAL if tpath is
+	 * descendant of "spath/sname". */
+
 	body = &req->r_body;
 
 	m0_md_cob_wire2mem(&attr, body);
-- 
1.8.3.2

