From d16cb346f5d83aee14389f8950fea9e2586d2204 Mon Sep 17 00:00:00 2001
From: Rajanikant Chirmade <rajanikant_chirmade@xyratex.com>
Date: Tue, 10 Apr 2012 16:26:30 +0530
Subject: [PATCH 05/11] - Some style related fixes.

---
 addb/addb.h         |  4 ++--
 ioservice/io_foms.c | 40 ++++++++++++++++++++++------------------
 2 files changed, 24 insertions(+), 20 deletions(-)

diff --git a/addb/addb.h b/addb/addb.h
index b7a4839..5740356 100644
--- a/addb/addb.h
+++ b/addb/addb.h
@@ -103,8 +103,8 @@ int c2_addb_net_add(struct c2_addb_dp *dp, struct c2_net_conn *);
 int c2_addb_choose_store_media(enum c2_addb_rec_store_type type, ...);
 
 /**
- * Adds variable addb messages.
- * ADDB message is fist formatted into the pre-allocated buffer according
+ * Adds a formatted message to addb stream.
+ * ADDB message is first formatted into the pre-allocated buffer according
  * to the custom defined format. Then if that succeeds, the formatted string
  * is added into addb as a trace message. Caller is supposed to allocate
  * msgbuf big enough to accommodate the formatted message. The output will
diff --git a/ioservice/io_foms.c b/ioservice/io_foms.c
index 055e6fa..edbd72b 100644
--- a/ioservice/io_foms.c
+++ b/ioservice/io_foms.c
@@ -565,7 +565,9 @@
 #define IOSERVICE_NAME "ioservice"
 
 enum {
-	IO_FOM_ADDB_MSG_SIZE = 256,
+	IO_FOM_ADDB_MSG_SMALL_SIZE  = 64,
+	IO_FOM_ADDB_MSG_MEDIUM_SIZE = 128,
+	IO_FOM_ADDB_MSG_LARGE_SIZE  = 256,
 };
 
 C2_TL_DESCR_DEFINE(stobio, "STOB I/O", static, struct c2_stob_io_desc,
@@ -932,15 +934,17 @@ static int  io_fom_cob_rw_indexvec_wire2mem(struct c2_fom         *fom,
 
 static int io_fom_cob_rw_indexvec_size(struct c2_fop *fop)
 {
-	uint32_t                 i,j;
+	uint32_t                 i;
+	uint32_t		 j;
 	c2_bcount_t              size = 0;
 	struct c2_fop_cob_rw    *rwfop;
 
 	C2_PRE(fop != NULL);
 
 	rwfop = io_rw_get(fop);
-        for (i = 0; i < rwfop->crw_desc.id_nr; i++) {
-		struct c2_io_indexvec *wire_ivec = &rwfop->crw_ivecs.cis_ivecs[i];
+	for (i = 0; i < rwfop->crw_desc.id_nr; i++) {
+		struct c2_io_indexvec *wire_ivec =
+		&rwfop->crw_ivecs.cis_ivecs[i];
 		for (j = 0; j < wire_ivec->ci_nr; j++) {
 			size += wire_ivec->ci_iosegs[j].ci_count;
 		}
@@ -1041,7 +1045,7 @@ int c2_io_fom_cob_rw_create(struct c2_fop *fop, struct c2_fom **out)
         struct c2_fop_cob_rw    *rwfop;
         const char              *fom_type_string;
         c2_bcount_t              size;
-	char                     addb_msg[IO_FOM_ADDB_MSG_SIZE];
+	char                     addb_msg[IO_FOM_ADDB_MSG_MEDIUM_SIZE];
 
         C2_PRE(fop != NULL);
         C2_PRE(c2_is_io_fop(fop));
@@ -1091,7 +1095,7 @@ int c2_io_fom_cob_rw_create(struct c2_fop *fop, struct c2_fom **out)
 	fom_type_string = c2_is_read_fop(fop) ? "I/O READ" : "I/O WRITE";
         size = io_fom_cob_rw_indexvec_size(fop);
 	c2_addb_var_msg_add(&fom->fo_fop->f_addb, &io_fom_addb_loc,
-			    addb_msg, IO_FOM_ADDB_MSG_SIZE,
+			    addb_msg, IO_FOM_ADDB_MSG_MEDIUM_SIZE,
 			    "FOM: created/type=%s, fid=%lu:%lu, ndesc=%d, "
 			    "size=%d", fom_type_string,
 			    rwfop->crw_fid.f_seq, rwfop->crw_fid.f_oid,
@@ -1122,7 +1126,7 @@ static int io_fom_cob_rw_acquire_net_buffer(struct c2_fom *fom)
         struct c2_fop                *fop;
         struct c2_io_fom_cob_rw      *fom_obj = NULL;
         struct c2_net_transfer_mc     tm;
-	char                          addb_msg[IO_FOM_ADDB_MSG_SIZE];
+	char                          addb_msg[IO_FOM_ADDB_MSG_SMALL_SIZE];
 
         C2_PRE(fom != NULL);
         C2_PRE(c2_is_io_fop(fom->fo_fop));
@@ -1223,7 +1227,7 @@ static int io_fom_cob_rw_acquire_net_buffer(struct c2_fom *fom)
 
         fom_obj->fcrw_batch_size = acquired_net_bufs;
 	c2_addb_var_msg_add(&fom->fo_fop->f_addb, &io_fom_addb_loc,
-			    addb_msg, IO_FOM_ADDB_MSG_SIZE,
+			    addb_msg, IO_FOM_ADDB_MSG_SMALL_SIZE,
 			    "FOM: net_buffs_acquire/batch_size=%d\n",
 			    fom_obj->fcrw_batch_size);
         return FSO_AGAIN;
@@ -1249,7 +1253,7 @@ static int io_fom_cob_rw_release_net_buffer(struct c2_fom *fom)
         struct c2_fop                  *fop;
         struct c2_io_fom_cob_rw        *fom_obj = NULL;
         struct c2_net_transfer_mc       tm;
-	char                            addb_msg[IO_FOM_ADDB_MSG_SIZE];
+	char                            addb_msg[IO_FOM_ADDB_MSG_SMALL_SIZE];
 
 
         C2_PRE(fom != NULL);
@@ -1283,7 +1287,7 @@ static int io_fom_cob_rw_release_net_buffer(struct c2_fom *fom)
 
         fom_obj->fcrw_batch_size = acquired_net_bufs;
 	c2_addb_var_msg_add(&fom->fo_fop->f_addb, &io_fom_addb_loc,
-			    addb_msg, IO_FOM_ADDB_MSG_SIZE,
+			    addb_msg, IO_FOM_ADDB_MSG_SMALL_SIZE,
 			    "FOM: net_buffs_release/batch_size=%d\n",
 			    fom_obj->fcrw_batch_size);
 
@@ -1316,7 +1320,7 @@ static int io_fom_cob_rw_initiate_zero_copy(struct c2_fom *fom)
         struct c2_net_buf_desc    *net_desc;
         struct c2_net_domain      *dom;
         uint32_t                   buffers_added = 0;
-	char                       addb_msg[IO_FOM_ADDB_MSG_SIZE];
+	char                       addb_msg[IO_FOM_ADDB_MSG_SMALL_SIZE];
 
         C2_PRE(fom != NULL);
         C2_PRE(c2_is_io_fop(fom->fo_fop));
@@ -1393,7 +1397,7 @@ static int io_fom_cob_rw_initiate_zero_copy(struct c2_fom *fom)
         }
 
 	c2_addb_var_msg_add(&fom->fo_fop->f_addb, &io_fom_addb_loc,
-			    addb_msg, IO_FOM_ADDB_MSG_SIZE,
+			    addb_msg, IO_FOM_ADDB_MSG_SMALL_SIZE,
 			    "FOM: zero_copy_initiated/batch_size=%d\n",
 			    fom_obj->fcrw_batch_size);
         return FSO_WAIT;
@@ -1472,7 +1476,7 @@ static int io_fom_cob_rw_io_launch(struct c2_fom *fom)
 	struct c2_io_indexvec    wire_ivec;
 	struct c2_stob_domain	*fom_stdom;
 	struct c2_fop_file_fid	*ffid;
-	char                     addb_msg[IO_FOM_ADDB_MSG_SIZE];
+	char                     addb_msg[IO_FOM_ADDB_MSG_SMALL_SIZE];
 
 	C2_PRE(fom != NULL);
         C2_PRE(c2_is_io_fop(fom->fo_fop));
@@ -1632,7 +1636,7 @@ static int io_fom_cob_rw_io_launch(struct c2_fom *fom)
                 c2_mutex_unlock(&fom_obj->fcrw_stio_mutex);
 
 		c2_addb_var_msg_add(&fom->fo_fop->f_addb, &io_fom_addb_loc,
-				    addb_msg, IO_FOM_ADDB_MSG_SIZE,
+				    addb_msg, IO_FOM_ADDB_MSG_SMALL_SIZE,
 				    "FOM: stob_io_launched/n_iocount=%d\n",
 				    fom_obj->fcrw_num_stobio_launched);
 
@@ -1729,7 +1733,7 @@ static int c2_io_fom_cob_rw_state(struct c2_fom *fom)
         int                                      rc = 0;
         struct c2_io_fom_cob_rw                 *fom_obj;
         struct c2_io_fom_cob_rw_state_transition st;
-	char                                     addb_msg[IO_FOM_ADDB_MSG_SIZE];
+	char                                     addb_msg[IO_FOM_ADDB_MSG_SMALL_SIZE];
 
         C2_PRE(fom != NULL);
         C2_PRE(c2_is_io_fop(fom->fo_fop));
@@ -1760,7 +1764,7 @@ static int c2_io_fom_cob_rw_state(struct c2_fom *fom)
                 rwrep->rwr_rc = fom->fo_rc;
                 rwrep->rwr_count = fom_obj->fcrw_bytes_transfered;
 		c2_addb_var_msg_add(&fom->fo_fop->f_addb, &io_fom_addb_loc,
-				    addb_msg, IO_FOM_ADDB_MSG_SIZE,
+				    addb_msg, IO_FOM_ADDB_MSG_SMALL_SIZE,
 				    "FOM: set reply/io_count=%lu, rc=%d\n",
 				    rwrep->rwr_count, rwrep->rwr_rc);
                 return rc;
@@ -1790,7 +1794,7 @@ static void c2_io_fom_cob_rw_fini(struct c2_fom *fom)
         struct c2_net_buffer      *nb = NULL;
         struct c2_stob_io_desc    *stio_desc = NULL;
         struct c2_net_transfer_mc  tm;
-	char                       addb_msg[IO_FOM_ADDB_MSG_SIZE];
+	char                       addb_msg[IO_FOM_ADDB_MSG_SMALL_SIZE];
 
         C2_PRE(fom != NULL);
 
@@ -1837,7 +1841,7 @@ static void c2_io_fom_cob_rw_fini(struct c2_fom *fom)
         c2_mutex_fini(&fom_obj->fcrw_stio_mutex);
 
 	c2_addb_var_msg_add(&fom->fo_fop->f_addb, &io_fom_addb_loc,
-			    addb_msg, IO_FOM_ADDB_MSG_SIZE,
+			    addb_msg, IO_FOM_ADDB_MSG_SMALL_SIZE,
 			    "FOM: finished/size=%lu, rc=%d",
 			    fom_obj->fcrw_bytes_transfered, fom->fo_rc);
         c2_fom_fini(fom);
-- 
1.8.3.2

