From fc4cfa88600d4b1fded1fbb3bfded3b80f160e48 Mon Sep 17 00:00:00 2001
From: Anatoliy Bilenko <anatoliy_bilenko@xyratex.com>
Date: Tue, 1 Apr 2014 15:05:44 -0700
Subject: [PATCH 08/11] seg0: minor changes.

---
 be/domain.c | 3 +--
 be/seg0.c   | 3 ++-
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/be/domain.c b/be/domain.c
index b762775..04c75bd 100644
--- a/be/domain.c
+++ b/be/domain.c
@@ -252,8 +252,7 @@ struct m0_be_seg *m0_be_domain_seg0_get(const struct m0_be_domain *dom)
 
 M0_INTERNAL bool m0_be_domain_is_locked(const struct m0_be_domain *dom)
 {
-	/* XXX: return m0_mutex_is_locked(&dom->bd_engine.eng_lock); */
-	return true;
+	return m0_mutex_is_locked(&dom->bd_engine.eng_lock);
 }
 
 #undef M0_TRACE_SUBSYSTEM
diff --git a/be/seg0.c b/be/seg0.c
index bc72e78..9f5d914 100644
--- a/be/seg0.c
+++ b/be/seg0.c
@@ -36,7 +36,8 @@ static bool be_0type_invarinat(const struct m0_be_0type *zt, bool registered)
 
 static bool dom_is_locked(const struct m0_be_domain *dom)
 {
-	return m0_be_domain_is_locked(dom);
+	return true;
+	/* XXX: return m0_be_domain_is_locked(dom); */
 }
 
 static struct m0_be_seg *seg0_get(const struct m0_be_domain *dom)
-- 
1.8.3.2

