From 5400366f2450cfb5eb1d2813712e4b484173f823 Mon Sep 17 00:00:00 2001
From: Anatoliy Bilenko <anatoliy_bilenko@xyratex.com>
Date: Wed, 2 Apr 2014 11:24:02 -0700
Subject: [PATCH 10/11] seg0: fake_mkfs() updated.

---
 be/ut/seg0.c | 15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)

diff --git a/be/ut/seg0.c b/be/ut/seg0.c
index 6910ddf..f536383 100644
--- a/be/ut/seg0.c
+++ b/be/ut/seg0.c
@@ -22,6 +22,7 @@
 #include "lib/types.h"		/* m0_uint128_eq */
 #include "lib/misc.h"		/* M0_BITS */
 #include "lib/memory.h"         /* M0_ALLOC_PTR, m0_free */
+#include "lib/errno.h"          /* program_invocation_name */
 #include "be/ut/helper.h"	/* m0_be_ut_backend */
 #include "be/seg.h"
 #include "be/seg0.h"
@@ -29,21 +30,23 @@
 #include "stob/linux.h"
 #include "ut/ut.h"
 
-#include <unistd.h>
-#include <stdlib.h>
+#include <unistd.h>		/* chdir, get_current_dir_name */
+#include <stdlib.h>		/* system */
 
 void fake_mkfs(void)
 {
+	extern char *program_invocation_name;
 	char *ut_dir;
-	int   rc;
+	char cmd[512] = {};
+	int rc;
 
 	ut_dir = get_current_dir_name();
-
 	rc = chdir("..");
 	M0_UT_ASSERT(rc == 0);
 
-	rc = system("./utils/.libs/lt-ut -t be-ut:fake_mkfs -k"
-		    " > /dev/null 2>&1");
+	snprintf(cmd, ARRAY_SIZE(cmd), "%s -t be-ut:fake_mkfs -k > "
+		 "/dev/null 2>&1", program_invocation_name);
+	rc = system(cmd);
 	M0_UT_ASSERT(rc == 0);
 
 	rc = chdir(ut_dir);
-- 
1.8.3.2

