From 66862138c7cf177891a2364db62ada34f0ecc7da Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Thu, 1 Mar 2012 19:55:14 +0200
Subject: [PATCH 029/311] add confc state diagram

+ add forgotten C2_EXPORTED()
---
 conf/confc.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/conf/confc.c b/conf/confc.c
index f91b7f1..21b7f8e 100644
--- a/conf/confc.c
+++ b/conf/confc.c
@@ -167,6 +167,29 @@
  *
  * @subsection confc-dld-lspec-state State Specification
  *
+ * @dot
+ * digraph confd_states {
+ * 	node [fontsize=11];
+ * 	edge [fontsize=11];
+ * 	INITIAL  [style=filled, fillcolor=lightgrey];
+ * 	TERMINAL [style=filled, fillcolor=lightgrey];
+ * 	FAILURE  [style=filled, fillcolor=lightgrey];
+ *
+ *   	INITIAL -> CHECK;
+ *   	CHECK -> TERMINAL [label="all \"needed\" obj-s\nare cached"];
+ *   	CHECK -> FAILURE [label="error"];
+ *   	CHECK -> WAIT_REPLY [label=
+ *   "at least one\nof the \"needed\" obj-s\nis MISSING\n{ set status(es) to LOADING;\nsend request }"];
+ *   	WAIT_REPLY -> FAILURE [label="timeout\nor\nrc != 0"];
+ * 	WAIT_REPLY -> GROW_CACHE [label="response received,\nrc == 0"];
+ * 	GROW_CACHE -> FAILURE [label="error"];
+ * 	GROW_CACHE -> CHECK [label="done"];
+ *   	CHECK -> WAIT_STATUS [label=
+ *   "no MISSING obj-s, but\nat least one is LOADING\n{ clink_add(&obj->co_chan) }"];
+ *   	WAIT_STATUS -> CHECK [label="\"status updated\" event"];
+ * }
+ * @enddot
+ *
  * XXX
  */
 
@@ -363,6 +386,7 @@ int c2_confc_init(const char *profile, const char *confd_addr /*XXX*/,
 	XXX;
 	/* XXX TODO: Consume `confd_addr'. */
 }
+C2_EXPORTED(c2_confc_init);
 
 void c2_confc_fini(void)
 {
@@ -389,6 +413,7 @@ void c2_confc_fini(void)
 	 */
 	XXX;
 }
+C2_EXPORTED(c2_confc_fini);
 
 int c2_confc_open(struct c2_confc_ctx *ctx, struct c2_conf_obj *path_origin,
 		  const struct c2_conf_pathcomp *path, size_t path_len)
-- 
1.8.3.2

