From 52c441930433fc15704539ae9bc23aaa3c2b304d Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Fri, 9 Mar 2012 21:57:31 +0200
Subject: [PATCH 049/311] lspec-state: describe states and transitions

---
 conf/confc.c | 31 ++++++++++++++++++++++++++++++-
 1 file changed, 30 insertions(+), 1 deletion(-)

diff --git a/conf/confc.c b/conf/confc.c
index 3a0d2b9..3cc4ee7 100644
--- a/conf/confc.c
+++ b/conf/confc.c
@@ -40,6 +40,7 @@
  * - @ref confc-dld-lspec
  *   - @ref confc-dld-lspec-comps
  *   - @ref confc-dld-lspec-state
+ *   - @ref confc-dld-lspec-thread
  * - XXX
  *
  * <hr> <!------------------------------------------------------------->
@@ -152,7 +153,7 @@
  *
  * Besides the DAG of configuration objects, confc keeps a @b registry
  * of cached objects --- an in-memory database that maps object
- * identities (identity of an object is its key and type) to memory
+ * @e identities (identity of an object is its key and type) to memory
  * addresses of these objects.  When new configuration data arrives
  * from the confd, confc consults the registry prior to adding new
  * objects to the cache.
@@ -182,7 +183,35 @@
  * }
  * @enddot
  *
+ * - When S_CHECK state is entered, check_st_in() is invoked.  This
+ *   function calls path_walk() and, depending on the result of this
+ *   call, moves the state machine to S_TERMINAL, S_WAIT_REPLY,
+ *   S_WAIT_STATUS, or S_FAILURE state.
+ *
+ * - When S_WAIT_REPLY is entered, wait_reply_st_in() is invoked.
+ *   All this function does is c2_rpc_post()ing c2_confc_ctx::fc_fop.
+ *
+ * - A state machine in S_WAIT_STATUS state waits for
+ *   c2_confc_ctx::fc_clink to be signaled.  on_unpinned() call-back,
+ *   associated with the clink, posts an AST that will move the
+ *   state machine to S_CHECK state.
+ *
+ * - A state machine in S_WAIT_REPLY state waits for a response from
+ *   confd to arrive.  Upon receiving a response, on_replied() is
+ *   called. This function moves the state machine to S_GROW_CACHE
+ *   state if c2_rpc_item::ri_error is zero, or to S_FAILURE if the
+ *   value is non-zero.
+ *
+ * - When S_GROW_CACHE is entered, grow_cache_st_in() is invoked.
+ *   This function updates the configuration cache using received
+ *   configuration data, which is contained in c2_conf_fetch_resp.
+ *   grow_cache_st_in() moves the state machine to S_CHECK state upon
+ *   success and to S_FAILURE in case of error.
+ *
+ * @subsection confc-dld-lspec-thread Threading and Concurrency Model
+ *
  * XXX
+ *
  */
 
 static struct {
-- 
1.8.3.2

