From 4857ea3d9d4fdbe3cf46d89034df7ceda53758f7 Mon Sep 17 00:00:00 2001
From: Anatoliy <anatoliy.bilenko@xyratex.com>
Date: Fri, 16 Mar 2012 10:55:54 +0200
Subject: [PATCH 068/311] Branched from confc-user

---
 conf/ut/README | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/conf/ut/README b/conf/ut/README
index 3bcdca3..b041d5f 100644
--- a/conf/ut/README
+++ b/conf/ut/README
@@ -8,3 +8,7 @@ pointless in such circumstances.
 I will update conf/ut/* when conf/* are more stable.
 
  --vvv, 2012-02-07
+
+
+So, confd uts live here.
+ --abilenko Fri Mar 16 10:54:04 EET 2012
-- 
1.8.3.2

