From f9b698d6e1a06baa935c98c12601d2611da24f38 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Wed, 21 Mar 2012 22:25:00 +0200
Subject: [PATCH 084/311] enumerate unit tests

---
 conf/confc.c | 37 ++++++++++++++++++++++++++++++++++++-
 1 file changed, 36 insertions(+), 1 deletion(-)

diff --git a/conf/confc.c b/conf/confc.c
index d87ecfb..2f7741e 100644
--- a/conf/confc.c
+++ b/conf/confc.c
@@ -44,6 +44,8 @@
  *   - @ref confc-dld-lspec-numa
  * - @ref confc-dld-conformance
  * - @ref confc-dld-ut
+ *   - @ref confc-dld-ut-infra
+ *   - @ref confc-dld-ut-confc
  * - @ref confc-dld-st
  * - @ref confc-dld-O
  * - @ref confc-dld-ref
@@ -280,7 +282,40 @@
  * <hr> <!------------------------------------------------------------->
  * @section confc-dld-ut Unit Tests
  *
- * XXX
+ * Fault Injection mechanism (lib/finject.h) will be used to test
+ * handling of "rare" errors (e.g., allocation errors) and to disable
+ * some of external modules' functionality (e.g., to make
+ * c2_rpc_post() a noop).
+ *
+ * @subsection confc-dld-ut-infra Infrastructure Test Suite
+ *
+ *     @test c2_conf_map operations will be tested.
+ *
+ *     @test c2_conf_path operations will be tested. This includes
+ *           checking validity of various paths, testing success and
+ *           failure of c2_conf_downlink().
+ *
+ *     @test Object operations will be tested. This includes
+ *           allocation, comparison with on-wire representation, stub
+ *           enrichment.
+ *
+ *     @test c2_conf__parse() will be tested.
+ *
+ *     @test c2_conf_buf operations will be tested.
+ *
+ * @subsection confc-dld-ut-confc Confc Test Suite
+ *
+ *     Test suite's init routine will create an "ast" thread (search
+ *     for `ast_thread' in sm/ut/sm.c).  This thread will process ASTs
+ *     as they are posted by confc functions.
+ *
+ *     @test path_walk() will be tested.
+ *
+ *     @test c2_confc_*open*() and c2_confc_*close() will be tested.
+ *
+ *     @test Cache operations will be tested. This includes
+ *           cache_add(), object_enrich(), cache_grow(), and
+ *           cache__preload().
  *
  * <hr> <!------------------------------------------------------------->
  * @section confc-dld-st System Tests
-- 
1.8.3.2

