From 2ab74fd0bcc10f3d73c7ccca746499555a907eab Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Thu, 26 Apr 2012 18:04:32 +0300
Subject: [PATCH 150/311] confc.c: check path validity in ctx_invariant()

---
 conf/confc.c | 10 ++++------
 1 file changed, 4 insertions(+), 6 deletions(-)

diff --git a/conf/confc.c b/conf/confc.c
index 9b83175..6e0a912 100644
--- a/conf/confc.c
+++ b/conf/confc.c
@@ -530,9 +530,8 @@ static int act(struct c2_confc_ctx *cx, const struct walk_ctx *wx,
 /**
  * Follows the path, checking statuses of met objects.
  *
- * @pre   c2_conf_path_is_valid(ctx->fc_origin, ctx->fc_path, NULL)
- * @pre   cache's sm_group is locked
- * @post  cache's sm_group is locked
+ * @pre  cache's sm_group is locked
+ * @post cache's sm_group is locked
  *
  * @retval C2_CS_READY    Path target is reachable.
  *
@@ -561,7 +560,6 @@ static int path_walk(struct c2_confc_ctx *ctx)
 	struct walk_ctx wx = { .w_atdir = false, .w_nload = 0 };
 
 	C2_PRE(ctx_invariant(ctx));
-	C2_PRE(c2_conf_path_is_valid(ctx->fc_origin, ctx->fc_path, NULL));
 	C2_PRE(c2_mutex_is_locked(group_lock(ctx)));
 
 	wx.w_obj  = ctx->fc_origin;
@@ -747,7 +745,6 @@ request_fill(struct c2_conf_ctx *ctx, const struct c2_conf_obj *origin,
 	struct c2_conf_fetch *req = &ctx->fc_req;
 
 	C2_PRE(ctx_invariant(ctx));
-	C2_PRE(c2_conf_path_is_valid(origin, path, NULL));
 	C2_PRE(c2_mutex_is_locked(group_lock(ctx)));
 	C2_PRE(c2_mutex_is_locked(cache_lock(ctx)));
 
@@ -1144,7 +1141,8 @@ void c2_confc_ctx_fini(struct c2_confc_ctx *ctx)
 static bool ctx_invariant(const struct c2_confc_ctx *ctx)
 {
 	const struct c2_rpc_item *item = &ctx->fc_fop.f_item;
-	return XXX /* more checks */ &&
+	return XXX_bob_check(ctx->fc_client) &&
+		c2_conf_path_is_valid(ctx->fc_origin, ctx->fc_path, NULL) &&
 		item->ri_type == &request_item_type &&
 		item->ri_ops != NULL &&
 		item->ri_ops->rio_replied == on_replied &&
-- 
1.8.3.2

