From 5f0dc73bdbe321bd60ca77edf2cd8fe47d26335f Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Wed, 4 Jul 2012 01:12:54 +0300
Subject: [PATCH 212/311] confc.h: remove locking from startup/shutdown example

---
 conf/confc.h | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/conf/confc.h b/conf/confc.h
index 1f763e0..c4ab948 100644
--- a/conf/confc.h
+++ b/conf/confc.h
@@ -112,19 +112,15 @@ struct c2_mutex;
  * struct c2_confc     confc;
  *
  * startup(...) {
- *     c2_mutex_lock(confc_lock);
  *     rc = c2_confc_init(&confc, "profile-name", "confd-endpoint", group,
  *                        confc_lock);
- *     c2_mutex_unlock(confc_lock);
  *     ...
  * }
  *
  * ... Access configuration objects using confc interfaces. ...
  *
  * shutdown(...) {
- *     c2_mutex_lock(confc_lock);
  *     c2_confc_fini(confc);
- *     c2_mutex_unlock(confc_lock);
  * }
  * @endcode
  *
-- 
1.8.3.2

