From e2c6ae48763ea0447279e0bab85182f86820b501 Mon Sep 17 00:00:00 2001
From: Anatoliy Bilenko <anatoliy_bilenko@xyratex.com>
Date: Mon, 3 Sep 2012 18:04:29 +0300
Subject: [PATCH 310/311] Fixed field names and grammar.

---
 conf/preload.c | 25 +++++++++++++------------
 1 file changed, 13 insertions(+), 12 deletions(-)

diff --git a/conf/preload.c b/conf/preload.c
index dd0296e..5d6a2b7 100644
--- a/conf/preload.c
+++ b/conf/preload.c
@@ -40,16 +40,16 @@
  * ------------------------------------------------------------------ */
 
 struct stack {
-	char  *stack;
-	size_t size;
-	size_t end; /* top + 1 */
+	char  *s_stack;
+	size_t s_size;
+	size_t s_end; /* top + 1 */
 };
 
-#define STACK_INIT(s) { .stack = s, .size = ARRAY_SIZE(s), .end = 0 }
+#define STACK_INIT(s) { .s_stack = s, .s_size = ARRAY_SIZE(s), .s_end = 0 }
 
 static bool stack_is_empty(const struct stack *stack)
 {
-	return stack->end == 0;
+	return stack->s_end == 0;
 }
 
 static int stack_pop(struct stack *stack)
@@ -57,24 +57,24 @@ static int stack_pop(struct stack *stack)
 	if (stack_is_empty(stack))
 		return -EINVAL; /* underflow! */
 
-	stack->end--;
+	stack->s_end--;
 	return 0;
 }
 
 static int stack_push(struct stack *stack, char c)
 {
-	if (stack->end >= stack->size)
+	if (stack->s_end >= stack->s_size)
 		return -ENOMEM; /* overflow! */
 
-	stack->stack[stack->end++] = c;
+	stack->s_stack[stack->s_end++] = c;
 	return 0;
 }
 
 static char stack_top(const struct stack *stack)
 {
-	C2_PRE(stack->end > 0 && stack->end <= stack->size);
+	C2_PRE(stack->s_end > 0 && stack->s_end <= stack->s_size);
 
-	return stack->stack[stack->end - 1];
+	return stack->s_stack[stack->s_end - 1];
 }
 
 
@@ -358,13 +358,14 @@ static int seq_iterate(struct hunk *unit, const struct hunk *buffer, char ss)
  */
 static int seq_length(const struct hunk *unit, char ss)
 {
+	int rc;
 	int length = 0;
 	struct hunk contents = HUNK_INITIAL;
 
-	while (seq_iterate(&contents, unit, ss) == 0)
+	while ((rc = seq_iterate(&contents, unit, ss)) == 0)
 		length++;
 
-	return length;
+	return rc == 0 ? length : rc;
 }
 
 
-- 
1.8.3.2

