From 0f79d6e0980d0e4d978082a3165b5aac91420a16 Mon Sep 17 00:00:00 2001
From: Rajesh Bhalerao <Rajesh_Bhalerao@xyratex.com>
Date: Wed, 10 Apr 2013 15:33:03 +0530
Subject: [PATCH 13/25] Fixed white space problems.

---
 rm/rm.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/rm/rm.c b/rm/rm.c
index 7b6c5b3..19403fc 100644
--- a/rm/rm.c
+++ b/rm/rm.c
@@ -1581,7 +1581,7 @@ static void incoming_check(struct m0_rm_incoming *in)
 			in->rin_rc = rc;
 			incoming_state_set(in, RI_WAIT);
 		}
-		
+
 	}
 	M0_LEAVE();
 }
@@ -1687,7 +1687,7 @@ static int incoming_check_with(struct m0_rm_incoming *in,
 			 * @todo - Revoke entire loan?? rest could be subset
 			 * of r
 			 */
-			M0_ASSERT(!credit_is_empty(r));	
+			M0_ASSERT(!credit_is_empty(r));
 			rc = revoke_send(in, loan, r) ?: credit_diff(rest, r);
 			if (rc != 0)
 				M0_RETURN(rc);
-- 
1.8.3.2

