From 00f173e41f91a41f3e8495fb5dd0141ee5f8b6d0 Mon Sep 17 00:00:00 2001
From: Rajesh Bhalerao <Rajesh_Bhalerao@xyratex.com>
Date: Fri, 12 Apr 2013 14:05:40 +0530
Subject: [PATCH 16/25] Removed print statements.

---
 rm/ut/file.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/rm/ut/file.c b/rm/ut/file.c
index 5cac6ce..0ea6707 100644
--- a/rm/ut/file.c
+++ b/rm/ut/file.c
@@ -413,9 +413,7 @@ static void client_tests(void)
 	m0_chan_wait(&tests_clink[TEST1]);
 
 	/* Test encode/decode for code coverage */
-	printf("Encode/Decode test started\n");
 	file_encdec_test(&rm_ctx[SERVER_1].rc_test_data);
-	printf("Encode/Decode test completed\n");
 
 	/* Now start the use-cases */
 	test1_run();
-- 
1.8.3.2

