From e50a11751fd658ebc74492846b2219e9eeef3e29 Mon Sep 17 00:00:00 2001
From: Rajesh Bhalerao <Rajesh_Bhalerao@xyratex.com>
Date: Tue, 30 Apr 2013 17:11:47 +0530
Subject: [PATCH 20/25] Changed UT as per changed in the code after the
 inspection.

---
 rm/ut/file.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/rm/ut/file.c b/rm/ut/file.c
index 3224fce..06401d7 100644
--- a/rm/ut/file.c
+++ b/rm/ut/file.c
@@ -210,6 +210,7 @@ static void file_encdec_test(struct rm_ut_data *utdata)
 	rc = utdata->rd_rt->rt_ops->rto_encode(&cur, utdata->rd_res);
 	M0_UT_ASSERT(rc == 0);
 
+	m0_bufvec_cursor_init(&cur, &bufvec);
 	rc = utdata->rd_rt->rt_ops->rto_decode(&cur, &dec_res);
 	M0_UT_ASSERT(rc == 0);
 	M0_UT_ASSERT(dec_res != NULL);
-- 
1.8.3.2

