From 2dfa2207627310636879f10f55a44d84f4eb6ef9 Mon Sep 17 00:00:00 2001
From: Rajesh Bhalerao <Rajesh_Bhalerao@xyratex.com>
Date: Tue, 30 Apr 2013 20:47:12 +0530
Subject: [PATCH 21/25] Corrected spelling mistake.

---
 rm/file.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/rm/file.c b/rm/file.c
index a6c0371..02b4cd7 100644
--- a/rm/file.c
+++ b/rm/file.c
@@ -355,7 +355,7 @@ static int file_lock_cr_diff(struct m0_rm_credit *this,
 	M0_ASSERT(c1 != NULL);
 
 	/*
-	 * Make sure that there is no diff of a valid credit agaist
+	 * Make sure that there is no diff of a valid credit against
 	 * the empty credit.
 	 */
 	if (this->cr_datum == 0 && c1->cr_datum == RM_FILE_LOCK)
-- 
1.8.3.2

