From fd9f6eb4c0adb1170ddf6a9e3b1de3c37175b6b1 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Mon, 11 Feb 2013 12:39:10 +0200
Subject: [PATCH 25/25] m0t1fs/st: do "sstart" on demand during "st mount"

---
 m0t1fs/linux_kernel/st/st | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/m0t1fs/linux_kernel/st/st b/m0t1fs/linux_kernel/st/st
index 5b986d3..68626da 100755
--- a/m0t1fs/linux_kernel/st/st
+++ b/m0t1fs/linux_kernel/st/st
@@ -34,6 +34,14 @@ export_lnet_nid_endpoints() {
     export CONFD_ENDPOINT="$IOS0_ENDPOINT"
 }
 
+init_endpoints() {
+    if ps -ef | grep '\<m[0]d\>' > /dev/null; then
+	export_lnet_nid_endpoints
+    else
+	_init
+    fi
+}
+
 ### Starts LNET. (Note that the script never stops LNET.)
 lnet_up() {
     modprobe lnet
@@ -239,7 +247,7 @@ case "${1:-}" in
     rmmod) modules_remove; exit;;
     sstart) _init; exit;;
     sstop) _fini; exit;;
-    mount) export_lnet_nid_endpoints; _mount local; exit;;
+    mount) init_endpoints; _mount local; exit;;
     umount) umount $SANDBOX_DIR/mnt; exit;;
     help) usage; exit;;
     *) usage >&2; die;;
-- 
1.8.3.2

