From d9e1ded882f5d1d407619db441fca2f790e64605 Mon Sep 17 00:00:00 2001
From: "trupti.patil" <trupti_patil@xyratex.com>
Date: Tue, 30 Oct 2012 11:33:58 +0530
Subject: [PATCH 015/157] A few cosmetic changes.

---
 layout/composite.c | 42 +++++++++++++++---------------------------
 layout/ut/layout.c |  4 +---
 2 files changed, 16 insertions(+), 30 deletions(-)

diff --git a/layout/composite.c b/layout/composite.c
index 02c4789..834273f 100644
--- a/layout/composite.c
+++ b/layout/composite.c
@@ -194,7 +194,7 @@ int c2_composite_build(struct c2_layout_domain *dom,
 	struct c2_layout               *l;
 	struct c2_composite_layout     *cl;
 	struct c2_composite_sub_layout *sl;
-	uint32_t                        sublayouts_nr = 0;
+	uint32_t                        sublayouts_nr;
 	int                             rc;
 
 	C2_PRE(out != NULL);
@@ -210,6 +210,7 @@ int c2_composite_build(struct c2_layout_domain *dom,
 	 * addition of reference and incrementing if user count will be undone
 	 * composite_fini().
 	 */
+	sublayouts_nr = 0;
 	c2_tl_for(c2_sub_layout, sub_layouts, sl) {
 		C2_ASSERT(c2_layout__invariant(sl->csl_l));
 		l = c2_layout_find(dom, sl->csl_l->l_id);
@@ -380,16 +381,6 @@ static int sublayouts_read(struct c2_composite_layout *cl,
 	uint32_t                        sublayouts_nr;//todo rm since in invar
 	int                             rc;
 
-#if 0
-	struct c2_indexvec vec = {
-		.iv_vec = {
-			.v_nr    = ARRAY_SIZE(len),
-			.v_count = len
-		},
-		.iv_index = val
-	};
-#endif
-
 	csd  = cl->cl_base.l_dom->ld_type_data[c2_composite_layout_type.lt_id];
 	emap = &csd->csd_comp_layout_ext_map;
 	prefix.lp_l_id   = cl->cl_base.l_id;
@@ -400,7 +391,7 @@ static int sublayouts_read(struct c2_composite_layout *cl,
 	C2_ASSERT(rc == 0); //todo handle err
 
 	/*
-	 * Read all the sub-layouts from the emap and store them in the
+	 * Read all the sub-layouts from the emap cursor and store them in the
 	 * sub_layouts list.
 	 */
 	sublayouts_nr = 0;
@@ -425,7 +416,7 @@ static int sublayouts_read(struct c2_composite_layout *cl,
 #endif
 		sl->csl_l = c2_layout_find(cl->cl_base.l_dom, seg->ee_val);
 		C2_ASSERT(sl->csl_l != NULL); //todo Handle err
-		c2_layout_user_count_inc(sl->csl_l); //todo Optimise
+		c2_layout_user_count_inc(sl->csl_l); //todo Optimise code
 		C2_ASSERT(!c2_ext_is_empty(&seg->ee_ext));
 		sl->csl_ext = seg->ee_ext;
 		c2_sub_layout_tlink_init_at_tail(sl, sub_layouts);
@@ -496,6 +487,10 @@ static int composite_decode(struct c2_layout *l,
 	c2_sub_layout_tlist_init(sub_layouts);
 
 	if (op == C2_LXO_BUFFER_OP) {
+		/*
+		 * Parse the sub-layout information from the buffer pointed by
+		 * cur and store it in the sub_layouts list.
+		 */
 		for (i = 0; i < sl_header->slh_nr; ++i) {
 			sl_entry = c2_bufvec_cursor_addr(cur);
 			c2_bufvec_cursor_move(cur, sizeof *sl_entry);
@@ -536,22 +531,15 @@ static int composite_decode(struct c2_layout *l,
 		C2_ASSERT(i == sl_header->slh_nr);
 		nr = sl_header->slh_nr; //todo Optimize code
 	} else {
+		/*
+		 * Read all the sub-layouts from the layout DB and store them
+		 * in the sub_layouts list.
+		 */
 		rc = sublayouts_read(cl, sub_layouts, &nr, tx);
 		C2_ASSERT(rc == 0); //todo Handle err
 	}
 
 	composite_populate(cl, sub_layouts, nr, user_count);
-#if 0
-	if (op == C2_LXO_DB_LOOKUP) {
-		Read all the segments from the comp_layout_ext_map table,
-		belonging to composite layout with layout id 'lid' and store
-		them in the cl->cl_sub_layouts.
-
-	} else {
-		Parse the sub-layout information from the buffer pointed by
-		cur and store it in cl->cl_sub_layouts.
-	}
-#endif
 	rc = 0; //todo
 	C2_POST(ergo(rc == 0, composite_invariant(cl)));
 	C2_POST(ergo(rc != 0, composite_allocated_invariant(cl)));
@@ -588,7 +576,7 @@ static int sublayouts_write(struct c2_layout *l,
 
 	csd  = l->l_dom->ld_type_data[c2_composite_layout_type.lt_id];
 	emap = &csd->csd_comp_layout_ext_map;
-	prefix.lp_l_id = l->l_id;
+	prefix.lp_l_id   = l->l_id;
 	prefix.lp_filler = 0; //todo Add a wrapper to set prefix
 
 	if (op == C2_LXO_DB_ADD) {
@@ -598,7 +586,7 @@ static int sublayouts_write(struct c2_layout *l,
 		if (rc == -EEXIST)
 			return rc;
 
-		total = 0;
+		total         = 0;
 		sublayouts_nr = 0;
 		C2_LOG(C2_DEBUG, "Composite lid %llu, sub-layouts: ",
 		       (unsigned long long)l->l_id);
@@ -614,7 +602,7 @@ static int sublayouts_write(struct c2_layout *l,
 			len[sublayouts_nr] = sl->csl_ext.e_end -
 					     sl->csl_ext.e_start + 1;
 			val[sublayouts_nr] = sl->csl_l->l_id;
-			total = total + len[sublayouts_nr];
+			total              = total + len[sublayouts_nr];
 			++sublayouts_nr;
 		} c2_tl_endfor;
 		C2_ASSERT(sublayouts_nr == cl->cl_nr);
diff --git a/layout/ut/layout.c b/layout/ut/layout.c
index 5a593b7..9d571f1 100644
--- a/layout/ut/layout.c
+++ b/layout/ut/layout.c
@@ -1772,9 +1772,7 @@ static int test_encode_composite(uint64_t lid, bool failure_test)
 #if 0
 	/* todo Verify the layout buffer produced by c2_layout_encode(). */
 	if (!failure_test)
-		composite_layout_buf_verify(enum_id, lid,
-					  N, K, P, &seed,
-					  11, 21, &cur);
+		composite_layout_buf_verify(lid, sublayouts_nr, &cur);
 #endif
 
 	/*
-- 
1.8.3.2

