From cc43cdd74577e47be24e51fa5237253ebafa1f87 Mon Sep 17 00:00:00 2001
From: "trupti.patil" <trupti_patil@xyratex.com>
Date: Fri, 8 Mar 2013 12:21:31 +0530
Subject: [PATCH 078/157] Updating the documentation

---
 layout/composite.h | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/layout/composite.h b/layout/composite.h
index a6a56d0..30e029b 100644
--- a/layout/composite.h
+++ b/layout/composite.h
@@ -290,6 +290,12 @@ M0_INTERNAL int m0_composite_layer_delete(struct m0_composite_layout *cl,
 					  struct m0_db_tx *tx);
 
 /**
+ * @todo Shall we add an API that will delete 'the second from the top layer'
+ * and will rename the top-most as the second from the top (now top-most)?
+ * If yes, shall we call it m0_composite_layer_delete_second_from_top()?
+ */
+
+/**
  * Looks up for a valid extent containing the given offset, starting with the
  * top-most layer
  * - in the in-meory layout.
-- 
1.8.3.2

