From 03928249fb1634ec17f83efe42c0af527f92eb39 Mon Sep 17 00:00:00 2001
From: "trupti.patil" <trupti_patil@xyratex.com>
Date: Thu, 14 Mar 2013 17:43:50 +0530
Subject: [PATCH 089/157] Using _0C in the newly added layer_invariant()

---
 layout/composite.c | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/layout/composite.c b/layout/composite.c
index 2674d53..c39b6ab 100644
--- a/layout/composite.c
+++ b/layout/composite.c
@@ -210,14 +210,14 @@ static bool composite_invariant_internal(const struct m0_composite_layout *cl)
 		_0C(m0_composite_layout_bob_check(cl)) &&
 		_0C(m0_layout__invariant(&cl->cl_base)) &&
 		_0C(cl->cl_layers_nr ==
-			comp_layer_tlist_length(&cl->cl_layers)) &&
+		    comp_layer_tlist_length(&cl->cl_layers)) &&
 		m0_tl_forall(comp_layer, layer, &cl->cl_layers,
 			     _0C(layer->clr_cl == &cl->cl_base) &&
 			     _0C(m0_layout__invariant(layer->clr_sl)) &&
 			     _0C(layer->clr_idx == i++) &&
 			     _0C(layer->clr_extents_nr > 0) &&
 			     _0C(layer->clr_extents_nr ==
-			     m0_composite_layer_ext_tlist_length(
+				 m0_composite_layer_ext_tlist_length(
 							&layer->clr_extents)));
 }
 
@@ -238,16 +238,17 @@ static bool layer_invariant(const struct m0_composite_layer *layer)
 	m0_bindex_t addr = 0;
 
 	return
-		layer->clr_extents_nr ==
-		m0_composite_layer_ext_tlist_length(&layer->clr_extents) &&
+		_0C(layer->clr_extents_nr ==
+		    m0_composite_layer_ext_tlist_length(&layer->clr_extents)) &&
 		m0_tl_forall(m0_composite_layer_ext, lr_ext,
 			     &layer->clr_extents,
-			     lr_ext->cle_ext.e_start == addr &&
+			     _0C(lr_ext->cle_ext.e_start == addr) &&
 			     /* Intentional '=' below. */
 			     (addr = lr_ext->cle_ext.e_end) &&
-			     ergo(i == 0, lr_ext->cle_ext.e_start == 0) &&
-			     ergo(i == layer->clr_extents_nr - 1,
-				  lr_ext->cle_ext.e_end == M0_BINDEX_MAX + 1) &&
+			     _0C(ergo(i == 0, lr_ext->cle_ext.e_start == 0)) &&
+			     _0C(ergo(i == layer->clr_extents_nr - 1,
+				      lr_ext->cle_ext.e_end ==
+				      M0_BINDEX_MAX + 1)) &&
 			     ++i);
 }
 
-- 
1.8.3.2

