From b89e7a876ebbd7132280f0f07a40f38078b52143 Mon Sep 17 00:00:00 2001
From: Rajanikant Chirmade <rajanikant_chirmade@xyratex.com>
Date: Wed, 12 Feb 2014 14:47:32 +0530
Subject: [PATCH 50/50]  - ut was failing @ assert in m0_time_sub().

---
 fop/fom.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/fop/fom.c b/fop/fom.c
index c18ca37..6afbe14 100644
--- a/fop/fom.c
+++ b/fop/fom.c
@@ -1199,6 +1199,7 @@ void m0_fom_init(struct m0_fom *fom, struct m0_fom_type *fom_type,
 	}
 	if (m0_addb_ctx_is_initialized(&fom->fo_addb_ctx))
 		M0_FOM_ADDB_POST(fom, &reqh->rh_addb_mc, &m0_addb_rt_fom_init);
+	fom->fo_sm_phase.sm_state_epoch = 0;
 }
 M0_EXPORTED(m0_fom_init);
 
-- 
1.8.3.2

