From 107747cb170c71122d363fa4a1a3ddcbddeaa415 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Wed, 20 Nov 2013 02:06:02 +0200
Subject: [PATCH 19/96] be/alloc: extra comments removed

---
 be/alloc.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/be/alloc.c b/be/alloc.c
index 72724b7..757f702 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -402,20 +402,16 @@ be_alloc_chunk_add_after(struct m0_be_allocator *a,
 	if (c != NULL) {
 		M0_BE_OP_SYNC(op,
 			      m0_be_list_add_after(chunks, &op, tx, c, new));
-		// chunks_all_tlist_add_after_c(a, tx, c, new);
 	} else {
 		M0_BE_OP_SYNC(op, m0_be_list_add(chunks, &op, tx, new));
-		// chunks_all_tlist_add_c(a, tx, new);
 	}
 	if (free) {
 		/** add chunk to m0_be_allocator_header.bac_free list */
 		if (f != NULL) {
 			M0_BE_OP_SYNC(op, m0_be_list_add_after(cfree, &op,
 							       tx, f, new));
-			// chunks_free_tlist_add_after_c(a, tx, f, new);
 		} else {
 			M0_BE_OP_SYNC(op, m0_be_list_add(cfree, &op, tx, new));
-			// chunks_free_tlist_add_c(a, tx, new);
 		}
 	}
 	M0_POST(be_alloc_chunk_invariant(a, new));
-- 
1.8.3.2

