From 1e8968557c6cdac18e5d4a6a0e8608e581e54ab2 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Wed, 20 Nov 2013 02:41:16 +0200
Subject: [PATCH 24/96] be/allocator: removed obsolete comment

---
 be/alloc.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/be/alloc.c b/be/alloc.c
index 530702e..0f0503f 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -720,7 +720,6 @@ M0_INTERNAL void m0_be_allocator_credit(struct m0_be_allocator *a,
 	m0_be_list_credit(free, M0_BLO_INSERT, 1, &cred_mark_free);
 	m0_be_tx_credit_add(&cred_mark_free, &cred_free_flag);
 
-	/** @todo TODO XXX add list credits instead of entire header */
 	switch (optype) {
 		case M0_BAO_CREATE:
 			m0_be_tx_credit_mac(accum, &cred_list_create, 2);
-- 
1.8.3.2

