From 1ab025a09aa6a88c7e2fcbf16caf15dd6b19d472 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Thu, 12 Dec 2013 05:28:29 +0200
Subject: [PATCH 26/96] be/allocator: added workaround for credit bugs

---
 be/alloc.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/be/alloc.c b/be/alloc.c
index 3ae20ee..532204b 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -755,6 +755,9 @@ M0_INTERNAL void m0_be_allocator_credit(struct m0_be_allocator *a,
 		default:
 			M0_IMPOSSIBLE("Invalid allocator operation type");
 	}
+	/* XXX FIXME ASAP workaround for allocator/btree/etc. credit bugs */
+	/* 640 bytes ought to be enough for anybody */
+	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT(40, 640));
 }
 
 M0_INTERNAL void m0_be_alloc_aligned(struct m0_be_allocator *a,
-- 
1.8.3.2

