From 21c45648dbaede313393765079f78d438336d055 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Thu, 9 Jan 2014 05:04:03 +0200
Subject: [PATCH 34/96] be/alloc: s/alloation/allocation/

---
 be/alloc.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/be/alloc.c b/be/alloc.c
index 5e45d45..4a6d2de 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -58,7 +58,7 @@
  * m0_be_alloc(): first-fit algorithm is used.
  * Time complexity O(N), I/O complexity O(N).
  *
- * List of free chunks is scanned until first chunk that fit the alloation
+ * List of free chunks is scanned until first chunk that fit the allocation
  * requirements (chunk can be split somehow to obtain the chunk with
  * size >= requested size and for this chunk start of user data should be
  * alighed according to "shift" parameter) found. Then free chunk is split
-- 
1.8.3.2

