From bbd49dfcb40e179fced26b6ae7ec9809650e1542 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Thu, 9 Jan 2014 05:21:06 +0200
Subject: [PATCH 36/96] be/alloc: no doxy-comments inside routines

---
 be/alloc.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/be/alloc.c b/be/alloc.c
index 6f48d0f..f3fe040 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -648,7 +648,6 @@ be_alloc_chunk_add_after(struct m0_be_allocator *a,
 			  (struct be_alloc_chunk *) be_alloc_chunk_after(a, c);
 	be_alloc_chunk_init(a, tx, new, size_total - sizeof(*new), free);
 
-	/** add chunk to m0_be_allocator.bac_chunks list */
 	if (c != NULL) {
 		M0_BE_OP_SYNC(op,
 			      m0_be_list_add_after(chunks, &op, tx, c, new));
-- 
1.8.3.2

