From b68e34773b086288fe660a7ae8c70ed4bc247622 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Thu, 9 Jan 2014 06:14:36 +0200
Subject: [PATCH 41/96] be/alloc: fixed shift param description for alloc()

---
 be/alloc.h | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/be/alloc.h b/be/alloc.h
index 19560cc..168a453 100644
--- a/be/alloc.h
+++ b/be/alloc.h
@@ -227,10 +227,8 @@ M0_INTERNAL void m0_be_allocator_credit(struct m0_be_allocator *a,
  *	      *ptr shouldn't be used before completion of the operation.
  *	      Can be NULL.
  * @param size Allocated memory size in bytes
- * @param shift Memory will be aligned on (2^shift)-byte boundary.
- *		It can be less than M0_BE_ALLOC_SHIFT_MIN - in this case
- *		allocation will be done as if it is equal to
- *		M0_BE_ALLOC_SHIFT_MIN.
+ * @param shift Memory will be aligned on 2^max(shift, M0_BE_ALLOC_SHIFT_MIN)
+ *		byte boundary.
  *
  * This function will return pointer to memory allocated in
  * m0_be_op.bo_u.u_allocator.a_ptr (always) and in *ptr (if *ptr != NULL)
-- 
1.8.3.2

