From 9f6ae1486c1cf7e82f0ef165df3af5174a15c278 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Thu, 9 Jan 2014 06:22:45 +0200
Subject: [PATCH 42/96] be/alloc: grammar fix

---
 be/alloc.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/be/alloc.c b/be/alloc.c
index 93f4736..94aea9d 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -92,7 +92,7 @@
  * - all chunks in a list are ordered by address;
  * - every chunk is in the chunks list;
  * - every free chunk is in the free list;
- * - free list don't contains used chunk;
+ * - free list doesn't contain used chunk;
  * - any two unequal to each other chunks in the free list are not adjacent;
  *
  * Special cases
@@ -150,7 +150,7 @@
  * - If tx parameter is NULL then allocator will not capture segment updates.
  *   It might be useful in the allocator UT.
  *
- * Know issues:
+ * Known issues:
  * - op is unconditionally transitioned to state M0_BOS_SUCCESS in m0_be_alloc()
  *   and m0_be_free().
  */
-- 
1.8.3.2

