From 39bcdb3388d65ecfa0b2fe2e909158e23afe46ff Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Thu, 9 Jan 2014 07:05:36 +0200
Subject: [PATCH 44/96] be/alloc: s/m0_tl_for/m0_tl_forall/ in invariant

---
 be/alloc.c | 13 ++-----------
 1 file changed, 2 insertions(+), 11 deletions(-)

diff --git a/be/alloc.c b/be/alloc.c
index 7fed892..826acfd 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -814,17 +814,8 @@ M0_INTERNAL void m0_be_allocator_fini(struct m0_be_allocator *a)
 
 M0_INTERNAL bool m0_be_allocator__invariant(const struct m0_be_allocator *a)
 {
-	struct be_alloc_chunk *iter;
-	bool		       success = true;
-
-	m0_tl_for(chunks_all, &a->ba_chunks.bl_list, iter) {
-		if (!be_alloc_chunk_invariant(a, iter)) {
-			success = false;
-			break;
-		}
-	} m0_tl_endfor;
-
-	return success;
+	return m0_tl_forall(chunks_all, iter, &a->ba_chunks.bl_list,
+			    be_alloc_chunk_invariant(a, iter));
 }
 
 M0_INTERNAL int m0_be_allocator_create(struct m0_be_allocator *a,
-- 
1.8.3.2

