From a2e317166b7f677a5fb70f634210d9c247cc2ebc Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Thu, 9 Jan 2014 07:12:17 +0200
Subject: [PATCH 45/96] be/alloc: improved be_alloc_chunk::bac_size doc

---
 be/alloc_internal.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/be/alloc_internal.h b/be/alloc_internal.h
index 76e8867..905b679 100644
--- a/be/alloc_internal.h
+++ b/be/alloc_internal.h
@@ -55,7 +55,7 @@ struct be_alloc_chunk {
 	struct m0_tlink bac_linkage_free;
 	/** magic for bac_linkage_free */
 	uint64_t	bac_magic_free;
-	/** size of chunk */
+	/** size of allocated memory */
 	m0_bcount_t	bac_size;
 	/** is chunk free? */
 	bool		bac_free;
-- 
1.8.3.2

