From d99bb422aca4767a525e032da48621c8d726ae7d Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Tue, 14 Jan 2014 17:39:37 +0200
Subject: [PATCH 49/96] be/fl: doc typo fix

---
 be/fl.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/be/fl.h b/be/fl.h
index e3bb3cc..517a0c0 100644
--- a/be/fl.h
+++ b/be/fl.h
@@ -33,7 +33,7 @@
 
 enum {
 	M0_BE_FL_STEP = 8,	/**< each size is increased to this boundary */
-	M0_BE_FL_NR = 128,	/*<* number of free lists for each size */
+	M0_BE_FL_NR = 128,	/**< number of free lists for each size */
 	M0_BE_FL_PICK_SCAN_LIMIT = 0x10, /**< scan limit for searching in
 					      chunks with size >=
 					      M0_BE_FL_NR * M0_BE_FL_STEP */
-- 
1.8.3.2

