From 1d90d40d29b6f4cd8d591f9bae358793514d89c3 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Tue, 14 Jan 2014 22:27:47 +0200
Subject: [PATCH 55/96] be/ut/helper: added workaround for wrong stob ref count

---
 be/ut/helper.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/be/ut/helper.c b/be/ut/helper.c
index 6603400..66195bf 100644
--- a/be/ut/helper.c
+++ b/be/ut/helper.c
@@ -468,6 +468,11 @@ struct m0_stob *m0_be_ut_stob_get_by_id(uint64_t id, bool stob_create)
 		rc = m0_stob_create_helper(h->buh_stob_dom, NULL, &stob_id,
 					   &stob);
 		M0_ASSERT(rc == 0);
+		/*
+		 * XXX workaround for bug(?) in m0_stob_find(): if it's a new
+		 * stob, then stob returned has ref count = 2.
+		 */
+		m0_stob_put(stob);
 	} else {
 		/* XXX memory leak here */
 		M0_ALLOC_PTR(stob);
-- 
1.8.3.2

