From ad3ec5a2dabe4fd413e8464ea2b9944b368c9677 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Tue, 14 Jan 2014 23:54:33 +0200
Subject: [PATCH 60/96] be/fl: credits for create()/destroy() implemented

---
 be/fl.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/be/fl.c b/be/fl.c
index 5fe222e..e4a351d 100644
--- a/be/fl.c
+++ b/be/fl.c
@@ -221,9 +221,11 @@ M0_INTERNAL void m0_be_fl_credit(struct m0_be_fl *fl,
 {
 	switch (fl_op) {
 	case M0_BFL_CREATE:
+		m0_be_list_credit(be_fl_list(fl, 0), M0_BLO_CREATE,
+				  ARRAY_SIZE(fl->bfl_free), accum);
 	case M0_BFL_DESTROY:
-		/* XXX make proper credit calculation */
-		m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT(0x1000, 0x10000));
+		m0_be_list_credit(be_fl_list(fl, 0), M0_BLO_DESTROY,
+				  ARRAY_SIZE(fl->bfl_free), accum);
 		break;
 	case M0_BFL_ADD:
 		m0_be_list_credit(be_fl_list(fl, 0), M0_BLO_TLINK_CREATE, 1,
-- 
1.8.3.2

