From b545210b36427ac2a98f42af404cc77962d06ea5 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Tue, 14 Jan 2014 23:59:41 +0200
Subject: [PATCH 61/96] be/fl: unnecessary braces removed from init()/fini()

---
 be/fl.c | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/be/fl.c b/be/fl.c
index e4a351d..66b14a0 100644
--- a/be/fl.c
+++ b/be/fl.c
@@ -62,9 +62,8 @@ M0_INTERNAL void m0_be_fl_init(struct m0_be_fl *fl, struct m0_be_seg *seg)
 {
 	int i;
 
-	for (i = 0; i < ARRAY_SIZE(fl->bfl_free); ++i) {
+	for (i = 0; i < ARRAY_SIZE(fl->bfl_free); ++i)
 		m0_be_list_init(be_fl_list(fl, i), seg);
-	}
 
 	/* XXX temporary solution to make capturing checkers pass */
 	m0_be_reg__write(&M0_BE_REG_PTR(seg, fl));	/* XXX */
@@ -75,9 +74,8 @@ M0_INTERNAL void m0_be_fl_fini(struct m0_be_fl *fl)
 	struct m0_be_seg *seg = fl->bfl_free[0].bfs_list.bl_seg;
 	int		  i;
 
-	for (i = 0; i < ARRAY_SIZE(fl->bfl_free); ++i) {
+	for (i = 0; i < ARRAY_SIZE(fl->bfl_free); ++i)
 		m0_be_list_fini(be_fl_list(fl, i));
-	}
 
 	/* XXX temporary solution to make capturing checkers pass */
 	m0_be_reg__write(&M0_BE_REG_PTR(seg, fl));	/* XXX */
-- 
1.8.3.2

