From df9b3f46443dfb6b82614b73347444fc01109943 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Wed, 15 Jan 2014 00:11:12 +0200
Subject: [PATCH 62/96] be/alloc: merged M0_BE_OP_SYNC() in chunk_add_after()

---
 be/alloc.c | 9 +++------
 1 file changed, 3 insertions(+), 6 deletions(-)

diff --git a/be/alloc.c b/be/alloc.c
index b6b82ad..ee4142c 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -346,12 +346,9 @@ be_alloc_chunk_add_after(struct m0_be_allocator *a,
 			  (struct be_alloc_chunk *) be_alloc_chunk_after(a, c);
 	be_alloc_chunk_init(a, tx, new, size_total - sizeof(*new), free);
 
-	if (c != NULL) {
-		M0_BE_OP_SYNC(op,
-			      m0_be_list_add_after(chunks, &op, tx, c, new));
-	} else {
-		M0_BE_OP_SYNC(op, m0_be_list_add(chunks, &op, tx, new));
-	}
+	M0_BE_OP_SYNC(op, c != NULL ?
+			  m0_be_list_add_after(chunks, &op, tx, c, new) :
+			  m0_be_list_add(chunks, &op, tx, new));
 	if (free) {
 		m0_be_fl_add(&a->ba_fl, tx, new);
 	}
-- 
1.8.3.2

