From fcd5a036d24a7a432e7b800a53b9d74b1442a915 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Wed, 15 Jan 2014 00:51:05 +0200
Subject: [PATCH 64/96] be/alloc: struct m0_be_allocator * is const wherever
 possible

---
 be/alloc.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/be/alloc.c b/be/alloc.c
index c0a3625..f28740b 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -169,7 +169,7 @@ M0_TL_DESCR_DEFINE(chunks_all, "list of all chunks in m0_be_allocator",
 		   M0_BE_ALLOC_ALL_LINK_MAGIC, M0_BE_ALLOC_ALL_MAGIC);
 M0_TL_DEFINE(chunks_all, static, struct be_alloc_chunk);
 
-static void be_alloc_chunk_capture(struct m0_be_allocator *a,
+static void be_alloc_chunk_capture(const struct m0_be_allocator *a,
 				   struct m0_be_tx *tx,
 				   struct be_alloc_chunk *c)
 {
@@ -186,14 +186,14 @@ static void be_alloc_head_capture(struct m0_be_allocator *a,
 	M0_BE_TX_CAPTURE_PTR(a->ba_seg, tx, &a->ba_stats);
 }
 
-static void be_alloc_free_flag_capture(struct m0_be_allocator *a,
+static void be_alloc_free_flag_capture(const struct m0_be_allocator *a,
 				       struct m0_be_tx *tx,
 				       struct be_alloc_chunk *c)
 {
 	M0_BE_TX_CAPTURE_PTR(a->ba_seg, tx, &c->bac_free);
 }
 
-static void be_alloc_size_capture(struct m0_be_allocator *a,
+static void be_alloc_size_capture(const struct m0_be_allocator *a,
 				  struct m0_be_tx *tx,
 				  struct be_alloc_chunk *c)
 {
-- 
1.8.3.2

