From 400ecd89287807727716c627036d919af918475e Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Wed, 15 Jan 2014 00:52:03 +0200
Subject: [PATCH 65/96] be/alloc: style fix in chunk_capture()

---
 be/alloc.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/be/alloc.c b/be/alloc.c
index f28740b..02935a4 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -173,9 +173,8 @@ static void be_alloc_chunk_capture(const struct m0_be_allocator *a,
 				   struct m0_be_tx *tx,
 				   struct be_alloc_chunk *c)
 {
-	if (c == NULL)
-		return;
-	M0_BE_TX_CAPTURE_PTR(a->ba_seg, tx, c);
+	if (c != NULL)
+		M0_BE_TX_CAPTURE_PTR(a->ba_seg, tx, c);
 }
 
 static void be_alloc_head_capture(struct m0_be_allocator *a,
-- 
1.8.3.2

