From f9bf327030996bc5aa0ed5d867a88956bb8bf8fc Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Wed, 15 Jan 2014 01:24:50 +0200
Subject: [PATCH 69/96] be/fl: conjuction style in _invariant()

---
 be/fl.c | 16 ++++------------
 1 file changed, 4 insertions(+), 12 deletions(-)

diff --git a/be/fl.c b/be/fl.c
index 66b14a0..83e18a2 100644
--- a/be/fl.c
+++ b/be/fl.c
@@ -119,18 +119,10 @@ M0_INTERNAL unsigned long m0_be_fl_index_round_down_chunk(struct m0_be_fl *fl,
 
 M0_INTERNAL bool m0_be_fl__invariant(struct m0_be_fl *fl)
 {
-	struct be_alloc_chunk *chunk;
-	int		       i;
-
-	/* check if each bfl_free contains chunks with appropriate size */
-	for (i = 0; i < ARRAY_SIZE(fl->bfl_free); ++i) {
-		m0_tl_for(chunks_free, &be_fl_list(fl, i)->bl_list, chunk) {
-			if (_0C(m0_be_fl_index_round_down_chunk(fl,
-								chunk) != i))
-				return false;
-		} m0_tl_endfor;
-	}
-	return true;
+	return m0_forall(i, ARRAY_SIZE(fl->bfl_free),
+		 m0_tl_forall(chunks_free, chunk, &be_fl_list(fl, i)->bl_list,
+			      _0C(m0_be_fl_index_round_down_chunk(fl, chunk) ==
+				  i)));
 }
 
 M0_INTERNAL void m0_be_fl_add(struct m0_be_fl *fl,
-- 
1.8.3.2

