From bfb4ea820e8268d9bd057cb4bbf666f07769fc6c Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Wed, 15 Jan 2014 01:37:28 +0200
Subject: [PATCH 70/96] be/fl: s/index ==/index </ in pick()

---
 be/fl.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/be/fl.c b/be/fl.c
index 83e18a2..5bbfedc 100644
--- a/be/fl.c
+++ b/be/fl.c
@@ -184,8 +184,7 @@ M0_INTERNAL struct be_alloc_chunk *m0_be_fl_pick(struct m0_be_fl *fl,
 			break;
 	}
 
-	flist = index == ARRAY_SIZE(fl->bfl_free) ? NULL :
-						    be_fl_list(fl, index);
+	flist = index < ARRAY_SIZE(fl->bfl_free) ? be_fl_list(fl, index) : NULL;
 	chunk = flist == NULL ? NULL : chunks_free_tlist_head(&flist->bl_list);
 	if (index == M0_BE_FL_NR && chunk != NULL) {
 		chunk = NULL;
-- 
1.8.3.2

