From 4a8e2fe8ce73d8c2203f33dc6890c1dc182caab5 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Sun, 16 Feb 2014 16:47:07 +0200
Subject: [PATCH 73/96] be/Makefile: whitespace fix

---
 be/Makefile.sub | 34 +++++++++++++++++-----------------
 1 file changed, 17 insertions(+), 17 deletions(-)

diff --git a/be/Makefile.sub b/be/Makefile.sub
index 36b6a6d..fcd987d 100644
--- a/be/Makefile.sub
+++ b/be/Makefile.sub
@@ -26,27 +26,27 @@ nobase_mero_include_HEADERS += be/alloc.h           \
                                be/tx_service.h	    \
 			       be/ut/helper.h
 
-mero_libmero_la_SOURCES += be/alloc.c     	\
-			   be/be.c        	\
-			   be/btree.c     	\
+mero_libmero_la_SOURCES += be/alloc.c		\
+			   be/be.c		\
+			   be/btree.c		\
 			   be/extmap.c		\
 			   be/extmap_xc.c	\
 			   be/domain.c		\
-			   be/engine.c       	\
-			   be/fl.c       	\
-			   be/io.c        	\
-			   be/list.c      	\
-			   be/log.c       	\
-			   be/log_store.c 	\
-			   be/op.c 		\
-			   be/seg.c       	\
-			   be/seg_dict.c       	\
-			   be/tx.c        	\
-			   be/tx_credit.c 	\
+			   be/engine.c		\
+			   be/fl.c		\
+			   be/io.c		\
+			   be/list.c		\
+			   be/log.c		\
+			   be/log_store.c	\
+			   be/op.c		\
+			   be/seg.c		\
+			   be/seg_dict.c	\
+			   be/tx.c		\
+			   be/tx_credit.c	\
 			   be/tx_group_fom.c	\
-			   be/tx_group.c  	\
-			   be/tx_group_ondisk.c \
-			   be/tx_regmap.c 	\
+			   be/tx_group.c	\
+			   be/tx_group_ondisk.c	\
+			   be/tx_regmap.c	\
 			   be/tx_service.c	\
                            be/ut/helper.c
 
-- 
1.8.3.2

