From 3207904fc5f2502a366b0863510189914eec85cf Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Sun, 16 Feb 2014 17:06:10 +0200
Subject: [PATCH 76/96] be/alloc: style fix - extra {} removed

---
 be/alloc.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/be/alloc.c b/be/alloc.c
index 8e0e43a..f66fb9f 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -337,9 +337,9 @@ be_alloc_chunk_add_after(struct m0_be_allocator *a,
 	M0_BE_OP_SYNC(op, c != NULL ?
 			  m0_be_list_add_after(&a->ba_chunks, &op, tx, c, new) :
 			  m0_be_list_add(&a->ba_chunks, &op, tx, new));
-	if (free) {
+	if (free)
 		m0_be_fl_add(&a->ba_fl, tx, new);
-	}
+
 	M0_POST(be_alloc_chunk_invariant(a, new));
 	M0_PRE(ergo(c != NULL, be_alloc_chunk_invariant(a, c)));
 	return new;
-- 
1.8.3.2

