From 41fd2a02446a66a13835ab6e7c3b015071e7f53b Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Sun, 16 Feb 2014 18:16:01 +0200
Subject: [PATCH 77/96] be/fl: made const parameter

---
 be/fl.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/be/fl.c b/be/fl.c
index 5bbfedc..28bd164 100644
--- a/be/fl.c
+++ b/be/fl.c
@@ -111,7 +111,7 @@ M0_INTERNAL unsigned long m0_be_fl_index_round_up(struct m0_be_fl *fl,
 }
 
 M0_INTERNAL unsigned long m0_be_fl_index_round_down_chunk(struct m0_be_fl *fl,
-						  struct be_alloc_chunk *chunk)
+				  const struct be_alloc_chunk *chunk)
 {
 	return min_type(unsigned long, M0_BE_FL_NR,
 			chunk->bac_size / M0_BE_FL_STEP);
-- 
1.8.3.2

