From 9f64f1fa11fef9ccb101ef23d3231a6c8043d37a Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Sun, 16 Feb 2014 18:16:35 +0200
Subject: [PATCH 78/96] be/list: made const parameter

---
 be/list.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/be/list.c b/be/list.c
index 04df7e7..6820d28 100644
--- a/be/list.c
+++ b/be/list.c
@@ -225,7 +225,8 @@ M0_INTERNAL void *m0_be_list_prev(struct m0_be_list *list, struct m0_be_op *op,
 	return list_iter(list, op, obj, m0_tlist_prev);
 }
 
-static struct m0_tlink *be_tlink_from_obj(void *obj, struct m0_be_list *list)
+static struct m0_tlink *be_tlink_from_obj(void *obj,
+					  const struct m0_be_list *list)
 {
 	return (struct m0_tlink *)
 	       (obj == NULL ? NULL : obj + list->bl_descr->td_link_offset);
-- 
1.8.3.2

