From b7f14b8b6da82ed52dda861570549ccf9d5479b3 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Tue, 18 Feb 2014 13:56:03 +0200
Subject: [PATCH 84/96] be/alloc: s/ENOMEM/ENOSPC/ if allocation failed

---
 be/alloc.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/be/alloc.c b/be/alloc.c
index 0d3c07e..45df118 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -703,7 +703,7 @@ M0_INTERNAL void m0_be_alloc_aligned(struct m0_be_allocator *a,
 		m0_be_tx_capture(tx, &M0_BE_REG(a->ba_seg, size, &c->bac_mem));
 	}
 	op->bo_u.u_allocator.a_ptr = c == NULL ?    NULL : &c->bac_mem;
-	op->bo_u.u_allocator.a_rc  = c == NULL ? -ENOMEM : 0;
+	op->bo_u.u_allocator.a_rc  = c == NULL ? -ENOSPC : 0;
 	if (ptr != NULL)
 		*ptr = op->bo_u.u_allocator.a_ptr;
 	/* and ends here */
-- 
1.8.3.2

