From 2b4fb516becba6ed39c55232e168a9c80c17c6f4 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Tue, 18 Feb 2014 15:29:01 +0200
Subject: [PATCH 86/96] be/alloc: removed wrong TODOs

---
 be/alloc.c | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/be/alloc.c b/be/alloc.c
index c639740..47d8023 100644
--- a/be/alloc.c
+++ b/be/alloc.c
@@ -541,7 +541,6 @@ M0_INTERNAL int m0_be_allocator_create(struct m0_be_allocator *a,
 
 	be_alloc_head_capture(a, tx);
 
-	/** @todo PUT_PTR h */
 	M0_POST_EX(m0_be_allocator__invariant(a));
 	m0_mutex_unlock(&a->ba_lock);
 
@@ -553,7 +552,6 @@ M0_INTERNAL void m0_be_allocator_destroy(struct m0_be_allocator *a,
 {
 	struct be_alloc_chunk *c;
 
-	/** @todo GET_PTR h */
 	m0_mutex_lock(&a->ba_lock);
 	M0_PRE_EX(m0_be_allocator__invariant(a));
 
@@ -568,7 +566,6 @@ M0_INTERNAL void m0_be_allocator_destroy(struct m0_be_allocator *a,
 	be_alloc_head_capture(a, tx);
 
 	m0_mutex_unlock(&a->ba_lock);
-	/** @todo PUT_PTR h */
 }
 
 M0_INTERNAL void m0_be_allocator_credit(struct m0_be_allocator *a,
@@ -689,7 +686,6 @@ M0_INTERNAL void m0_be_alloc_aligned(struct m0_be_allocator *a,
 
 	shift = max_check(shift, (unsigned) M0_BE_ALLOC_SHIFT_MIN);
 
-	/* XXX */
 	m0_be_op_state_set(op, M0_BOS_ACTIVE);
 
 	m0_mutex_lock(&a->ba_lock);
@@ -799,12 +795,10 @@ M0_INTERNAL void m0_be_free(struct m0_be_allocator *a,
 M0_INTERNAL void m0_be_alloc_stats(struct m0_be_allocator *a,
 				   struct m0_be_allocator_stats *out)
 {
-	/** @todo GET_PTR a->ba_h */
 	m0_mutex_lock(&a->ba_lock);
 	M0_PRE_EX(m0_be_allocator__invariant(a));
 	*out = a->ba_stats;
 	m0_mutex_unlock(&a->ba_lock);
-	/** @todo PUT_PTR a->ba_h */
 }
 
 #undef M0_TRACE_SUBSYSTEM
-- 
1.8.3.2

