From 8d6e58f074b452e963c36a6ad72fc9066e924c0c Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Tue, 18 Feb 2014 15:29:11 +0200
Subject: [PATCH 87/96] be/fl: removed copy-paste from credit()

---
 be/fl.c | 19 +++++++++----------
 1 file changed, 9 insertions(+), 10 deletions(-)

diff --git a/be/fl.c b/be/fl.c
index 8957b06..ac06407 100644
--- a/be/fl.c
+++ b/be/fl.c
@@ -198,23 +198,22 @@ M0_INTERNAL void m0_be_fl_credit(struct m0_be_fl *fl,
 				 enum m0_be_fl_op fl_op,
 				 struct m0_be_tx_credit *accum)
 {
+	const struct m0_be_list *list	 = be_fl_list(fl, 0);
+	size_t			 list_nr = ARRAY_SIZE(fl->bfl_free);
+
 	switch (fl_op) {
 	case M0_BFL_CREATE:
-		m0_be_list_credit(be_fl_list(fl, 0), M0_BLO_CREATE,
-				  ARRAY_SIZE(fl->bfl_free), accum);
+		m0_be_list_credit(list, M0_BLO_CREATE, list_nr, accum);
 	case M0_BFL_DESTROY:
-		m0_be_list_credit(be_fl_list(fl, 0), M0_BLO_DESTROY,
-				  ARRAY_SIZE(fl->bfl_free), accum);
+		m0_be_list_credit(list, M0_BLO_DESTROY, list_nr, accum);
 		break;
 	case M0_BFL_ADD:
-		m0_be_list_credit(be_fl_list(fl, 0), M0_BLO_TLINK_CREATE, 1,
-				  accum);
-		m0_be_list_credit(be_fl_list(fl, 0), M0_BLO_INSERT, 1, accum);
+		m0_be_list_credit(list, M0_BLO_TLINK_CREATE, 1, accum);
+		m0_be_list_credit(list, M0_BLO_INSERT, 1, accum);
 		break;
 	case M0_BFL_DEL:
-		m0_be_list_credit(be_fl_list(fl, 0), M0_BLO_DELETE, 1, accum);
-		m0_be_list_credit(be_fl_list(fl, 0), M0_BLO_TLINK_DESTROY, 1,
-				  accum);
+		m0_be_list_credit(list, M0_BLO_DELETE, 1, accum);
+		m0_be_list_credit(list, M0_BLO_TLINK_DESTROY, 1, accum);
 		break;
 	default:
 		M0_ASSERT_INFO(false, "fl_op = %d", fl_op);
-- 
1.8.3.2

