From 7f8ac185860312d39a42e3cee6f8cecee4de2741 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Tue, 25 Feb 2014 19:53:23 +0200
Subject: [PATCH 93/96] be/fl: [doc] s/item/chunk/

---
 be/fl.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/be/fl.c b/be/fl.c
index 7826347..4ac4bc4 100644
--- a/be/fl.c
+++ b/be/fl.c
@@ -46,7 +46,7 @@
  * - for each subrange doubly-linked LRU list is maintained;
  * - pick() implementation
  *   - find what range the requested size is in, select corresponding list;
- *   - if size is in [0, M0_BE_FL_NR * M0_BE_FL_STEP) then select first item
+ *   - if size is in [0, M0_BE_FL_NR * M0_BE_FL_STEP) then select first chunk
  *     from the list;
  *   - otherwise scan the list to find first suitable chunk.
  *
-- 
1.8.3.2

