From a8074e3842260fd0584b9c57612ae8f6152b0887 Mon Sep 17 00:00:00 2001
From: Mandar Sawant <mandar_sawant@xyratex.com>
Date: Wed, 30 Apr 2014 17:38:01 +0530
Subject: [PATCH 6/6] stob-unlink, - Fold stob_ad_map_ext_delete() into
 stob_ad_destroy().

---
 stob/ad.c | 57 +++++++++++++++++++++++----------------------------------
 1 file changed, 23 insertions(+), 34 deletions(-)

diff --git a/stob/ad.c b/stob/ad.c
index ff486e1..440d6d1 100644
--- a/stob/ad.c
+++ b/stob/ad.c
@@ -577,39 +577,6 @@ static int stob_ad_destroy_credit(struct m0_stob *stob,
 	return M0_RC(rc);
 }
 
-static int stob_ad_map_ext_delete(struct m0_stob_ad_domain *adom,
-				  struct m0_be_emap_cursor *it,
-				  struct m0_dtx *tx)
-{
-	struct m0_be_op *it_op;
-	struct m0_ext   *ext = &it->ec_seg.ee_ext;
-	struct m0_ext    todo = {
-				.e_start = 0,
-				.e_end   = M0_BCOUNT_MAX
-			 };
-	int              rc = 0;
-
-	M0_LOG(M0_DEBUG, "ext="EXT_F" val=%llu",
-		EXT_P(ext), (unsigned long long)it->ec_seg.ee_val);
-
-	it_op = &it->ec_op;
-	m0_be_op_init(it_op);
-	m0_be_emap_paste(it, &tx->tx_betx, &todo, AET_HOLE,
-		 LAMBDA(void, (struct m0_be_emap_seg *seg) {
-				/* handle extent deletion. */
-				M0_LOG(M0_DEBUG, "del: val=%llu",
-					(unsigned long long)seg->ee_val);
-				rc = rc ?: stob_ad_seg_free(tx, adom, seg,
-							    &seg->ee_ext,
-							    seg->ee_val);
-		}), NULL, NULL);
-	M0_ASSERT(m0_be_op_state(it_op) == M0_BOS_SUCCESS);
-	rc = it_op->bo_u.u_emap.e_rc;
-	m0_be_op_fini(it_op);
-
-	return M0_RC(rc);
-}
-
 /**
  * Destroys ad stob ext map and releases underlying storage object's
  * extents.
@@ -620,6 +587,12 @@ static int stob_ad_destroy(struct m0_stob *stob, struct m0_dtx *tx)
 	struct m0_be_emap_seg    *seg;
 	struct m0_be_emap_cursor  it;
 	struct m0_uint128         prefix;
+	struct m0_be_op          *it_op;
+	struct m0_ext            *ext;
+	struct m0_ext             todo = {
+					.e_start = 0,
+					.e_end   = M0_BCOUNT_MAX
+				  };
 	int                       rc;
 
 	adom   = stob_ad_domain2ad(m0_stob_dom_get(stob));
@@ -627,7 +600,23 @@ static int stob_ad_destroy(struct m0_stob *stob, struct m0_dtx *tx)
 	rc = stob_ad_cursor(adom, stob, 0, &it);
 	M0_LOG(M0_DEBUG, U128D_F, U128_P(&it.ec_prefix));
 	seg = m0_be_emap_seg_get(&it);
-	rc = stob_ad_map_ext_delete(adom, &it, tx);
+	ext = &it.ec_seg.ee_ext;
+	M0_LOG(M0_DEBUG, "ext="EXT_F" val=%llu",
+		EXT_P(ext), (unsigned long long)it.ec_seg.ee_val);
+	it_op = &it.ec_op;
+	m0_be_op_init(it_op);
+	m0_be_emap_paste(&it, &tx->tx_betx, &todo, AET_HOLE,
+		 LAMBDA(void, (struct m0_be_emap_seg *seg) {
+				/* handle extent deletion. */
+				M0_LOG(M0_DEBUG, "del: val=%llu",
+					(unsigned long long)seg->ee_val);
+				rc = rc ?: stob_ad_seg_free(tx, adom, seg,
+							    &seg->ee_ext,
+							    seg->ee_val);
+		}), NULL, NULL);
+	M0_ASSERT(m0_be_op_state(it_op) == M0_BOS_SUCCESS);
+	rc = m0_be_emap_op_rc(&it);
+	m0_be_op_fini(it_op);
 	if (rc == 0)
 		rc = M0_BE_OP_SYNC_RET(op,
 				       m0_be_emap_obj_delete(&adom->sad_adata,
-- 
1.8.3.2

