From 5b91a74e4f3d19fbbb1481d5d3dd7c5b924848bf Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Thu, 26 Sep 2013 05:12:08 +0300
Subject: [PATCH 2/5] be/list,seg_dict: includes fixed

---
 be/list.c     | 1 +
 be/seg_dict.c | 1 +
 2 files changed, 2 insertions(+)

diff --git a/be/list.c b/be/list.c
index a6eeade..15f8b34 100644
--- a/be/list.c
+++ b/be/list.c
@@ -24,6 +24,7 @@
 #include "be/list.h"
 #include "be/alloc.h"
 #include "be/seg.h"
+#include "be/tx.h"		/* m0_be_tx */
 #include "be/tx_credit.h"
 #include "be/be.h"  /* m0_be_op_state */
 
diff --git a/be/seg_dict.c b/be/seg_dict.c
index 8d46b54..dc4c788 100644
--- a/be/seg_dict.c
+++ b/be/seg_dict.c
@@ -25,6 +25,7 @@
 
 #include "be/seg.h"
 #include "be/seg_internal.h"  /* m0_be_seg_hdr */
+#include "be/tx.h"	      /* m0_be_tx */
 
 /**
  * @addtogroup be
-- 
1.8.3.2

