From f248b96a0e1a92ce1965c6a518d2572e9c248249 Mon Sep 17 00:00:00 2001
From: Maxim Medved <max_medved@xyratex.com>
Date: Thu, 26 Sep 2013 05:13:49 +0300
Subject: [PATCH 4/5] be/tx: added missing forward declaration

---
 be/tx.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/be/tx.h b/be/tx.h
index 4974f95..644839a 100644
--- a/be/tx.h
+++ b/be/tx.h
@@ -29,6 +29,7 @@
 struct m0_ref;
 struct m0_be_domain;
 struct m0_be_tx_group;
+struct m0_be_tx;
 
 /**
  * @defgroup be
-- 
1.8.3.2

