From 5c9e9c8be2357a433def5b1535c2486e09a72deb Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Mon, 17 Jan 2011 06:18:36 -0700
Subject: [PATCH 004/177] - some more mdstore code.

---
 mdstore/mdstore.c | 5 +++++
 mdstore/mdstore.h | 2 ++
 reqh/reqh.c       | 2 ++
 3 files changed, 9 insertions(+)

diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index d6a9a21..d24d745 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -17,6 +17,11 @@ int c2_md_store_init(struct c2_md_store *store, struct c2_cob_domain *cob)
         return 0;
 }
 
+int c2_md_store_handle(struct c2_md_store *store, struct c2_fop *fop)
+{
+        return 0;
+}
+
 void c2_md_store_fini(struct c2_md_store *store)
 {
 }
diff --git a/mdstore/mdstore.h b/mdstore/mdstore.h
index 222ba31..cdb70c5 100644
--- a/mdstore/mdstore.h
+++ b/mdstore/mdstore.h
@@ -4,12 +4,14 @@
 #define __COLIBRI_MDSTORE_MDSTORE_H__
 
 struct c2_cob_domain;
+struct c2_fop;
 
 struct c2_md_store {
         struct c2_cob_domain  *md_cob;
 };
 
 int c2_md_store_init(struct c2_md_store *store, struct c2_cob_domain *cob);
+int c2_md_store_handle(struct c2_md_store *store, struct c2_fop *fop);
 void c2_md_store_fini(struct c2_md_store *store);
 
 /* __COLIBRI_MDSTORE_MDSTORE_H__ */
diff --git a/reqh/reqh.c b/reqh/reqh.c
index f643f2c..ba3fa21 100644
--- a/reqh/reqh.c
+++ b/reqh/reqh.c
@@ -194,6 +194,8 @@ void c2_reqh_fop_handle(struct c2_reqh *reqh, struct c2_fop *fop)
 	fop->f_type->ft_ops->fto_fom_init(fop, &fom);
 	fom->fm_ops->fmo_run(fom);
 #endif
+        if (reqh->rh_md)
+                c2_md_store_handle(reqh->rh_md, fop);
 }
 
 void c2_reqh_fop_sortkey_get(struct c2_reqh *reqh, struct c2_fop *fop,
-- 
1.8.3.2

