From a2123ce46d7b736c9caa1af12c3bef4dc5dbb8e6 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Tue, 22 Mar 2011 06:52:27 -0600
Subject: [PATCH 024/177] - added comments to mdstore API

---
 mdstore/mdstore.h | 77 +++++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 77 insertions(+)

diff --git a/mdstore/mdstore.h b/mdstore/mdstore.h
index 7ffbb21..0c0c8d8 100644
--- a/mdstore/mdstore.h
+++ b/mdstore/mdstore.h
@@ -16,71 +16,148 @@ struct c2_md_store {
         struct c2_fid         md_rootid;
 };
 
+/**
+   Make in-memory fid from fop fid.
+*/
 void md_store_make_fid(struct c2_fid *fid, 
                        struct c2_fop_fid *ffid);
 
+/**
+   Make nskey from passed parent fid and child name.
+*/
 void md_store_make_nskey(struct c2_cob_nskey **keyh, 
                          struct c2_fop_fid *fid, 
                          struct c2_fop_name *name);
 
+/**
+   Make oikey from passed child fid and liunk number.
+*/
 void md_store_make_oikey(struct c2_cob_oikey *oikey,
                          struct c2_fop_fid *fid,
                          int linkno);
 
+/**
+   Initialize mdstore with passed cob comain id and database env.
+*/
 int c2_md_store_init(struct c2_md_store *md, 
                      struct c2_cob_domain_id *id,
                      struct c2_dbenv *db);
 
+/**
+   Finalize mdstore instance.
+*/
 void c2_md_store_fini(struct c2_md_store *md);
 
+/**
+   Handle link operation described by @req on @cob. Input @cob
+   is so called statdata cob and returned by c2_cob_locate(). 
+   Save result to @rep.
+   
+   Error code is returned in error case or zero otherwise.
+*/
 int c2_md_store_link(struct c2_md_store *md, 
                      struct c2_cob *cob, 
                      struct c2_fop_link *req, 
                      struct c2_fop_link_rep *rep, 
                      struct c2_fop_ctx *ctx);
                      
+/**
+   Handle unlink operation described by @req on @cob. Input @cob
+   is so called statdata cob and returned by c2_cob_locate(). 
+   Save result to @rep.
+   
+   Error code is returned in error case or zero otherwise.
+*/
 int c2_md_store_unlink(struct c2_md_store *md, 
                        struct c2_cob *cob,
                        struct c2_fop_unink *req, 
                        struct c2_fop_unink_rep *rep, 
                        struct c2_fop_ctx *ctx);
                        
+/**
+   Handle rename operation described by @req on @cob. Input @cob
+   is so called statdata cob of source and returned by c2_cob_locate(). 
+   Save result to @rep.
+   
+   Error code is returned in error case or zero otherwise.
+*/
 int c2_md_store_rename(struct c2_md_store *md, 
                        struct c2_cob *cob,
                        struct c2_fop_rename *req, 
                        struct c2_fop_rename_rep *rep, 
                        struct c2_fop_ctx *ctx);
 
+/**
+   Handle create operation described by @req on @cob. Input @cob
+   is returned by c2_cob_alloc(). Save result to @rep.
+   
+   Error code is returned in error case or zero otherwise.
+*/
 int c2_md_store_create(struct c2_md_store *md, 
                        struct c2_cob *cob,
                        struct c2_fop_create *req, 
                        struct c2_fop_create_rep *rep, 
                        struct c2_fop_ctx *ctx);
 
+/**
+   Handle open operation described by @req on @cob. Input @cob
+   is so called statdata cob and returned by c2_cob_locate(). 
+   Save result to @rep.
+   
+   Error code is returned in error case or zero otherwise.
+*/
 int c2_md_store_open(struct c2_md_store *md, 
                      struct c2_cob *cob,
                      struct c2_fop_open *req, 
                      struct c2_fop_open_rep *rep, 
                      struct c2_fop_ctx *ctx);
 
+/**
+   Handle close operation described by @req on @cob. Input @cob
+   is so called statdata cob and returned by c2_cob_locate(). 
+   Save result to @rep. Orphans handled.
+   
+   Error code is returned in error case or zero otherwise.
+*/
 int c2_md_store_close(struct c2_md_store *md, 
                       struct c2_cob *cob,
                       struct c2_fop_close *req, 
                       struct c2_fop_close_rep *rep, 
                       struct c2_fop_ctx *ctx);
 
+/**
+   Handle setattr operation described by @req on @cob. Input @cob
+   is so called statdata cob and returned by c2_cob_locate(). 
+   Save result to @rep.
+   
+   Error code is returned in error case or zero otherwise.
+*/
 int c2_md_store_setattr(struct c2_md_store *md, 
                         struct c2_cob *cob,
                         struct c2_fop_setattr *req, 
                         struct c2_fop_setattr_rep *rep, 
                         struct c2_fop_ctx *ctx);
 
+/**
+   Handle getattr operation described by @req on @cob. Input @cob
+   is so called statdata cob and returned by c2_cob_locate(). 
+   Save result to @rep.
+   
+   Error code is returned in error case or zero otherwise.
+*/
 int c2_md_store_getattr(struct c2_md_store *md, 
                         struct c2_cob *cob,
                         struct c2_fop_getattr *req, 
                         struct c2_fop_getattr_rep *rep, 
                         struct c2_fop_ctx *ctx);
 
+/**
+   Handle readdir operation described by @req on @cob. Input @cob
+   is so called statdata cob and returned by c2_cob_locate(). 
+   Save result to @rep.
+   
+   Error code is returned in error case or zero otherwise.
+*/
 int c2_md_store_readdir(struct c2_md_store *md, 
                         struct c2_cob *cob,
                         struct c2_fop_readdir *req, 
-- 
1.8.3.2

