From f5e365821bfae60f0d3246c9343a17e415024bec Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Wed, 6 Apr 2011 05:34:20 -0600
Subject: [PATCH 039/177] - open root cob in md_store init time.

---
 mdstore/mdstore.c | 56 ++++++++++++++++++++++++++++++++++++++++++-------------
 mdstore/mdstore.h |  4 +++-
 2 files changed, 46 insertions(+), 14 deletions(-)

diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 47be046..f2cd35d 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -31,7 +31,7 @@ static const struct c2_addb_loc mdstore_addb_loc = {
 	.al_name = "mdstore"
 };
 
-static const struct c2_fid C2_MD_ROOTID = {
+static struct c2_fid C2_MD_ROOTID = {
         .f_container = 1024ULL, 
         .f_key       = 1ULL
 };
@@ -56,23 +56,55 @@ static void c2_md_store_make_nskey(struct c2_cob_nskey **keyh,
         *keyh = key;
 }
 
+static int md_store_add_name(struct c2_cob *cob,
+                             struct c2_fid *pfid,
+                             const char *name,
+                             int nlen,
+                             struct c2_db_tx *tx)
+{
+        struct c2_cob_nskey   *nskey;
+        int                    rc;
+        
+        c2_md_store_make_nskey(&nskey, pfid, name, nlen);
+        rc = c2_cob_add_name(cob, nskey, tx);
+        c2_free(nskey);
+        return rc;
+}
+
 int c2_md_store_init(struct c2_md_store *md, struct c2_cob_domain_id *id,
                      struct c2_dbenv *db)
 {
-        int rc;
+        struct c2_db_tx tx;
+        int             rc;
         
         C2_SET0(md);
-        rc = c2_cob_domain_init(&md->md_dom, db, id);
-        if (rc)
-                return rc;
         c2_addb_ctx_init(&md->md_addb, &mdstore_addb_ctx, 
                          &md->md_dom.cd_dbenv->d_addb);
-        md->md_rootid = C2_MD_ROOTID;
+        rc = c2_cob_domain_init(&md->md_dom, db, id);
+        if (rc) {
+                C2_ADDB_ADD(&md->md_addb, &mdstore_addb_loc, 
+                            c2_addb_func_fail, "cob_domain_init", rc);
+                return rc;
+        }
+        c2_db_tx_init(&tx, db, 0);
+        rc = c2_md_store_locate(md, &C2_MD_ROOTID, &md->md_root, 0, &tx);
+        if (rc) {
+                c2_db_tx_abort(&tx);
+        } else {
+                c2_db_tx_commit(&tx);
+        }
+        C2_ADDB_ADD(&md->md_addb, &mdstore_addb_loc, 
+                    c2_addb_func_fail, "md_root_locate", rc);
+        if (rc) {
+	        c2_addb_ctx_fini(&md->md_addb);
+                c2_cob_domain_fini(&md->md_dom);
+        }
         return rc;
 }
 
 void c2_md_store_fini(struct c2_md_store *md)
 {
+        c2_cob_put(md->md_root);
         c2_cob_domain_fini(&md->md_dom);
 	c2_addb_ctx_fini(&md->md_addb);
 }
@@ -152,9 +184,8 @@ int c2_md_store_create(struct c2_md_store *md,
                 /*
                  * Add direntry "." to empty dir
                  */
-                c2_md_store_make_nskey(&nskey, &cob->co_fid, ".", 1);
-                rc = c2_cob_add_name(cob, nskey, &ctx->fc_tx->tx_dbtx);
-                c2_free(nskey);
+                rc = md_store_add_name(cob, &cob->co_fid, ".", 1, 
+                                       &ctx->fc_tx->tx_dbtx);
                 if (rc)
                         goto out;
 
@@ -167,9 +198,8 @@ int c2_md_store_create(struct c2_md_store *md,
                 if (rc)
                         goto out;
 
-                c2_md_store_make_nskey(&nskey, &cob->co_fid, "..", 2);
-                rc = c2_cob_add_name(pcob, nskey, &ctx->fc_tx->tx_dbtx);
-                c2_free(nskey);
+                rc = md_store_add_name(pcob, &cob->co_fid, "..", 2, 
+                                       &ctx->fc_tx->tx_dbtx);
                 if (rc) {
                         c2_cob_put(pcob);
                         goto out;
@@ -366,7 +396,7 @@ static int c2_md_store_rename_sanity(struct c2_md_store *md,
          */
         do {
                 /* Is root reached? */
-                if (c2_fid_eq(&md->md_rootid, &tgtfid))
+                if (c2_fid_eq(&md->md_root->co_fid, &tgtfid))
                         return 0;
                 
                 /*
diff --git a/mdstore/mdstore.h b/mdstore/mdstore.h
index cce9560..fad5ba5 100644
--- a/mdstore/mdstore.h
+++ b/mdstore/mdstore.h
@@ -12,10 +12,12 @@ struct c2_fop_ctx;
 struct c2_dbenv;
 struct c2_fid;
 struct c2_fop;
+struct c2_cob;
 
 struct c2_md_store {
         struct c2_cob_domain  md_dom;
-        struct c2_fid         md_rootid;
+        struct c2_cob        *md_root;
+
         /** 
            An ADDB context for events related to this store.
         */
-- 
1.8.3.2

