From 20e15d65cf48b1e5b45166b224b915d940e31e1e Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Thu, 14 Apr 2011 05:44:00 -0600
Subject: [PATCH 062/177] - more todo comments.

---
 mdservice/md_foms.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mdservice/md_foms.c b/mdservice/md_foms.c
index c1ee3ef..779172b 100644
--- a/mdservice/md_foms.c
+++ b/mdservice/md_foms.c
@@ -172,7 +172,8 @@ static int c2_md_link_fom_sanity(struct c2_fom *fom)
                 
         /*
          * @todo:
-         * 1. Check for MAY_WRITE and MAY_EXEC in parent.
+         * 1. Check for MAY_WRITE and MAY_EXEC in parent;
+         * 2. Check for immutable and append target.
          */
         c2_cob_put(cob);
         return rc;
-- 
1.8.3.2

