From 96a699be89f49d3e08c716745eea9418d3e17816 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Mon, 13 Jun 2011 07:18:05 -0600
Subject: [PATCH 127/177] - fixed unlink

---
 mdservice/md_foms.c | 15 +++++++--------
 1 file changed, 7 insertions(+), 8 deletions(-)

diff --git a/mdservice/md_foms.c b/mdservice/md_foms.c
index d5c33c6..a1fa32b 100644
--- a/mdservice/md_foms.c
+++ b/mdservice/md_foms.c
@@ -420,14 +420,13 @@ static int c2_md_unlink_fom_state(struct c2_fom *fom)
                 /*
                  * Name exists but it has older version. Let's kill it.
                  */
-                if (attr.ca_version > scob->co_nsrec.cnr_version) {
-                        rc = c2_md_store_unlink(md, &pfid, scob, 
-                                                req->u_name.s_buf,
-                                                req->u_name.s_len,
-                                                tx);
-                        if (rc == 0 && scob->co_nsrec.cnr_nlink > 0)
-                                rc = c2_md_store_setattr(md, scob, &attr, tx);
-                }
+                rc = c2_md_store_unlink(md, &pfid, scob, 
+                                        req->u_name.s_buf,
+                                        req->u_name.s_len,
+                                        tx);
+                if (rc == 0 && scob->co_nsrec.cnr_nlink > 0 &&
+                    attr.ca_version > scob->co_nsrec.cnr_version)
+                        rc = c2_md_store_setattr(md, scob, &attr, tx);
                 c2_cob_put(scob);
         }
         
-- 
1.8.3.2

