From 298f49d045d83ea8059926ccdf52c6d2631dcc05 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Thu, 16 Jun 2011 04:59:53 -0600
Subject: [PATCH 141/177] - wake up nrs on shutdown

---
 nrs/nrs.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/nrs/nrs.c b/nrs/nrs.c
index 4419089..bbace76 100644
--- a/nrs/nrs.c
+++ b/nrs/nrs.c
@@ -69,14 +69,13 @@ void c2_nrs_fini(struct c2_nrs *nrs)
 {
 	c2_mutex_lock(&nrs->s_guard);
 	nrs->s_shutdown = true;
+        c2_cond_signal(&nrs->s_gotwork, &nrs->s_guard);
 	c2_mutex_unlock(&nrs->s_guard);
 
         if (nrs->s_thread.t_func != NULL) {
                 c2_thread_join(&nrs->s_thread);
                 c2_thread_fini(&nrs->s_thread);
         }
-        nrs->s_reqh = NULL;
-        nrs->s_pol = NULL;
 }
 
 void c2_nrs_enqueue(struct c2_nrs *nrs, struct c2_fop *fop, void *cookie)
-- 
1.8.3.2

