From f6fd259cf0d26df4a12190d3b7960bacacd00f53 Mon Sep 17 00:00:00 2001
From: Andrew Perepechko <andrew_perepechko@xyratex.com>
Date: Thu, 14 Jul 2011 20:06:28 +0400
Subject: [PATCH 163/177] disabled invariant assertion due to serious
 performance reasons

---
 lib/queue.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/lib/queue.c b/lib/queue.c
index b1c28cc..dfb2a1f 100644
--- a/lib/queue.c
+++ b/lib/queue.c
@@ -50,18 +50,24 @@ const struct c2_queue C2_QUEUE_INIT = {
 void c2_queue_init(struct c2_queue *q)
 {
 	q->q_head = q->q_tail = EOQ;
+#if 0
 	C2_ASSERT(c2_queue_invariant(q));
+#endif
 }
 
 void c2_queue_fini(struct c2_queue *q)
 {
+#if 0
 	C2_ASSERT(c2_queue_invariant(q));
+#endif
 	C2_ASSERT(c2_queue_is_empty(q));
 }
 
 bool c2_queue_is_empty(const struct c2_queue *q)
 {
+#if 0
 	C2_ASSERT(c2_queue_invariant(q));
+#endif
 	return q->q_head == EOQ;
 }
 
@@ -85,7 +91,9 @@ bool c2_queue_contains(const struct c2_queue *q,
 {
 	struct c2_queue_link *scan;
 
+#if 0
 	C2_ASSERT(c2_queue_invariant(q));
+#endif
 	for (scan = q->q_head; scan != EOQ; scan = scan->ql_next) {
 		C2_ASSERT(scan != NULL);
 		if (scan == ql)
@@ -99,7 +107,9 @@ size_t c2_queue_length(const struct c2_queue *q)
 	size_t length;
 	struct c2_queue_link *scan;
 
+#if 0
 	C2_ASSERT(c2_queue_invariant(q));
+#endif
 
 	for (length = 0, scan = q->q_head; scan != EOQ; scan = scan->ql_next)
 		++length;
@@ -120,7 +130,9 @@ struct c2_queue_link *c2_queue_get(struct c2_queue *q)
 			q->q_tail = EOQ;
 		head->ql_next = NULL;
 	}
+#if 0
 	C2_ASSERT(c2_queue_invariant(q));
+#endif
 	return head;
 		
 }
@@ -134,7 +146,9 @@ void c2_queue_put(struct c2_queue *q, struct c2_queue_link *ql)
 		q->q_tail->ql_next = ql;
 	q->q_tail = ql;
 	ql->ql_next = EOQ;
+#if 0
 	C2_ASSERT(c2_queue_invariant(q));
+#endif
 }
 
 bool c2_queue_invariant(const struct c2_queue *q)
-- 
1.8.3.2

