From 8fa42444ac5f96bf33e8a82e5521e9cc95ecd750 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Tue, 19 Jul 2011 07:40:01 -0600
Subject: [PATCH 165/177] - fixed returning error on unexpended case.

---
 mdservice/md_foms.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mdservice/md_foms.c b/mdservice/md_foms.c
index 6a4de8a..98fe6ee 100644
--- a/mdservice/md_foms.c
+++ b/mdservice/md_foms.c
@@ -1133,7 +1133,7 @@ int c2_md_req_fom_init(struct c2_fop *fop,
 	        break;
         default:
                 c2_free(fom_obj);
-                return rc;
+                return -EOPNOTSUPP;
         }
 
         fom_obj->fm_fop_rep = c2_fop_alloc(fop_type, NULL);
-- 
1.8.3.2

