From 5bca3a5f32537d848b8f7d92060faaf5f3c8994f Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Wed, 22 Feb 2012 02:44:53 +0400
Subject: [PATCH 173/177] more merge changes

---
 lib/Makefile.am | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/lib/Makefile.am b/lib/Makefile.am
index 53f88e6..f75cdc4 100644
--- a/lib/Makefile.am
+++ b/lib/Makefile.am
@@ -33,12 +33,6 @@ libc2_HEADERS        = \
                        ut.h \
                        vec.h
 
-if HAVE_CUNIT_SUPPORT
-UT_SOURCES           = ut.c ut.h ub.c ub.h 
-else
-UT_SOURCES           =
-endif
-
 noinst_LTLIBRARIES   = libc2.la
 libc2_la_SOURCES     = \
                        adt.h \
-- 
1.8.3.2

