From 57c6079c4aaf91c41ce825b0671d34b453e10e51 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Wed, 22 Feb 2012 13:42:22 +0400
Subject: [PATCH 175/177] - more merge changes

---
 ioservice/io_foms.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ioservice/io_foms.c b/ioservice/io_foms.c
index 2f8711d..e6b58d6 100644
--- a/ioservice/io_foms.c
+++ b/ioservice/io_foms.c
@@ -321,7 +321,7 @@ static int io_fom_cob_rwv_state(struct c2_fom *fom)
 
 static void io_fom_cob_rwv_fini(struct c2_fom *fom)
 {
-	struct c2_io_fom_cob_rwv *fom_obj;
+	struct c2_io_fom_cob_rwv *fom_ctx;
 
 	fom_ctx = container_of(fom, struct c2_io_fom_cob_rwv, fcrw_gen);
 	c2_free(fom_ctx);
-- 
1.8.3.2

