From 0644f3d5332169d588c91938058e333864772fc2 Mon Sep 17 00:00:00 2001
From: Nikita Danilov <nikita_danilov@xyratex.com>
Date: Tue, 28 May 2013 22:35:22 +0300
Subject: [PATCH 015/161] be: add m0_be_op_tick_ret() prototype.

---
 be/seg.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/be/seg.h b/be/seg.h
index cece0b4..b7f8901 100644
--- a/be/seg.h
+++ b/be/seg.h
@@ -322,5 +322,12 @@ void m0_be_btree_release(struct m0_be_btree *tree,
 /** Waits for the operation to complete and returns its rc. */
 int m0_be_op_wait(struct m0_be_op *op);
 
+/**
+ * Moves the fom to the "next_state" and arranges for state transitions to
+ * continue when "op" completes. Returns value suitable to be returned from
+ * m0_fom_ops::fo_tick() implementation.
+ */
+int m0_be_op_tick_ret(struct m0_be_op *op, struct m0_fom *fom, int next_state);
+
 /** @} end of seg group */
 #endif /* __MERO_BE_SEG_H__ */
-- 
1.8.3.2

