From b924de20231edfc775c826ce34e710369ff579ad Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Wed, 29 May 2013 13:01:20 +0300
Subject: [PATCH 023/161] be: remove M0_UT_ASSERT() from seg_header_write()

---
 be/seg.c | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/be/seg.c b/be/seg.c
index f4d4b96..344e693 100644
--- a/be/seg.c
+++ b/be/seg.c
@@ -50,19 +50,18 @@ static void seg_header_prepare(struct m0_be_seg *seg, struct m0_stob_io *io)
 
 static int seg_header_write(struct m0_be_seg *seg)
 {
-	int		   result;
-	struct m0_stob_io  io;
-	struct m0_clink    clink;
+	struct m0_stob_io io;
+	struct m0_clink   clink;
+	int               rc;
 
 	m0_stob_io_init(&io);
-
 	seg_header_prepare(seg, &io);
 
 	m0_clink_init(&clink, NULL);
 	m0_clink_add(&io.si_wait, &clink);
 
-	result = m0_stob_io_launch(&io, seg->bs_stob, NULL, NULL);
-	M0_UT_ASSERT(result == 0);
+	rc = m0_stob_io_launch(&io, seg->bs_stob, NULL, NULL);
+	M0_ASSERT(rc != 0); /* XXX FIXME */
 
 	m0_chan_wait(&clink);
 	m0_clink_del(&clink);
-- 
1.8.3.2

