From 4c5f1d9bac145af4289542d521a82a5614f21208 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Wed, 29 May 2013 14:03:30 +0300
Subject: [PATCH 036/161] changed m0_be_seg_create() in UT

---
 be/ut/seg.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/be/ut/seg.c b/be/ut/seg.c
index f8cfae4..a849b27 100644
--- a/be/ut/seg.c
+++ b/be/ut/seg.c
@@ -28,6 +28,7 @@
 enum {
 	BE_SEG_UT_DOM_ID = 42,
 	BE_SEG_UT_STOB_ID = 42,
+	BE_SEG_UT_SEG_SIZE = 0x10000,
 };
 
 static struct m0_stob_domain	dom;
@@ -71,7 +72,7 @@ static void test_create_destroy(void)
 	int		 rc;
 
 	seg_init_helper(&seg);
-	rc = m0_be_seg_create(&seg);
+	rc = m0_be_seg_create(&seg, BE_SEG_UT_SEG_SIZE);
 	M0_UT_ASSERT(rc == 0);
 	rc = m0_be_seg_destroy(&seg);
 	M0_UT_ASSERT(rc == 0);
@@ -84,7 +85,7 @@ static void test_open_close(void)
 	int		 rc;
 
 	seg_init_helper(&seg);
-	rc = m0_be_seg_create(&seg);
+	rc = m0_be_seg_create(&seg, BE_SEG_UT_SEG_SIZE);
 	M0_UT_ASSERT(rc == 0);
 	rc = m0_be_seg_open(&seg);
 	M0_UT_ASSERT(rc == 0);
-- 
1.8.3.2

