From 519490e9f015a5a48a777d50f1579532825b6901 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Wed, 29 May 2013 17:24:59 +0300
Subject: [PATCH 046/161] be/seg: use offs parameter in stob_io_single_read()

---
 be/seg.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/be/seg.c b/be/seg.c
index 95077dd..0494e57 100644
--- a/be/seg.c
+++ b/be/seg.c
@@ -158,7 +158,7 @@ static int stob_io_single_read(void *buf, m0_bcount_t bufsz,
 	io.si_user.ov_buf         = rdbufpk;
 	io.si_user.ov_vec.v_nr    = 1;
 	io.si_user.ov_vec.v_count = rdcnt;
-	rdoffs[0]                 = 0;  /* header offset in STOB */
+	rdoffs[0]                 = m0_stob_addr_pack(offs, pgshift);
 	io.si_stob.iv_index       = rdoffs;
 	io.si_stob.iv_vec.v_nr    = 1;
 	io.si_stob.iv_vec.v_count = rdcnt;
-- 
1.8.3.2

