From aad05ffda9d446446ff6e900b82430dcc12d9a29 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Thu, 30 May 2013 19:18:14 +0300
Subject: [PATCH 073/161] be: seg UT working now

---
 be/seg.c    | 2 ++
 be/ut/seg.c | 1 +
 2 files changed, 3 insertions(+)

diff --git a/be/seg.c b/be/seg.c
index b3772f5..92d4472 100644
--- a/be/seg.c
+++ b/be/seg.c
@@ -250,6 +250,8 @@ M0_INTERNAL int m0_be_seg_open(struct m0_be_seg *seg)
 	/* Read whole segment from storage. */
 	rc = stob_io_single_read(seg_addr0, seg_size, seg->bs_stob,
 				 0, st_block_shift);
+	if (rc == 0)
+		seg->bs_state = M0_BSS_OPENED;
 	return rc;
 #else
 	return -EOPNOTSUPP;
diff --git a/be/ut/seg.c b/be/ut/seg.c
index 6b756c4..16a5eff 100644
--- a/be/ut/seg.c
+++ b/be/ut/seg.c
@@ -94,6 +94,7 @@ static void seg_fini_helper(struct m0_be_seg *seg, bool stob_put)
 	if (stob_put)
 		m0_stob_put(stob);
 	m0_dtx_fini(&dtx);
+	m0_stob_domain_fini(dom);
 }
 
 static void test_init_fini(void)
-- 
1.8.3.2

