From c6ec3aa9f0a3dab97b15a4ba6653f38fc9d81e81 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Tue, 4 Jun 2013 19:16:06 +0300
Subject: [PATCH 100/161] be/seg: document m0_be__reg_is_pinned() function.

---
 be/seg.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/be/seg.h b/be/seg.h
index 8fa00ba..3fe9619 100644
--- a/be/seg.h
+++ b/be/seg.h
@@ -119,6 +119,7 @@ M0_INTERNAL void m0_be_reg_get_fast(const struct m0_be_reg *reg);
  */
 M0_INTERNAL void m0_be_reg_put(const struct m0_be_reg *reg);
 
+/* Returns true iff all region's pages are pinned. */
 M0_INTERNAL bool m0_be__reg_is_pinned(const struct m0_be_reg *reg);
 
 M0_INTERNAL bool m0_be__reg_invariant(const struct m0_be_reg *reg);
-- 
1.8.3.2

