From 5a5e753f7afd5a4e982a2d2d8096eba5b426aa5a Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Thu, 6 Jun 2013 15:40:15 +0300
Subject: [PATCH 110/161] be/list: implement m0_be_list_credit()

---
 be/list.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/be/list.c b/be/list.c
index 8a4230f..d1b93db 100644
--- a/be/list.c
+++ b/be/list.c
@@ -123,6 +123,32 @@ M0_INTERNAL void m0_be_list_put(const struct m0_be_list *list,
 	}
 }
 
+M0_INTERNAL void m0_be_list_credit(const struct m0_be_list *list,
+				   enum m0_be_list_op optype,
+				   m0_bcount_t nr,
+				   struct m0_be_tx_credit *accum)
+{
+	struct m0_be_tx_credit k;
+	M0_PRE(list != NULL && list->bl_descr != NULL);
+
+	switch (optype) {
+	case M0_BLO_INSERT:
+	case M0_BLO_DELETE:
+		k.tc_reg_nr = 4;
+		k.tc_reg_size = sizeof(struct m0_be_list) +
+			2*sizeof(void *) + list->bl_descr->td_container_size;
+		m0_be_tx_credit_add(accum, &k);
+		break;
+	case M0_BLO_MOVE:
+	default:
+		k.tc_reg_nr = 6;
+		k.tc_reg_size = sizeof(struct m0_be_list) +
+			4*sizeof(void *) + list->bl_descr->td_container_size;
+		m0_be_tx_credit_add(accum, &k);
+		break;
+	}
+}
+
 /*
  * Neither m0_be_list_init() nor m0_be_list_fini() will ever do anything.
  * We keep the interfaces for consistency.
-- 
1.8.3.2

