From 0227009ee85366b18a656cf4c5817a974a0d1533 Mon Sep 17 00:00:00 2001
From: Maxim Medved <Max_Medved@xyratex.com>
Date: Thu, 6 Jun 2013 05:02:47 +0300
Subject: [PATCH 112/161] added op to alloc in be btree UT

---
 be/ut/btree.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/be/ut/btree.c b/be/ut/btree.c
index 699c7f0..a309689 100644
--- a/be/ut/btree.c
+++ b/be/ut/btree.c
@@ -23,6 +23,7 @@
 #include "be/seg.h"
 #include "be/btree.h"
 #include "be/alloc.h"
+#include "be/ut/helper.h"
 #include "lib/misc.h"
 #include "lib/memory.h"
 
@@ -162,10 +163,14 @@ void test_insert(void)
 	m0_be_tx_open(&tx);
 
 	for (i = 0; i < INSERT_NR; ++i) {
-		m0_buf_init(&key, m0_be_alloc(alloc, &tx, ALLOC_SIZE,
+		m0_buf_init(&key, m0_be_alloc(alloc, &tx, &op, ALLOC_SIZE,
 					      ALLOC_SHIFT), ALLOC_SIZE);
-		m0_buf_init(&val, m0_be_alloc(alloc, &tx, ALLOC_SIZE,
+		M0_UT_ASSERT(M0_IN(m0_be_op_state(&op), (M0_BOS_SUCCESS,
+							 M0_BOS_FAILURE)));
+		m0_buf_init(&val, m0_be_alloc(alloc, &tx, &op, ALLOC_SIZE,
 					      ALLOC_SHIFT), ALLOC_SIZE);
+		M0_UT_ASSERT(M0_IN(m0_be_op_state(&op), (M0_BOS_SUCCESS,
+							 M0_BOS_FAILURE)));
 		M0_UT_ASSERT(key.b_addr != NULL);
 		M0_UT_ASSERT(val.b_addr != NULL);
 
@@ -211,8 +216,10 @@ void test_update(void)
 	m0_be_tx_open(&tx);
 
 	for (i = 0; i < UPDATE_NR; ++i) {
-		m0_buf_init(&key, m0_be_alloc(alloc, &tx, ALLOC_SIZE,
+		m0_buf_init(&key, m0_be_alloc(alloc, &tx, &op, ALLOC_SIZE,
 					      ALLOC_SHIFT), ALLOC_SIZE);
+		M0_UT_ASSERT(M0_IN(m0_be_op_state(&op), (M0_BOS_SUCCESS,
+							 M0_BOS_FAILURE)));
 		M0_UT_ASSERT(key.b_addr != NULL);
 		sprintf(key.b_addr, "%03d", i);
 
-- 
1.8.3.2

