From ad613ed7a3893104b6119de622bdd4199e51b601 Mon Sep 17 00:00:00 2001
From: Anatoliy Bilenko <anatoliy_bilenko@xyratex.com>
Date: Wed, 12 Jun 2013 11:18:40 +0300
Subject: [PATCH 141/161] be: tx: Minor fixes

---
 be/tx.c | 4 +++-
 be/tx.h | 2 --
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/be/tx.c b/be/tx.c
index 06593a7..9bbccf8 100644
--- a/be/tx.c
+++ b/be/tx.c
@@ -289,7 +289,7 @@ m0_be_tx_capture(struct m0_be_tx *tx, const struct m0_be_reg *reg)
 	m0_bindex_t             idx = pos->tc_reg_nr;
 	struct m0_be_reg_d     *new = &tx->t_reg_d_area[idx];
 	struct m0_be_reg_d     *old;
-	struct m0_be_reg_d     *prev;
+	struct m0_be_reg_d     *prev = NULL;
 
 	M0_PRE(m0_be__tx_invariant(tx));
 	M0_PRE(tx_state(tx) == M0_BTS_ACTIVE);
@@ -565,6 +565,8 @@ static bool credit_le(const struct m0_be_tx_credit *c0,
 static void credit_mod(struct m0_be_tx_credit *cr,
 		       const struct m0_be_reg_d *rd, int sgn)
 {
+	M0_PRE(M0_IN(sgn, (+1, -1)));
+
 	cr->tc_reg_nr += sgn;
 	cr->tc_reg_size += sgn * rd->rd_reg.br_size;
 
diff --git a/be/tx.h b/be/tx.h
index 9ab78e0..e4e135f 100644
--- a/be/tx.h
+++ b/be/tx.h
@@ -586,8 +586,6 @@ struct m0_be_tx {
 	 *
 	 * The tree is maintained by glibc tree functions from <search.h>.
 	 */
-	void                  *t_tree;
-	/*XXX*/
 	void                  *t_root;
 };
 
-- 
1.8.3.2

